// @aztec/simulator/dest/public/avm/opcodes/contract.js
import { Field, TypeTag, Uint1 } from '../avm_memory_types.js';
import { InstructionExecutionError } from '../errors.js';
import { Opcode, OperandType } from '../serialization/instruction_serialization.js';
import { Addressing } from './addressing_mode.js';
import { Instruction } from './instruction.js';
export var ContractInstanceMember = /*#__PURE__*/ function(ContractInstanceMember) {
    ContractInstanceMember[ContractInstanceMember["DEPLOYER"] = 0] = "DEPLOYER";
    ContractInstanceMember[ContractInstanceMember["CLASS_ID"] = 1] = "CLASS_ID";
    ContractInstanceMember[ContractInstanceMember["INIT_HASH"] = 2] = "INIT_HASH";
    return ContractInstanceMember;
}({});
export class GetContractInstance extends Instruction {
    addressingMode;
    addressOffset;
    dstOffset;
    memberEnum;
    static type = 'GETCONTRACTINSTANCE';
    static opcode = Opcode.GETCONTRACTINSTANCE;
    // Informs (de)serialization. See Instruction.deserialize.
    static wireFormat = [
        OperandType.UINT8,
        OperandType.UINT8,
        OperandType.UINT16,
        OperandType.UINT16,
        OperandType.UINT8
    ];
    constructor(addressingMode, addressOffset, dstOffset, memberEnum){
        super(), this.addressingMode = addressingMode, this.addressOffset = addressOffset, this.dstOffset = dstOffset, this.memberEnum = memberEnum;
    }
    async execute(context) {
        const memory = context.machineState.memory;
        const addressing = Addressing.fromWire(this.addressingMode);
        context.machineState.consumeGas(this.baseGasCost(addressing.indirectOperandsCount(), addressing.relativeOperandsCount()));
        if (!(this.memberEnum in ContractInstanceMember)) {
            throw new InstructionExecutionError(`Invalid GETCONSTRACTINSTANCE member enum ${this.memberEnum}`);
        }
        const operands = [
            this.addressOffset,
            this.dstOffset
        ];
        const [addressOffset, dstOffset] = addressing.resolve(operands, memory);
        memory.checkTag(TypeTag.FIELD, addressOffset);
        const address = memory.get(addressOffset).toAztecAddress();
        const instance = await context.persistableState.getContractInstance(address);
        const exists = instance !== undefined;
        let memberValue = new Field(0);
        if (exists) {
            switch(this.memberEnum){
                case 0:
                    memberValue = new Field(instance.deployer.toField());
                    break;
                case 1:
                    memberValue = new Field(instance.currentContractClassId.toField());
                    break;
                case 2:
                    memberValue = new Field(instance.initializationHash);
                    break;
            }
        }
        memory.setSlice(dstOffset, [
            new Uint1(exists ? 1 : 0),
            memberValue
        ]);
    }
}
