// @aztec/simulator/dest/public/avm/opcodes/instruction_impl.js
import { OperandType } from '../serialization/instruction_serialization.js';
import { Instruction } from './instruction.js';
/** Wire format that informs deserialization for instructions with three operands. */ export const ThreeOperandWireFormat8 = [
    OperandType.UINT8,
    OperandType.UINT8,
    OperandType.UINT8,
    OperandType.UINT8,
    OperandType.UINT8
];
export const ThreeOperandWireFormat16 = [
    OperandType.UINT8,
    OperandType.UINT8,
    OperandType.UINT16,
    OperandType.UINT16,
    OperandType.UINT16
];
/**
 * Covers (de)serialization for an instruction with:
 * addressing mode, inTag, and three operands.
 */ export class ThreeOperandInstruction extends Instruction {
    addressingMode;
    aOffset;
    bOffset;
    dstOffset;
    static wireFormat8 = ThreeOperandWireFormat8;
    static wireFormat16 = ThreeOperandWireFormat16;
    constructor(addressingMode, aOffset, bOffset, dstOffset){
        super(), this.addressingMode = addressingMode, this.aOffset = aOffset, this.bOffset = bOffset, this.dstOffset = dstOffset;
    }
}
