// @aztec/validator-ha-signer/dest/config.js
import { booleanConfigHelper, getConfigFromMappings, getDefaultConfig, numberConfigHelper, optionalNumberConfigHelper } from '@aztec/foundation/config';
import { EthAddress } from '@aztec/foundation/eth-address';
import { z } from 'zod';
export const validatorHASignerConfigMappings = {
    haSigningEnabled: {
        env: 'VALIDATOR_HA_SIGNING_ENABLED',
        description: 'Whether HA signing / slashing protection is enabled',
        ...booleanConfigHelper(false)
    },
    l1Contracts: {
        description: 'L1 contract addresses (rollup address required)',
        nested: {
            rollupAddress: {
                description: 'The Ethereum address of the rollup contract (must be set programmatically)',
                parseEnv: (val)=>EthAddress.fromString(val)
            }
        }
    },
    nodeId: {
        env: 'VALIDATOR_HA_NODE_ID',
        description: 'The unique identifier for this node',
        defaultValue: ''
    },
    pollingIntervalMs: {
        env: 'VALIDATOR_HA_POLLING_INTERVAL_MS',
        description: 'The number of ms to wait between polls when a duty is being signed',
        ...numberConfigHelper(100)
    },
    signingTimeoutMs: {
        env: 'VALIDATOR_HA_SIGNING_TIMEOUT_MS',
        description: 'The maximum time to wait for a duty being signed to complete',
        ...numberConfigHelper(3_000)
    },
    maxStuckDutiesAgeMs: {
        env: 'VALIDATOR_HA_MAX_STUCK_DUTIES_AGE_MS',
        description: 'The maximum age of a stuck duty in ms (defaults to 2x Aztec slot duration)',
        ...optionalNumberConfigHelper()
    },
    cleanupOldDutiesAfterHours: {
        env: 'VALIDATOR_HA_OLD_DUTIES_MAX_AGE_H',
        description: 'Optional: clean up old duties after this many hours (disabled if not set)',
        ...optionalNumberConfigHelper()
    },
    databaseUrl: {
        env: 'VALIDATOR_HA_DATABASE_URL',
        description: 'PostgreSQL connection string for validator HA signer (format: postgresql://user:password@host:port/database)'
    },
    poolMaxCount: {
        env: 'VALIDATOR_HA_POOL_MAX',
        description: 'Maximum number of clients in the pool',
        ...numberConfigHelper(10)
    },
    poolMinCount: {
        env: 'VALIDATOR_HA_POOL_MIN',
        description: 'Minimum number of clients in the pool',
        ...numberConfigHelper(0)
    },
    poolIdleTimeoutMs: {
        env: 'VALIDATOR_HA_POOL_IDLE_TIMEOUT_MS',
        description: 'Idle timeout in milliseconds',
        ...numberConfigHelper(10_000)
    },
    poolConnectionTimeoutMs: {
        env: 'VALIDATOR_HA_POOL_CONNECTION_TIMEOUT_MS',
        description: 'Connection timeout in milliseconds (0 means no timeout)',
        ...numberConfigHelper(0)
    }
};
export const defaultValidatorHASignerConfig = getDefaultConfig(validatorHASignerConfigMappings);
/**
 * Returns the validator HA signer configuration from environment variables.
 * Note: If an environment variable is not set, the default value is used.
 * @returns The validator HA signer configuration.
 */ export function getConfigEnvVars() {
    return getConfigFromMappings(validatorHASignerConfigMappings);
}
export const ValidatorHASignerConfigSchema = z.object({
    haSigningEnabled: z.boolean(),
    l1Contracts: z.object({
        rollupAddress: z.instanceof(EthAddress)
    }),
    nodeId: z.string(),
    pollingIntervalMs: z.number().min(0),
    signingTimeoutMs: z.number().min(0),
    maxStuckDutiesAgeMs: z.number().min(0).optional(),
    cleanupOldDutiesAfterHours: z.number().min(0).optional(),
    databaseUrl: z.string().optional(),
    poolMaxCount: z.number().min(0).optional(),
    poolMinCount: z.number().min(0).optional(),
    poolIdleTimeoutMs: z.number().min(0).optional(),
    poolConnectionTimeoutMs: z.number().min(0).optional()
});
