// @aztec/validator-ha-signer/dest/types.js
import { BlockNumber } from '@aztec/foundation/branded-types';
import { DutyType } from './db/types.js';
export { DutyStatus, DutyType, getBlockIndexFromDutyIdentifier, normalizeBlockIndex } from './db/types.js';
/**
 * Type guard to check if a SigningContext requires HA protection.
 * Returns true for contexts that need HA protection, false for AUTH_REQUEST and TXS.
 */ export function isHAProtectedContext(context) {
    return context.dutyType !== DutyType.AUTH_REQUEST && context.dutyType !== DutyType.TXS;
}
/**
 * Gets the block number from a signing context.
 * - Vote duties (GOVERNANCE_VOTE, SLASHING_VOTE): returns BlockNumber(0)
 * - Other duties: returns the blockNumber from the context
 */ export function getBlockNumberFromSigningContext(context) {
    // Check for duty types that have blockNumber
    if (context.dutyType === DutyType.BLOCK_PROPOSAL || context.dutyType === DutyType.CHECKPOINT_PROPOSAL || context.dutyType === DutyType.ATTESTATION || context.dutyType === DutyType.ATTESTATIONS_AND_SIGNERS) {
        return context.blockNumber;
    }
    // Vote duties (GOVERNANCE_VOTE, SLASHING_VOTE) don't have blockNumber
    return BlockNumber(0);
}
