// @aztec/stdlib/dest/vks/vk_data.js
import { VK_TREE_HEIGHT } from '@aztec/constants';
import { makeTuple } from '@aztec/foundation/array';
import { Fr } from '@aztec/foundation/curves/bn254';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { bufferToHex } from '@aztec/foundation/string';
import { VerificationKeyData } from './verification_key.js';
export class VkData {
    vk;
    leafIndex;
    siblingPath;
    constructor(vk, /**
     * Index of the vk in the vk tree.
     */ leafIndex, /**
     * Sibling path of the vk in the vk tree.
     */ siblingPath){
        this.vk = vk;
        this.leafIndex = leafIndex;
        this.siblingPath = siblingPath;
    }
    static empty() {
        return new VkData(VerificationKeyData.empty(), 0, makeTuple(VK_TREE_HEIGHT, Fr.zero));
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new VkData(reader.readObject(VerificationKeyData), reader.readNumber(), reader.readArray(VK_TREE_HEIGHT, Fr));
    }
    toBuffer() {
        return serializeToBuffer(this.vk, this.leafIndex, this.siblingPath);
    }
    toString() {
        return bufferToHex(this.toBuffer());
    }
}
