// @aztec/stdlib/dest/checkpoint/checkpoint_data.js
import { BlockNumberSchema, CheckpointNumberSchema } from '@aztec/foundation/branded-types';
import { schemas } from '@aztec/foundation/schemas';
import { z } from 'zod';
import { CommitteeAttestation } from '../block/proposal/committee_attestation.js';
import { CheckpointHeader } from '../rollup/checkpoint_header.js';
import { AppendOnlyTreeSnapshot } from '../trees/append_only_tree_snapshot.js';
import { L1PublishedData } from './published_checkpoint.js';
export const CheckpointDataSchema = z.object({
    checkpointNumber: CheckpointNumberSchema,
    header: CheckpointHeader.schema,
    archive: AppendOnlyTreeSnapshot.schema,
    checkpointOutHash: schemas.Fr,
    startBlock: BlockNumberSchema,
    blockCount: schemas.Integer,
    attestations: z.array(CommitteeAttestation.schema),
    l1: L1PublishedData.schema
}).transform((obj)=>({
        checkpointNumber: obj.checkpointNumber,
        header: obj.header,
        archive: obj.archive,
        checkpointOutHash: obj.checkpointOutHash,
        startBlock: obj.startBlock,
        blockCount: obj.blockCount,
        attestations: obj.attestations,
        l1: obj.l1
    }));
