// @aztec/protocol-contracts/dest/protocol_contract_data.js
// GENERATED FILE - DO NOT EDIT. RUN `yarn generate` or `yarn generate:data`
import { Fr } from '@aztec/foundation/curves/bn254';
import { AztecAddress } from '@aztec/stdlib/aztec-address';
import { ProtocolContracts } from '@aztec/stdlib/tx';
export const protocolContractNames = [
    'AuthRegistry',
    'ContractInstanceRegistry',
    'ContractClassRegistry',
    'MultiCallEntrypoint',
    'FeeJuice',
    'PublicChecks'
];
export const ProtocolContractSalt = {
    AuthRegistry: new Fr(1),
    ContractInstanceRegistry: new Fr(1),
    ContractClassRegistry: new Fr(1),
    MultiCallEntrypoint: new Fr(1),
    FeeJuice: new Fr(1),
    PublicChecks: new Fr(1)
};
export const ProtocolContractAddress = {
    AuthRegistry: AztecAddress.fromBigInt(1n),
    ContractInstanceRegistry: AztecAddress.fromBigInt(2n),
    ContractClassRegistry: AztecAddress.fromBigInt(3n),
    MultiCallEntrypoint: AztecAddress.fromBigInt(4n),
    FeeJuice: AztecAddress.fromBigInt(5n),
    PublicChecks: AztecAddress.fromBigInt(6n)
};
export const ProtocolContractDerivedAddress = {
    AuthRegistry: AztecAddress.fromString('0x139f8eb6d6e7e7a7c0322c3b7558687a7e24201f11bf2c4cb2fe56c18d363695'),
    ContractInstanceRegistry: AztecAddress.fromString('0x1254246c88aca5a66fa66f3aa78c408a698ebca3b713120497c7555dfc718592'),
    ContractClassRegistry: AztecAddress.fromString('0x14d670efa326a07b99777b01fb706427ca776095246569150f2a3f17a7d4dc66'),
    MultiCallEntrypoint: AztecAddress.fromString('0x230d0b47ba6d5ed99afb89d584f32ff33438b64f51000f252a140cf995781628'),
    FeeJuice: AztecAddress.fromString('0x204913186c0dd70015d05bf9100a12e31ccb7cc2527aacdfae0c19ad6439fcf4'),
    PublicChecks: AztecAddress.fromString('0x1198142fd84a58c0ab22d5fde371ce527042db49487e05206a326ad154952ac8')
};
export const ProtocolContractsList = new ProtocolContracts([
    AztecAddress.fromString('0x139f8eb6d6e7e7a7c0322c3b7558687a7e24201f11bf2c4cb2fe56c18d363695'),
    AztecAddress.fromString('0x1254246c88aca5a66fa66f3aa78c408a698ebca3b713120497c7555dfc718592'),
    AztecAddress.fromString('0x14d670efa326a07b99777b01fb706427ca776095246569150f2a3f17a7d4dc66'),
    AztecAddress.fromString('0x230d0b47ba6d5ed99afb89d584f32ff33438b64f51000f252a140cf995781628'),
    AztecAddress.fromString('0x204913186c0dd70015d05bf9100a12e31ccb7cc2527aacdfae0c19ad6439fcf4'),
    AztecAddress.fromString('0x1198142fd84a58c0ab22d5fde371ce527042db49487e05206a326ad154952ac8'),
    AztecAddress.fromString('0x0000000000000000000000000000000000000000000000000000000000000000'),
    AztecAddress.fromString('0x0000000000000000000000000000000000000000000000000000000000000000'),
    AztecAddress.fromString('0x0000000000000000000000000000000000000000000000000000000000000000'),
    AztecAddress.fromString('0x0000000000000000000000000000000000000000000000000000000000000000'),
    AztecAddress.fromString('0x0000000000000000000000000000000000000000000000000000000000000000')
]);
export const protocolContractsHash = Fr.fromString('0x2672340d9a0107a7b81e6d10d25b854debe613f3272e8738e8df0ca2ff297141');
export const CONTRACT_INSTANCE_PUBLISHED_EVENT_TAG = Fr.fromHexString('0x1a7e1badb79abdd38c684b3c8306ffe7ecb33c69e3380d9855730aaaa83a21a8');
