// @aztec/stdlib/dest/abi/function_call.js
import { z } from 'zod';
import { AztecAddress } from '../aztec-address/index.js';
import { schemas } from '../schemas/index.js';
import { AbiTypeSchema, FunctionType } from './abi.js';
import { FunctionSelector } from './function_selector.js';
/** A request to call a function on a contract. */ export class FunctionCall {
    name;
    to;
    selector;
    type;
    hideMsgSender;
    isStatic;
    args;
    returnTypes;
    constructor(/** The name of the function to call */ name, /** The recipient contract */ to, /** The function being called */ selector, /** Type of the function */ type, /** Only applicable for enqueued public function calls. `hideMsgSender = true` will set the msg_sender field (the caller's address) to "null", meaning the public function (and observers around the world) won't know which smart contract address made the call. */ hideMsgSender, /** Whether this call can make modifications to state or not */ isStatic, /** The encoded args */ args, /** The return type for decoding */ returnTypes){
        this.name = name;
        this.to = to;
        this.selector = selector;
        this.type = type;
        this.hideMsgSender = hideMsgSender;
        this.isStatic = isStatic;
        this.args = args;
        this.returnTypes = returnTypes;
    }
    static getFields(fields) {
        return [
            fields.name,
            fields.to,
            fields.selector,
            fields.type,
            fields.hideMsgSender,
            fields.isStatic,
            fields.args,
            fields.returnTypes
        ];
    }
    static from(fields) {
        return new FunctionCall(...FunctionCall.getFields(fields));
    }
    static get schema() {
        return z.object({
            name: z.string(),
            to: schemas.AztecAddress,
            selector: schemas.FunctionSelector,
            type: z.nativeEnum(FunctionType),
            isStatic: z.boolean(),
            hideMsgSender: z.boolean(),
            args: z.array(schemas.Fr),
            returnTypes: z.array(AbiTypeSchema)
        }).transform(FunctionCall.from);
    }
    isPublicStatic() {
        return this.type === FunctionType.PUBLIC && this.isStatic;
    }
    /**
   * Creates an empty function call.
   * @returns an empty function call.
   */ static empty() {
        return FunctionCall.from({
            name: '',
            to: AztecAddress.ZERO,
            selector: FunctionSelector.empty(),
            type: FunctionType.PUBLIC,
            hideMsgSender: false,
            isStatic: false,
            args: [],
            returnTypes: []
        });
    }
}
