// @aztec/stdlib/dest/block/l2_block_stream/l2_tips_memory_store.js
import { L2TipsStoreBase } from './l2_tips_store_base.js';
/**
 * In-memory implementation of L2 tips store. Useful for testing and lightweight clients.
 * @dev Tests in kv-store/src/stores/l2_tips_memory_store.test.ts
 */ export class L2TipsMemoryStore extends L2TipsStoreBase {
    tips = new Map();
    blockHashes = new Map();
    blockToCheckpoint = new Map();
    checkpoints = new Map();
    getTip(tag) {
        return Promise.resolve(this.tips.get(tag));
    }
    setTip(tag, blockNumber) {
        this.tips.set(tag, blockNumber);
        return Promise.resolve();
    }
    getStoredBlockHash(blockNumber) {
        return Promise.resolve(this.blockHashes.get(blockNumber));
    }
    setBlockHash(blockNumber, hash) {
        this.blockHashes.set(blockNumber, hash);
        return Promise.resolve();
    }
    deleteBlockHashesBefore(blockNumber) {
        for (const key of this.blockHashes.keys()){
            if (key < blockNumber) {
                this.blockHashes.delete(key);
            }
        }
        return Promise.resolve();
    }
    getCheckpointNumberForBlock(blockNumber) {
        return Promise.resolve(this.blockToCheckpoint.get(blockNumber));
    }
    setCheckpointNumberForBlock(blockNumber, checkpointNumber) {
        this.blockToCheckpoint.set(blockNumber, checkpointNumber);
        return Promise.resolve();
    }
    deleteBlockToCheckpointBefore(blockNumber) {
        for (const key of this.blockToCheckpoint.keys()){
            if (key < blockNumber) {
                this.blockToCheckpoint.delete(key);
            }
        }
        return Promise.resolve();
    }
    getCheckpoint(checkpointNumber) {
        return Promise.resolve(this.checkpoints.get(checkpointNumber));
    }
    saveCheckpointData(checkpoint) {
        this.checkpoints.set(checkpoint.checkpoint.number, checkpoint);
        return Promise.resolve();
    }
    deleteCheckpointsBefore(checkpointNumber) {
        for (const key of this.checkpoints.keys()){
            if (key < checkpointNumber) {
                this.checkpoints.delete(key);
            }
        }
        return Promise.resolve();
    }
    runInTransaction(fn) {
        // Memory store doesn't need transactions - just execute immediately
        return fn();
    }
}
