// @aztec/constants/dest/constants.gen.js
// GENERATED FILE - DO NOT EDIT, RUN yarn remake-constants
export const MAX_FIELD_VALUE = 21888242871839275222246405745257275088548364400416034343698204186575808495616n;
export const MAX_ETH_ADDRESS_BIT_SIZE = 160;
export const MAX_ETH_ADDRESS_VALUE = 1461501637330902918203684832716283019655932542975n;
export const MAX_U64_VALUE = 18446744073709551615n;
export const MAX_U32_VALUE = 4294967295;
export const MAX_FR_CALLDATA_TO_ALL_ENQUEUED_CALLS = 16000;
export const ARCHIVE_HEIGHT = 30;
export const VK_TREE_HEIGHT = 7;
export const FUNCTION_TREE_HEIGHT = 7;
export const NOTE_HASH_TREE_HEIGHT = 42;
export const PUBLIC_DATA_TREE_HEIGHT = 40;
export const NULLIFIER_TREE_HEIGHT = 42;
export const L1_TO_L2_MSG_TREE_HEIGHT = 36;
export const OUT_HASH_TREE_HEIGHT = 5;
export const ARTIFACT_FUNCTION_TREE_MAX_HEIGHT = 7;
export const NULLIFIER_TREE_ID = 0;
export const NOTE_HASH_TREE_ID = 1;
export const PUBLIC_DATA_TREE_ID = 2;
export const L1_TO_L2_MESSAGE_TREE_ID = 3;
export const ARCHIVE_TREE_ID = 4;
export const NOTE_HASH_TREE_LEAF_COUNT = 4398046511104;
export const L1_TO_L2_MSG_TREE_LEAF_COUNT = 68719476736;
export const OUT_HASH_TREE_LEAF_COUNT = 32;
export const NOTE_HASH_SUBTREE_HEIGHT = 6;
export const NULLIFIER_SUBTREE_HEIGHT = 6;
export const PUBLIC_DATA_SUBTREE_HEIGHT = 6;
export const L1_TO_L2_MSG_SUBTREE_HEIGHT = 10;
export const NOTE_HASH_SUBTREE_ROOT_SIBLING_PATH_LENGTH = 36;
export const NULLIFIER_SUBTREE_ROOT_SIBLING_PATH_LENGTH = 36;
export const L1_TO_L2_MSG_SUBTREE_ROOT_SIBLING_PATH_LENGTH = 26;
export const MAX_L2_TO_L1_MSG_SUBTREES_PER_TX = 3;
export const MAX_NOTE_HASHES_PER_TX = 64;
export const MAX_NULLIFIERS_PER_TX = 64;
export const MAX_PRIVATE_CALL_STACK_LENGTH_PER_TX = 16;
export const MAX_ENQUEUED_CALLS_PER_TX = 32;
export const PROTOCOL_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX = 1;
export const MAX_TOTAL_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX = 64;
export const MAX_PUBLIC_DATA_UPDATE_REQUESTS_PER_TX = 63;
export const MAX_PUBLIC_DATA_READS_PER_TX = 64;
export const MAX_L2_TO_L1_MSGS_PER_TX = 8;
export const MAX_PRIVATE_LOGS_PER_TX = 64;
export const MAX_CONTRACT_CLASS_LOGS_PER_TX = 1;
export const MAX_NOTE_HASH_READ_REQUESTS_PER_TX = 64;
export const MAX_NULLIFIER_READ_REQUESTS_PER_TX = 64;
export const MAX_KEY_VALIDATION_REQUESTS_PER_TX = 64;
export const MAX_NOTE_HASHES_PER_CALL = 16;
export const MAX_NULLIFIERS_PER_CALL = 16;
export const MAX_PRIVATE_CALL_STACK_LENGTH_PER_CALL = 8;
export const MAX_ENQUEUED_CALLS_PER_CALL = 32;
export const MAX_L2_TO_L1_MSGS_PER_CALL = 8;
export const MAX_PRIVATE_LOGS_PER_CALL = 16;
export const MAX_CONTRACT_CLASS_LOGS_PER_CALL = 1;
export const MAX_NOTE_HASH_READ_REQUESTS_PER_CALL = 16;
export const MAX_NULLIFIER_READ_REQUESTS_PER_CALL = 16;
export const MAX_KEY_VALIDATION_REQUESTS_PER_CALL = 16;
export const NUMBER_OF_L1_L2_MESSAGES_PER_ROLLUP = 1024;
export const PRIVATE_KERNEL_INIT_VK_INDEX = 0;
export const PRIVATE_KERNEL_INNER_VK_INDEX = 1;
export const PRIVATE_KERNEL_TAIL_VK_INDEX = 2;
export const PRIVATE_KERNEL_TAIL_TO_PUBLIC_VK_INDEX = 3;
export const HIDING_KERNEL_TO_ROLLUP_VK_INDEX = 4;
export const HIDING_KERNEL_TO_PUBLIC_VK_INDEX = 5;
export const PUBLIC_CHONK_VERIFIER_VK_INDEX = 6;
export const PRIVATE_TX_BASE_ROLLUP_VK_INDEX = 7;
export const PUBLIC_TX_BASE_ROLLUP_VK_INDEX = 8;
export const TX_MERGE_ROLLUP_VK_INDEX = 9;
export const BLOCK_ROOT_FIRST_ROLLUP_VK_INDEX = 10;
export const BLOCK_ROOT_SINGLE_TX_FIRST_ROLLUP_VK_INDEX = 11;
export const BLOCK_ROOT_EMPTY_TX_FIRST_ROLLUP_VK_INDEX = 12;
export const BLOCK_ROOT_ROLLUP_VK_INDEX = 13;
export const BLOCK_ROOT_SINGLE_TX_ROLLUP_VK_INDEX = 14;
export const BLOCK_MERGE_ROLLUP_VK_INDEX = 15;
export const CHECKPOINT_ROOT_ROLLUP_VK_INDEX = 16;
export const CHECKPOINT_ROOT_SINGLE_BLOCK_ROLLUP_VK_INDEX = 17;
export const CHECKPOINT_PADDING_ROLLUP_VK_INDEX = 18;
export const CHECKPOINT_MERGE_ROLLUP_VK_INDEX = 19;
export const ROOT_ROLLUP_VK_INDEX = 20;
export const PARITY_BASE_VK_INDEX = 21;
export const PARITY_ROOT_VK_INDEX = 22;
export const PRIVATE_KERNEL_RESET_VK_INDEX = 23;
export const GLOBAL_INDEX_NOTE_HASH_READ_REQUEST_OFFSET = 0;
export const GLOBAL_INDEX_NULLIFIER_READ_REQUEST_OFFSET = 16;
export const GLOBAL_INDEX_NOTE_HASH_OFFSET = 32;
export const GLOBAL_INDEX_NULLIFIER_OFFSET = 48;
export const GLOBAL_INDEX_PRIVATE_CALL_REQUEST_START_OFFSET = 64;
export const GLOBAL_INDEX_PRIVATE_CALL_REQUEST_END_OFFSET = 72;
export const GLOBAL_INDEX_PUBLIC_CALL_REQUEST_OFFSET = 80;
export const GLOBAL_INDEX_L2_TO_L1_MSG_OFFSET = 112;
export const GLOBAL_INDEX_PRIVATE_LOG_OFFSET = 120;
export const GLOBAL_INDEX_CONTRACT_CLASS_LOG_HASH_OFFSET = 136;
export const GLOBAL_INDEX_CONTRACT_MIN_REVERTIBLE_SIDE_EFFECT_COUNTER_OFFSET = 137;
export const TOTAL_COUNTED_SIDE_EFFECTS_PER_CALL = 138;
export const FUNCTION_SELECTOR_NUM_BYTES = 4;
export const INITIAL_CHECKPOINT_NUMBER = 1;
export const INITIAL_L2_BLOCK_NUM = 1;
export const FIELDS_PER_BLOB = 4096;
export const BLOBS_PER_CHECKPOINT = 6;
export const MAX_CHECKPOINTS_PER_EPOCH = 32;
export const MAX_TX_LIFETIME = 86400;
export const GENESIS_BLOCK_HEADER_HASH = 21694244557328181932449230888455310614586084993178835904749768133365362325861n;
export const GENESIS_ARCHIVE_ROOT = 9682850228538071369704502076456077473410427336083826595120404283897422804423n;
export const EMPTY_EPOCH_OUT_HASH = 355785372471781095838790036702437931769306153278986832745847530947941691539n;
export const MAX_PACKED_PUBLIC_BYTECODE_SIZE_IN_FIELDS = 3000;
export const MAX_PACKED_BYTECODE_SIZE_PER_PRIVATE_FUNCTION_IN_FIELDS = 3000;
export const MAX_PACKED_BYTECODE_SIZE_PER_UTILITY_FUNCTION_IN_FIELDS = 3000;
export const CLASS_REGISTRY_PRIVATE_FUNCTION_BROADCASTED_ADDITIONAL_FIELDS = 23;
export const CLASS_REGISTRY_UTILITY_FUNCTION_BROADCASTED_ADDITIONAL_FIELDS = 14;
export const CONTRACT_CLASS_PUBLISHED_MAGIC_VALUE = 14907836659448373306373608164128296596488873008270905914742877230397948162628n;
export const CONTRACT_CLASS_REGISTRY_PRIVATE_FUNCTION_BROADCASTED_MAGIC_VALUE = 6632427679359753950122686582223287785801228472790519808071945107172278003224n;
export const CONTRACT_CLASS_REGISTRY_UTILITY_FUNCTION_BROADCASTED_MAGIC_VALUE = 2863388930963370662477859081312917957786876946407365497018765359301074232169n;
export const CONTRACT_INSTANCE_PUBLISHED_MAGIC_VALUE = 10538216027419913765597387738085647348651103543680388181336823392401502757423n;
export const CONTRACT_INSTANCE_UPDATED_MAGIC_VALUE = 20721543224513346060908370400407150739273836456436647488068002302723900469047n;
export const MAX_PROTOCOL_CONTRACTS = 11;
export const CANONICAL_AUTH_REGISTRY_ADDRESS = 1;
export const CONTRACT_INSTANCE_REGISTRY_CONTRACT_ADDRESS = 2;
export const CONTRACT_CLASS_REGISTRY_CONTRACT_ADDRESS = 3;
export const MULTI_CALL_ENTRYPOINT_ADDRESS = 4;
export const FEE_JUICE_ADDRESS = 5;
export const PUBLIC_CHECKS_ADDRESS = 6;
export const SIDE_EFFECT_MASKING_ADDRESS = 19523154334483583633304358390644137470227519736821975910774528428729027989987n;
export const NULL_MSG_SENDER_CONTRACT_ADDRESS = 21888242871839275222246405745257275088548364400416034343698204186575808495616n;
export const CONTRACT_CLASS_REGISTRY_BYTECODE_CAPSULE_SLOT = 14193106819744442689484501689686180698286338820089744102289275815863062044599n;
export const FEE_JUICE_BALANCES_SLOT = 1;
export const UPDATED_CLASS_IDS_SLOT = 1;
export const DEFAULT_NPK_M_X = 582240093077765400562621227108555700500271598878376310175765873770292988861n;
export const DEFAULT_NPK_M_Y = 10422444662424639723529825114205836958711284159673861467999592572974769103684n;
export const DEFAULT_IVPK_M_X = 339708709767762472786445938838804872781183545349360029270386718856175781484n;
export const DEFAULT_IVPK_M_Y = 12719619215050539905199178334954929730355853796706924300730604757520758976849n;
export const DEFAULT_OVPK_M_X = 12212787719617305570587928860288475454328008955283046946846066128763901043335n;
export const DEFAULT_OVPK_M_Y = 3646747884782549389807830220601404629716007431341772952958971658285958854707n;
export const DEFAULT_TPK_M_X = 728059161893070741164607238299536939695876538801885465230641192969135857403n;
export const DEFAULT_TPK_M_Y = 14575718736702206050102425029229426215631664471161015518982549597389390371695n;
export const AZTEC_ADDRESS_LENGTH = 1;
export const GAS_FEES_LENGTH = 2;
export const GAS_LENGTH = 2;
export const GAS_SETTINGS_LENGTH = 8;
export const CALL_CONTEXT_LENGTH = 4;
export const CONTRACT_INSTANCE_LENGTH = 16;
export const CONTRACT_STORAGE_READ_LENGTH = 3;
export const CONTRACT_STORAGE_UPDATE_REQUEST_LENGTH = 3;
export const ETH_ADDRESS_LENGTH = 1;
export const FUNCTION_DATA_LENGTH = 2;
export const FUNCTION_LEAF_PREIMAGE_LENGTH = 5;
export const GLOBAL_VARIABLES_LENGTH = 9;
export const APPEND_ONLY_TREE_SNAPSHOT_LENGTH = 2;
export const APPEND_ONLY_TREE_SNAPSHOT_LENGTH_BYTES = 36;
export const SPONGE_BLOB_LENGTH = 10;
export const BLS12_FR_LIMBS = 3;
export const BLS12_FQ_LIMBS = 4;
export const BLS12_POINT_LENGTH = 9;
export const BLS12_POINT_COMPRESSED_BYTES = 48;
export const BLOB_ACCUMULATOR_LENGTH = 18;
export const FINAL_BLOB_BATCHING_CHALLENGES_LENGTH = 4;
export const FINAL_BLOB_ACCUMULATOR_LENGTH = 7;
export const L1_TO_L2_MESSAGE_LENGTH = 6;
export const L2_TO_L1_MESSAGE_LENGTH = 2;
export const COUNTED_L2_TO_L1_MESSAGE_LENGTH = 3;
export const SCOPED_L2_TO_L1_MESSAGE_LENGTH = 3;
export const SCOPED_COUNTED_L2_TO_L1_MESSAGE_LENGTH = 4;
export const KEY_VALIDATION_REQUEST_LENGTH = 4;
export const KEY_VALIDATION_REQUEST_AND_GENERATOR_LENGTH = 5;
export const SCOPED_KEY_VALIDATION_REQUEST_AND_GENERATOR_LENGTH = 6;
export const PARTIAL_STATE_REFERENCE_LENGTH = 6;
export const TREE_LEAF_READ_REQUEST_LENGTH = 2;
export const PRIVATE_LOG_SIZE_IN_FIELDS = 16;
export const PRIVATE_LOG_LENGTH = 17;
export const PRIVATE_LOG_DATA_LENGTH = 19;
export const PRIVATE_LOG_CIPHERTEXT_LEN = 15;
export const SCOPED_PRIVATE_LOG_DATA_LENGTH = 20;
export const FLAT_PUBLIC_LOGS_HEADER_LENGTH = 1;
export const FLAT_PUBLIC_LOGS_PAYLOAD_LENGTH = 4096;
export const PUBLIC_LOGS_LENGTH = 4097;
export const PUBLIC_LOG_HEADER_LENGTH = 2;
export const MAX_PUBLIC_LOG_SIZE_IN_FIELDS = 4094;
export const CONTRACT_CLASS_LOG_SIZE_IN_FIELDS = 3023;
export const CONTRACT_CLASS_LOG_LENGTH = 3025;
export const LOG_HASH_LENGTH = 2;
export const COUNTED_LOG_HASH_LENGTH = 3;
export const SCOPED_LOG_HASH_LENGTH = 3;
export const SCOPED_COUNTED_LOG_HASH_LENGTH = 4;
export const NOTE_HASH_LENGTH = 2;
export const SCOPED_NOTE_HASH_LENGTH = 3;
export const NULLIFIER_LENGTH = 3;
export const SCOPED_NULLIFIER_LENGTH = 4;
export const PUBLIC_DATA_WRITE_LENGTH = 2;
export const PUBLIC_CALL_STACK_ITEM_COMPRESSED_LENGTH = 12;
export const PRIVATE_CALL_REQUEST_LENGTH = 8;
export const PUBLIC_CALL_REQUEST_LENGTH = 4;
export const COUNTED_PUBLIC_CALL_REQUEST_LENGTH = 5;
export const PUBLIC_INNER_CALL_REQUEST_LENGTH = 13;
export const STATE_REFERENCE_LENGTH = 8;
export const TREE_SNAPSHOTS_LENGTH = 8;
export const TX_CONTEXT_LENGTH = 10;
export const TX_REQUEST_LENGTH = 15;
export const TOTAL_FEES_LENGTH = 1;
export const TOTAL_MANA_USED_LENGTH = 1;
export const BLOCK_HEADER_LENGTH = 22;
export const CHECKPOINT_HEADER_LENGTH = 12;
export const CHECKPOINT_HEADER_SIZE_IN_BYTES = 316;
export const SCOPED_READ_REQUEST_LEN = 3;
export const PRIVATE_CIRCUIT_PUBLIC_INPUTS_LENGTH = 870;
export const PRIVATE_CONTEXT_INPUTS_LENGTH = 37;
export const FEE_RECIPIENT_LENGTH = 2;
export const HIDING_KERNEL_IO_PUBLIC_INPUTS_SIZE = 28;
export const PAIRING_POINTS_SIZE = 8;
export const IPA_CLAIM_SIZE = 6;
export const PUBLIC_DATA_READ_LENGTH = 3;
export const PRIVATE_VALIDATION_REQUESTS_LENGTH = 771;
export const PRIVATE_TO_ROLLUP_ACCUMULATED_DATA_LENGTH = 1243;
export const TX_CONSTANT_DATA_LENGTH = 34;
export const COMBINED_CONSTANT_DATA_LENGTH = 43;
export const PRIVATE_ACCUMULATED_DATA_LENGTH = 2059;
export const PRIVATE_KERNEL_CIRCUIT_PUBLIC_INPUTS_LENGTH = 2873;
export const PRIVATE_TO_PUBLIC_ACCUMULATED_DATA_LENGTH = 1371;
export const PRIVATE_TO_AVM_ACCUMULATED_DATA_LENGTH = 152;
export const NUM_PRIVATE_TO_AVM_ACCUMULATED_DATA_ARRAYS = 3;
export const NUM_AVM_ACCUMULATED_DATA_ARRAYS = 4;
export const NUM_PUBLIC_CALL_REQUEST_ARRAYS = 3;
export const AVM_ACCUMULATED_DATA_LENGTH = 4377;
export const PRIVATE_TO_PUBLIC_KERNEL_CIRCUIT_PUBLIC_INPUTS_LENGTH = 2784;
export const PRIVATE_TO_ROLLUP_KERNEL_CIRCUIT_PUBLIC_INPUTS_LENGTH = 1281;
export const AVM_CIRCUIT_PUBLIC_INPUTS_LENGTH = 5008;
export const BLOCK_CONSTANT_DATA_LENGTH = 16;
export const CHECKPOINT_CONSTANT_DATA_LENGTH = 10;
export const EPOCH_CONSTANT_DATA_LENGTH = 5;
export const TX_ROLLUP_PUBLIC_INPUTS_LENGTH = 52;
export const BLOCK_ROLLUP_PUBLIC_INPUTS_LENGTH = 56;
export const CHECKPOINT_ROLLUP_PUBLIC_INPUTS_LENGTH = 149;
export const ROOT_ROLLUP_PUBLIC_INPUTS_LENGTH = 111;
export const NUM_MSGS_PER_BASE_PARITY = 256;
export const NUM_BASE_PARITY_PER_ROOT_PARITY = 4;
export const RECURSIVE_PROOF_LENGTH = 449;
export const NESTED_RECURSIVE_PROOF_LENGTH = 449;
export const IPA_PROOF_LENGTH = 64;
export const ULTRA_KECCAK_PROOF_LENGTH = 331;
export const RECURSIVE_ROLLUP_HONK_PROOF_LENGTH = 519;
export const NESTED_RECURSIVE_ROLLUP_HONK_PROOF_LENGTH = 519;
export const CHONK_PROOF_LENGTH = 1935;
export const ULTRA_VK_LENGTH_IN_FIELDS = 115;
export const MEGA_VK_LENGTH_IN_FIELDS = 127;
export const CHONK_VK_LENGTH_IN_FIELDS = 127;
export const AVM_VERIFICATION_KEY_LENGTH_IN_FIELDS = 86;
export const MAX_PUBLIC_BYTECODE_SIZE_IN_BYTES = 96000;
export const MAX_PUBLIC_CALLS_TO_UNIQUE_CONTRACT_CLASS_IDS = 21;
export const MEM_TAG_FF = 0;
export const MEM_TAG_U1 = 1;
export const MEM_TAG_U8 = 2;
export const MEM_TAG_U16 = 3;
export const MEM_TAG_U32 = 4;
export const MEM_TAG_U64 = 5;
export const MEM_TAG_U128 = 6;
export const AVM_HIGHEST_MEM_ADDRESS = 4294967295;
export const AVM_MEMORY_NUM_BITS = 32;
export const AVM_MEMORY_SIZE = 4294967296;
export const AVM_BITWISE_AND_OP_ID = 1;
export const AVM_BITWISE_OR_OP_ID = 2;
export const AVM_BITWISE_XOR_OP_ID = 4;
export const AVM_KECCAKF1600_NUM_ROUNDS = 24;
export const AVM_KECCAKF1600_STATE_SIZE = 25;
export const AVM_TX_PHASE_VALUE_START = 0;
export const AVM_TX_PHASE_VALUE_SETUP = 3;
export const AVM_TX_PHASE_VALUE_LAST = 11;
export const AVM_SUBTRACE_ID_EXECUTION = 1;
export const AVM_SUBTRACE_ID_ALU = 2;
export const AVM_SUBTRACE_ID_BITWISE = 4;
export const AVM_SUBTRACE_ID_CAST = 8;
export const AVM_SUBTRACE_ID_CALLDATA_COPY = 16;
export const AVM_SUBTRACE_ID_RETURNDATA_COPY = 32;
export const AVM_SUBTRACE_ID_SET = 64;
export const AVM_SUBTRACE_ID_GETCONTRACTINSTANCE = 128;
export const AVM_SUBTRACE_ID_EMITPUBLICLOG = 256;
export const AVM_SUBTRACE_ID_POSEIDON2_PERM = 512;
export const AVM_SUBTRACE_ID_SHA256_COMPRESSION = 1024;
export const AVM_SUBTRACE_ID_KECCAKF1600 = 2048;
export const AVM_SUBTRACE_ID_ECC = 4096;
export const AVM_SUBTRACE_ID_TO_RADIX = 8192;
export const AVM_DYN_GAS_ID_CALLDATACOPY = 1;
export const AVM_DYN_GAS_ID_RETURNDATACOPY = 2;
export const AVM_DYN_GAS_ID_TORADIX = 4;
export const AVM_DYN_GAS_ID_BITWISE = 8;
export const AVM_DYN_GAS_ID_EMITPUBLICLOG = 16;
export const AVM_DYN_GAS_ID_SSTORE = 32;
export const AVM_EXEC_OP_ID_GETENVVAR = 1;
export const AVM_EXEC_OP_ID_MOV = 2;
export const AVM_EXEC_OP_ID_JUMP = 4;
export const AVM_EXEC_OP_ID_JUMPI = 8;
export const AVM_EXEC_OP_ID_CALL = 16;
export const AVM_EXEC_OP_ID_STATICCALL = 32;
export const AVM_EXEC_OP_ID_INTERNALCALL = 64;
export const AVM_EXEC_OP_ID_INTERNALRETURN = 128;
export const AVM_EXEC_OP_ID_RETURN = 256;
export const AVM_EXEC_OP_ID_REVERT = 512;
export const AVM_EXEC_OP_ID_SUCCESSCOPY = 1024;
export const AVM_EXEC_OP_ID_RETURNDATASIZE = 2048;
export const AVM_EXEC_OP_ID_DEBUGLOG = 4096;
export const AVM_EXEC_OP_ID_SLOAD = 8192;
export const AVM_EXEC_OP_ID_SSTORE = 16384;
export const AVM_EXEC_OP_ID_NOTEHASH_EXISTS = 32768;
export const AVM_EXEC_OP_ID_EMIT_NOTEHASH = 65536;
export const AVM_EXEC_OP_ID_L1_TO_L2_MESSAGE_EXISTS = 131072;
export const AVM_EXEC_OP_ID_NULLIFIER_EXISTS = 262144;
export const AVM_EXEC_OP_ID_EMIT_NULLIFIER = 524288;
export const AVM_EXEC_OP_ID_SENDL2TOL1MSG = 1048576;
export const AVM_EXEC_OP_ID_ALU_ADD = 1;
export const AVM_EXEC_OP_ID_ALU_SUB = 2;
export const AVM_EXEC_OP_ID_ALU_MUL = 4;
export const AVM_EXEC_OP_ID_ALU_DIV = 8;
export const AVM_EXEC_OP_ID_ALU_FDIV = 16;
export const AVM_EXEC_OP_ID_ALU_EQ = 32;
export const AVM_EXEC_OP_ID_ALU_LT = 64;
export const AVM_EXEC_OP_ID_ALU_LTE = 128;
export const AVM_EXEC_OP_ID_ALU_NOT = 256;
export const AVM_EXEC_OP_ID_ALU_SHL = 512;
export const AVM_EXEC_OP_ID_ALU_SHR = 1024;
export const AVM_EXEC_OP_ID_ALU_TRUNCATE = 2048;
export const AVM_PUBLIC_INPUTS_GLOBAL_VARIABLES_ROW_IDX = 0;
export const AVM_PUBLIC_INPUTS_GLOBAL_VARIABLES_CHAIN_ID_ROW_IDX = 0;
export const AVM_PUBLIC_INPUTS_GLOBAL_VARIABLES_VERSION_ROW_IDX = 1;
export const AVM_PUBLIC_INPUTS_GLOBAL_VARIABLES_BLOCK_NUMBER_ROW_IDX = 2;
export const AVM_PUBLIC_INPUTS_GLOBAL_VARIABLES_SLOT_NUMBER_ROW_IDX = 3;
export const AVM_PUBLIC_INPUTS_GLOBAL_VARIABLES_TIMESTAMP_ROW_IDX = 4;
export const AVM_PUBLIC_INPUTS_GLOBAL_VARIABLES_COINBASE_ROW_IDX = 5;
export const AVM_PUBLIC_INPUTS_GLOBAL_VARIABLES_FEE_RECIPIENT_ROW_IDX = 6;
export const AVM_PUBLIC_INPUTS_GLOBAL_VARIABLES_GAS_FEES_ROW_IDX = 7;
export const AVM_PUBLIC_INPUTS_PROTOCOL_CONTRACTS_ROW_IDX = 8;
export const AVM_PUBLIC_INPUTS_START_TREE_SNAPSHOTS_ROW_IDX = 19;
export const AVM_PUBLIC_INPUTS_START_TREE_SNAPSHOTS_L1_TO_L2_MESSAGE_TREE_ROW_IDX = 19;
export const AVM_PUBLIC_INPUTS_START_TREE_SNAPSHOTS_NOTE_HASH_TREE_ROW_IDX = 20;
export const AVM_PUBLIC_INPUTS_START_TREE_SNAPSHOTS_NULLIFIER_TREE_ROW_IDX = 21;
export const AVM_PUBLIC_INPUTS_START_TREE_SNAPSHOTS_PUBLIC_DATA_TREE_ROW_IDX = 22;
export const AVM_PUBLIC_INPUTS_START_GAS_USED_ROW_IDX = 23;
export const AVM_PUBLIC_INPUTS_GAS_SETTINGS_ROW_IDX = 24;
export const AVM_PUBLIC_INPUTS_GAS_SETTINGS_GAS_LIMITS_ROW_IDX = 24;
export const AVM_PUBLIC_INPUTS_GAS_SETTINGS_TEARDOWN_GAS_LIMITS_ROW_IDX = 25;
export const AVM_PUBLIC_INPUTS_GAS_SETTINGS_MAX_FEES_PER_GAS_ROW_IDX = 26;
export const AVM_PUBLIC_INPUTS_GAS_SETTINGS_MAX_PRIORITY_FEES_PER_GAS_ROW_IDX = 27;
export const AVM_PUBLIC_INPUTS_EFFECTIVE_GAS_FEES_ROW_IDX = 28;
export const AVM_PUBLIC_INPUTS_FEE_PAYER_ROW_IDX = 29;
export const AVM_PUBLIC_INPUTS_PROVER_ID_ROW_IDX = 30;
export const AVM_PUBLIC_INPUTS_PUBLIC_CALL_REQUEST_ARRAY_LENGTHS_ROW_IDX = 31;
export const AVM_PUBLIC_INPUTS_PUBLIC_CALL_REQUEST_ARRAY_LENGTHS_SETUP_CALLS_ROW_IDX = 31;
export const AVM_PUBLIC_INPUTS_PUBLIC_CALL_REQUEST_ARRAY_LENGTHS_APP_LOGIC_CALLS_ROW_IDX = 32;
export const AVM_PUBLIC_INPUTS_PUBLIC_CALL_REQUEST_ARRAY_LENGTHS_TEARDOWN_CALL_ROW_IDX = 33;
export const AVM_PUBLIC_INPUTS_PUBLIC_SETUP_CALL_REQUESTS_ROW_IDX = 34;
export const AVM_PUBLIC_INPUTS_PUBLIC_APP_LOGIC_CALL_REQUESTS_ROW_IDX = 66;
export const AVM_PUBLIC_INPUTS_PUBLIC_TEARDOWN_CALL_REQUEST_ROW_IDX = 98;
export const AVM_PUBLIC_INPUTS_PREVIOUS_NON_REVERTIBLE_ACCUMULATED_DATA_ARRAY_LENGTHS_ROW_IDX = 99;
export const AVM_PUBLIC_INPUTS_PREVIOUS_NON_REVERTIBLE_ACCUMULATED_DATA_ARRAY_LENGTHS_NOTE_HASHES_ROW_IDX = 99;
export const AVM_PUBLIC_INPUTS_PREVIOUS_NON_REVERTIBLE_ACCUMULATED_DATA_ARRAY_LENGTHS_NULLIFIERS_ROW_IDX = 100;
export const AVM_PUBLIC_INPUTS_PREVIOUS_NON_REVERTIBLE_ACCUMULATED_DATA_ARRAY_LENGTHS_L2_TO_L1_MSGS_ROW_IDX = 101;
export const AVM_PUBLIC_INPUTS_PREVIOUS_REVERTIBLE_ACCUMULATED_DATA_ARRAY_LENGTHS_ROW_IDX = 102;
export const AVM_PUBLIC_INPUTS_PREVIOUS_REVERTIBLE_ACCUMULATED_DATA_ARRAY_LENGTHS_NOTE_HASHES_ROW_IDX = 102;
export const AVM_PUBLIC_INPUTS_PREVIOUS_REVERTIBLE_ACCUMULATED_DATA_ARRAY_LENGTHS_NULLIFIERS_ROW_IDX = 103;
export const AVM_PUBLIC_INPUTS_PREVIOUS_REVERTIBLE_ACCUMULATED_DATA_ARRAY_LENGTHS_L2_TO_L1_MSGS_ROW_IDX = 104;
export const AVM_PUBLIC_INPUTS_PREVIOUS_NON_REVERTIBLE_ACCUMULATED_DATA_ROW_IDX = 105;
export const AVM_PUBLIC_INPUTS_PREVIOUS_NON_REVERTIBLE_ACCUMULATED_DATA_NOTE_HASHES_ROW_IDX = 105;
export const AVM_PUBLIC_INPUTS_PREVIOUS_NON_REVERTIBLE_ACCUMULATED_DATA_NULLIFIERS_ROW_IDX = 169;
export const AVM_PUBLIC_INPUTS_PREVIOUS_NON_REVERTIBLE_ACCUMULATED_DATA_L2_TO_L1_MSGS_ROW_IDX = 233;
export const AVM_PUBLIC_INPUTS_PREVIOUS_REVERTIBLE_ACCUMULATED_DATA_ROW_IDX = 241;
export const AVM_PUBLIC_INPUTS_PREVIOUS_REVERTIBLE_ACCUMULATED_DATA_NOTE_HASHES_ROW_IDX = 241;
export const AVM_PUBLIC_INPUTS_PREVIOUS_REVERTIBLE_ACCUMULATED_DATA_NULLIFIERS_ROW_IDX = 305;
export const AVM_PUBLIC_INPUTS_PREVIOUS_REVERTIBLE_ACCUMULATED_DATA_L2_TO_L1_MSGS_ROW_IDX = 369;
export const AVM_PUBLIC_INPUTS_END_TREE_SNAPSHOTS_ROW_IDX = 377;
export const AVM_PUBLIC_INPUTS_END_TREE_SNAPSHOTS_L1_TO_L2_MESSAGE_TREE_ROW_IDX = 377;
export const AVM_PUBLIC_INPUTS_END_TREE_SNAPSHOTS_NOTE_HASH_TREE_ROW_IDX = 378;
export const AVM_PUBLIC_INPUTS_END_TREE_SNAPSHOTS_NULLIFIER_TREE_ROW_IDX = 379;
export const AVM_PUBLIC_INPUTS_END_TREE_SNAPSHOTS_PUBLIC_DATA_TREE_ROW_IDX = 380;
export const AVM_PUBLIC_INPUTS_END_GAS_USED_ROW_IDX = 381;
export const AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_ARRAY_LENGTHS_ROW_IDX = 382;
export const AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_ARRAY_LENGTHS_NOTE_HASHES_ROW_IDX = 382;
export const AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_ARRAY_LENGTHS_NULLIFIERS_ROW_IDX = 383;
export const AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_ARRAY_LENGTHS_L2_TO_L1_MSGS_ROW_IDX = 384;
export const AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_ARRAY_LENGTHS_PUBLIC_DATA_WRITES_ROW_IDX = 385;
export const AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_ROW_IDX = 386;
export const AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_NOTE_HASHES_ROW_IDX = 386;
export const AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_NULLIFIERS_ROW_IDX = 450;
export const AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_L2_TO_L1_MSGS_ROW_IDX = 514;
export const AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_PUBLIC_LOGS_ROW_IDX = 522;
export const AVM_PUBLIC_INPUTS_AVM_ACCUMULATED_DATA_PUBLIC_DATA_WRITES_ROW_IDX = 4619;
export const AVM_PUBLIC_INPUTS_TRANSACTION_FEE_ROW_IDX = 4683;
export const AVM_PUBLIC_INPUTS_REVERTED_ROW_IDX = 4684;
export const AVM_PUBLIC_INPUTS_COLUMNS_MAX_LENGTH = 4685;
export const AVM_NUM_PUBLIC_INPUT_COLUMNS = 4;
export const AVM_PUBLIC_INPUTS_COLUMNS_COMBINED_LENGTH = 18740;
export const AVM_V2_PROOF_LENGTH_IN_FIELDS_PADDED = 16400;
export const AVM_V2_VERIFICATION_KEY_LENGTH_IN_FIELDS_PADDED = 1000;
export const AVM_MAX_PROCESSABLE_L2_GAS = 6000000;
export const DA_BYTES_PER_FIELD = 32;
export const DA_GAS_PER_BYTE = 1;
export const DA_GAS_PER_FIELD = 32;
export const FIXED_DA_GAS = 96;
export const FIXED_L2_GAS = 512;
export const FIXED_AVM_STARTUP_L2_GAS = 20000;
export const MAX_PROCESSABLE_DA_GAS_PER_CHECKPOINT = 786432;
export const GAS_ESTIMATION_TEARDOWN_L2_GAS_LIMIT = 6000000;
export const GAS_ESTIMATION_L2_GAS_LIMIT = 12000000;
export const GAS_ESTIMATION_TEARDOWN_DA_GAS_LIMIT = 786432;
export const GAS_ESTIMATION_DA_GAS_LIMIT = 1572864;
export const DEFAULT_TEARDOWN_L2_GAS_LIMIT = 1000000;
export const DEFAULT_L2_GAS_LIMIT = 6000000;
export const DEFAULT_TEARDOWN_DA_GAS_LIMIT = 393216;
export const DEFAULT_DA_GAS_LIMIT = 786432;
export const L2_GAS_DISTRIBUTED_STORAGE_PREMIUM = 1024;
export const AVM_PC_SIZE_IN_BITS = 32;
export const AVM_MAX_OPERANDS = 7;
export const AVM_MAX_REGISTERS = 6;
export const AVM_ADDRESSING_BASE_RESOLUTION_L2_GAS = 3;
export const AVM_ADDRESSING_INDIRECT_L2_GAS = 3;
export const AVM_ADDRESSING_RELATIVE_L2_GAS = 3;
export const L2_GAS_PER_NOTE_HASH = 0;
export const L2_GAS_PER_NULLIFIER = 0;
export const L2_GAS_PER_L2_TO_L1_MSG = 200;
export const L2_GAS_PER_PRIVATE_LOG = 0;
export const L2_GAS_PER_CONTRACT_CLASS_LOG = 0;
export const AVM_ADD_BASE_L2_GAS = 12;
export const AVM_SUB_BASE_L2_GAS = 12;
export const AVM_MUL_BASE_L2_GAS = 27;
export const AVM_DIV_BASE_L2_GAS = 27;
export const AVM_FDIV_BASE_L2_GAS = 225;
export const AVM_EQ_BASE_L2_GAS = 12;
export const AVM_LT_BASE_L2_GAS = 42;
export const AVM_LTE_BASE_L2_GAS = 42;
export const AVM_AND_BASE_L2_GAS = 12;
export const AVM_OR_BASE_L2_GAS = 12;
export const AVM_XOR_BASE_L2_GAS = 12;
export const AVM_NOT_BASE_L2_GAS = 12;
export const AVM_SHL_BASE_L2_GAS = 18;
export const AVM_SHR_BASE_L2_GAS = 18;
export const AVM_CAST_BASE_L2_GAS = 27;
export const AVM_GETENVVAR_BASE_L2_GAS = 12;
export const AVM_CALLDATACOPY_BASE_L2_GAS = 18;
export const AVM_SUCCESSCOPY_BASE_L2_GAS = 12;
export const AVM_RETURNDATASIZE_BASE_L2_GAS = 12;
export const AVM_RETURNDATACOPY_BASE_L2_GAS = 18;
export const AVM_JUMP_BASE_L2_GAS = 9;
export const AVM_JUMPI_BASE_L2_GAS = 9;
export const AVM_INTERNALCALL_BASE_L2_GAS = 9;
export const AVM_INTERNALRETURN_BASE_L2_GAS = 9;
export const AVM_SET_BASE_L2_GAS = 27;
export const AVM_MOV_BASE_L2_GAS = 12;
export const AVM_SLOAD_BASE_L2_GAS = 1290;
export const AVM_SSTORE_BASE_L2_GAS = 33140;
export const AVM_NOTEHASHEXISTS_BASE_L2_GAS = 504;
export const AVM_EMITNOTEHASH_BASE_L2_GAS = 19275;
export const AVM_NULLIFIEREXISTS_BASE_L2_GAS = 903;
export const AVM_EMITNULLIFIER_BASE_L2_GAS = 30800;
export const AVM_L1TOL2MSGEXISTS_BASE_L2_GAS = 540;
export const AVM_GETCONTRACTINSTANCE_BASE_L2_GAS = 6108;
export const AVM_EMITPUBLICLOG_BASE_L2_GAS = 15;
export const AVM_SENDL2TOL1MSG_BASE_L2_GAS = 478;
export const AVM_CALL_BASE_L2_GAS = 9936;
export const AVM_STATICCALL_BASE_L2_GAS = 9936;
export const AVM_RETURN_BASE_L2_GAS = 9;
export const AVM_REVERT_BASE_L2_GAS = 9;
export const AVM_DEBUGLOG_BASE_L2_GAS = 9;
export const AVM_POSEIDON2_BASE_L2_GAS = 360;
export const AVM_SHA256COMPRESSION_BASE_L2_GAS = 12288;
export const AVM_KECCAKF1600_BASE_L2_GAS = 58176;
export const AVM_ECADD_BASE_L2_GAS = 270;
export const AVM_TORADIXBE_BASE_L2_GAS = 24;
export const AVM_CALLDATACOPY_DYN_L2_GAS = 3;
export const AVM_RETURNDATACOPY_DYN_L2_GAS = 3;
export const AVM_TORADIXBE_DYN_L2_GAS = 3;
export const AVM_BITWISE_DYN_L2_GAS = 3;
export const AVM_EMITPUBLICLOG_DYN_L2_GAS = 3;
export const AVM_EMITNOTEHASH_BASE_DA_GAS = 32;
export const AVM_EMITNULLIFIER_BASE_DA_GAS = 32;
export const AVM_SENDL2TOL1MSG_BASE_DA_GAS = 32;
export const AVM_EMITPUBLICLOG_BASE_DA_GAS = 64;
export const AVM_EMITPUBLICLOG_DYN_DA_GAS = 32;
export const AVM_SSTORE_DYN_DA_GAS = 64;
export const TX_START_PREFIX = 2624615704;
export const BLOCK_END_PREFIX = 3951939007;
export const CHECKPOINT_END_PREFIX = 2355328067;
export const PROOF_TYPE_HONK = 0;
export const PROOF_TYPE_OINK = 1;
export const PROOF_TYPE_HN = 2;
export const PROOF_TYPE_AVM = 3;
export const PROOF_TYPE_ROLLUP_HONK = 4;
export const PROOF_TYPE_ROOT_ROLLUP_HONK = 5;
export const PROOF_TYPE_HN_FINAL = 7;
export const PROOF_TYPE_HN_TAIL = 8;
export const PROOF_TYPE_CHONK = 9;
export const TWO_POW_64 = 18446744073709551616n;
export const AVM_WRITTEN_PUBLIC_DATA_SLOTS_TREE_HEIGHT = 6;
export const AVM_WRITTEN_PUBLIC_DATA_SLOTS_TREE_INITIAL_ROOT = 18291678969210913367302010540259942201271604198321103848479209155223586227821n;
export const AVM_WRITTEN_PUBLIC_DATA_SLOTS_TREE_INITIAL_SIZE = 1;
export const AVM_RETRIEVED_BYTECODES_TREE_HEIGHT = 5;
export const AVM_RETRIEVED_BYTECODES_TREE_INITIAL_ROOT = 7257575663883662864904159007845791361042428565864275462740313586853981161757n;
export const AVM_RETRIEVED_BYTECODES_TREE_INITIAL_SIZE = 1;
export const TIMESTAMP_OF_CHANGE_BIT_SIZE = 32;
export const DEFAULT_UPDATE_DELAY = 86400;
export const MINIMUM_UPDATE_DELAY = 600;
export const UPDATES_VALUE_SIZE = 1;
export const UPDATES_DELAYED_PUBLIC_MUTABLE_VALUES_LEN = 3;
export const UPDATES_DELAYED_PUBLIC_MUTABLE_SDC_DELAY_BIT_SIZE = 32;
export const UPDATES_DELAYED_PUBLIC_MUTABLE_SDC_IS_SOME_BIT_SIZE = 8;
export const UPDATES_DELAYED_PUBLIC_MUTABLE_SDC_OPTION_DELAY_BIT_SIZE = 40;
export const UPDATES_DELAYED_PUBLIC_MUTABLE_METADATA_BIT_SIZE = 144;
export const GRUMPKIN_ONE_X = 1;
export const GRUMPKIN_ONE_Y = 17631683881184975370165255887551781615748388533673675138860n;
export const DEFAULT_MAX_DEBUG_LOG_MEMORY_READS = 125000;
export var DomainSeparator = /*#__PURE__*/ function(DomainSeparator) {
    DomainSeparator[DomainSeparator["NOTE_HASH"] = 116501019] = "NOTE_HASH";
    DomainSeparator[DomainSeparator["NOTE_HASH_NONCE"] = 1721808740] = "NOTE_HASH_NONCE";
    DomainSeparator[DomainSeparator["UNIQUE_NOTE_HASH"] = 226850429] = "UNIQUE_NOTE_HASH";
    DomainSeparator[DomainSeparator["SILOED_NOTE_HASH"] = 3361878420] = "SILOED_NOTE_HASH";
    DomainSeparator[DomainSeparator["NOTE_NULLIFIER"] = 50789342] = "NOTE_NULLIFIER";
    DomainSeparator[DomainSeparator["MESSAGE_NULLIFIER"] = 3754509616] = "MESSAGE_NULLIFIER";
    DomainSeparator[DomainSeparator["SILOED_NULLIFIER"] = 57496191] = "SILOED_NULLIFIER";
    DomainSeparator[DomainSeparator["PRIVATE_LOG_FIRST_FIELD"] = 2769976252] = "PRIVATE_LOG_FIRST_FIELD";
    DomainSeparator[DomainSeparator["PUBLIC_LEAF_SLOT"] = 1247650290] = "PUBLIC_LEAF_SLOT";
    DomainSeparator[DomainSeparator["PUBLIC_STORAGE_MAP_SLOT"] = 4015149901] = "PUBLIC_STORAGE_MAP_SLOT";
    DomainSeparator[DomainSeparator["PRIVATE_FUNCTION_LEAF"] = 1389398688] = "PRIVATE_FUNCTION_LEAF";
    DomainSeparator[DomainSeparator["PUBLIC_BYTECODE"] = 260313585] = "PUBLIC_BYTECODE";
    DomainSeparator[DomainSeparator["CONTRACT_CLASS_ID"] = 3923495515] = "CONTRACT_CLASS_ID";
    DomainSeparator[DomainSeparator["INITIALIZER"] = 385396519] = "INITIALIZER";
    DomainSeparator[DomainSeparator["NHK_M"] = 242137788] = "NHK_M";
    DomainSeparator[DomainSeparator["IVSK_M"] = 2747825907] = "IVSK_M";
    DomainSeparator[DomainSeparator["OVSK_M"] = 4272201051] = "OVSK_M";
    DomainSeparator[DomainSeparator["TSK_M"] = 1546190975] = "TSK_M";
    DomainSeparator[DomainSeparator["PUBLIC_KEYS_HASH"] = 777457226] = "PUBLIC_KEYS_HASH";
    DomainSeparator[DomainSeparator["PARTIAL_ADDRESS"] = 2103633018] = "PARTIAL_ADDRESS";
    DomainSeparator[DomainSeparator["CONTRACT_ADDRESS_V1"] = 1788365517] = "CONTRACT_ADDRESS_V1";
    DomainSeparator[DomainSeparator["BLOCK_HEADER_HASH"] = 4195546849] = "BLOCK_HEADER_HASH";
    DomainSeparator[DomainSeparator["TX_REQUEST"] = 3763737512] = "TX_REQUEST";
    DomainSeparator[DomainSeparator["PUBLIC_TX_HASH"] = 1630108851] = "PUBLIC_TX_HASH";
    DomainSeparator[DomainSeparator["PRIVATE_TX_HASH"] = 1971680439] = "PRIVATE_TX_HASH";
    DomainSeparator[DomainSeparator["PUBLIC_CALLDATA"] = 2760353947] = "PUBLIC_CALLDATA";
    DomainSeparator[DomainSeparator["FUNCTION_ARGS"] = 3576554347] = "FUNCTION_ARGS";
    DomainSeparator[DomainSeparator["PROTOCOL_CONTRACTS"] = 3904434327] = "PROTOCOL_CONTRACTS";
    DomainSeparator[DomainSeparator["EVENT_COMMITMENT"] = 2517418573] = "EVENT_COMMITMENT";
    DomainSeparator[DomainSeparator["AUTHWIT_INNER"] = 221354163] = "AUTHWIT_INNER";
    DomainSeparator[DomainSeparator["AUTHWIT_OUTER"] = 3283595782] = "AUTHWIT_OUTER";
    DomainSeparator[DomainSeparator["AUTHWIT_NULLIFIER"] = 1239150694] = "AUTHWIT_NULLIFIER";
    DomainSeparator[DomainSeparator["SYMMETRIC_KEY"] = 3882206064] = "SYMMETRIC_KEY";
    DomainSeparator[DomainSeparator["SYMMETRIC_KEY_2"] = 4129434989] = "SYMMETRIC_KEY_2";
    DomainSeparator[DomainSeparator["PARTIAL_NOTE_VALIDITY_COMMITMENT"] = 623934423] = "PARTIAL_NOTE_VALIDITY_COMMITMENT";
    DomainSeparator[DomainSeparator["INITIALIZATION_NULLIFIER"] = 1653084894] = "INITIALIZATION_NULLIFIER";
    DomainSeparator[DomainSeparator["SECRET_HASH"] = 4199652938] = "SECRET_HASH";
    DomainSeparator[DomainSeparator["TX_NULLIFIER"] = 1025801951] = "TX_NULLIFIER";
    DomainSeparator[DomainSeparator["SIGNATURE_PAYLOAD"] = 463525807] = "SIGNATURE_PAYLOAD";
    return DomainSeparator;
}({});
