// @aztec/simulator/dest/public/avm/serialization/buffer_cursor.js
/*
 * A Buffer-like class that automatically advances the position.
 */ export class BufferCursor {
    _buffer;
    _position;
    constructor(_buffer, _position = 0){
        this._buffer = _buffer;
        this._position = _position;
    }
    position() {
        return this._position;
    }
    buffer() {
        return this._buffer;
    }
    eof() {
        return this._position === this._buffer.length;
    }
    bufferAtPosition() {
        return this._buffer.subarray(this._position);
    }
    advance(n) {
        this._position += n;
    }
    peekUint8() {
        const ret = this._buffer.readUint8(this._position);
        return ret;
    }
    readUint8() {
        const ret = this._buffer.readUint8(this._position);
        this._position += 1;
        return ret;
    }
    readUint16LE() {
        const ret = this._buffer.readUint16LE(this._position);
        this._position += 2;
        return ret;
    }
    readUint16BE() {
        const ret = this._buffer.readUint16BE(this._position);
        this._position += 2;
        return ret;
    }
    readUint32LE() {
        const ret = this._buffer.readUint32LE(this._position);
        this._position += 4;
        return ret;
    }
    readUint32BE() {
        const ret = this._buffer.readUint32BE(this._position);
        this._position += 4;
        return ret;
    }
    readBigInt64LE() {
        const ret = this._buffer.readBigInt64LE(this._position);
        this._position += 8;
        return ret;
    }
    readBigInt64BE() {
        const ret = this._buffer.readBigInt64BE(this._position);
        this._position += 8;
        return ret;
    }
    writeUint8(v) {
        const ret = this._buffer.writeUint8(v, this._position);
        this._position += 1;
        return ret;
    }
    writeUint16LE(v) {
        const ret = this._buffer.writeUint16LE(v, this._position);
        this._position += 2;
        return ret;
    }
    writeUint16BE(v) {
        const ret = this._buffer.writeUint16BE(v, this._position);
        this._position += 2;
        return ret;
    }
    writeUint32LE(v) {
        const ret = this._buffer.writeUint32LE(v, this._position);
        this._position += 4;
        return ret;
    }
    writeUint32BE(v) {
        const ret = this._buffer.writeUint32BE(v, this._position);
        this._position += 4;
        return ret;
    }
    writeBigInt64LE(v) {
        const ret = this._buffer.writeBigInt64LE(v, this._position);
        this._position += 8;
        return ret;
    }
    writeBigInt64BE(v) {
        const ret = this._buffer.writeBigInt64BE(v, this._position);
        this._position += 8;
        return ret;
    }
}
