// @aztec/aztec.js/src/contract/protocol_contracts/fee-juice.ts
/* Autogenerated file, do not edit! */
/* eslint-disable */
import { ProtocolContractAddress } from '@aztec/protocol-contracts';
import { FunctionType } from '@aztec/stdlib/abi';

import type { ContractArtifact } from '../../api/abi.js';
import { PublicKeys } from '../../api/keys.js';
import type {
  AztecAddressLike,
  EthAddressLike,
  FieldLike,
  FunctionSelectorLike,
  WrappedFieldLike,
} from '../../utils/abi_types.js';
import type { Wallet } from '../../wallet/wallet.js';
import { ContractBase, type ContractMethod } from '../contract_base.js';
import { ContractFunctionInteraction } from '../contract_function_interaction.js';

const FeeJuiceContractArtifact: ContractArtifact = {
  name: 'FeeJuice',
  functions: [
    {
      ...{
        functionType: FunctionType.PRIVATE,
        name: 'claim',
        isOnlySelf: false,
        isStatic: false,
        isInitializer: false,
        parameters: [
          {
            name: 'to',
            type: {
              kind: 'struct',
              fields: [{ name: 'inner', type: { kind: 'field' } }],
              path: 'aztec::protocol_types::address::aztec_address::AztecAddress',
            },
            visibility: 'private',
          },
          { name: 'amount', type: { kind: 'integer', sign: 'unsigned', width: 128 }, visibility: 'private' },
          { name: 'secret', type: { kind: 'field' }, visibility: 'private' },
          { name: 'message_leaf_index', type: { kind: 'field' }, visibility: 'private' },
        ],
        returnTypes: [],
        errorTypes: {
          '361444214588792908': { error_kind: 'string', string: 'attempt to multiply with overflow' },
          '1998584279744703196': { error_kind: 'string', string: 'attempt to subtract with overflow' },
          '6609888949476313245': { error_kind: 'string', string: 'Message not in state' },
          '11019205087382408538': { error_kind: 'string', string: 'Field failed to decompose into specified 4 limbs' },
          '12370419938245003393': { error_kind: 'string', string: 'Field failed to decompose into specified 36 limbs' },
          '14990209321349310352': { error_kind: 'string', string: 'attempt to add with overflow' },
          '15764276373176857197': { error_kind: 'string', string: 'Stack too deep' },
          '15835548349546956319': { error_kind: 'string', string: 'Field failed to decompose into specified 32 limbs' },
          '16431471497789672479': { error_kind: 'string', string: 'Index out of bounds' },
        },
      },
      bytecode: Buffer.from([]),
      debugSymbols: '',
    },
    {
      ...{
        functionType: FunctionType.PRIVATE,
        name: 'claim_and_end_setup',
        isOnlySelf: false,
        isStatic: false,
        isInitializer: false,
        parameters: [
          {
            name: 'to',
            type: {
              kind: 'struct',
              fields: [{ name: 'inner', type: { kind: 'field' } }],
              path: 'aztec::protocol_types::address::aztec_address::AztecAddress',
            },
            visibility: 'private',
          },
          { name: 'amount', type: { kind: 'integer', sign: 'unsigned', width: 128 }, visibility: 'private' },
          { name: 'secret', type: { kind: 'field' }, visibility: 'private' },
          { name: 'message_leaf_index', type: { kind: 'field' }, visibility: 'private' },
        ],
        returnTypes: [],
        errorTypes: {
          '361444214588792908': { error_kind: 'string', string: 'attempt to multiply with overflow' },
          '1998584279744703196': { error_kind: 'string', string: 'attempt to subtract with overflow' },
          '6609888949476313245': { error_kind: 'string', string: 'Message not in state' },
          '11019205087382408538': { error_kind: 'string', string: 'Field failed to decompose into specified 4 limbs' },
          '12370419938245003393': { error_kind: 'string', string: 'Field failed to decompose into specified 36 limbs' },
          '14990209321349310352': { error_kind: 'string', string: 'attempt to add with overflow' },
          '15764276373176857197': { error_kind: 'string', string: 'Stack too deep' },
          '15835548349546956319': { error_kind: 'string', string: 'Field failed to decompose into specified 32 limbs' },
          '16431471497789672479': { error_kind: 'string', string: 'Index out of bounds' },
        },
      },
      bytecode: Buffer.from([]),
      debugSymbols: '',
    },
    {
      ...{
        functionType: FunctionType.PUBLIC,
        name: 'public_dispatch',
        isOnlySelf: false,
        isStatic: false,
        isInitializer: false,
        parameters: [{ name: 'selector', type: { kind: 'field' }, visibility: 'private' }],
        returnTypes: [],
        errorTypes: {
          '361444214588792908': { error_kind: 'string', string: 'attempt to multiply with overflow' },
          '1080336462466469436': {
            error_kind: 'string',
            string: 'Function check_balance can only be called statically',
          },
          '1998584279744703196': { error_kind: 'string', string: 'attempt to subtract with overflow' },
          '8637882143979902881': {
            error_kind: 'string',
            string: 'Function _increase_public_balance can only be called by the same contract',
          },
          '9543832133280402246': { error_kind: 'string', string: 'Balance too low' },
          '12511970388699677811': { error_kind: 'fmtstring', length: 27, item_types: [{ kind: 'field' }] },
          '13455385521185560676': {
            error_kind: 'string',
            string: 'Storage slot 0 not allowed. Storage slots must start from 1.',
          },
          '14990209321349310352': { error_kind: 'string', string: 'attempt to add with overflow' },
          '15764276373176857197': { error_kind: 'string', string: 'Stack too deep' },
          '15860392472534760515': {
            error_kind: 'string',
            string: 'Function balance_of_public can only be called statically',
          },
          '16431471497789672479': { error_kind: 'string', string: 'Index out of bounds' },
        },
      },
      bytecode: Buffer.from([]),
      debugSymbols: '',
    },
  ],
  nonDispatchPublicFunctions: [
    {
      functionType: FunctionType.PUBLIC,
      name: '_increase_public_balance',
      isOnlySelf: true,
      isStatic: false,
      isInitializer: false,
      parameters: [
        {
          name: 'to',
          type: {
            kind: 'struct',
            fields: [{ name: 'inner', type: { kind: 'field' } }],
            path: 'aztec::protocol_types::address::aztec_address::AztecAddress',
          },
          visibility: 'private',
        },
        { name: 'amount', type: { kind: 'integer', sign: 'unsigned', width: 128 }, visibility: 'private' },
      ],
      returnTypes: [],
      errorTypes: {
        '361444214588792908': { error_kind: 'string', string: 'attempt to multiply with overflow' },
        '1998584279744703196': { error_kind: 'string', string: 'attempt to subtract with overflow' },
        '8637882143979902881': {
          error_kind: 'string',
          string: 'Function _increase_public_balance can only be called by the same contract',
        },
        '13455385521185560676': {
          error_kind: 'string',
          string: 'Storage slot 0 not allowed. Storage slots must start from 1.',
        },
        '14990209321349310352': { error_kind: 'string', string: 'attempt to add with overflow' },
        '15764276373176857197': { error_kind: 'string', string: 'Stack too deep' },
        '16431471497789672479': { error_kind: 'string', string: 'Index out of bounds' },
      },
    },
    {
      functionType: FunctionType.PUBLIC,
      name: 'balance_of_public',
      isOnlySelf: false,
      isStatic: true,
      isInitializer: false,
      parameters: [
        {
          name: 'owner',
          type: {
            kind: 'struct',
            fields: [{ name: 'inner', type: { kind: 'field' } }],
            path: 'aztec::protocol_types::address::aztec_address::AztecAddress',
          },
          visibility: 'private',
        },
      ],
      returnTypes: [{ kind: 'integer', sign: 'unsigned', width: 128 }],
      errorTypes: {
        '13455385521185560676': {
          error_kind: 'string',
          string: 'Storage slot 0 not allowed. Storage slots must start from 1.',
        },
        '15764276373176857197': { error_kind: 'string', string: 'Stack too deep' },
        '15860392472534760515': {
          error_kind: 'string',
          string: 'Function balance_of_public can only be called statically',
        },
      },
    },
    {
      functionType: FunctionType.PUBLIC,
      name: 'check_balance',
      isOnlySelf: false,
      isStatic: true,
      isInitializer: false,
      parameters: [
        { name: 'fee_limit', type: { kind: 'integer', sign: 'unsigned', width: 128 }, visibility: 'private' },
      ],
      returnTypes: [],
      errorTypes: {
        '1080336462466469436': { error_kind: 'string', string: 'Function check_balance can only be called statically' },
        '9543832133280402246': { error_kind: 'string', string: 'Balance too low' },
        '13455385521185560676': {
          error_kind: 'string',
          string: 'Storage slot 0 not allowed. Storage slots must start from 1.',
        },
        '15764276373176857197': { error_kind: 'string', string: 'Stack too deep' },
      },
    },
  ],
  outputs: { structs: {}, globals: {} },
  storageLayout: {},
  fileMap: {},
};

export class FeeJuiceContract extends ContractBase {
  private constructor(wallet: Wallet) {
    super(ProtocolContractAddress.FeeJuice, FeeJuiceContractArtifact, wallet);
  }

  public static at(wallet: Wallet): FeeJuiceContract {
    return new FeeJuiceContract(wallet);
  }

  declare public methods: {
    /** balance_of_public(owner: struct) */
    balance_of_public: ((owner: AztecAddressLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** check_balance(fee_limit: integer) */
    check_balance: ((fee_limit: bigint | number) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;

    /** claim(to: struct, amount: integer, secret: field, message_leaf_index: field) */
    claim: ((
      to: AztecAddressLike,
      amount: bigint | number,
      secret: FieldLike,
      message_leaf_index: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** claim_and_end_setup(to: struct, amount: integer, secret: field, message_leaf_index: field) */
    claim_and_end_setup: ((
      to: AztecAddressLike,
      amount: bigint | number,
      secret: FieldLike,
      message_leaf_index: FieldLike,
    ) => ContractFunctionInteraction) &
      Pick<ContractMethod, 'selector'>;

    /** public_dispatch(selector: field) */
    public_dispatch: ((selector: FieldLike) => ContractFunctionInteraction) & Pick<ContractMethod, 'selector'>;
  };
}
