// @aztec/pxe/dest/entrypoints/client/bundle/utils.js
import { BBPrivateKernelProver } from '@aztec/bb-prover/client';
import { BBBundlePrivateKernelProver } from '@aztec/bb-prover/client/bundle';
import { createLogger } from '@aztec/foundation/log';
import { createStore } from '@aztec/kv-store/indexeddb';
import { BundledProtocolContractsProvider } from '@aztec/protocol-contracts/providers/bundle';
import { WASMSimulator } from '@aztec/simulator/client';
import { PXE } from '../../../pxe.js';
import { PXE_DATA_SCHEMA_VERSION } from '../../../storage/metadata.js';
/**
 * Create and start an PXE instance with the given AztecNode.
 * If no keyStore or database is provided, it will use KeyStore and MemoryDB as default values.
 * Returns a Promise that resolves to the started PXE instance.
 *
 * @param aztecNode - The AztecNode instance to be used by the server.
 * @param config - The PXE Config to use
 * @param options - (Optional) Optional information for creating an PXE.
 * @returns A Promise that resolves to the started PXE instance.
 */ export async function createPXE(aztecNode, config, options = {
    loggers: {}
}) {
    const actor = options.loggerActorLabel;
    const loggers = options.loggers ?? {};
    const l1Contracts = await aztecNode.getL1ContractAddresses();
    const configWithContracts = {
        ...config,
        l1Contracts
    };
    const storeLogger = loggers.store ?? createLogger('pxe:data:idb', {
        actor
    });
    const store = options.store ?? await createStore('pxe_data', configWithContracts, PXE_DATA_SCHEMA_VERSION, storeLogger);
    const simulator = options.simulator ?? new WASMSimulator();
    const proverLogger = loggers.prover ?? createLogger('pxe:bb:wasm:bundle', {
        actor
    });
    let prover;
    if (options.proverOrOptions instanceof BBPrivateKernelProver) {
        prover = options.proverOrOptions;
    } else {
        prover = new BBBundlePrivateKernelProver(simulator, {
            ...options.proverOrOptions,
            logger: proverLogger
        });
    }
    const protocolContractsProvider = new BundledProtocolContractsProvider();
    const pxeLogger = loggers.pxe ?? createLogger('pxe:service', {
        actor
    });
    const pxe = await PXE.create({
        node: aztecNode,
        store,
        proofCreator: prover,
        simulator,
        protocolContractsProvider,
        config,
        loggerOrSuffix: pxeLogger
    });
    return pxe;
}
