// @aztec/stdlib/dest/kernel/hints/nullifier_read_request_hints.js
import { MAX_NULLIFIER_READ_REQUESTS_PER_TX, NULLIFIER_TREE_HEIGHT } from '@aztec/constants';
import { makeTuple } from '@aztec/foundation/array';
import { NullifierLeafPreimage } from '../../trees/index.js';
import { PendingReadHint, ReadRequestAction, ReadRequestResetHints, SettledReadHint } from './read_request_hints.js';
export function nullifierReadRequestHintsFromBuffer(buffer, numPendingReads, numSettledReads) {
    return ReadRequestResetHints.fromBuffer(buffer, MAX_NULLIFIER_READ_REQUESTS_PER_TX, numPendingReads, numSettledReads, NULLIFIER_TREE_HEIGHT, NullifierLeafPreimage);
}
export class NullifierReadRequestHintsBuilder {
    maxPending;
    maxSettled;
    hints;
    numPendingReadHints;
    numSettledReadHints;
    constructor(maxPending, maxSettled){
        this.maxPending = maxPending;
        this.maxSettled = maxSettled;
        this.numPendingReadHints = 0;
        this.numSettledReadHints = 0;
        this.hints = new ReadRequestResetHints(makeTuple(MAX_NULLIFIER_READ_REQUESTS_PER_TX, ReadRequestAction.skip), makeTuple(maxPending, ()=>PendingReadHint.nada(MAX_NULLIFIER_READ_REQUESTS_PER_TX)), makeTuple(maxSettled, ()=>SettledReadHint.nada(MAX_NULLIFIER_READ_REQUESTS_PER_TX, NULLIFIER_TREE_HEIGHT, NullifierLeafPreimage.empty)));
    }
    static empty(maxPending, maxSettled) {
        return new NullifierReadRequestHintsBuilder(maxPending, maxSettled).toHints();
    }
    addPendingReadRequest(readRequestIndex, nullifierIndex) {
        if (this.numPendingReadHints === this.maxPending) {
            throw new Error('Cannot add more pending read request.');
        }
        this.hints.readRequestActions[readRequestIndex] = ReadRequestAction.readAsPending(this.numPendingReadHints);
        this.hints.pendingReadHints[this.numPendingReadHints] = new PendingReadHint(readRequestIndex, nullifierIndex);
        this.numPendingReadHints++;
    }
    addSettledReadRequest(readRequestIndex, membershipWitness, leafPreimage) {
        if (this.numSettledReadHints === this.maxSettled) {
            throw new Error('Cannot add more settled read request.');
        }
        this.hints.readRequestActions[readRequestIndex] = ReadRequestAction.readAsSettled(this.numSettledReadHints);
        this.hints.settledReadHints[this.numSettledReadHints] = new SettledReadHint(readRequestIndex, membershipWitness, leafPreimage);
        this.numSettledReadHints++;
    }
    toHints() {
        return this.hints;
    }
}
