// @aztec/stdlib/dest/kernel/hints/read_request_hints.js
import { makeTuple } from '@aztec/foundation/array';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { MembershipWitness } from '@aztec/foundation/trees';
export var ReadRequestActionEnum = /*#__PURE__*/ function(ReadRequestActionEnum) {
    ReadRequestActionEnum[ReadRequestActionEnum["SKIP"] = 0] = "SKIP";
    ReadRequestActionEnum[ReadRequestActionEnum["READ_AS_PENDING"] = 1] = "READ_AS_PENDING";
    ReadRequestActionEnum[ReadRequestActionEnum["READ_AS_SETTLED"] = 2] = "READ_AS_SETTLED";
    return ReadRequestActionEnum;
}({});
export class ReadRequestAction {
    action;
    hintIndex;
    constructor(action, hintIndex){
        this.action = action;
        this.hintIndex = hintIndex;
    }
    static skip() {
        return new ReadRequestAction(0, 0);
    }
    static readAsPending(hintIndex) {
        return new ReadRequestAction(1, hintIndex);
    }
    static readAsSettled(hintIndex) {
        return new ReadRequestAction(2, hintIndex);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new ReadRequestAction(reader.readNumber(), reader.readNumber());
    }
    toBuffer() {
        return serializeToBuffer(this.action, this.hintIndex);
    }
}
export class PendingReadHint {
    readRequestIndex;
    pendingValueIndex;
    constructor(readRequestIndex, pendingValueIndex){
        this.readRequestIndex = readRequestIndex;
        this.pendingValueIndex = pendingValueIndex;
    }
    static nada(readRequestLen) {
        return new PendingReadHint(readRequestLen, 0);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PendingReadHint(reader.readNumber(), reader.readNumber());
    }
    toBuffer() {
        return serializeToBuffer(this.readRequestIndex, this.pendingValueIndex);
    }
}
export class SettledReadHint {
    readRequestIndex;
    membershipWitness;
    leafPreimage;
    constructor(readRequestIndex, membershipWitness, leafPreimage){
        this.readRequestIndex = readRequestIndex;
        this.membershipWitness = membershipWitness;
        this.leafPreimage = leafPreimage;
    }
    static nada(readRequestLen, treeHeight, emptyLeafPreimage) {
        return new SettledReadHint(readRequestLen, MembershipWitness.empty(treeHeight), emptyLeafPreimage());
    }
    static fromBuffer(buffer, treeHeight, leafPreimage) {
        const reader = BufferReader.asReader(buffer);
        return new SettledReadHint(reader.readNumber(), MembershipWitness.fromBuffer(reader, treeHeight), reader.readObject(leafPreimage));
    }
    toBuffer() {
        return serializeToBuffer(this.readRequestIndex, this.membershipWitness, this.leafPreimage);
    }
}
/**
 * Hints for read request reset circuit.
 */ export class ReadRequestResetHints {
    readRequestActions;
    pendingReadHints;
    settledReadHints;
    constructor(readRequestActions, /**
     * The hints for read requests reading pending values.
     */ pendingReadHints, /**
     * The hints for read requests reading settled values.
     */ settledReadHints){
        this.readRequestActions = readRequestActions;
        this.pendingReadHints = pendingReadHints;
        this.settledReadHints = settledReadHints;
    }
    trimToSizes(numPendingReads, numSettledReads) {
        return new ReadRequestResetHints(this.readRequestActions, this.pendingReadHints.slice(0, numPendingReads), this.settledReadHints.slice(0, numSettledReads));
    }
    /**
   * Deserializes from a buffer or reader.
   * @param buffer - Buffer or reader to read from.
   * @returns The deserialized instance.
   */ static fromBuffer(buffer, readRequestLen, numPendingReads, numSettledReads, treeHeight, leafPreimageFromBuffer) {
        const reader = BufferReader.asReader(buffer);
        return new ReadRequestResetHints(reader.readArray(readRequestLen, ReadRequestAction), reader.readArray(numPendingReads, PendingReadHint), reader.readArray(numSettledReads, {
            fromBuffer: (r)=>SettledReadHint.fromBuffer(r, treeHeight, leafPreimageFromBuffer)
        }));
    }
    toBuffer() {
        return serializeToBuffer(this.readRequestActions, this.pendingReadHints, this.settledReadHints);
    }
}
export class ReadRequestResetActions {
    actions;
    pendingReadHints;
    constructor(actions, pendingReadHints){
        this.actions = actions;
        this.pendingReadHints = pendingReadHints;
    }
    static empty(numReads) {
        return new ReadRequestResetActions(makeTuple(numReads, ()=>0), []);
    }
}
