// @aztec/stdlib/dest/kernel/hints/transient_data_squashing_hint.js
var _computedKey;
import { Fr } from '@aztec/foundation/curves/bn254';
import { BufferReader, FieldReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { inspect } from 'util';
_computedKey = inspect.custom;
export class TransientDataSquashingHint {
    nullifierIndex;
    noteHashIndex;
    constructor(nullifierIndex, noteHashIndex){
        this.nullifierIndex = nullifierIndex;
        this.noteHashIndex = noteHashIndex;
    }
    toFields() {
        return [
            new Fr(this.nullifierIndex),
            new Fr(this.noteHashIndex)
        ];
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new TransientDataSquashingHint(reader.readU32(), reader.readU32());
    }
    isEmpty() {
        return !this.nullifierIndex && !this.noteHashIndex;
    }
    static empty() {
        return new TransientDataSquashingHint(0, 0);
    }
    toBuffer() {
        return serializeToBuffer(this.nullifierIndex, this.noteHashIndex);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new TransientDataSquashingHint(reader.readNumber(), reader.readNumber());
    }
    toString() {
        return `nullifierIndex=${this.nullifierIndex} noteHashIndex=${this.noteHashIndex}`;
    }
    [_computedKey]() {
        return `TransientDataSquashingHint { ${this.toString()} }`;
    }
}
