// @aztec/stdlib/dest/kernel/hints/tree_leaf_read_request.js
import { Fr } from '@aztec/foundation/curves/bn254';
import { BufferReader, FieldReader, serializeToBuffer } from '@aztec/foundation/serialize';
export class TreeLeafReadRequest {
    value;
    leafIndex;
    constructor(value, leafIndex){
        this.value = value;
        this.leafIndex = leafIndex;
    }
    toBuffer() {
        return serializeToBuffer(this.value, this.leafIndex);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new TreeLeafReadRequest(Fr.fromBuffer(reader), Fr.fromBuffer(reader));
    }
    toFields() {
        return [
            this.value,
            this.leafIndex
        ];
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new TreeLeafReadRequest(reader.readField(), reader.readField());
    }
    isEmpty() {
        return this.value.isZero() && this.leafIndex.isZero();
    }
    static empty() {
        return new TreeLeafReadRequest(Fr.zero(), Fr.zero());
    }
}
