// @aztec/pxe/dest/debug/pxe_debug_utils.js
/**
 * Methods provided by this class might help debugging but must not be used in production.
 * No backwards compatibility or API stability should be expected. Use at your own risk.
 */ export class PXEDebugUtils {
    contractSyncService;
    noteStore;
    blockStateSynchronizer;
    anchorBlockStore;
    #putJobInQueue;
    #getSimulatorForTx;
    #simulateUtility;
    constructor(contractSyncService, noteStore, blockStateSynchronizer, anchorBlockStore){
        this.contractSyncService = contractSyncService;
        this.noteStore = noteStore;
        this.blockStateSynchronizer = blockStateSynchronizer;
        this.anchorBlockStore = anchorBlockStore;
    }
    /** Not injected through constructor since they're are co-dependant */ setPXEHelpers(putJobInQueue, getSimulatorForTx, simulateUtility) {
        this.#putJobInQueue = putJobInQueue;
        this.#getSimulatorForTx = getSimulatorForTx;
        this.#simulateUtility = simulateUtility;
    }
    /**
   * A debugging utility to get notes based on the provided filter.
   *
   * Note that this should not be used in production code because the structure of notes is considered to be
   * an implementation detail of contracts. This is only meant to be used for debugging purposes. If you need to obtain
   * note-related information in production code, please implement a custom utility function on your contract and call
   * that function instead (e.g. `get_balance(owner: AztecAddress) -> u128` utility function on a Token contract).
   *
   * @param filter - The filter to apply to the notes.
   * @returns The requested notes.
   */ getNotes(filter) {
        return this.#putJobInQueue(async (jobId)=>{
            await this.blockStateSynchronizer.sync();
            const anchorBlockHeader = await this.anchorBlockStore.getBlockHeader();
            const contractFunctionSimulator = this.#getSimulatorForTx();
            await this.contractSyncService.ensureContractSynced(filter.contractAddress, null, async (privateSyncCall, execScopes)=>await this.#simulateUtility(contractFunctionSimulator, privateSyncCall, [], execScopes, jobId), anchorBlockHeader, jobId, filter.scopes);
            return this.noteStore.getNotes(filter, jobId);
        });
    }
    /**
   * Triggers a sync of the PXE with the node.
   * Blocks until the sync is complete.
   */ sync() {
        return this.#putJobInQueue(()=>this.blockStateSynchronizer.sync());
    }
}
