// @aztec/bb.js/dest/node-cjs/barretenberg_wasm/helpers/browser/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Ready = void 0;
exports.getSharedMemoryAvailable = getSharedMemoryAvailable;
exports.getRemoteBarretenbergWasm = getRemoteBarretenbergWasm;
exports.getNumCpu = getNumCpu;
exports.threadLogger = threadLogger;
exports.killSelf = killSelf;
exports.getAvailableThreads = getAvailableThreads;
exports.readinessListener = readinessListener;
const comlink_1 = require("comlink");
function getSharedMemoryAvailable() {
    const globalScope = typeof window !== 'undefined' ? window : globalThis;
    return typeof SharedArrayBuffer !== 'undefined' && globalScope.crossOriginIsolated;
}
function getRemoteBarretenbergWasm(worker) {
    return (0, comlink_1.wrap)(worker);
}
function getNumCpu() {
    return navigator.hardwareConcurrency;
}
function threadLogger(useCustomLogger) {
    if (useCustomLogger) {
        // Post log messages back to main thread for routing through user-provided logger
        return (msg) => {
            postMessage({ type: 'log', msg });
        };
    }
    // Use console.log directly when no custom logger is provided
    return console.log;
}
function killSelf() {
    self.close();
}
function getAvailableThreads(logger) {
    if (typeof navigator !== 'undefined' && navigator.hardwareConcurrency) {
        return navigator.hardwareConcurrency;
    }
    else {
        logger(`Could not detect environment to query number of threads. Falling back to one thread.`);
        return 1;
    }
}
// Solution to async initialization of workers, taken from
// https://github.com/GoogleChromeLabs/comlink/issues/635#issuecomment-1598913044
/** The message expected by the `readinessListener`. */
exports.Ready = { ready: true };
/** Listen for the readiness message from the Worker and call the `callback` once. */
function readinessListener(worker, callback) {
    worker.addEventListener('message', function ready(event) {
        if (!!event.data && event.data.ready === true) {
            worker.removeEventListener('message', ready);
            callback();
        }
    });
}
