// @aztec/stdlib/src/proofs/proof_data.ts
import type { CHONK_PROOF_LENGTH, RECURSIVE_PROOF_LENGTH, RECURSIVE_ROLLUP_HONK_PROOF_LENGTH } from '@aztec/constants';
import { BufferReader, type Bufferable, serializeToBuffer } from '@aztec/foundation/serialize';

import { RecursiveProof } from '../proofs/recursive_proof.js';
import { VkData } from '../vks/index.js';

/**
 * Represents the data of a recursive proof.
 */
export class ProofData<T extends Bufferable, PROOF_LENGTH extends number> {
  constructor(
    public publicInputs: T,
    public proof: RecursiveProof<PROOF_LENGTH>,
    public vkData: VkData,
  ) {}

  public toBuffer(): Buffer {
    return serializeToBuffer(this.publicInputs, this.proof, this.vkData);
  }

  public static fromBuffer<T extends Bufferable, PROOF_LENGTH extends number>(
    buffer: Buffer | BufferReader,
    publicInputs: {
      fromBuffer: (reader: BufferReader) => T;
    },
  ): ProofData<T, PROOF_LENGTH> {
    const reader = BufferReader.asReader(buffer);
    return new ProofData(reader.readObject(publicInputs), RecursiveProof.fromBuffer(reader), reader.readObject(VkData));
  }
}

/**
 * Represents the data of a recursive proof for a circuit with a fixed verification key.
 */
export class ProofDataForFixedVk<T extends Bufferable, PROOF_LENGTH extends number> {
  constructor(
    public publicInputs: T,
    public proof: RecursiveProof<PROOF_LENGTH>,
  ) {}

  public toBuffer(): Buffer {
    return serializeToBuffer(this.publicInputs, this.proof);
  }

  public static fromBuffer<T extends Bufferable, PROOF_LENGTH extends number>(
    buffer: Buffer | BufferReader,
    publicInputs: {
      fromBuffer: (reader: BufferReader) => T;
    },
  ): ProofDataForFixedVk<T, PROOF_LENGTH> {
    const reader = BufferReader.asReader(buffer);
    return new ProofDataForFixedVk(reader.readObject(publicInputs), RecursiveProof.fromBuffer(reader));
  }
}

export type ChonkProofData<T extends Bufferable> = ProofData<T, typeof CHONK_PROOF_LENGTH>;

export type UltraHonkProofData<T extends Bufferable> = ProofData<T, typeof RECURSIVE_PROOF_LENGTH>;

export type RollupHonkProofData<T extends Bufferable> = ProofData<T, typeof RECURSIVE_ROLLUP_HONK_PROOF_LENGTH>;
