// @aztec/simulator/dest/private/acvm_wasm_with_blobs.js
import { Timer } from '@aztec/foundation/timer';
import { executeCircuit } from '@aztec/noir-acvm_js';
import { enrichNoirError } from './circuit_simulator.js';
/**
 * A circuit simulator that uses the WASM simulator with the ability to handle blobs via the foreign call handler.
 * This class is temporary while brillig cannot handle the blob math, and it is kept separate
 * because the zkg commitment library used in the blob code is not browser compatible.
 *
 * It is only used in the context of server-side code executing simulated protocol circuits.
 */ export class WASMSimulatorWithBlobs {
    async executeProtocolCircuit(input, artifact, callback) {
        // Decode the bytecode from base64 since the acvm does not know about base64 encoding
        const decodedBytecode = Buffer.from(artifact.bytecode, 'base64');
        //
        // Execute the circuit
        try {
            const timer = new Timer();
            const _witnessMap = await executeCircuit(decodedBytecode, input, callback);
            return {
                witness: _witnessMap,
                duration: timer.ms()
            };
        } catch (err) {
            // Typescript types caught errors as unknown or any, so we need to narrow its type to check if it has raw
            // assertion payload.
            if (typeof err === 'object' && err !== null && 'rawAssertionPayload' in err) {
                throw enrichNoirError(artifact, err);
            }
            throw new Error(`Circuit execution failed: ${err}`);
        }
    }
    executeUserCircuit(_input, _artifact, _callback) {
        throw new Error('Not implemented');
    }
}
