// @aztec/simulator/dest/private/acvm_wasm.js
import { resolveLogger } from '@aztec/foundation/log';
import { Timer } from '@aztec/foundation/timer';
import initACVM, { executeCircuit } from '@aztec/noir-acvm_js';
import initAbi from '@aztec/noir-noirc_abi';
import { acvm } from './acvm/acvm.js';
import { enrichNoirError } from './circuit_simulator.js';
export class WASMSimulator {
    log;
    constructor(loggerOrBindings){
        this.log = resolveLogger('wasm-simulator', loggerOrBindings);
    }
    async init() {
        // If these are available, then we are in the
        // web environment. For the node environment, this
        // is a no-op.
        if (typeof initAbi === 'function') {
            /** @ts-expect-error The node bundle doesn't include these default imports, so TS complains */ await Promise.all([
                initAbi(),
                initACVM()
            ]);
        }
    }
    async executeProtocolCircuit(input, artifact, callback) {
        this.log.debug('init', {
            hash: artifact.hash
        });
        await this.init();
        // Decode the bytecode from base64 since the acvm does not know about base64 encoding
        const decodedBytecode = Buffer.from(artifact.bytecode, 'base64');
        //
        // Execute the circuit
        try {
            const timer = new Timer();
            const result = await executeCircuit(decodedBytecode, input, callback);
            this.log.debug('execution successful', {
                hash: artifact.hash
            });
            return {
                witness: result,
                duration: timer.ms()
            };
        } catch (err) {
            // Typescript types caught errors as unknown or any, so we need to narrow its type to check if it has raw
            // assertion payload.
            if (typeof err === 'object' && err !== null && 'rawAssertionPayload' in err) {
                const parsed = enrichNoirError(artifact, err);
                this.log.debug('execution failed', {
                    hash: artifact.hash,
                    error: parsed,
                    message: parsed.message
                });
                throw parsed;
            }
            this.log.debug('execution failed', {
                hash: artifact.hash,
                error: err
            });
            throw new Error(`Circuit execution failed: ${err}`);
        }
    }
    async executeUserCircuit(input, artifact, callback) {
        await this.init();
        return acvm(artifact.bytecode, input, callback, this.log.createChild('acvm'));
    }
}
