// @aztec/bb.js/dest/node-cjs/crs/node/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GrumpkinCrs = exports.Crs = void 0;
const net_crs_js_1 = require("../net_crs.js");
const fs_1 = require("fs");
const promises_1 = require("fs/promises");
const stream_1 = require("stream");
const os_1 = require("os");
const promises_2 = require("stream/promises");
const path_1 = require("path");
/**
 * Generic CRS finder utility class.
 */
class Crs {
    numPoints;
    path;
    logger;
    constructor(numPoints, path, logger = () => { }) {
        this.numPoints = numPoints;
        this.path = path;
        this.logger = logger;
    }
    static async new(numPoints, crsPath = process.env.CRS_PATH ?? (0, path_1.join)((0, os_1.homedir)(), '.bb-crs'), logger = () => { }) {
        const crs = new Crs(numPoints, crsPath, logger);
        await crs.init();
        return crs;
    }
    async init() {
        (0, fs_1.mkdirSync)(this.path, { recursive: true });
        const g1FileSize = await (0, promises_1.stat)(this.path + '/bn254_g1.dat')
            .then(stats => stats.size)
            .catch(() => 0);
        const g2FileSize = await (0, promises_1.stat)(this.path + '/bn254_g2.dat')
            .then(stats => stats.size)
            .catch(() => 0);
        if (g1FileSize >= this.numPoints * 64 && g1FileSize % 64 == 0 && g2FileSize == 128) {
            this.logger(`Using cached CRS of size ${g1FileSize / 64}`);
            return;
        }
        this.logger(`Downloading CRS of size ${this.numPoints} into ${this.path}`);
        const crs = new net_crs_js_1.NetCrs(this.numPoints);
        const [g1, g2] = await Promise.all([crs.streamG1Data(), crs.streamG2Data()]);
        await Promise.all([
            (0, promises_2.finished)(stream_1.Readable.fromWeb(g1).pipe((0, fs_1.createWriteStream)(this.path + '/bn254_g1.dat'))),
            (0, promises_2.finished)(stream_1.Readable.fromWeb(g2).pipe((0, fs_1.createWriteStream)(this.path + '/bn254_g2.dat'))),
        ]);
    }
    /**
     * G1 points data for prover key.
     * @returns The points data.
     */
    getG1Data() {
        // Ensure length > 0, otherwise we might read a huge file.
        // This is a backup.
        const length = Math.max(this.numPoints, 1) * 64;
        const fd = (0, fs_1.openSync)(this.path + '/bn254_g1.dat', 'r');
        const buffer = new Uint8Array(length);
        (0, fs_1.readSync)(fd, buffer, 0, length, 0);
        (0, fs_1.closeSync)(fd);
        return buffer;
    }
    /**
     * G2 points data for verification key.
     * @returns The points data.
     */
    getG2Data() {
        return (0, fs_1.readFileSync)(this.path + '/bn254_g2.dat');
    }
}
exports.Crs = Crs;
/**
 * Generic Grumpkin CRS finder utility class.
 */
class GrumpkinCrs {
    numPoints;
    path;
    logger;
    constructor(numPoints, path, logger = () => { }) {
        this.numPoints = numPoints;
        this.path = path;
        this.logger = logger;
    }
    static async new(numPoints, crsPath = process.env.CRS_PATH ?? (0, path_1.join)((0, os_1.homedir)(), '.bb-crs'), logger = () => { }) {
        const crs = new GrumpkinCrs(numPoints, crsPath, logger);
        await crs.init();
        return crs;
    }
    async init() {
        (0, fs_1.mkdirSync)(this.path, { recursive: true });
        const g1FileSize = await (0, promises_1.stat)(this.path + '/grumpkin_g1.flat.dat')
            .then(stats => stats.size)
            .catch(() => 0);
        if (g1FileSize >= this.numPoints * 64 && g1FileSize % 64 == 0) {
            this.logger(`Using cached Grumpkin CRS of size ${g1FileSize / 64}`);
            return;
        }
        this.logger(`Downloading Grumpkin CRS of size ${this.numPoints} into ${this.path}`);
        const crs = new net_crs_js_1.NetGrumpkinCrs(this.numPoints);
        const stream = await crs.streamG1Data();
        await (0, promises_2.finished)(stream_1.Readable.fromWeb(stream).pipe((0, fs_1.createWriteStream)(this.path + '/grumpkin_g1.flat.dat')));
        (0, fs_1.writeFileSync)(this.path + '/grumpkin_size', String(crs.numPoints));
    }
    /**
     * G1 points data for prover key.
     * @returns The points data.
     */
    getG1Data() {
        return (0, fs_1.readFileSync)(this.path + '/grumpkin_g1.flat.dat');
    }
}
exports.GrumpkinCrs = GrumpkinCrs;
