// @aztec/stdlib/dest/contract/interfaces/contract_instance.js
import { Fr } from '@aztec/foundation/curves/bn254';
import { z } from 'zod';
import { AztecAddress } from '../../aztec-address/index.js';
import { PublicKeys } from '../../keys/public_keys.js';
import { schemas, zodFor } from '../../schemas/index.js';
const VERSION = 1;
export const ContractInstanceSchema = zodFor()(z.object({
    version: z.literal(VERSION),
    salt: schemas.Fr,
    deployer: schemas.AztecAddress,
    currentContractClassId: schemas.Fr,
    originalContractClassId: schemas.Fr,
    initializationHash: schemas.Fr,
    publicKeys: PublicKeys.schema
}));
export const ContractInstanceWithAddressSchema = zodFor()(ContractInstanceSchema.and(z.object({
    address: schemas.AztecAddress
})));
/**
 * Creates a ContractInstance from a plain object without Zod validation.
 * Suitable for deserializing trusted data (e.g., from C++ via MessagePack).
 */ export function contractInstanceFromPlainObject(obj) {
    return {
        version: 1,
        salt: Fr.fromPlainObject(obj.salt),
        deployer: AztecAddress.fromPlainObject(obj.deployer),
        currentContractClassId: Fr.fromPlainObject(obj.currentContractClassId),
        originalContractClassId: Fr.fromPlainObject(obj.originalContractClassId),
        initializationHash: Fr.fromPlainObject(obj.initializationHash),
        publicKeys: PublicKeys.fromPlainObject(obj.publicKeys)
    };
}
/**
 * Creates a ContractInstanceWithAddress from a plain object without Zod validation.
 * Suitable for deserializing trusted data (e.g., from C++ via MessagePack).
 */ export function contractInstanceWithAddressFromPlainObject(address, obj) {
    return {
        ...contractInstanceFromPlainObject(obj),
        address
    };
}
