// @aztec/foundation/dest/timer/date.js
import { createLogger } from '../log/pino-logger.js';
/** Returns current datetime. */ export class DateProvider {
    now() {
        return Date.now();
    }
    nowInSeconds() {
        return Math.floor(this.now() / 1000);
    }
    nowAsDate() {
        return new Date(this.now());
    }
}
/** Returns current datetime and allows to override it. */ export class TestDateProvider extends DateProvider {
    logger;
    offset;
    constructor(logger = createLogger('foundation:test-date-provider')){
        super(), this.logger = logger, this.offset = 0;
    }
    now() {
        return Date.now() + this.offset;
    }
    setTime(timeMs) {
        this.offset = timeMs - Date.now();
        this.logger.warn(`Time set to ${new Date(timeMs).toISOString()}`, {
            offset: this.offset,
            timeMs
        });
    }
    /** Advances the time by the given number of seconds. */ advanceTime(seconds) {
        this.offset += seconds * 1000;
    }
}
/**
 * A date provider for tests that only advances time via explicit advanceTime() calls.
 * Unlike TestDateProvider, this does NOT track real time progression - time is completely
 * frozen until explicitly advanced. This eliminates flakiness from tests taking
 * varying amounts of real time to execute.
 */ export class ManualDateProvider extends DateProvider {
    currentTimeMs;
    /**
   * @param initialTimeMs - Initial time in milliseconds. Defaults to a round timestamp for easy visualization.
   */ constructor(initialTimeMs = Date.UTC(2025, 0, 1, 0, 0, 0)){
        super();
        this.currentTimeMs = initialTimeMs;
    }
    now() {
        return this.currentTimeMs;
    }
    /** Sets the current time to the given timestamp in milliseconds. */ setTime(timeMs) {
        this.currentTimeMs = timeMs;
    }
    /** Advances the time by the given number of seconds. */ advanceTime(seconds) {
        this.currentTimeMs += seconds * 1000;
    }
    /** Advances the time by the given number of milliseconds. */ advanceTimeMs(ms) {
        this.currentTimeMs += ms;
    }
}
