// @aztec/bb.js/dest/node-cjs/barretenberg_wasm/helpers/node/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
exports.getSharedMemoryAvailable = getSharedMemoryAvailable;
exports.getRemoteBarretenbergWasm = getRemoteBarretenbergWasm;
exports.getNumCpu = getNumCpu;
exports.threadLogger = threadLogger;
exports.killSelf = killSelf;
exports.getAvailableThreads = getAvailableThreads;
const worker_threads_1 = require("worker_threads");
const os_1 = tslib_1.__importDefault(require("os"));
const comlink_1 = require("comlink");
const node_endpoint_js_1 = require("./node_endpoint.js");
const fs_1 = require("fs");
function getSharedMemoryAvailable() {
    return true;
}
/**
 * Comlink allows you to produce a Proxy to the worker, enabling you to call methods as if it were a normal class.
 * Note we give it the type information it needs so the returned Proxy object looks like that type.
 * Node has a different implementation, needing this nodeEndpoint wrapper, hence this function exists here.
 */
function getRemoteBarretenbergWasm(worker) {
    return (0, comlink_1.wrap)((0, node_endpoint_js_1.nodeEndpoint)(worker));
}
/**
 * Returns number of cpus as reported by the system, unless overriden by HARDWARE_CONCURRENCY env var.
 */
function getNumCpu() {
    return +process.env.HARDWARE_CONCURRENCY || os_1.default.cpus().length;
}
/**
 * Returns a logger function for worker threads.
 * When a custom logger is provided, posts messages back to the main thread.
 * Otherwise, writes directly to stdout.
 */
function threadLogger(useCustomLogger) {
    if (useCustomLogger) {
        return (msg) => {
            if (worker_threads_1.parentPort) {
                worker_threads_1.parentPort.postMessage({ type: 'log', msg });
            }
        };
    }
    // Write directly to stdout when no custom logger is provided
    return (msg) => {
        (0, fs_1.writeSync)(1, msg + '\n');
    };
}
function killSelf() {
    // Extordinarily hard process termination. Due to how parent threads block on child threads etc, even process.exit
    // doesn't seem to be able to abort the process. The following does.
    process.kill(process.pid);
    throw new Error();
}
function getAvailableThreads(logger) {
    try {
        return os_1.default.cpus().length;
    }
    catch (e) {
        logger(`Could not detect environment to query number of threads. Falling back to one thread. Error: ${e.message ?? e}`);
        return 1;
    }
}
