// @aztec/merkle-tree/dest/sparse_tree/sparse_tree.js
import { serializeToBuffer } from '@aztec/foundation/serialize';
import { FullTreeSnapshotBuilder } from '../snapshots/full_snapshot.js';
import { INITIAL_LEAF, TreeBase } from '../tree_base.js';
/**
 * A Merkle tree implementation that uses a LevelDB database to store the tree.
 */ export class SparseTree extends TreeBase {
    #snapshotBuilder = new FullTreeSnapshotBuilder(this.store, this, this.deserializer);
    /**
   * Updates a leaf in the tree.
   * @param leaf - New contents of the leaf.
   * @param index - Index of the leaf to be updated.
   */ updateLeaf(value, index) {
        if (index > this.maxIndex) {
            throw Error(`Index out of bounds. Index ${index}, max index: ${this.maxIndex}.`);
        }
        const leaf = serializeToBuffer(value);
        const insertingZeroElement = leaf.equals(INITIAL_LEAF);
        const originallyZeroElement = this.getLeafBuffer(index, true)?.equals(INITIAL_LEAF);
        if (insertingZeroElement && originallyZeroElement) {
            return Promise.resolve();
        }
        this.addLeafToCacheAndHashToRoot(leaf, index);
        if (insertingZeroElement) {
            // Deleting element (originally non-zero and new value is zero)
            this.cachedSize = (this.cachedSize ?? this.size) - 1n;
        } else if (originallyZeroElement) {
            // Inserting new element (originally zero and new value is non-zero)
            this.cachedSize = (this.cachedSize ?? this.size) + 1n;
        }
        return Promise.resolve();
    }
    snapshot(block) {
        return this.#snapshotBuilder.snapshot(block);
    }
    getSnapshot(block) {
        return this.#snapshotBuilder.getSnapshot(block);
    }
    findLeafIndex(_value, _includeUncommitted) {
        throw new Error('Finding leaf index is not supported for sparse trees');
    }
    findLeafIndexAfter(_value, _startIndex, _includeUncommitted) {
        throw new Error('Finding leaf index is not supported for sparse trees');
    }
}
