// @aztec/aztec.js/dest/account/account_with_secret_key.js
import { computeAddressSecret, deriveMasterIncomingViewingSecretKey } from '@aztec/stdlib/keys';
/**
 * Extends {@link BaseAccount} with the encryption private key. Not required for
 * implementing the wallet interface but useful for testing purposes or exporting
 * an account to another pxe.
 */ export class AccountWithSecretKey {
    account;
    secretKey;
    salt;
    constructor(account, secretKey, /** Deployment salt for this account contract. */ salt){
        this.account = account;
        this.secretKey = secretKey;
        this.salt = salt;
    }
    createTxExecutionRequest(exec, gasSettings, chainInfo, options) {
        return this.account.createTxExecutionRequest(exec, gasSettings, chainInfo, options);
    }
    wrapExecutionPayload(exec, options) {
        return this.account.wrapExecutionPayload(exec, options);
    }
    createAuthWit(intent, chainInfo) {
        return this.account.createAuthWit(intent, chainInfo);
    }
    getCompleteAddress() {
        return this.account.getCompleteAddress();
    }
    getAddress() {
        return this.account.getAddress();
    }
    /** Returns the encryption private key associated with this account. */ getSecretKey() {
        return this.secretKey;
    }
    /** Returns the encryption secret, the secret of the encryption point—the point that others use to encrypt messages to this account
   * note - this ensures that the address secret always corresponds to an address point with y being positive
   * dev - this is also referred to as the address secret, which decrypts payloads encrypted to an address point
   */ async getEncryptionSecret() {
        return computeAddressSecret(await this.getCompleteAddress().getPreaddress(), deriveMasterIncomingViewingSecretKey(this.getSecretKey()));
    }
}
