// @aztec/aztec.js/dest/account/account.js
import { computeAuthWitMessageHash } from '../utils/authwit.js';
// docs:end:account-interface
/**
 * An account implementation that uses authwits as an authentication mechanism
 * and can assemble transaction execution requests for an entrypoint.
 */ export class BaseAccount {
    entrypoint;
    authWitnessProvider;
    completeAddress;
    constructor(entrypoint, authWitnessProvider, completeAddress){
        this.entrypoint = entrypoint;
        this.authWitnessProvider = authWitnessProvider;
        this.completeAddress = completeAddress;
    }
    createTxExecutionRequest(exec, gasSettings, chainInfo, options) {
        return this.entrypoint.createTxExecutionRequest(exec, gasSettings, chainInfo, options);
    }
    wrapExecutionPayload(exec, options) {
        return this.entrypoint.wrapExecutionPayload(exec, options);
    }
    async createAuthWit(messageHashOrIntent, chainInfo) {
        const messageHash = await computeAuthWitMessageHash(messageHashOrIntent, chainInfo);
        return this.authWitnessProvider.createAuthWit(messageHash);
    }
    getCompleteAddress() {
        return this.completeAddress;
    }
    getAddress() {
        return this.completeAddress.address;
    }
}
