// @aztec/noir-protocol-circuits-types/dest/client_artifacts_helper.js
/* eslint-disable camelcase */ // GENERATED FILE - DO NOT EDIT. RUN `yarn generate` in the noir-protocol-circuits-types package to update.
import { abiToVKData } from './utils/vk_json.js';
export const ClientCircuitArtifactNames = {
    PrivateKernelInitArtifact: 'private_kernel_init',
    PrivateKernelInnerArtifact: 'private_kernel_inner',
    PrivateKernelTailArtifact: 'private_kernel_tail',
    PrivateKernelTailToPublicArtifact: 'private_kernel_tail_to_public',
    HidingKernelToRollup: 'hiding_kernel_to_rollup',
    HidingKernelToPublic: 'hiding_kernel_to_public',
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_4_4: 'private_kernel_reset_32_4_32_4_4_4_4_4_4',
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_4_64: 'private_kernel_reset_32_4_32_4_4_4_4_4_64',
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_64_4: 'private_kernel_reset_32_4_32_4_4_4_4_64_4',
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_64_64: 'private_kernel_reset_32_4_32_4_4_4_4_64_64',
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_4_4: 'private_kernel_reset_32_4_32_4_4_4_64_4_4',
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_4_64: 'private_kernel_reset_32_4_32_4_4_4_64_4_64',
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_64_4: 'private_kernel_reset_32_4_32_4_4_4_64_64_4',
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_64_64: 'private_kernel_reset_32_4_32_4_4_4_64_64_64',
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_4_4: 'private_kernel_reset_32_4_32_16_4_4_4_4_4',
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_4_64: 'private_kernel_reset_32_4_32_16_4_4_4_4_64',
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_64_4: 'private_kernel_reset_32_4_32_16_4_4_4_64_4',
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_64_64: 'private_kernel_reset_32_4_32_16_4_4_4_64_64',
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_4_4: 'private_kernel_reset_32_4_32_16_4_4_64_4_4',
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_4_64: 'private_kernel_reset_32_4_32_16_4_4_64_4_64',
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_64_4: 'private_kernel_reset_32_4_32_16_4_4_64_64_4',
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_64_64: 'private_kernel_reset_32_4_32_16_4_4_64_64_64',
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_4_4: 'private_kernel_reset_32_16_32_4_4_4_4_4_4',
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_4_64: 'private_kernel_reset_32_16_32_4_4_4_4_4_64',
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_64_4: 'private_kernel_reset_32_16_32_4_4_4_4_64_4',
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_64_64: 'private_kernel_reset_32_16_32_4_4_4_4_64_64',
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_4_4: 'private_kernel_reset_32_16_32_4_4_4_64_4_4',
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_4_64: 'private_kernel_reset_32_16_32_4_4_4_64_4_64',
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_64_4: 'private_kernel_reset_32_16_32_4_4_4_64_64_4',
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_64_64: 'private_kernel_reset_32_16_32_4_4_4_64_64_64',
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_4_4: 'private_kernel_reset_32_16_32_16_4_4_4_4_4',
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_4_64: 'private_kernel_reset_32_16_32_16_4_4_4_4_64',
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_64_4: 'private_kernel_reset_32_16_32_16_4_4_4_64_4',
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_64_64: 'private_kernel_reset_32_16_32_16_4_4_4_64_64',
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_4_4: 'private_kernel_reset_32_16_32_16_4_4_64_4_4',
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_4_64: 'private_kernel_reset_32_16_32_16_4_4_64_4_64',
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_64_4: 'private_kernel_reset_32_16_32_16_4_4_64_64_4',
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_64_64: 'private_kernel_reset_32_16_32_16_4_4_64_64_64',
    PrivateKernelResetArtifact_64_0_0_0_0_0_0_0_0: 'private_kernel_reset_64_0_0_0_0_0_0_0_0',
    PrivateKernelResetArtifact_0_64_0_0_0_0_0_0_0: 'private_kernel_reset_0_64_0_0_0_0_0_0_0',
    PrivateKernelResetArtifact_0_0_64_0_0_0_0_0_0: 'private_kernel_reset_0_0_64_0_0_0_0_0_0',
    PrivateKernelResetArtifact_0_0_0_64_0_0_0_0_0: 'private_kernel_reset_0_0_0_64_0_0_0_0_0',
    PrivateKernelResetArtifact_0_0_0_0_64_0_0_0_0: 'private_kernel_reset_0_0_0_0_64_0_0_0_0',
    PrivateKernelResetArtifact_0_0_0_0_0_64_0_0_0: 'private_kernel_reset_0_0_0_0_0_64_0_0_0',
    PrivateKernelResetArtifact_4_4_4_4_4_4_4_4_4: 'private_kernel_reset_4_4_4_4_4_4_4_4_4',
    PrivateKernelResetArtifact_16_16_16_16_16_16_16_16_16: 'private_kernel_reset_16_16_16_16_16_16_16_16_16',
    PrivateKernelResetArtifact_32_32_32_32_32_32_32_32_32: 'private_kernel_reset_32_32_32_32_32_32_32_32_32',
    PrivateKernelResetArtifact_64_64_64_64_64_64_64_64_64: 'private_kernel_reset'
};
export async function getClientCircuitArtifact(artifactName, simulated) {
    const isReset = artifactName.includes('private_kernel_reset');
    const normalizedArtifactName = isReset ? `${simulated ? artifactName.replace('private_kernel_reset', 'private_kernel_reset_simulated') : artifactName}` : `${artifactName}${simulated ? '_simulated' : ''}`;
    switch(normalizedArtifactName){
        case 'private_kernel_init':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_init.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_init'
                };
            }
        case 'private_kernel_init_simulated':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_init_simulated.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_init_simulated'
                };
            }
        case 'private_kernel_inner':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_inner.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_inner'
                };
            }
        case 'private_kernel_inner_simulated':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_inner_simulated.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_inner_simulated'
                };
            }
        case 'private_kernel_tail':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_tail.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_tail'
                };
            }
        case 'private_kernel_tail_simulated':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_tail_simulated.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_tail_simulated'
                };
            }
        case 'private_kernel_tail_to_public':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_tail_to_public.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_tail_to_public'
                };
            }
        case 'private_kernel_tail_to_public_simulated':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_tail_to_public_simulated.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_tail_to_public_simulated'
                };
            }
        case 'hiding_kernel_to_rollup':
            {
                const { default: compiledCircuit } = await import("../artifacts/hiding_kernel_to_rollup.json");
                return {
                    ...compiledCircuit,
                    name: 'hiding_kernel_to_rollup'
                };
            }
        case 'hiding_kernel_to_public':
            {
                const { default: compiledCircuit } = await import("../artifacts/hiding_kernel_to_public.json");
                return {
                    ...compiledCircuit,
                    name: 'hiding_kernel_to_public'
                };
            }
        case 'private_kernel_reset_32_4_32_4_4_4_4_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_4_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_4_4_4_4_4_4'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_4_4_4_4_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_4_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_4_4_4_4_4_4'
                };
            }
        case 'private_kernel_reset_32_4_32_4_4_4_4_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_4_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_4_4_4_4_4_64'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_4_4_4_4_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_4_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_4_4_4_4_4_64'
                };
            }
        case 'private_kernel_reset_32_4_32_4_4_4_4_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_4_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_4_4_4_4_64_4'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_4_4_4_4_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_4_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_4_4_4_4_64_4'
                };
            }
        case 'private_kernel_reset_32_4_32_4_4_4_4_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_4_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_4_4_4_4_64_64'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_4_4_4_4_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_4_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_4_4_4_4_64_64'
                };
            }
        case 'private_kernel_reset_32_4_32_4_4_4_64_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_64_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_4_4_4_64_4_4'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_4_4_4_64_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_64_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_4_4_4_64_4_4'
                };
            }
        case 'private_kernel_reset_32_4_32_4_4_4_64_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_64_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_4_4_4_64_4_64'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_4_4_4_64_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_64_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_4_4_4_64_4_64'
                };
            }
        case 'private_kernel_reset_32_4_32_4_4_4_64_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_64_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_4_4_4_64_64_4'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_4_4_4_64_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_64_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_4_4_4_64_64_4'
                };
            }
        case 'private_kernel_reset_32_4_32_4_4_4_64_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_64_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_4_4_4_64_64_64'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_4_4_4_64_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_64_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_4_4_4_64_64_64'
                };
            }
        case 'private_kernel_reset_32_4_32_16_4_4_4_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_4_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_16_4_4_4_4_4'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_16_4_4_4_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_4_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_16_4_4_4_4_4'
                };
            }
        case 'private_kernel_reset_32_4_32_16_4_4_4_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_4_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_16_4_4_4_4_64'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_16_4_4_4_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_4_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_16_4_4_4_4_64'
                };
            }
        case 'private_kernel_reset_32_4_32_16_4_4_4_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_4_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_16_4_4_4_64_4'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_16_4_4_4_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_4_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_16_4_4_4_64_4'
                };
            }
        case 'private_kernel_reset_32_4_32_16_4_4_4_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_4_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_16_4_4_4_64_64'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_16_4_4_4_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_4_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_16_4_4_4_64_64'
                };
            }
        case 'private_kernel_reset_32_4_32_16_4_4_64_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_64_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_16_4_4_64_4_4'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_16_4_4_64_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_64_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_16_4_4_64_4_4'
                };
            }
        case 'private_kernel_reset_32_4_32_16_4_4_64_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_64_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_16_4_4_64_4_64'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_16_4_4_64_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_64_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_16_4_4_64_4_64'
                };
            }
        case 'private_kernel_reset_32_4_32_16_4_4_64_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_64_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_16_4_4_64_64_4'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_16_4_4_64_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_64_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_16_4_4_64_64_4'
                };
            }
        case 'private_kernel_reset_32_4_32_16_4_4_64_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_64_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_4_32_16_4_4_64_64_64'
                };
            }
        case 'private_kernel_reset_simulated_32_4_32_16_4_4_64_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_64_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_4_32_16_4_4_64_64_64'
                };
            }
        case 'private_kernel_reset_32_16_32_4_4_4_4_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_4_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_4_4_4_4_4_4'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_4_4_4_4_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_4_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_4_4_4_4_4_4'
                };
            }
        case 'private_kernel_reset_32_16_32_4_4_4_4_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_4_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_4_4_4_4_4_64'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_4_4_4_4_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_4_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_4_4_4_4_4_64'
                };
            }
        case 'private_kernel_reset_32_16_32_4_4_4_4_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_4_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_4_4_4_4_64_4'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_4_4_4_4_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_4_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_4_4_4_4_64_4'
                };
            }
        case 'private_kernel_reset_32_16_32_4_4_4_4_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_4_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_4_4_4_4_64_64'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_4_4_4_4_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_4_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_4_4_4_4_64_64'
                };
            }
        case 'private_kernel_reset_32_16_32_4_4_4_64_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_64_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_4_4_4_64_4_4'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_4_4_4_64_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_64_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_4_4_4_64_4_4'
                };
            }
        case 'private_kernel_reset_32_16_32_4_4_4_64_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_64_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_4_4_4_64_4_64'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_4_4_4_64_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_64_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_4_4_4_64_4_64'
                };
            }
        case 'private_kernel_reset_32_16_32_4_4_4_64_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_64_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_4_4_4_64_64_4'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_4_4_4_64_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_64_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_4_4_4_64_64_4'
                };
            }
        case 'private_kernel_reset_32_16_32_4_4_4_64_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_64_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_4_4_4_64_64_64'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_4_4_4_64_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_64_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_4_4_4_64_64_64'
                };
            }
        case 'private_kernel_reset_32_16_32_16_4_4_4_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_4_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_16_4_4_4_4_4'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_16_4_4_4_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_4_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_16_4_4_4_4_4'
                };
            }
        case 'private_kernel_reset_32_16_32_16_4_4_4_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_4_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_16_4_4_4_4_64'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_16_4_4_4_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_4_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_16_4_4_4_4_64'
                };
            }
        case 'private_kernel_reset_32_16_32_16_4_4_4_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_4_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_16_4_4_4_64_4'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_16_4_4_4_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_4_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_16_4_4_4_64_4'
                };
            }
        case 'private_kernel_reset_32_16_32_16_4_4_4_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_4_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_16_4_4_4_64_64'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_16_4_4_4_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_4_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_16_4_4_4_64_64'
                };
            }
        case 'private_kernel_reset_32_16_32_16_4_4_64_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_64_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_16_4_4_64_4_4'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_16_4_4_64_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_64_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_16_4_4_64_4_4'
                };
            }
        case 'private_kernel_reset_32_16_32_16_4_4_64_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_64_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_16_4_4_64_4_64'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_16_4_4_64_4_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_64_4_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_16_4_4_64_4_64'
                };
            }
        case 'private_kernel_reset_32_16_32_16_4_4_64_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_64_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_16_4_4_64_64_4'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_16_4_4_64_64_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_64_64_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_16_4_4_64_64_4'
                };
            }
        case 'private_kernel_reset_32_16_32_16_4_4_64_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_64_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_16_32_16_4_4_64_64_64'
                };
            }
        case 'private_kernel_reset_simulated_32_16_32_16_4_4_64_64_64':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_64_64_64.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_16_32_16_4_4_64_64_64'
                };
            }
        case 'private_kernel_reset_64_0_0_0_0_0_0_0_0':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_64_0_0_0_0_0_0_0_0.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_64_0_0_0_0_0_0_0_0'
                };
            }
        case 'private_kernel_reset_simulated_64_0_0_0_0_0_0_0_0':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_64_0_0_0_0_0_0_0_0.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_64_0_0_0_0_0_0_0_0'
                };
            }
        case 'private_kernel_reset_0_64_0_0_0_0_0_0_0':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_0_64_0_0_0_0_0_0_0.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_0_64_0_0_0_0_0_0_0'
                };
            }
        case 'private_kernel_reset_simulated_0_64_0_0_0_0_0_0_0':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_0_64_0_0_0_0_0_0_0.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_0_64_0_0_0_0_0_0_0'
                };
            }
        case 'private_kernel_reset_0_0_64_0_0_0_0_0_0':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_0_0_64_0_0_0_0_0_0.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_0_0_64_0_0_0_0_0_0'
                };
            }
        case 'private_kernel_reset_simulated_0_0_64_0_0_0_0_0_0':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_0_0_64_0_0_0_0_0_0.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_0_0_64_0_0_0_0_0_0'
                };
            }
        case 'private_kernel_reset_0_0_0_64_0_0_0_0_0':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_0_0_0_64_0_0_0_0_0.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_0_0_0_64_0_0_0_0_0'
                };
            }
        case 'private_kernel_reset_simulated_0_0_0_64_0_0_0_0_0':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_0_0_0_64_0_0_0_0_0.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_0_0_0_64_0_0_0_0_0'
                };
            }
        case 'private_kernel_reset_0_0_0_0_64_0_0_0_0':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_0_0_0_0_64_0_0_0_0.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_0_0_0_0_64_0_0_0_0'
                };
            }
        case 'private_kernel_reset_simulated_0_0_0_0_64_0_0_0_0':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_0_0_0_0_64_0_0_0_0.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_0_0_0_0_64_0_0_0_0'
                };
            }
        case 'private_kernel_reset_0_0_0_0_0_64_0_0_0':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_0_0_0_0_0_64_0_0_0.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_0_0_0_0_0_64_0_0_0'
                };
            }
        case 'private_kernel_reset_simulated_0_0_0_0_0_64_0_0_0':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_0_0_0_0_0_64_0_0_0.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_0_0_0_0_0_64_0_0_0'
                };
            }
        case 'private_kernel_reset_4_4_4_4_4_4_4_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_4_4_4_4_4_4_4_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_4_4_4_4_4_4_4_4_4'
                };
            }
        case 'private_kernel_reset_simulated_4_4_4_4_4_4_4_4_4':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_4_4_4_4_4_4_4_4_4.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_4_4_4_4_4_4_4_4_4'
                };
            }
        case 'private_kernel_reset_16_16_16_16_16_16_16_16_16':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_16_16_16_16_16_16_16_16_16.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_16_16_16_16_16_16_16_16_16'
                };
            }
        case 'private_kernel_reset_simulated_16_16_16_16_16_16_16_16_16':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_16_16_16_16_16_16_16_16_16.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_16_16_16_16_16_16_16_16_16'
                };
            }
        case 'private_kernel_reset_32_32_32_32_32_32_32_32_32':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_32_32_32_32_32_32_32_32_32.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_32_32_32_32_32_32_32_32_32'
                };
            }
        case 'private_kernel_reset_simulated_32_32_32_32_32_32_32_32_32':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated_32_32_32_32_32_32_32_32_32.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated_32_32_32_32_32_32_32_32_32'
                };
            }
        case 'private_kernel_reset':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset'
                };
            }
        case 'private_kernel_reset_simulated':
            {
                const { default: compiledCircuit } = await import("../artifacts/private_kernel_reset_simulated.json");
                return {
                    ...compiledCircuit,
                    name: 'private_kernel_reset_simulated'
                };
            }
        default:
            throw new Error(`Unknown artifact: ${artifactName}`);
    }
}
export async function getClientCircuitVkData(artifactName) {
    switch(artifactName){
        case 'private_kernel_init':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_init.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_inner':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_inner.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_tail':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_tail.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_tail_to_public':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_tail_to_public.json");
                return abiToVKData(keyData);
            }
        case 'hiding_kernel_to_rollup':
            {
                const { default: keyData } = await import("../artifacts/hiding_kernel_to_rollup.json");
                return abiToVKData(keyData);
            }
        case 'hiding_kernel_to_public':
            {
                const { default: keyData } = await import("../artifacts/hiding_kernel_to_public.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_4_4_4_4_4_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_4_4_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_4_4_4_4_4_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_4_4_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_4_4_4_4_64_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_4_64_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_4_4_4_4_64_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_4_64_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_4_4_4_64_4_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_64_4_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_4_4_4_64_4_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_64_4_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_4_4_4_64_64_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_64_64_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_4_4_4_64_64_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_4_4_4_64_64_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_16_4_4_4_4_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_4_4_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_16_4_4_4_4_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_4_4_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_16_4_4_4_64_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_4_64_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_16_4_4_4_64_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_4_64_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_16_4_4_64_4_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_64_4_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_16_4_4_64_4_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_64_4_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_16_4_4_64_64_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_64_64_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_4_32_16_4_4_64_64_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_4_32_16_4_4_64_64_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_4_4_4_4_4_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_4_4_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_4_4_4_4_4_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_4_4_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_4_4_4_4_64_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_4_64_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_4_4_4_4_64_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_4_64_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_4_4_4_64_4_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_64_4_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_4_4_4_64_4_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_64_4_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_4_4_4_64_64_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_64_64_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_4_4_4_64_64_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_4_4_4_64_64_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_16_4_4_4_4_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_4_4_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_16_4_4_4_4_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_4_4_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_16_4_4_4_64_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_4_64_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_16_4_4_4_64_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_4_64_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_16_4_4_64_4_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_64_4_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_16_4_4_64_4_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_64_4_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_16_4_4_64_64_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_64_64_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_16_32_16_4_4_64_64_64':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_16_32_16_4_4_64_64_64.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_64_0_0_0_0_0_0_0_0':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_64_0_0_0_0_0_0_0_0.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_0_64_0_0_0_0_0_0_0':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_0_64_0_0_0_0_0_0_0.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_0_0_64_0_0_0_0_0_0':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_0_0_64_0_0_0_0_0_0.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_0_0_0_64_0_0_0_0_0':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_0_0_0_64_0_0_0_0_0.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_0_0_0_0_64_0_0_0_0':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_0_0_0_0_64_0_0_0_0.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_0_0_0_0_0_64_0_0_0':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_0_0_0_0_0_64_0_0_0.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_4_4_4_4_4_4_4_4_4':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_4_4_4_4_4_4_4_4_4.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_16_16_16_16_16_16_16_16_16':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_16_16_16_16_16_16_16_16_16.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset_32_32_32_32_32_32_32_32_32':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset_32_32_32_32_32_32_32_32_32.json");
                return abiToVKData(keyData);
            }
        case 'private_kernel_reset':
            {
                const { default: keyData } = await import("../artifacts/private_kernel_reset.json");
                return abiToVKData(keyData);
            }
        default:
            throw new Error(`Unknown artifact: ${artifactName}`);
    }
}
