// @aztec/noir-protocol-circuits-types/dest/private_kernel_reset_vks.js
/* eslint-disable camelcase */ // GENERATED FILE - DO NOT EDIT. RUN `yarn generate` or `yarn generate:reset-data`
import { abiToVKData } from './utils/vk_json.js';
import PrivateKernelResetJson_32_4_32_4_4_4_4_4_4 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_4_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_4_4_4_4_4_64 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_4_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_4_4_4_4_64_4 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_4_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_4_4_4_4_64_64 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_4_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_4_4_4_64_4_4 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_64_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_4_4_4_64_4_64 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_64_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_4_4_4_64_64_4 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_64_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_4_4_4_64_64_64 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_64_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_4_4_4 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_4_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_4_4_64 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_4_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_4_64_4 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_4_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_4_64_64 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_4_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_64_4_4 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_64_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_64_4_64 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_64_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_64_64_4 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_64_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_64_64_64 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_64_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_4_4_4 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_4_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_4_4_64 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_4_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_4_64_4 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_4_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_4_64_64 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_4_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_64_4_4 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_64_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_64_4_64 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_64_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_64_64_4 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_64_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_64_64_64 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_64_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_4_4_4 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_4_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_4_4_64 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_4_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_4_64_4 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_4_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_4_64_64 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_4_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_64_4_4 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_64_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_64_4_64 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_64_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_64_64_4 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_64_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_64_64_64 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_64_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_64_0_0_0_0_0_0_0_0 from '../artifacts/private_kernel_reset_64_0_0_0_0_0_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetJson_0_64_0_0_0_0_0_0_0 from '../artifacts/private_kernel_reset_0_64_0_0_0_0_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetJson_0_0_64_0_0_0_0_0_0 from '../artifacts/private_kernel_reset_0_0_64_0_0_0_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetJson_0_0_0_64_0_0_0_0_0 from '../artifacts/private_kernel_reset_0_0_0_64_0_0_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetJson_0_0_0_0_64_0_0_0_0 from '../artifacts/private_kernel_reset_0_0_0_0_64_0_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetJson_0_0_0_0_0_64_0_0_0 from '../artifacts/private_kernel_reset_0_0_0_0_0_64_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetJson_4_4_4_4_4_4_4_4_4 from '../artifacts/private_kernel_reset_4_4_4_4_4_4_4_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_16_16_16_16_16_16_16_16_16 from '../artifacts/private_kernel_reset_16_16_16_16_16_16_16_16_16.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_32_32_32_32_32_32_32_32 from '../artifacts/private_kernel_reset_32_32_32_32_32_32_32_32_32.json' with {
    type: 'json'
};
import PrivateKernelResetJson from '../artifacts/private_kernel_reset.json' with {
    type: 'json'
};
export const PrivateKernelResetVks = {
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_4_4: abiToVKData(PrivateKernelResetJson_32_4_32_4_4_4_4_4_4),
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_4_64: abiToVKData(PrivateKernelResetJson_32_4_32_4_4_4_4_4_64),
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_64_4: abiToVKData(PrivateKernelResetJson_32_4_32_4_4_4_4_64_4),
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_64_64: abiToVKData(PrivateKernelResetJson_32_4_32_4_4_4_4_64_64),
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_4_4: abiToVKData(PrivateKernelResetJson_32_4_32_4_4_4_64_4_4),
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_4_64: abiToVKData(PrivateKernelResetJson_32_4_32_4_4_4_64_4_64),
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_64_4: abiToVKData(PrivateKernelResetJson_32_4_32_4_4_4_64_64_4),
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_64_64: abiToVKData(PrivateKernelResetJson_32_4_32_4_4_4_64_64_64),
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_4_4: abiToVKData(PrivateKernelResetJson_32_4_32_16_4_4_4_4_4),
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_4_64: abiToVKData(PrivateKernelResetJson_32_4_32_16_4_4_4_4_64),
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_64_4: abiToVKData(PrivateKernelResetJson_32_4_32_16_4_4_4_64_4),
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_64_64: abiToVKData(PrivateKernelResetJson_32_4_32_16_4_4_4_64_64),
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_4_4: abiToVKData(PrivateKernelResetJson_32_4_32_16_4_4_64_4_4),
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_4_64: abiToVKData(PrivateKernelResetJson_32_4_32_16_4_4_64_4_64),
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_64_4: abiToVKData(PrivateKernelResetJson_32_4_32_16_4_4_64_64_4),
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_64_64: abiToVKData(PrivateKernelResetJson_32_4_32_16_4_4_64_64_64),
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_4_4: abiToVKData(PrivateKernelResetJson_32_16_32_4_4_4_4_4_4),
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_4_64: abiToVKData(PrivateKernelResetJson_32_16_32_4_4_4_4_4_64),
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_64_4: abiToVKData(PrivateKernelResetJson_32_16_32_4_4_4_4_64_4),
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_64_64: abiToVKData(PrivateKernelResetJson_32_16_32_4_4_4_4_64_64),
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_4_4: abiToVKData(PrivateKernelResetJson_32_16_32_4_4_4_64_4_4),
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_4_64: abiToVKData(PrivateKernelResetJson_32_16_32_4_4_4_64_4_64),
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_64_4: abiToVKData(PrivateKernelResetJson_32_16_32_4_4_4_64_64_4),
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_64_64: abiToVKData(PrivateKernelResetJson_32_16_32_4_4_4_64_64_64),
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_4_4: abiToVKData(PrivateKernelResetJson_32_16_32_16_4_4_4_4_4),
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_4_64: abiToVKData(PrivateKernelResetJson_32_16_32_16_4_4_4_4_64),
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_64_4: abiToVKData(PrivateKernelResetJson_32_16_32_16_4_4_4_64_4),
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_64_64: abiToVKData(PrivateKernelResetJson_32_16_32_16_4_4_4_64_64),
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_4_4: abiToVKData(PrivateKernelResetJson_32_16_32_16_4_4_64_4_4),
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_4_64: abiToVKData(PrivateKernelResetJson_32_16_32_16_4_4_64_4_64),
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_64_4: abiToVKData(PrivateKernelResetJson_32_16_32_16_4_4_64_64_4),
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_64_64: abiToVKData(PrivateKernelResetJson_32_16_32_16_4_4_64_64_64),
    PrivateKernelResetArtifact_64_0_0_0_0_0_0_0_0: abiToVKData(PrivateKernelResetJson_64_0_0_0_0_0_0_0_0),
    PrivateKernelResetArtifact_0_64_0_0_0_0_0_0_0: abiToVKData(PrivateKernelResetJson_0_64_0_0_0_0_0_0_0),
    PrivateKernelResetArtifact_0_0_64_0_0_0_0_0_0: abiToVKData(PrivateKernelResetJson_0_0_64_0_0_0_0_0_0),
    PrivateKernelResetArtifact_0_0_0_64_0_0_0_0_0: abiToVKData(PrivateKernelResetJson_0_0_0_64_0_0_0_0_0),
    PrivateKernelResetArtifact_0_0_0_0_64_0_0_0_0: abiToVKData(PrivateKernelResetJson_0_0_0_0_64_0_0_0_0),
    PrivateKernelResetArtifact_0_0_0_0_0_64_0_0_0: abiToVKData(PrivateKernelResetJson_0_0_0_0_0_64_0_0_0),
    PrivateKernelResetArtifact_4_4_4_4_4_4_4_4_4: abiToVKData(PrivateKernelResetJson_4_4_4_4_4_4_4_4_4),
    PrivateKernelResetArtifact_16_16_16_16_16_16_16_16_16: abiToVKData(PrivateKernelResetJson_16_16_16_16_16_16_16_16_16),
    PrivateKernelResetArtifact_32_32_32_32_32_32_32_32_32: abiToVKData(PrivateKernelResetJson_32_32_32_32_32_32_32_32_32),
    PrivateKernelResetArtifact_64_64_64_64_64_64_64_64_64: abiToVKData(PrivateKernelResetJson)
};
export const PrivateKernelResetVkIndexes = {
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_4_4: 23,
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_4_64: 24,
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_64_4: 25,
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_64_64: 26,
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_4_4: 27,
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_4_64: 28,
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_64_4: 29,
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_64_64: 30,
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_4_4: 31,
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_4_64: 32,
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_64_4: 33,
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_64_64: 34,
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_4_4: 35,
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_4_64: 36,
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_64_4: 37,
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_64_64: 38,
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_4_4: 39,
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_4_64: 40,
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_64_4: 41,
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_64_64: 42,
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_4_4: 43,
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_4_64: 44,
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_64_4: 45,
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_64_64: 46,
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_4_4: 47,
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_4_64: 48,
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_64_4: 49,
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_64_64: 50,
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_4_4: 51,
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_4_64: 52,
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_64_4: 53,
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_64_64: 54,
    PrivateKernelResetArtifact_64_0_0_0_0_0_0_0_0: 55,
    PrivateKernelResetArtifact_0_64_0_0_0_0_0_0_0: 56,
    PrivateKernelResetArtifact_0_0_64_0_0_0_0_0_0: 57,
    PrivateKernelResetArtifact_0_0_0_64_0_0_0_0_0: 58,
    PrivateKernelResetArtifact_0_0_0_0_64_0_0_0_0: 59,
    PrivateKernelResetArtifact_0_0_0_0_0_64_0_0_0: 60,
    PrivateKernelResetArtifact_4_4_4_4_4_4_4_4_4: 61,
    PrivateKernelResetArtifact_16_16_16_16_16_16_16_16_16: 62,
    PrivateKernelResetArtifact_32_32_32_32_32_32_32_32_32: 63,
    PrivateKernelResetArtifact_64_64_64_64_64_64_64_64_64: 64
};
