// @aztec/noir-protocol-circuits-types/dest/private_kernel_reset_data.js
/* eslint-disable camelcase */ // GENERATED FILE - DO NOT EDIT. RUN `yarn generate` or `yarn generate:reset-data`
import PrivateKernelResetJson_32_4_32_4_4_4_4_4_4 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_4_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_4_4_4_4_4_64 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_4_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_4_4_4_4_64_4 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_4_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_4_4_4_4_64_64 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_4_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_4_4_4_64_4_4 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_64_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_4_4_4_64_4_64 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_64_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_4_4_4_64_64_4 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_64_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_4_4_4_64_64_64 from '../artifacts/private_kernel_reset_32_4_32_4_4_4_64_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_4_4_4 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_4_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_4_4_64 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_4_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_4_64_4 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_4_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_4_64_64 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_4_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_64_4_4 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_64_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_64_4_64 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_64_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_64_64_4 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_64_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_4_32_16_4_4_64_64_64 from '../artifacts/private_kernel_reset_32_4_32_16_4_4_64_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_4_4_4 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_4_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_4_4_64 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_4_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_4_64_4 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_4_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_4_64_64 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_4_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_64_4_4 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_64_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_64_4_64 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_64_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_64_64_4 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_64_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_4_4_4_64_64_64 from '../artifacts/private_kernel_reset_32_16_32_4_4_4_64_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_4_4_4 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_4_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_4_4_64 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_4_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_4_64_4 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_4_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_4_64_64 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_4_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_64_4_4 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_64_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_64_4_64 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_64_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_64_64_4 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_64_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_16_32_16_4_4_64_64_64 from '../artifacts/private_kernel_reset_32_16_32_16_4_4_64_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetJson_64_0_0_0_0_0_0_0_0 from '../artifacts/private_kernel_reset_64_0_0_0_0_0_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetJson_0_64_0_0_0_0_0_0_0 from '../artifacts/private_kernel_reset_0_64_0_0_0_0_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetJson_0_0_64_0_0_0_0_0_0 from '../artifacts/private_kernel_reset_0_0_64_0_0_0_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetJson_0_0_0_64_0_0_0_0_0 from '../artifacts/private_kernel_reset_0_0_0_64_0_0_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetJson_0_0_0_0_64_0_0_0_0 from '../artifacts/private_kernel_reset_0_0_0_0_64_0_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetJson_0_0_0_0_0_64_0_0_0 from '../artifacts/private_kernel_reset_0_0_0_0_0_64_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetJson_4_4_4_4_4_4_4_4_4 from '../artifacts/private_kernel_reset_4_4_4_4_4_4_4_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetJson_16_16_16_16_16_16_16_16_16 from '../artifacts/private_kernel_reset_16_16_16_16_16_16_16_16_16.json' with {
    type: 'json'
};
import PrivateKernelResetJson_32_32_32_32_32_32_32_32_32 from '../artifacts/private_kernel_reset_32_32_32_32_32_32_32_32_32.json' with {
    type: 'json'
};
import PrivateKernelResetJson from '../artifacts/private_kernel_reset.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_4_4_4_4_4_4 from '../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_4_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_4_4_4_4_4_64 from '../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_4_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_4_4_4_4_64_4 from '../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_4_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_4_4_4_4_64_64 from '../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_4_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_4_4_4_64_4_4 from '../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_64_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_4_4_4_64_4_64 from '../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_64_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_4_4_4_64_64_4 from '../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_64_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_4_4_4_64_64_64 from '../artifacts/private_kernel_reset_simulated_32_4_32_4_4_4_64_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_16_4_4_4_4_4 from '../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_4_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_16_4_4_4_4_64 from '../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_4_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_16_4_4_4_64_4 from '../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_4_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_16_4_4_4_64_64 from '../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_4_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_16_4_4_64_4_4 from '../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_64_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_16_4_4_64_4_64 from '../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_64_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_16_4_4_64_64_4 from '../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_64_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_4_32_16_4_4_64_64_64 from '../artifacts/private_kernel_reset_simulated_32_4_32_16_4_4_64_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_4_4_4_4_4_4 from '../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_4_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_4_4_4_4_4_64 from '../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_4_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_4_4_4_4_64_4 from '../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_4_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_4_4_4_4_64_64 from '../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_4_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_4_4_4_64_4_4 from '../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_64_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_4_4_4_64_4_64 from '../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_64_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_4_4_4_64_64_4 from '../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_64_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_4_4_4_64_64_64 from '../artifacts/private_kernel_reset_simulated_32_16_32_4_4_4_64_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_16_4_4_4_4_4 from '../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_4_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_16_4_4_4_4_64 from '../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_4_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_16_4_4_4_64_4 from '../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_4_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_16_4_4_4_64_64 from '../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_4_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_16_4_4_64_4_4 from '../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_64_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_16_4_4_64_4_64 from '../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_64_4_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_16_4_4_64_64_4 from '../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_64_64_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_16_32_16_4_4_64_64_64 from '../artifacts/private_kernel_reset_simulated_32_16_32_16_4_4_64_64_64.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_64_0_0_0_0_0_0_0_0 from '../artifacts/private_kernel_reset_simulated_64_0_0_0_0_0_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_0_64_0_0_0_0_0_0_0 from '../artifacts/private_kernel_reset_simulated_0_64_0_0_0_0_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_0_0_64_0_0_0_0_0_0 from '../artifacts/private_kernel_reset_simulated_0_0_64_0_0_0_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_0_0_0_64_0_0_0_0_0 from '../artifacts/private_kernel_reset_simulated_0_0_0_64_0_0_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_0_0_0_0_64_0_0_0_0 from '../artifacts/private_kernel_reset_simulated_0_0_0_0_64_0_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_0_0_0_0_0_64_0_0_0 from '../artifacts/private_kernel_reset_simulated_0_0_0_0_0_64_0_0_0.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_4_4_4_4_4_4_4_4_4 from '../artifacts/private_kernel_reset_simulated_4_4_4_4_4_4_4_4_4.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_16_16_16_16_16_16_16_16_16 from '../artifacts/private_kernel_reset_simulated_16_16_16_16_16_16_16_16_16.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson_32_32_32_32_32_32_32_32_32 from '../artifacts/private_kernel_reset_simulated_32_32_32_32_32_32_32_32_32.json' with {
    type: 'json'
};
import PrivateKernelResetSimulatedJson from '../artifacts/private_kernel_reset_simulated.json' with {
    type: 'json'
};
export const PrivateKernelResetArtifacts = {
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_4_4: PrivateKernelResetJson_32_4_32_4_4_4_4_4_4,
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_4_64: PrivateKernelResetJson_32_4_32_4_4_4_4_4_64,
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_64_4: PrivateKernelResetJson_32_4_32_4_4_4_4_64_4,
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_64_64: PrivateKernelResetJson_32_4_32_4_4_4_4_64_64,
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_4_4: PrivateKernelResetJson_32_4_32_4_4_4_64_4_4,
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_4_64: PrivateKernelResetJson_32_4_32_4_4_4_64_4_64,
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_64_4: PrivateKernelResetJson_32_4_32_4_4_4_64_64_4,
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_64_64: PrivateKernelResetJson_32_4_32_4_4_4_64_64_64,
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_4_4: PrivateKernelResetJson_32_4_32_16_4_4_4_4_4,
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_4_64: PrivateKernelResetJson_32_4_32_16_4_4_4_4_64,
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_64_4: PrivateKernelResetJson_32_4_32_16_4_4_4_64_4,
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_64_64: PrivateKernelResetJson_32_4_32_16_4_4_4_64_64,
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_4_4: PrivateKernelResetJson_32_4_32_16_4_4_64_4_4,
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_4_64: PrivateKernelResetJson_32_4_32_16_4_4_64_4_64,
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_64_4: PrivateKernelResetJson_32_4_32_16_4_4_64_64_4,
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_64_64: PrivateKernelResetJson_32_4_32_16_4_4_64_64_64,
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_4_4: PrivateKernelResetJson_32_16_32_4_4_4_4_4_4,
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_4_64: PrivateKernelResetJson_32_16_32_4_4_4_4_4_64,
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_64_4: PrivateKernelResetJson_32_16_32_4_4_4_4_64_4,
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_64_64: PrivateKernelResetJson_32_16_32_4_4_4_4_64_64,
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_4_4: PrivateKernelResetJson_32_16_32_4_4_4_64_4_4,
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_4_64: PrivateKernelResetJson_32_16_32_4_4_4_64_4_64,
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_64_4: PrivateKernelResetJson_32_16_32_4_4_4_64_64_4,
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_64_64: PrivateKernelResetJson_32_16_32_4_4_4_64_64_64,
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_4_4: PrivateKernelResetJson_32_16_32_16_4_4_4_4_4,
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_4_64: PrivateKernelResetJson_32_16_32_16_4_4_4_4_64,
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_64_4: PrivateKernelResetJson_32_16_32_16_4_4_4_64_4,
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_64_64: PrivateKernelResetJson_32_16_32_16_4_4_4_64_64,
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_4_4: PrivateKernelResetJson_32_16_32_16_4_4_64_4_4,
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_4_64: PrivateKernelResetJson_32_16_32_16_4_4_64_4_64,
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_64_4: PrivateKernelResetJson_32_16_32_16_4_4_64_64_4,
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_64_64: PrivateKernelResetJson_32_16_32_16_4_4_64_64_64,
    PrivateKernelResetArtifact_64_0_0_0_0_0_0_0_0: PrivateKernelResetJson_64_0_0_0_0_0_0_0_0,
    PrivateKernelResetArtifact_0_64_0_0_0_0_0_0_0: PrivateKernelResetJson_0_64_0_0_0_0_0_0_0,
    PrivateKernelResetArtifact_0_0_64_0_0_0_0_0_0: PrivateKernelResetJson_0_0_64_0_0_0_0_0_0,
    PrivateKernelResetArtifact_0_0_0_64_0_0_0_0_0: PrivateKernelResetJson_0_0_0_64_0_0_0_0_0,
    PrivateKernelResetArtifact_0_0_0_0_64_0_0_0_0: PrivateKernelResetJson_0_0_0_0_64_0_0_0_0,
    PrivateKernelResetArtifact_0_0_0_0_0_64_0_0_0: PrivateKernelResetJson_0_0_0_0_0_64_0_0_0,
    PrivateKernelResetArtifact_4_4_4_4_4_4_4_4_4: PrivateKernelResetJson_4_4_4_4_4_4_4_4_4,
    PrivateKernelResetArtifact_16_16_16_16_16_16_16_16_16: PrivateKernelResetJson_16_16_16_16_16_16_16_16_16,
    PrivateKernelResetArtifact_32_32_32_32_32_32_32_32_32: PrivateKernelResetJson_32_32_32_32_32_32_32_32_32,
    PrivateKernelResetArtifact_64_64_64_64_64_64_64_64_64: PrivateKernelResetJson
};
export const PrivateKernelResetSimulatedArtifacts = {
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_4_4: PrivateKernelResetSimulatedJson_32_4_32_4_4_4_4_4_4,
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_4_64: PrivateKernelResetSimulatedJson_32_4_32_4_4_4_4_4_64,
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_64_4: PrivateKernelResetSimulatedJson_32_4_32_4_4_4_4_64_4,
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_64_64: PrivateKernelResetSimulatedJson_32_4_32_4_4_4_4_64_64,
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_4_4: PrivateKernelResetSimulatedJson_32_4_32_4_4_4_64_4_4,
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_4_64: PrivateKernelResetSimulatedJson_32_4_32_4_4_4_64_4_64,
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_64_4: PrivateKernelResetSimulatedJson_32_4_32_4_4_4_64_64_4,
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_64_64: PrivateKernelResetSimulatedJson_32_4_32_4_4_4_64_64_64,
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_4_4: PrivateKernelResetSimulatedJson_32_4_32_16_4_4_4_4_4,
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_4_64: PrivateKernelResetSimulatedJson_32_4_32_16_4_4_4_4_64,
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_64_4: PrivateKernelResetSimulatedJson_32_4_32_16_4_4_4_64_4,
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_64_64: PrivateKernelResetSimulatedJson_32_4_32_16_4_4_4_64_64,
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_4_4: PrivateKernelResetSimulatedJson_32_4_32_16_4_4_64_4_4,
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_4_64: PrivateKernelResetSimulatedJson_32_4_32_16_4_4_64_4_64,
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_64_4: PrivateKernelResetSimulatedJson_32_4_32_16_4_4_64_64_4,
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_64_64: PrivateKernelResetSimulatedJson_32_4_32_16_4_4_64_64_64,
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_4_4: PrivateKernelResetSimulatedJson_32_16_32_4_4_4_4_4_4,
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_4_64: PrivateKernelResetSimulatedJson_32_16_32_4_4_4_4_4_64,
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_64_4: PrivateKernelResetSimulatedJson_32_16_32_4_4_4_4_64_4,
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_64_64: PrivateKernelResetSimulatedJson_32_16_32_4_4_4_4_64_64,
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_4_4: PrivateKernelResetSimulatedJson_32_16_32_4_4_4_64_4_4,
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_4_64: PrivateKernelResetSimulatedJson_32_16_32_4_4_4_64_4_64,
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_64_4: PrivateKernelResetSimulatedJson_32_16_32_4_4_4_64_64_4,
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_64_64: PrivateKernelResetSimulatedJson_32_16_32_4_4_4_64_64_64,
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_4_4: PrivateKernelResetSimulatedJson_32_16_32_16_4_4_4_4_4,
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_4_64: PrivateKernelResetSimulatedJson_32_16_32_16_4_4_4_4_64,
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_64_4: PrivateKernelResetSimulatedJson_32_16_32_16_4_4_4_64_4,
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_64_64: PrivateKernelResetSimulatedJson_32_16_32_16_4_4_4_64_64,
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_4_4: PrivateKernelResetSimulatedJson_32_16_32_16_4_4_64_4_4,
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_4_64: PrivateKernelResetSimulatedJson_32_16_32_16_4_4_64_4_64,
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_64_4: PrivateKernelResetSimulatedJson_32_16_32_16_4_4_64_64_4,
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_64_64: PrivateKernelResetSimulatedJson_32_16_32_16_4_4_64_64_64,
    PrivateKernelResetArtifact_64_0_0_0_0_0_0_0_0: PrivateKernelResetSimulatedJson_64_0_0_0_0_0_0_0_0,
    PrivateKernelResetArtifact_0_64_0_0_0_0_0_0_0: PrivateKernelResetSimulatedJson_0_64_0_0_0_0_0_0_0,
    PrivateKernelResetArtifact_0_0_64_0_0_0_0_0_0: PrivateKernelResetSimulatedJson_0_0_64_0_0_0_0_0_0,
    PrivateKernelResetArtifact_0_0_0_64_0_0_0_0_0: PrivateKernelResetSimulatedJson_0_0_0_64_0_0_0_0_0,
    PrivateKernelResetArtifact_0_0_0_0_64_0_0_0_0: PrivateKernelResetSimulatedJson_0_0_0_0_64_0_0_0_0,
    PrivateKernelResetArtifact_0_0_0_0_0_64_0_0_0: PrivateKernelResetSimulatedJson_0_0_0_0_0_64_0_0_0,
    PrivateKernelResetArtifact_4_4_4_4_4_4_4_4_4: PrivateKernelResetSimulatedJson_4_4_4_4_4_4_4_4_4,
    PrivateKernelResetArtifact_16_16_16_16_16_16_16_16_16: PrivateKernelResetSimulatedJson_16_16_16_16_16_16_16_16_16,
    PrivateKernelResetArtifact_32_32_32_32_32_32_32_32_32: PrivateKernelResetSimulatedJson_32_32_32_32_32_32_32_32_32,
    PrivateKernelResetArtifact_64_64_64_64_64_64_64_64_64: PrivateKernelResetSimulatedJson
};
