// @aztec/noir-protocol-circuits-types/dest/private_kernel_reset_types.js
/* eslint-disable camelcase */ // GENERATED FILE - DO NOT EDIT. RUN `yarn generate` or `yarn generate:reset-data`
import { PrivateKernelResetDimensions } from '@aztec/stdlib/kernel';
export const PrivateKernelResetArtifactFileNames = {
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_4_4: 'private_kernel_reset_32_4_32_4_4_4_4_4_4',
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_4_64: 'private_kernel_reset_32_4_32_4_4_4_4_4_64',
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_64_4: 'private_kernel_reset_32_4_32_4_4_4_4_64_4',
    PrivateKernelResetArtifact_32_4_32_4_4_4_4_64_64: 'private_kernel_reset_32_4_32_4_4_4_4_64_64',
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_4_4: 'private_kernel_reset_32_4_32_4_4_4_64_4_4',
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_4_64: 'private_kernel_reset_32_4_32_4_4_4_64_4_64',
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_64_4: 'private_kernel_reset_32_4_32_4_4_4_64_64_4',
    PrivateKernelResetArtifact_32_4_32_4_4_4_64_64_64: 'private_kernel_reset_32_4_32_4_4_4_64_64_64',
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_4_4: 'private_kernel_reset_32_4_32_16_4_4_4_4_4',
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_4_64: 'private_kernel_reset_32_4_32_16_4_4_4_4_64',
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_64_4: 'private_kernel_reset_32_4_32_16_4_4_4_64_4',
    PrivateKernelResetArtifact_32_4_32_16_4_4_4_64_64: 'private_kernel_reset_32_4_32_16_4_4_4_64_64',
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_4_4: 'private_kernel_reset_32_4_32_16_4_4_64_4_4',
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_4_64: 'private_kernel_reset_32_4_32_16_4_4_64_4_64',
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_64_4: 'private_kernel_reset_32_4_32_16_4_4_64_64_4',
    PrivateKernelResetArtifact_32_4_32_16_4_4_64_64_64: 'private_kernel_reset_32_4_32_16_4_4_64_64_64',
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_4_4: 'private_kernel_reset_32_16_32_4_4_4_4_4_4',
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_4_64: 'private_kernel_reset_32_16_32_4_4_4_4_4_64',
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_64_4: 'private_kernel_reset_32_16_32_4_4_4_4_64_4',
    PrivateKernelResetArtifact_32_16_32_4_4_4_4_64_64: 'private_kernel_reset_32_16_32_4_4_4_4_64_64',
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_4_4: 'private_kernel_reset_32_16_32_4_4_4_64_4_4',
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_4_64: 'private_kernel_reset_32_16_32_4_4_4_64_4_64',
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_64_4: 'private_kernel_reset_32_16_32_4_4_4_64_64_4',
    PrivateKernelResetArtifact_32_16_32_4_4_4_64_64_64: 'private_kernel_reset_32_16_32_4_4_4_64_64_64',
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_4_4: 'private_kernel_reset_32_16_32_16_4_4_4_4_4',
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_4_64: 'private_kernel_reset_32_16_32_16_4_4_4_4_64',
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_64_4: 'private_kernel_reset_32_16_32_16_4_4_4_64_4',
    PrivateKernelResetArtifact_32_16_32_16_4_4_4_64_64: 'private_kernel_reset_32_16_32_16_4_4_4_64_64',
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_4_4: 'private_kernel_reset_32_16_32_16_4_4_64_4_4',
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_4_64: 'private_kernel_reset_32_16_32_16_4_4_64_4_64',
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_64_4: 'private_kernel_reset_32_16_32_16_4_4_64_64_4',
    PrivateKernelResetArtifact_32_16_32_16_4_4_64_64_64: 'private_kernel_reset_32_16_32_16_4_4_64_64_64',
    PrivateKernelResetArtifact_64_0_0_0_0_0_0_0_0: 'private_kernel_reset_64_0_0_0_0_0_0_0_0',
    PrivateKernelResetArtifact_0_64_0_0_0_0_0_0_0: 'private_kernel_reset_0_64_0_0_0_0_0_0_0',
    PrivateKernelResetArtifact_0_0_64_0_0_0_0_0_0: 'private_kernel_reset_0_0_64_0_0_0_0_0_0',
    PrivateKernelResetArtifact_0_0_0_64_0_0_0_0_0: 'private_kernel_reset_0_0_0_64_0_0_0_0_0',
    PrivateKernelResetArtifact_0_0_0_0_64_0_0_0_0: 'private_kernel_reset_0_0_0_0_64_0_0_0_0',
    PrivateKernelResetArtifact_0_0_0_0_0_64_0_0_0: 'private_kernel_reset_0_0_0_0_0_64_0_0_0',
    PrivateKernelResetArtifact_4_4_4_4_4_4_4_4_4: 'private_kernel_reset_4_4_4_4_4_4_4_4_4',
    PrivateKernelResetArtifact_16_16_16_16_16_16_16_16_16: 'private_kernel_reset_16_16_16_16_16_16_16_16_16',
    PrivateKernelResetArtifact_32_32_32_32_32_32_32_32_32: 'private_kernel_reset_32_32_32_32_32_32_32_32_32',
    PrivateKernelResetArtifact_64_64_64_64_64_64_64_64_64: 'private_kernel_reset'
};
export const privateKernelResetDimensionsConfig = {
    "dimensions": {
        "NOTE_HASH_PENDING_READ": {
            "variants": [
                32
            ],
            "standalone": [
                64
            ],
            "cost": 100
        },
        "NOTE_HASH_SETTLED_READ": {
            "variants": [
                4,
                16
            ],
            "standalone": [
                64
            ],
            "cost": 3000
        },
        "NULLIFIER_PENDING_READ": {
            "variants": [
                32
            ],
            "standalone": [
                64
            ],
            "cost": 100
        },
        "NULLIFIER_SETTLED_READ": {
            "variants": [
                4,
                16
            ],
            "standalone": [
                64
            ],
            "cost": 3000
        },
        "KEY_VALIDATION": {
            "variants": [
                4
            ],
            "standalone": [
                64
            ],
            "cost": 2500
        },
        "TRANSIENT_DATA_SQUASHING": {
            "variants": [
                4
            ],
            "standalone": [
                64
            ],
            "cost": 100
        },
        "NOTE_HASH_SILOING": {
            "variants": [
                4,
                64
            ],
            "standalone": [],
            "cost": 250
        },
        "NULLIFIER_SILOING": {
            "variants": [
                4,
                64
            ],
            "standalone": [],
            "cost": 150
        },
        "PRIVATE_LOG_SILOING": {
            "variants": [
                4,
                64
            ],
            "standalone": [],
            "cost": 150
        }
    },
    "specialCases": [
        [
            4,
            4,
            4,
            4,
            4,
            4,
            4,
            4,
            4
        ],
        [
            16,
            16,
            16,
            16,
            16,
            16,
            16,
            16,
            16
        ],
        [
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32,
            32
        ],
        [
            64,
            64,
            64,
            64,
            64,
            64,
            64,
            64,
            64
        ]
    ]
};
export const maxPrivateKernelResetDimensions = PrivateKernelResetDimensions.fromValues([
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64,
    64
]);
