// @aztec/stdlib/dest/tx/public_simulation_output.js
import { Fr } from '@aztec/foundation/curves/bn254';
import { schemas } from '@aztec/foundation/schemas';
import times from 'lodash.times';
import { z } from 'zod';
import { SimulationError } from '../errors/simulation_error.js';
import { Gas } from '../gas/gas.js';
import { NullishToUndefined } from '../schemas/schemas.js';
import { TxEffect } from '../tx/tx_effect.js';
import { GlobalVariables } from './global_variables.js';
/** Return values of simulating complete callstack. */ export class NestedProcessReturnValues {
    values;
    nested;
    constructor(values, nested){
        this.values = values;
        this.nested = nested ?? [];
    }
    equals(other) {
        return this.values?.length === other.values?.length && this.nested.length === other.nested.length && (this.values === undefined || this.values.every((v, i)=>v.equals(other.values[i]))) && this.nested.every((n, i)=>n.equals(other.nested[i]));
    }
    static get schema() {
        return z.object({
            values: NullishToUndefined(z.array(schemas.Fr)),
            nested: z.array(z.lazy(()=>NestedProcessReturnValues.schema))
        }).transform(({ values, nested })=>new NestedProcessReturnValues(values, nested));
    }
    static fromPlainObject(obj) {
        return new NestedProcessReturnValues(obj.values?.map(Fr.fromPlainObject), obj.nested?.map(NestedProcessReturnValues.fromPlainObject));
    }
    static empty() {
        return new NestedProcessReturnValues([]);
    }
    static random(depth = 1) {
        return new NestedProcessReturnValues(times(3, Fr.random), depth > 0 ? [
            NestedProcessReturnValues.random(depth - 1)
        ] : []);
    }
}
/**
 * Outputs of processing the public component of a transaction.
 */ export class PublicSimulationOutput {
    revertReason;
    globalVariables;
    txEffect;
    publicReturnValues;
    gasUsed;
    constructor(revertReason, globalVariables, txEffect, publicReturnValues, gasUsed){
        this.revertReason = revertReason;
        this.globalVariables = globalVariables;
        this.txEffect = txEffect;
        this.publicReturnValues = publicReturnValues;
        this.gasUsed = gasUsed;
    }
    static get schema() {
        return z.object({
            revertReason: SimulationError.schema.optional(),
            globalVariables: GlobalVariables.schema,
            txEffect: TxEffect.schema,
            publicReturnValues: z.array(NestedProcessReturnValues.schema),
            gasUsed: z.object({
                totalGas: Gas.schema,
                teardownGas: Gas.schema,
                publicGas: Gas.schema,
                billedGas: Gas.schema
            })
        }).transform((fields)=>new PublicSimulationOutput(fields.revertReason, fields.globalVariables, fields.txEffect, fields.publicReturnValues, fields.gasUsed));
    }
    static async random() {
        return new PublicSimulationOutput(await SimulationError.random(), GlobalVariables.empty(), TxEffect.empty(), times(2, NestedProcessReturnValues.random), {
            teardownGas: Gas.random(),
            totalGas: Gas.random(),
            publicGas: Gas.random(),
            billedGas: Gas.random()
        });
    }
}
