// @aztec/stdlib/dest/tx/tx_hash.js
import { Fr } from '@aztec/foundation/curves/bn254';
import { BufferReader, serializeArrayOfBufferableToVector } from '@aztec/foundation/serialize';
import { schemas } from '../schemas/index.js';
/**
 * A class representing hash of Aztec transaction.
 * @dev Computed by hashing the public inputs of the private kernel tail circuit (see Tx::getTxHash function).
 */ export class TxHash {
    hash;
    constructor(/** A field representing the tx hash (tx hash is an output of poseidon hash hence it's a field). */ hash){
        this.hash = hash;
    }
    static random() {
        return new TxHash(Fr.random());
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new this(reader.readObject(Fr));
    }
    static fromString(str) {
        return new TxHash(Fr.fromString(str));
    }
    static fromBigInt(value) {
        return new TxHash(new Fr(value));
    }
    static fromField(value) {
        return new TxHash(value);
    }
    toBuffer() {
        return this.hash.toBuffer();
    }
    toString() {
        return this.hash.toString();
    }
    toBigInt() {
        return this.hash.toBigInt();
    }
    toJSON() {
        return this.toString();
    }
    equals(other) {
        return this.hash.equals(other.hash);
    }
    static get schema() {
        return schemas.BufferHex.transform((value)=>new TxHash(Fr.fromBuffer(value)));
    }
    static zero() {
        return new TxHash(Fr.ZERO);
    }
    static get SIZE() {
        return Fr.SIZE_IN_BYTES;
    }
}
/**
 * Helper class to handle Serialization and Deserialization of TxHashes array.
 **/ export class TxHashArray extends Array {
    static fromBuffer(buffer) {
        try {
            const reader = BufferReader.asReader(buffer);
            const hashes = reader.readVector(TxHash);
            return new TxHashArray(...hashes);
        } catch  {
            return new TxHashArray();
        }
    }
    toBuffer() {
        return serializeArrayOfBufferableToVector([
            ...this
        ]);
    }
}
