// @aztec/ethereum/dest/l1_tx_utils/forwarder_l1_tx_utils.js
import { encodeFunctionData } from 'viem';
import { FORWARDER_ABI } from '../forwarder_proxy.js';
import { resolveSignerSource } from './factory.js';
import { L1TxUtils } from './l1_tx_utils.js';
/**
 * Extends L1TxUtils to wrap all transactions through a forwarder contract.
 * This is mainly used for testing the archiver's ability to decode transactions that go through proxies.
 */ export class ForwarderL1TxUtils extends L1TxUtils {
    forwarderAddress;
    constructor(client, senderAddress, signingCallback, logger, dateProvider, config, debugMaxGasLimit, store, metrics, kzg, delayer, forwarderAddress){
        super(client, senderAddress, signingCallback, logger, dateProvider, config, debugMaxGasLimit, store, metrics, kzg, delayer), this.forwarderAddress = forwarderAddress;
    }
    /**
   * Wraps the transaction request in a call to the forwarder contract.
   */ wrapInForwarder(request) {
        const forwarderCalldata = encodeFunctionData({
            abi: FORWARDER_ABI,
            functionName: 'forward',
            args: [
                request.to,
                request.data
            ]
        });
        return {
            to: this.forwarderAddress.toString(),
            data: forwarderCalldata,
            value: request.value,
            abi: request.abi
        };
    }
    /**
   * Override sendAndMonitorTransaction to wrap the request in a forwarder call.
   */ sendAndMonitorTransaction(request, gasConfig, blobInputs) {
        this.logger.debug(`Wrapping transaction to ${request.to} in forwarder at ${this.forwarderAddress.toString()}`);
        const wrappedRequest = this.wrapInForwarder(request);
        return super.sendAndMonitorTransaction(wrappedRequest, gasConfig, blobInputs);
    }
}
export function createForwarderL1TxUtils(source, forwarderAddress, deps, config) {
    const { client, address, signingCallback } = resolveSignerSource(source);
    return new ForwarderL1TxUtils(client, address, signingCallback, deps?.logger, deps?.dateProvider, config ?? {}, config?.debugMaxGasLimit ?? false, deps?.store, deps?.metrics, deps?.kzg, deps?.delayer, forwarderAddress);
}
