// @aztec/noir-protocol-circuits-types/dest/artifacts/server.js
import PublicChonkVerifierJson from '../../artifacts/chonk_verifier_public.json' with {
    type: 'json'
};
import ParityBaseJson from '../../artifacts/parity_base.json' with {
    type: 'json'
};
import ParityRootJson from '../../artifacts/parity_root.json' with {
    type: 'json'
};
import BlockMergeRollupJson from '../../artifacts/rollup_block_merge.json' with {
    type: 'json'
};
import BlockRootRollupJson from '../../artifacts/rollup_block_root.json' with {
    type: 'json'
};
import BlockRootFirstRollupJson from '../../artifacts/rollup_block_root_first.json' with {
    type: 'json'
};
import BlockRootEmptyTxFirstRollupJson from '../../artifacts/rollup_block_root_first_empty_tx.json' with {
    type: 'json'
};
import BlockRootFirstRollupSimulatedJson from '../../artifacts/rollup_block_root_first_simulated.json' with {
    type: 'json'
};
import BlockRootSingleTxFirstRollupJson from '../../artifacts/rollup_block_root_first_single_tx.json' with {
    type: 'json'
};
import BlockRootSingleTxFirstRollupSimulatedJson from '../../artifacts/rollup_block_root_first_single_tx_simulated.json' with {
    type: 'json'
};
import BlockRootRollupSimulatedJson from '../../artifacts/rollup_block_root_simulated.json' with {
    type: 'json'
};
import BlockRootSingleTxRollupJson from '../../artifacts/rollup_block_root_single_tx.json' with {
    type: 'json'
};
import BlockRootSingleTxRollupSimulatedJson from '../../artifacts/rollup_block_root_single_tx_simulated.json' with {
    type: 'json'
};
import CheckpointMergeRollupJson from '../../artifacts/rollup_checkpoint_merge.json' with {
    type: 'json'
};
import CheckpointPaddingRollupJson from '../../artifacts/rollup_checkpoint_padding.json' with {
    type: 'json'
};
import CheckpointRootRollupJson from '../../artifacts/rollup_checkpoint_root.json' with {
    type: 'json'
};
import CheckpointRootRollupSimulatedJson from '../../artifacts/rollup_checkpoint_root_simulated.json' with {
    type: 'json'
};
import CheckpointRootSingleBlockRollupJson from '../../artifacts/rollup_checkpoint_root_single_block.json' with {
    type: 'json'
};
import CheckpointRootSingleBlockRollupSimulatedJson from '../../artifacts/rollup_checkpoint_root_single_block_simulated.json' with {
    type: 'json'
};
import RootRollupJson from '../../artifacts/rollup_root.json' with {
    type: 'json'
};
import PrivateTxBaseRollupJson from '../../artifacts/rollup_tx_base_private.json' with {
    type: 'json'
};
import PrivateTxBaseRollupSimulatedJson from '../../artifacts/rollup_tx_base_private_simulated.json' with {
    type: 'json'
};
import PublicTxBaseRollupJson from '../../artifacts/rollup_tx_base_public.json' with {
    type: 'json'
};
import PublicTxBaseRollupSimulatedJson from '../../artifacts/rollup_tx_base_public_simulated.json' with {
    type: 'json'
};
import TxMergeRollupJson from '../../artifacts/rollup_tx_merge.json' with {
    type: 'json'
};
export const ServerCircuitArtifacts = {
    ParityBaseArtifact: ParityBaseJson,
    ParityRootArtifact: ParityRootJson,
    PublicChonkVerifier: PublicChonkVerifierJson,
    PrivateTxBaseRollupArtifact: PrivateTxBaseRollupJson,
    PublicTxBaseRollupArtifact: PublicTxBaseRollupJson,
    TxMergeRollupArtifact: TxMergeRollupJson,
    BlockRootFirstRollupArtifact: BlockRootFirstRollupJson,
    BlockRootSingleTxFirstRollupArtifact: BlockRootSingleTxFirstRollupJson,
    BlockRootEmptyTxFirstRollupArtifact: BlockRootEmptyTxFirstRollupJson,
    BlockRootRollupArtifact: BlockRootRollupJson,
    BlockRootSingleTxRollupArtifact: BlockRootSingleTxRollupJson,
    BlockMergeRollupArtifact: BlockMergeRollupJson,
    CheckpointRootRollupArtifact: CheckpointRootRollupJson,
    CheckpointRootSingleBlockRollupArtifact: CheckpointRootSingleBlockRollupJson,
    CheckpointPaddingRollupArtifact: CheckpointPaddingRollupJson,
    CheckpointMergeRollupArtifact: CheckpointMergeRollupJson,
    RootRollupArtifact: RootRollupJson
};
export const SimulatedServerCircuitArtifacts = {
    ParityBaseArtifact: ParityBaseJson,
    ParityRootArtifact: ParityRootJson,
    PublicChonkVerifier: PublicChonkVerifierJson,
    PrivateTxBaseRollupArtifact: PrivateTxBaseRollupSimulatedJson,
    PublicTxBaseRollupArtifact: PublicTxBaseRollupSimulatedJson,
    TxMergeRollupArtifact: TxMergeRollupJson,
    BlockRootFirstRollupArtifact: BlockRootFirstRollupSimulatedJson,
    BlockRootSingleTxFirstRollupArtifact: BlockRootSingleTxFirstRollupSimulatedJson,
    BlockRootEmptyTxFirstRollupArtifact: BlockRootEmptyTxFirstRollupJson,
    BlockRootRollupArtifact: BlockRootRollupSimulatedJson,
    BlockRootSingleTxRollupArtifact: BlockRootSingleTxRollupSimulatedJson,
    BlockMergeRollupArtifact: BlockMergeRollupJson,
    CheckpointRootRollupArtifact: CheckpointRootRollupSimulatedJson,
    CheckpointRootSingleBlockRollupArtifact: CheckpointRootSingleBlockRollupSimulatedJson,
    CheckpointPaddingRollupArtifact: CheckpointPaddingRollupJson,
    CheckpointMergeRollupArtifact: CheckpointMergeRollupJson,
    RootRollupArtifact: RootRollupJson
};
export function getServerCircuitArtifact(name) {
    return {
        ...ServerCircuitArtifacts[name],
        name
    };
}
export function getSimulatedServerCircuitArtifact(name) {
    return {
        ...SimulatedServerCircuitArtifacts[name],
        name
    };
}
