// @aztec/protocol-contracts/dest/instance-registry/contract_instance_updated_event.js
import { CONTRACT_INSTANCE_UPDATED_MAGIC_VALUE } from '@aztec/constants';
import { FieldReader } from '@aztec/foundation/serialize';
import { AztecAddress } from '@aztec/stdlib/aztec-address';
import { ProtocolContractAddress } from '../protocol_contract_data.js';
/** Event emitted from the ContractInstanceRegistry. */ export class ContractInstanceUpdatedEvent {
    address;
    prevContractClassId;
    newContractClassId;
    timestampOfChange;
    constructor(address, prevContractClassId, newContractClassId, timestampOfChange){
        this.address = address;
        this.prevContractClassId = prevContractClassId;
        this.newContractClassId = newContractClassId;
        this.timestampOfChange = timestampOfChange;
    }
    static isContractInstanceUpdatedEvent(log) {
        return log.contractAddress.equals(ProtocolContractAddress.ContractInstanceRegistry) && log.fields[0].toBigInt() === CONTRACT_INSTANCE_UPDATED_MAGIC_VALUE;
    }
    static fromLog(log) {
        const reader = new FieldReader(log.fields.slice(1));
        const address = reader.readObject(AztecAddress);
        const prevContractClassId = reader.readField();
        const newContractClassId = reader.readField();
        const timestampOfChange = reader.readField().toBigInt();
        return new ContractInstanceUpdatedEvent(address, prevContractClassId, newContractClassId, timestampOfChange);
    }
    toContractInstanceUpdate() {
        return {
            address: this.address,
            prevContractClassId: this.prevContractClassId,
            newContractClassId: this.newContractClassId,
            timestampOfChange: this.timestampOfChange
        };
    }
}
