// @aztec/kv-store/dest/lmdb-v2/message.js
export var Database = /*#__PURE__*/ function(Database) {
    Database["DATA"] = "data";
    Database["INDEX"] = "index";
    return Database;
}({});
export const CURSOR_PAGE_SIZE = 10;
export var LMDBMessageType = /*#__PURE__*/ function(LMDBMessageType) {
    LMDBMessageType[LMDBMessageType["OPEN_DATABASE"] = 100] = "OPEN_DATABASE";
    LMDBMessageType[LMDBMessageType["GET"] = 101] = "GET";
    LMDBMessageType[LMDBMessageType["HAS"] = 102] = "HAS";
    LMDBMessageType[LMDBMessageType["START_CURSOR"] = 103] = "START_CURSOR";
    LMDBMessageType[LMDBMessageType["ADVANCE_CURSOR"] = 104] = "ADVANCE_CURSOR";
    LMDBMessageType[LMDBMessageType["ADVANCE_CURSOR_COUNT"] = 105] = "ADVANCE_CURSOR_COUNT";
    LMDBMessageType[LMDBMessageType["CLOSE_CURSOR"] = 106] = "CLOSE_CURSOR";
    LMDBMessageType[LMDBMessageType["BATCH"] = 107] = "BATCH";
    LMDBMessageType[LMDBMessageType["STATS"] = 108] = "STATS";
    LMDBMessageType[LMDBMessageType["CLOSE"] = 109] = "CLOSE";
    LMDBMessageType[LMDBMessageType["COPY_STORE"] = 110] = "COPY_STORE";
    return LMDBMessageType;
}({});
