// @aztec/accounts/dest/schnorr/account_contract.js
import { Schnorr } from '@aztec/foundation/crypto/schnorr';
import { AuthWitness } from '@aztec/stdlib/auth-witness';
import { DefaultAccountContract } from '../defaults/account_contract.js';
/**
 * Account contract that authenticates transactions using Schnorr signatures
 * verified against a Grumpkin public key stored in an immutable encrypted note.
 * This abstract version does not provide a way to retrieve the artifact, as it
 * can be implemented with or without lazy loading.
 */ export class SchnorrBaseAccountContract extends DefaultAccountContract {
    signingPrivateKey;
    constructor(signingPrivateKey){
        super(), this.signingPrivateKey = signingPrivateKey;
    }
    async getInitializationFunctionAndArgs() {
        const signingPublicKey = await new Schnorr().computePublicKey(this.signingPrivateKey);
        return {
            constructorName: 'constructor',
            constructorArgs: [
                signingPublicKey.x,
                signingPublicKey.y
            ]
        };
    }
    getAuthWitnessProvider(_address) {
        return new SchnorrAuthWitnessProvider(this.signingPrivateKey);
    }
}
/** Creates auth witnesses using Schnorr signatures. */ export class SchnorrAuthWitnessProvider {
    signingPrivateKey;
    constructor(signingPrivateKey){
        this.signingPrivateKey = signingPrivateKey;
    }
    async createAuthWit(messageHash) {
        const schnorr = new Schnorr();
        const signature = await schnorr.constructSignature(messageHash.toBuffer(), this.signingPrivateKey);
        return new AuthWitness(messageHash, [
            ...signature.toBuffer()
        ]);
    }
}
