// @aztec/stdlib/dest/block/l2_block_source.js
import { BlockNumberSchema, CheckpointNumberSchema } from '@aztec/foundation/branded-types';
import { z } from 'zod';
import { CheckpointHeader } from '../rollup/checkpoint_header.js';
export const GENESIS_CHECKPOINT_HEADER_HASH = CheckpointHeader.empty().hash();
/** Creates an L2 block id */ export function makeL2BlockId(number, hash) {
    if (number !== 0 && !hash) {
        throw new Error(`Hash is required for non-genesis blocks (got block number ${number})`);
    }
    return {
        number,
        hash: hash
    };
}
/** Creates an L2 checkpoint id */ export function makeL2CheckpointId(number, hash) {
    return {
        number,
        hash
    };
}
const L2BlockIdSchema = z.object({
    number: BlockNumberSchema,
    hash: z.string()
});
const L2CheckpointIdSchema = z.object({
    number: CheckpointNumberSchema,
    hash: z.string()
});
const L2TipIdSchema = z.object({
    block: L2BlockIdSchema,
    checkpoint: L2CheckpointIdSchema
});
export const L2TipsSchema = z.object({
    proposed: L2BlockIdSchema,
    checkpointed: L2TipIdSchema,
    proven: L2TipIdSchema,
    finalized: L2TipIdSchema
});
export var L2BlockSourceEvents = /*#__PURE__*/ function(L2BlockSourceEvents) {
    L2BlockSourceEvents["L2PruneUnproven"] = "l2PruneUnproven";
    L2BlockSourceEvents["L2PruneUncheckpointed"] = "l2PruneUncheckpointed";
    L2BlockSourceEvents["L2BlockProven"] = "l2BlockProven";
    L2BlockSourceEvents["L2BlocksCheckpointed"] = "l2BlocksCheckpointed";
    L2BlockSourceEvents["InvalidAttestationsCheckpointDetected"] = "invalidCheckpointDetected";
    return L2BlockSourceEvents;
}({});
