// @aztec/stdlib/dest/block/validate_block_result.js
import { EpochNumber, EpochNumberSchema } from '@aztec/foundation/branded-types';
import { EthAddress } from '@aztec/foundation/eth-address';
import { schemas } from '@aztec/foundation/schemas';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { z } from 'zod';
import { CheckpointInfoSchema, deserializeCheckpointInfo, serializeCheckpointInfo } from '../checkpoint/checkpoint_info.js';
import { MAX_COMMITTEE_SIZE } from '../deserialization/index.js';
import { CommitteeAttestation } from './proposal/committee_attestation.js';
export const ValidateCheckpointResultSchema = z.union([
    z.object({
        valid: z.literal(true)
    }),
    z.object({
        valid: z.literal(false),
        checkpoint: CheckpointInfoSchema,
        committee: z.array(schemas.EthAddress),
        epoch: EpochNumberSchema,
        seed: schemas.BigInt,
        attestors: z.array(schemas.EthAddress),
        attestations: z.array(CommitteeAttestation.schema),
        reason: z.literal('insufficient-attestations')
    }),
    z.object({
        valid: z.literal(false),
        checkpoint: CheckpointInfoSchema,
        committee: z.array(schemas.EthAddress),
        epoch: EpochNumberSchema,
        seed: schemas.BigInt,
        attestors: z.array(schemas.EthAddress),
        attestations: z.array(CommitteeAttestation.schema),
        reason: z.literal('invalid-attestation'),
        invalidIndex: z.number()
    })
]);
export function serializeValidateCheckpointResult(result) {
    if (result.valid) {
        return serializeToBuffer(true);
    }
    const checkpointBuffer = serializeCheckpointInfo(result.checkpoint);
    return serializeToBuffer(result.valid, result.reason, checkpointBuffer.length, checkpointBuffer, result.committee.length, result.committee, result.epoch, result.seed ?? 0n, result.attestors.length, result.attestors, result.attestations.length, result.attestations, result.reason === 'invalid-attestation' ? result.invalidIndex : 0);
}
export function deserializeValidateCheckpointResult(bufferOrReader) {
    const reader = BufferReader.asReader(bufferOrReader);
    const valid = reader.readBoolean();
    if (valid) {
        return {
            valid
        };
    }
    const reason = reader.readString(64);
    const checkpoint = deserializeCheckpointInfo(reader.readBuffer());
    const committee = reader.readVector(EthAddress, MAX_COMMITTEE_SIZE);
    const epoch = EpochNumber(reader.readNumber());
    const seed = reader.readBigInt();
    const attestors = reader.readVector(EthAddress, MAX_COMMITTEE_SIZE);
    const attestations = reader.readVector(CommitteeAttestation, MAX_COMMITTEE_SIZE);
    const invalidIndex = reader.readNumber();
    if (reason === 'insufficient-attestations') {
        return {
            valid,
            reason,
            checkpoint,
            committee,
            epoch,
            seed,
            attestors,
            attestations
        };
    } else if (reason === 'invalid-attestation') {
        return {
            valid,
            reason,
            checkpoint,
            committee,
            epoch,
            seed,
            attestors,
            invalidIndex,
            attestations
        };
    } else {
        const _ = reason;
        throw new Error(`Unknown reason: ${reason}`);
    }
}
