// @aztec/stdlib/dest/kernel/claimed_length_array.js
var _computedKey;
import { BufferReader, FieldReader, serializeToBuffer, serializeToFields } from '@aztec/foundation/serialize';
import { inspect } from 'util';
_computedKey = inspect.custom;
export class ClaimedLengthArray {
    array;
    claimedLength;
    constructor(array, // Named `claimedLength` instead of `length` to avoid being confused with array.length.
    claimedLength){
        this.array = array;
        this.claimedLength = claimedLength;
    }
    static fromBuffer(buffer, deserializer, arrayLength) {
        const reader = BufferReader.asReader(buffer);
        const array = reader.readArray(arrayLength, deserializer);
        const claimedLength = reader.readNumber();
        return new ClaimedLengthArray(array, claimedLength);
    }
    toBuffer() {
        return serializeToBuffer(this.array, this.claimedLength);
    }
    static fromFields(fields, deserializer, arrayLength) {
        const reader = FieldReader.asReader(fields);
        const array = reader.readArray(arrayLength, deserializer);
        const claimedLength = reader.readU32();
        return new ClaimedLengthArray(array, claimedLength);
    }
    toFields() {
        return serializeToFields(this.array, this.claimedLength);
    }
    static empty(elem, arraySize) {
        const array = Array(arraySize).fill(elem.empty());
        return new ClaimedLengthArray(array, 0);
    }
    isEmpty() {
        return this.claimedLength === 0;
    }
    getActiveItems() {
        return this.array.slice(0, this.claimedLength);
    }
    getSize() {
        return this.toBuffer().length;
    }
    [_computedKey]() {
        return `ClaimedLengthArray {
      array: [${this.getActiveItems().map((x)=>inspect(x)).join(', ')}],
      claimedLength: ${this.claimedLength},
    `;
    }
}
export function ClaimedLengthArrayFromBuffer(deserializer, arrayLength) {
    return {
        fromBuffer: (reader)=>ClaimedLengthArray.fromBuffer(reader, deserializer, arrayLength)
    };
}
export function ClaimedLengthArrayFromFields(deserializer, arrayLength) {
    return {
        fromFields: (reader)=>ClaimedLengthArray.fromFields(reader, deserializer, arrayLength)
    };
}
