// @aztec/stdlib/dest/kernel/hiding_kernel_private_inputs.js
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { VkData } from '../vks/index.js';
import { PrivateToPublicKernelCircuitPublicInputs } from './private_to_public_kernel_circuit_public_inputs.js';
import { PrivateToRollupKernelCircuitPublicInputs } from './private_to_rollup_kernel_circuit_public_inputs.js';
export class HidingKernelToRollupPrivateInputs {
    previousKernelPublicInputs;
    previousKernelVkData;
    constructor(previousKernelPublicInputs, previousKernelVkData){
        this.previousKernelPublicInputs = previousKernelPublicInputs;
        this.previousKernelVkData = previousKernelVkData;
    }
    toBuffer() {
        return serializeToBuffer(this.previousKernelPublicInputs, this.previousKernelVkData);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new this(reader.readObject(PrivateToRollupKernelCircuitPublicInputs), reader.readObject(VkData));
    }
}
export class HidingKernelToPublicPrivateInputs {
    previousKernelPublicInputs;
    previousKernelVkData;
    constructor(previousKernelPublicInputs, previousKernelVkData){
        this.previousKernelPublicInputs = previousKernelPublicInputs;
        this.previousKernelVkData = previousKernelVkData;
    }
    toBuffer() {
        return serializeToBuffer(this.previousKernelPublicInputs, this.previousKernelVkData);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new this(reader.readObject(PrivateToPublicKernelCircuitPublicInputs), reader.readObject(VkData));
    }
}
