// @aztec/stdlib/dest/kernel/log_hash.js
var _computedKey;
import { Fr } from '@aztec/foundation/curves/bn254';
import { BufferReader, FieldReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { inspect } from 'util';
import { AztecAddress } from '../aztec-address/index.js';
_computedKey = inspect.custom;
export class LogHash {
    value;
    length;
    constructor(value, length){
        this.value = value;
        this.length = length;
    }
    static from(fields) {
        return new LogHash(fields.value, fields.length);
    }
    toFields() {
        return [
            this.value,
            new Fr(this.length)
        ];
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new LogHash(reader.readField(), reader.readU32());
    }
    isEmpty() {
        return this.value.isZero() && !this.length;
    }
    static empty() {
        return new LogHash(Fr.zero(), 0);
    }
    toBuffer() {
        return serializeToBuffer(this.value, this.length);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new LogHash(Fr.fromBuffer(reader), reader.readNumber());
    }
    toString() {
        return `value=${this.value} length=${this.length}`;
    }
    scope(contractAddress) {
        return new ScopedLogHash(this, contractAddress);
    }
    [_computedKey]() {
        return `LogHash { ${this.toString()} }`;
    }
}
export class CountedLogHash {
    logHash;
    counter;
    constructor(logHash, counter){
        this.logHash = logHash;
        this.counter = counter;
    }
    toFields() {
        return [
            ...this.logHash.toFields(),
            new Fr(this.counter)
        ];
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new CountedLogHash(reader.readObject(LogHash), reader.readU32());
    }
    isEmpty() {
        return this.logHash.isEmpty() && !this.counter;
    }
    static empty() {
        return new CountedLogHash(LogHash.empty(), 0);
    }
    toBuffer() {
        return serializeToBuffer(this.logHash, this.counter);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new CountedLogHash(LogHash.fromBuffer(reader), reader.readNumber());
    }
}
export class ScopedLogHash {
    logHash;
    contractAddress;
    constructor(logHash, contractAddress){
        this.logHash = logHash;
        this.contractAddress = contractAddress;
    }
    get value() {
        return this.logHash.value;
    }
    toFields() {
        return [
            ...this.logHash.toFields(),
            this.contractAddress.toField()
        ];
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new ScopedLogHash(reader.readObject(LogHash), AztecAddress.fromField(reader.readField()));
    }
    isEmpty() {
        return this.logHash.isEmpty() && this.contractAddress.isZero();
    }
    static empty() {
        return new ScopedLogHash(LogHash.empty(), AztecAddress.ZERO);
    }
    toBuffer() {
        return serializeToBuffer(this.logHash, this.contractAddress);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new ScopedLogHash(LogHash.fromBuffer(reader), AztecAddress.fromBuffer(reader));
    }
    toString() {
        return `logHash=${this.logHash} contractAddress=${this.contractAddress}`;
    }
}
export class ScopedCountedLogHash {
    inner;
    contractAddress;
    constructor(inner, contractAddress){
        this.inner = inner;
        this.contractAddress = contractAddress;
    }
    toFields() {
        return [
            ...this.inner.toFields(),
            this.contractAddress.toField()
        ];
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new ScopedCountedLogHash(reader.readObject(CountedLogHash), AztecAddress.fromField(reader.readField()));
    }
    toBuffer() {
        return serializeToBuffer(this.inner, this.contractAddress);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new ScopedCountedLogHash(CountedLogHash.fromBuffer(reader), AztecAddress.fromBuffer(reader));
    }
    isEmpty() {
        return this.inner.isEmpty() && this.contractAddress.isZero();
    }
    static empty() {
        return new ScopedCountedLogHash(CountedLogHash.empty(), AztecAddress.ZERO);
    }
}
