// @aztec/stdlib/dest/kernel/nullifier.js
import { Fr } from '@aztec/foundation/curves/bn254';
import { BufferReader, FieldReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { AztecAddress } from '../aztec-address/index.js';
export class Nullifier {
    value;
    noteHash;
    counter;
    constructor(value, noteHash, counter){
        this.value = value;
        this.noteHash = noteHash;
        this.counter = counter;
    }
    toFields() {
        return [
            this.value,
            this.noteHash,
            new Fr(this.counter)
        ];
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new Nullifier(reader.readField(), reader.readField(), reader.readU32());
    }
    isEmpty() {
        return this.value.isZero() && this.noteHash.isZero() && !this.counter;
    }
    static empty() {
        return new Nullifier(Fr.zero(), Fr.zero(), 0);
    }
    toBuffer() {
        return serializeToBuffer(this.value, this.noteHash, this.counter);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new Nullifier(Fr.fromBuffer(reader), Fr.fromBuffer(reader), reader.readNumber());
    }
    toString() {
        return `value=${this.value} noteHash=${this.noteHash} counter=${this.counter}`;
    }
    scope(contractAddress) {
        return new ScopedNullifier(this, contractAddress);
    }
}
export class ScopedNullifier {
    nullifier;
    contractAddress;
    constructor(nullifier, contractAddress){
        this.nullifier = nullifier;
        this.contractAddress = contractAddress;
    }
    get counter() {
        return this.nullifier.counter;
    }
    get value() {
        return this.nullifier.value;
    }
    get nullifiedNoteHash() {
        return this.nullifier.noteHash;
    }
    toFields() {
        return [
            ...this.nullifier.toFields(),
            this.contractAddress.toField()
        ];
    }
    static fromFields(fields) {
        const reader = FieldReader.asReader(fields);
        return new ScopedNullifier(reader.readObject(Nullifier), AztecAddress.fromField(reader.readField()));
    }
    isEmpty() {
        return this.nullifier.isEmpty() && this.contractAddress.isZero();
    }
    static empty() {
        return new ScopedNullifier(Nullifier.empty(), AztecAddress.ZERO);
    }
    toBuffer() {
        return serializeToBuffer(this.nullifier, this.contractAddress);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new ScopedNullifier(Nullifier.fromBuffer(reader), AztecAddress.fromBuffer(reader));
    }
    toString() {
        return `nullifier=${this.nullifier} contractAddress=${this.contractAddress}`;
    }
}
