// @aztec/stdlib/dest/kernel/padded_side_effects.js
import { MAX_NOTE_HASHES_PER_TX, MAX_NULLIFIERS_PER_TX, MAX_PRIVATE_LOGS_PER_TX } from '@aztec/constants';
import { makeTuple } from '@aztec/foundation/array';
import { Fr } from '@aztec/foundation/curves/bn254';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { PrivateLog } from '../logs/index.js';
export class PaddedSideEffects {
    noteHashes;
    nullifiers;
    privateLogs;
    constructor(noteHashes, nullifiers, privateLogs){
        this.noteHashes = noteHashes;
        this.nullifiers = nullifiers;
        this.privateLogs = privateLogs;
    }
    toBuffer() {
        return serializeToBuffer(this.noteHashes, this.nullifiers, this.privateLogs);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PaddedSideEffects(reader.readArray(MAX_NOTE_HASHES_PER_TX, Fr), reader.readArray(MAX_NULLIFIERS_PER_TX, Fr), reader.readArray(MAX_PRIVATE_LOGS_PER_TX, PrivateLog));
    }
    static empty() {
        return new PaddedSideEffects(makeTuple(MAX_NOTE_HASHES_PER_TX, Fr.zero), makeTuple(MAX_NULLIFIERS_PER_TX, Fr.zero), makeTuple(MAX_PRIVATE_LOGS_PER_TX, PrivateLog.empty));
    }
}
export class PaddedSideEffectAmounts {
    nonRevertibleNoteHashes;
    revertibleNoteHashes;
    nonRevertibleNullifiers;
    revertibleNullifiers;
    nonRevertiblePrivateLogs;
    revertiblePrivateLogs;
    constructor(nonRevertibleNoteHashes, revertibleNoteHashes, nonRevertibleNullifiers, revertibleNullifiers, nonRevertiblePrivateLogs, revertiblePrivateLogs){
        this.nonRevertibleNoteHashes = nonRevertibleNoteHashes;
        this.revertibleNoteHashes = revertibleNoteHashes;
        this.nonRevertibleNullifiers = nonRevertibleNullifiers;
        this.revertibleNullifiers = revertibleNullifiers;
        this.nonRevertiblePrivateLogs = nonRevertiblePrivateLogs;
        this.revertiblePrivateLogs = revertiblePrivateLogs;
    }
    toBuffer() {
        return serializeToBuffer(this.nonRevertibleNoteHashes, this.revertibleNoteHashes, this.nonRevertibleNullifiers, this.revertibleNullifiers, this.nonRevertiblePrivateLogs, this.revertiblePrivateLogs);
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PaddedSideEffectAmounts(reader.readNumber(), reader.readNumber(), reader.readNumber(), reader.readNumber(), reader.readNumber(), reader.readNumber());
    }
    static empty() {
        return new PaddedSideEffectAmounts(0, 0, 0, 0, 0, 0);
    }
}
