// @aztec/stdlib/dest/kernel/private_kernel_reset_dimensions.js
import { serializeToBuffer } from '@aztec/foundation/serialize';
export class PrivateKernelResetDimensions {
    NOTE_HASH_PENDING_READ;
    NOTE_HASH_SETTLED_READ;
    NULLIFIER_PENDING_READ;
    NULLIFIER_SETTLED_READ;
    KEY_VALIDATION;
    TRANSIENT_DATA_SQUASHING;
    NOTE_HASH_SILOING;
    NULLIFIER_SILOING;
    PRIVATE_LOG_SILOING;
    constructor(NOTE_HASH_PENDING_READ, NOTE_HASH_SETTLED_READ, NULLIFIER_PENDING_READ, NULLIFIER_SETTLED_READ, KEY_VALIDATION, TRANSIENT_DATA_SQUASHING, NOTE_HASH_SILOING, NULLIFIER_SILOING, PRIVATE_LOG_SILOING){
        this.NOTE_HASH_PENDING_READ = NOTE_HASH_PENDING_READ;
        this.NOTE_HASH_SETTLED_READ = NOTE_HASH_SETTLED_READ;
        this.NULLIFIER_PENDING_READ = NULLIFIER_PENDING_READ;
        this.NULLIFIER_SETTLED_READ = NULLIFIER_SETTLED_READ;
        this.KEY_VALIDATION = KEY_VALIDATION;
        this.TRANSIENT_DATA_SQUASHING = TRANSIENT_DATA_SQUASHING;
        this.NOTE_HASH_SILOING = NOTE_HASH_SILOING;
        this.NULLIFIER_SILOING = NULLIFIER_SILOING;
        this.PRIVATE_LOG_SILOING = PRIVATE_LOG_SILOING;
    }
    toBuffer() {
        return serializeToBuffer(this.NOTE_HASH_PENDING_READ, this.NOTE_HASH_SETTLED_READ, this.NULLIFIER_PENDING_READ, this.NULLIFIER_SETTLED_READ, this.KEY_VALIDATION, this.TRANSIENT_DATA_SQUASHING, this.NOTE_HASH_SILOING, this.NULLIFIER_SILOING, this.PRIVATE_LOG_SILOING);
    }
    static empty() {
        return new PrivateKernelResetDimensions(0, 0, 0, 0, 0, 0, 0, 0, 0);
    }
    static from(fields) {
        const dimensions = PrivateKernelResetDimensions.empty();
        privateKernelResetDimensionNames.forEach((name)=>dimensions[name] = fields[name] ?? 0);
        return dimensions;
    }
    static fromValues(values) {
        if (values.length !== privateKernelResetDimensionNames.length) {
            throw new Error(`Not enough values for all dimensions. Required ${privateKernelResetDimensionNames.length}. Got ${values.length}.`);
        }
        const dimensions = PrivateKernelResetDimensions.empty();
        privateKernelResetDimensionNames.forEach((name, i)=>dimensions[name] = values[i]);
        return dimensions;
    }
    toValues() {
        return privateKernelResetDimensionNames.map((name)=>this[name]);
    }
}
export const privateKernelResetDimensionNames = [
    'NOTE_HASH_PENDING_READ',
    'NOTE_HASH_SETTLED_READ',
    'NULLIFIER_PENDING_READ',
    'NULLIFIER_SETTLED_READ',
    'KEY_VALIDATION',
    'TRANSIENT_DATA_SQUASHING',
    'NOTE_HASH_SILOING',
    'NULLIFIER_SILOING',
    'PRIVATE_LOG_SILOING'
];
