// @aztec/stdlib/dest/kernel/private_kernel_reset_circuit_private_inputs.js
import { serializeToBuffer } from '@aztec/foundation/serialize';
export class PrivateKernelResetCircuitPrivateInputsVariants {
    previousKernel;
    paddedSideEffects;
    hints;
    constructor(previousKernel, paddedSideEffects, hints){
        this.previousKernel = previousKernel;
        this.paddedSideEffects = paddedSideEffects;
        this.hints = hints;
    }
    toBuffer() {
        return serializeToBuffer(this.previousKernel, this.paddedSideEffects, this.hints);
    }
}
/**
 * Input to the private kernel circuit - reset call.
 */ export class PrivateKernelResetCircuitPrivateInputs {
    previousKernel;
    paddedSideEffects;
    hints;
    dimensions;
    constructor(/**
     * The previous kernel data
     */ previousKernel, paddedSideEffects, hints, dimensions){
        this.previousKernel = previousKernel;
        this.paddedSideEffects = paddedSideEffects;
        this.hints = hints;
        this.dimensions = dimensions;
    }
    /**
   * Serialize this as a buffer.
   * @returns The buffer.
   */ toBuffer() {
        return serializeToBuffer(this.previousKernel, this.paddedSideEffects, this.hints, this.dimensions);
    }
    trimToSizes() {
        const hints = this.hints.trimToSizes(this.dimensions.NOTE_HASH_PENDING_READ, this.dimensions.NOTE_HASH_SETTLED_READ, this.dimensions.NULLIFIER_PENDING_READ, this.dimensions.NULLIFIER_SETTLED_READ, this.dimensions.KEY_VALIDATION, this.dimensions.TRANSIENT_DATA_SQUASHING);
        return new PrivateKernelResetCircuitPrivateInputsVariants(this.previousKernel, this.paddedSideEffects, hints);
    }
}
