// @aztec/stdlib/dest/kernel/private_kernel_init_circuit_private_inputs.js
import { Fr } from '@aztec/foundation/curves/bn254';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { ProtocolContracts } from '../tx/protocol_contracts.js';
import { TxRequest } from '../tx/tx_request.js';
import { PrivateCallData } from './private_call_data.js';
/**
 * Input to the private kernel circuit - initial call.
 */ export class PrivateKernelInitCircuitPrivateInputs {
    txRequest;
    vkTreeRoot;
    protocolContracts;
    privateCall;
    isPrivateOnly;
    firstNullifierHint;
    revertibleCounterHint;
    constructor(/**
     * The transaction request which led to the creation of these inputs.
     */ txRequest, /**
     * The root of the vk tree.
     */ vkTreeRoot, /**
     * The protocol contracts.
     */ protocolContracts, /**
     * Private calldata corresponding to this iteration of the kernel.
     */ privateCall, /**
     * Wether this tx will make public calls or not.
     */ isPrivateOnly, /**
     * A hint to what will be the first nullifier of the transaction, used for nonce generation.
     */ firstNullifierHint, /**
     * A claim to the final min revertible side effect counter of a tx.
     */ revertibleCounterHint){
        this.txRequest = txRequest;
        this.vkTreeRoot = vkTreeRoot;
        this.protocolContracts = protocolContracts;
        this.privateCall = privateCall;
        this.isPrivateOnly = isPrivateOnly;
        this.firstNullifierHint = firstNullifierHint;
        this.revertibleCounterHint = revertibleCounterHint;
    }
    /**
   * Serialize this as a buffer.
   * @returns The buffer.
   */ toBuffer() {
        return serializeToBuffer(this.txRequest, this.vkTreeRoot, this.protocolContracts, this.privateCall, this.firstNullifierHint, this.revertibleCounterHint);
    }
    /**
   * Deserializes from a buffer or reader.
   * @param buffer - Buffer or reader to read from.
   * @returns The deserialized instance.
   */ static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PrivateKernelInitCircuitPrivateInputs(reader.readObject(TxRequest), Fr.fromBuffer(reader), reader.readObject(ProtocolContracts), reader.readObject(PrivateCallData), reader.readBoolean(), Fr.fromBuffer(reader), reader.readNumber());
    }
}
