// @aztec/stdlib/dest/kernel/private_to_public_accumulated_data_builder.js
import { MAX_CONTRACT_CLASS_LOGS_PER_TX, MAX_ENQUEUED_CALLS_PER_TX, MAX_L2_TO_L1_MSGS_PER_TX, MAX_NOTE_HASHES_PER_TX, MAX_NULLIFIERS_PER_TX, MAX_PRIVATE_LOGS_PER_TX } from '@aztec/constants';
import { padArrayEnd } from '@aztec/foundation/collection';
import { Fr } from '@aztec/foundation/curves/bn254';
import { PrivateLog } from '../logs/private_log.js';
import { ScopedL2ToL1Message } from '../messaging/l2_to_l1_message.js';
import { ScopedLogHash } from './log_hash.js';
import { PrivateToPublicAccumulatedData } from './private_to_public_accumulated_data.js';
import { PublicCallRequest } from './public_call_request.js';
/**
 * TESTS-ONLY CLASS
 * Builder for PrivateToPublicAccumulatedData, used to conveniently construct instances for testing,
 * as PrivateToPublicAccumulatedData is (or will shortly be) immutable.
 *
 */ export class PrivateToPublicAccumulatedDataBuilder {
    noteHashes = [];
    nullifiers = [];
    l2ToL1Msgs = [];
    privateLogs = [];
    contractClassLogsHashes = [];
    publicCallRequests = [];
    pushNoteHash(newNoteHash) {
        this.noteHashes.push(newNoteHash);
        return this;
    }
    withNoteHashes(noteHashes) {
        this.noteHashes = noteHashes;
        return this;
    }
    pushNullifier(newNullifier) {
        this.nullifiers.push(newNullifier);
        return this;
    }
    withNullifiers(nullifiers) {
        this.nullifiers = nullifiers;
        return this;
    }
    pushL2ToL1Msg(newL2ToL1Msg) {
        this.l2ToL1Msgs.push(newL2ToL1Msg);
        return this;
    }
    withL2ToL1Msgs(l2ToL1Msgs) {
        this.l2ToL1Msgs = l2ToL1Msgs;
        return this;
    }
    pushPrivateLog(privateLog) {
        this.privateLogs.push(privateLog);
        return this;
    }
    withPrivateLogs(privateLogs) {
        this.privateLogs = privateLogs;
        return this;
    }
    pushContractClassLogsHash(contractClassLogsHash) {
        this.contractClassLogsHashes.push(contractClassLogsHash);
        return this;
    }
    withContractClassLogsHashes(contractClassLogsHashes) {
        this.contractClassLogsHashes = contractClassLogsHashes;
        return this;
    }
    pushPublicCall(publicCall) {
        this.publicCallRequests.push(publicCall);
        return this;
    }
    withPublicCallRequests(publicCallRequests) {
        this.publicCallRequests = publicCallRequests;
        return this;
    }
    build() {
        return new PrivateToPublicAccumulatedData(padArrayEnd(this.noteHashes, Fr.ZERO, MAX_NOTE_HASHES_PER_TX), padArrayEnd(this.nullifiers, Fr.ZERO, MAX_NULLIFIERS_PER_TX), padArrayEnd(this.l2ToL1Msgs, ScopedL2ToL1Message.empty(), MAX_L2_TO_L1_MSGS_PER_TX), padArrayEnd(this.privateLogs, PrivateLog.empty(), MAX_PRIVATE_LOGS_PER_TX), padArrayEnd(this.contractClassLogsHashes, ScopedLogHash.empty(), MAX_CONTRACT_CLASS_LOGS_PER_TX), padArrayEnd(this.publicCallRequests, PublicCallRequest.empty(), MAX_ENQUEUED_CALLS_PER_TX));
    }
}
