// @aztec/bb.js/dest/node-cjs/barretenberg/testing/fields.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Fr = void 0;
const index_js_1 = require("../../random/index.js");
const bigint_buffer_js_1 = require("./bigint-buffer.js");
class Fr {
    static ZERO = new Fr(0n);
    static MODULUS = 0x30644e72e131a029b85045b68181585d2833e84879b9709143e1f593f0000001n;
    static MAX_VALUE = this.MODULUS - 1n;
    static SIZE_IN_BYTES = 32;
    value;
    constructor(value) {
        // We convert buffer value to bigint to be able to check it fits within modulus
        const valueBigInt = typeof value === 'bigint'
            ? value
            : value instanceof Buffer
                ? (0, bigint_buffer_js_1.buffer32BytesToBigIntBE)(value)
                : (0, bigint_buffer_js_1.uint8ArrayToBigIntBE)(value);
        if (valueBigInt > Fr.MAX_VALUE) {
            throw new Error(`Value 0x${valueBigInt.toString(16)} is greater or equal to field modulus.`);
        }
        this.value =
            typeof value === 'bigint' ? (0, bigint_buffer_js_1.bigIntToUint8ArrayBE)(value) : value instanceof Buffer ? new Uint8Array(value) : value;
    }
    static random() {
        const r = (0, bigint_buffer_js_1.uint8ArrayToBigIntBE)((0, index_js_1.randomBytes)(64)) % Fr.MODULUS;
        return new this(r);
    }
    toBuffer() {
        return this.value;
    }
    toString() {
        return '0x' + this.toBuffer().reduce((accumulator, byte) => accumulator + byte.toString(16).padStart(2, '0'), '');
    }
    equals(rhs) {
        return this.value.every((v, i) => v === rhs.value[i]);
    }
    isZero() {
        return this.value.every(v => v === 0);
    }
    static fromBuffer(value) {
        return Fr.fromBufferReduce(value);
    }
    static fromBufferReduce(value) {
        const valueBigInt = (0, bigint_buffer_js_1.uint8ArrayToBigIntBE)(value);
        const reducedValue = valueBigInt % Fr.MODULUS;
        return new Fr(reducedValue);
    }
}
exports.Fr = Fr;
