// @aztec/simulator/dest/public/avm/avm_execution_environment.js
import { Fr } from '@aztec/foundation/curves/bn254';
/**
 * Contains variables that remain constant during AVM execution
 * These variables are provided by the public kernel circuit
 */ export class AvmExecutionEnvironment {
    address;
    sender;
    contractCallDepth;
    transactionFee;
    globals;
    isStaticCall;
    calldata;
    config;
    constructor(address, sender, contractCallDepth, transactionFee, globals, isStaticCall, calldata, config){
        this.address = address;
        this.sender = sender;
        this.contractCallDepth = contractCallDepth;
        this.transactionFee = transactionFee;
        this.globals = globals;
        this.isStaticCall = isStaticCall;
        this.calldata = calldata;
        this.config = config;
    }
    deriveEnvironmentForNestedCallInternal(targetAddress, calldata, isStaticCall) {
        return new AvmExecutionEnvironment(/*address=*/ targetAddress, /*sender=*/ this.address, this.contractCallDepth.add(Fr.ONE), this.transactionFee, this.globals, isStaticCall, calldata, this.config);
    }
    deriveEnvironmentForNestedCall(targetAddress, calldata) {
        return this.deriveEnvironmentForNestedCallInternal(targetAddress, calldata, /*isStaticCall=*/ false);
    }
    deriveEnvironmentForNestedStaticCall(targetAddress, calldata) {
        return this.deriveEnvironmentForNestedCallInternal(targetAddress, calldata, /*isStaticCall=*/ true);
    }
}
