// @aztec/simulator/dest/public/avm/avm_contract_call_result.js
import { Gas } from '@aztec/stdlib/gas';
import { inspect } from 'util';
import { createSimulationError } from '../../common/errors.js';
/**
 * Results of an contract call's execution in the AVM.
 */ export class AvmContractCallResult {
    reverted;
    output;
    gasLeft;
    revertReason;
    totalInstructions;
    constructor(reverted, output, gasLeft, revertReason, totalInstructions = 0){
        this.reverted = reverted;
        this.output = output;
        this.gasLeft = gasLeft;
        this.revertReason = revertReason;
        this.totalInstructions = totalInstructions;
    }
    toString() {
        let resultsStr = `reverted: ${this.reverted}, output: ${this.output.bestEffortReadAll(10)}${this.output.length() > 10 ? ' ...' : ''}, gasLeft: ${inspect(this.gasLeft)}, totalInstructions: ${this.totalInstructions}`;
        if (this.revertReason) {
            resultsStr += `, revertReason: ${this.revertReason.message}`;
        }
        return resultsStr;
    }
    finalize() {
        const revertReason = this.revertReason ? createSimulationError(this.revertReason, this.output.bestEffortReadAll()) : undefined;
        return new AvmFinalizedCallResult(this.reverted, this.output, Gas.from(this.gasLeft), revertReason, this.totalInstructions);
    }
}
/**
 * Similar to AvmContractCallResult, but with a SimulationError and standard Gas
 * which are useful for consumption external to core AVM simulation.
 */ export class AvmFinalizedCallResult {
    reverted;
    output;
    gasLeft;
    revertReason;
    totalInstructions;
    constructor(reverted, output, gasLeft, revertReason, totalInstructions = 0){
        this.reverted = reverted;
        this.output = output;
        this.gasLeft = gasLeft;
        this.revertReason = revertReason;
        this.totalInstructions = totalInstructions;
    }
    toString() {
        let resultsStr = `reverted: ${this.reverted}, output: ${this.output.bestEffortReadAll(10)}${this.output.length() > 10 ? ' ...' : ''}, gasLeft: ${inspect(this.gasLeft)}, totalInstructions: ${this.totalInstructions}`;
        if (this.revertReason) {
            resultsStr += `, revertReason: ${this.revertReason.message}`;
        }
        return resultsStr;
    }
}
