// @aztec/accounts/dest/stub/account_contract.js
import { AuthWitness } from '@aztec/stdlib/auth-witness';
import { DefaultAccountContract } from '../defaults/account_contract.js';
/**
 * An Account contract that does not perform any authentication of authwits
 * (`is_valid` always returns `true`)
 * It is used to capture authorization data during simulations
 */ export class StubBaseAccountContract extends DefaultAccountContract {
    constructor(){
        super();
    }
    getInitializationFunctionAndArgs() {
        return Promise.resolve({
            constructorName: 'constructor',
            constructorArgs: []
        });
    }
    getAuthWitnessProvider(_address) {
        return new StubAuthWitnessProvider();
    }
}
/** Creates auth witnesses using Stub signatures. */ export class StubAuthWitnessProvider {
    constructor(){}
    /**
   * A fake account always returns an empty authwitness, since it doesn't
   * perform any verification whatsoever
   * @param messageHash - The outer hash of the message for which the auth witness is created
   */ createAuthWit(messageHash) {
        return Promise.resolve(new AuthWitness(messageHash, []));
    }
}
