// @aztec/aztec.js/src/api/wallet.ts
export {
  type Aliased,
  type SimulateOptions,
  type SimulateUtilityOptions,
  type ProfileOptions,
  type SendOptions,
  type BatchableMethods,
  type BatchedMethod,
  type BatchedMethodResult,
  type BatchedMethodResultWrapper,
  type BatchResults,
  type Wallet,
  type PrivateEvent,
  type PrivateEventFilter,
  type PublicEvent,
  type PublicEventFilter,
  type ContractMetadata,
  type ContractClassMetadata,
  AppCapabilitiesSchema,
  WalletCapabilitiesSchema,
  ExecutionPayloadSchema,
  GasSettingsOptionSchema,
  WalletSimulationFeeOptionSchema,
  WaitOptsSchema,
  SendOptionsSchema,
  SimulateOptionsSchema,
  ProfileOptionsSchema,
  MessageHashOrIntentSchema,
  BatchedMethodSchema,
  EventMetadataDefinitionSchema,
  PrivateEventSchema,
  PrivateEventFilterSchema,
  PublicEventSchema,
  PublicEventFilterSchema,
  ContractClassMetadataSchema,
  ContractMetadataSchema,
  WalletSchema,
  ContractFunctionPatternSchema,
  AccountsCapabilitySchema,
  GrantedAccountsCapabilitySchema,
  ContractsCapabilitySchema,
  GrantedContractsCapabilitySchema,
  ContractClassesCapabilitySchema,
  GrantedContractClassesCapabilitySchema,
  SimulationCapabilitySchema,
  GrantedSimulationCapabilitySchema,
  TransactionCapabilitySchema,
  GrantedTransactionCapabilitySchema,
  DataCapabilitySchema,
  GrantedDataCapabilitySchema,
  CapabilitySchema,
  GrantedCapabilitySchema,
} from '../wallet/wallet.js';

export {
  type AppCapabilities,
  type WalletCapabilities,
  CAPABILITY_VERSION,
  type Capability,
  type GrantedCapability,
  type ContractFunctionPattern,
  type AccountsCapability,
  type GrantedAccountsCapability,
  type ContractsCapability,
  type GrantedContractsCapability,
  type ContractClassesCapability,
  type GrantedContractClassesCapability,
  type SimulationCapability,
  type GrantedSimulationCapability,
  type TransactionCapability,
  type GrantedTransactionCapability,
  type DataCapability,
  type GrantedDataCapability,
} from '../wallet/capabilities.js';

export { AccountManager } from '../wallet/account_manager.js';

export { type DeployAccountOptions, DeployAccountMethod } from '../wallet/deploy_account_method.js';
