// @aztec/validator-ha-signer/dest/db/types.js
/**
 * Type of validator duty being performed
 */ export var DutyType = /*#__PURE__*/ function(DutyType) {
    DutyType["BLOCK_PROPOSAL"] = "BLOCK_PROPOSAL";
    DutyType["CHECKPOINT_PROPOSAL"] = "CHECKPOINT_PROPOSAL";
    DutyType["ATTESTATION"] = "ATTESTATION";
    DutyType["ATTESTATIONS_AND_SIGNERS"] = "ATTESTATIONS_AND_SIGNERS";
    DutyType["GOVERNANCE_VOTE"] = "GOVERNANCE_VOTE";
    DutyType["SLASHING_VOTE"] = "SLASHING_VOTE";
    DutyType["AUTH_REQUEST"] = "AUTH_REQUEST";
    DutyType["TXS"] = "TXS";
    return DutyType;
}({});
/**
 * Status of a duty in the database
 */ export var DutyStatus = /*#__PURE__*/ function(DutyStatus) {
    DutyStatus["SIGNING"] = "signing";
    DutyStatus["SIGNED"] = "signed";
    return DutyStatus;
}({});
/**
 * Validates and normalizes the block index for a duty.
 * - BLOCK_PROPOSAL: validates blockIndexWithinCheckpoint is provided and >= 0
 * - Other duty types: always returns -1
 *
 * @throws Error if BLOCK_PROPOSAL is missing blockIndexWithinCheckpoint or has invalid value
 */ export function normalizeBlockIndex(dutyType, blockIndexWithinCheckpoint) {
    if (dutyType === "BLOCK_PROPOSAL") {
        if (blockIndexWithinCheckpoint === undefined) {
            throw new Error('BLOCK_PROPOSAL duties require blockIndexWithinCheckpoint to be specified');
        }
        if (blockIndexWithinCheckpoint < 0) {
            throw new Error(`BLOCK_PROPOSAL duties require blockIndexWithinCheckpoint >= 0, got ${blockIndexWithinCheckpoint}`);
        }
        return blockIndexWithinCheckpoint;
    }
    // For all other duty types, always use -1
    return -1;
}
/**
 * Gets the block index from a DutyIdentifier.
 * - BLOCK_PROPOSAL: returns the blockIndexWithinCheckpoint
 * - Other duty types: returns -1
 */ export function getBlockIndexFromDutyIdentifier(duty) {
    if (duty.dutyType === "BLOCK_PROPOSAL") {
        return duty.blockIndexWithinCheckpoint;
    }
    return -1;
}
