// @aztec/simulator/dest/public/contracts_db_checkpoint.js
export class ContractsDbCheckpoint {
    instances = new Map();
    classes = new Map();
    bytecodeCommitments = new Map();
    addInstance(address, instance) {
        this.instances.set(address.toString(), instance);
    }
    addClass(classId, contractClass) {
        this.classes.set(classId.toString(), contractClass);
    }
    addBytecodeCommitment(classId, commitment) {
        this.bytecodeCommitments.set(classId.toString(), commitment);
    }
    getInstance(address) {
        return this.instances.get(address.toString());
    }
    getClass(classId) {
        return this.classes.get(classId.toString());
    }
    getBytecodeCommitment(classId) {
        return this.bytecodeCommitments.get(classId.toString());
    }
    deepCopy() {
        const copy = new ContractsDbCheckpoint();
        this.instances.forEach((value, key)=>copy.instances.set(key, value));
        this.classes.forEach((value, key)=>copy.classes.set(key, value));
        this.bytecodeCommitments.forEach((value, key)=>copy.bytecodeCommitments.set(key, value));
        return copy;
    }
}
