// @aztec/simulator/dest/public/test_executor_metrics.js
import { sum } from '@aztec/foundation/collection';
import { createLogger } from '@aztec/foundation/log';
import { Timer } from '@aztec/foundation/timer';
import { strict as assert } from 'assert';
const NUM_SPACES = 4;
const H1 = '# ';
const H2 = '\n## ';
const INDENT = ' '.repeat(NUM_SPACES);
const INDENT0 = '- ';
const INDENT1 = INDENT + '- ';
const INDENT2 = INDENT + INDENT + '- ';
const H_LINE = '\n---------------------------------------------------------------------';
export var PublicTxMetricsFilter = /*#__PURE__*/ function(PublicTxMetricsFilter) {
    PublicTxMetricsFilter[PublicTxMetricsFilter["ALL"] = 0] = "ALL";
    PublicTxMetricsFilter[PublicTxMetricsFilter["TOTALS"] = 1] = "TOTALS";
    PublicTxMetricsFilter[PublicTxMetricsFilter["DURATIONS"] = 2] = "DURATIONS";
    PublicTxMetricsFilter[PublicTxMetricsFilter["INSTRUCTIONS"] = 3] = "INSTRUCTIONS";
    PublicTxMetricsFilter[PublicTxMetricsFilter["PROVING"] = 4] = "PROVING";
    return PublicTxMetricsFilter;
}({});
function createEmptyTxMetrics() {
    return {
        // TS simulation
        totalDurationMs: 0,
        manaUsed: 0,
        totalInstructionsExecuted: 0,
        nonRevertiblePrivateInsertionsUs: undefined,
        revertiblePrivateInsertionsUs: undefined,
        enqueuedCalls: [],
        revertedCode: undefined,
        // Proving
        proverSimulationStepMs: undefined,
        proverProvingStepMs: undefined,
        proverTraceGenerationStepMs: undefined,
        // Proving (detail)
        traceGenerationInteractionsMs: undefined,
        traceGenerationTracesMs: undefined,
        provingSumcheckMs: undefined,
        provingPcsMs: undefined,
        provingLogDerivativeInverseMs: undefined,
        provingLogDerivativeInverseCommitmentsMs: undefined,
        provingWireCommitmentsMs: undefined
    };
}
export class TestExecutorMetrics {
    logger;
    // tx label -> tx metrics
    txMetrics = new Map();
    currentTxLabel;
    txTimer;
    constructor(bindings){
        this.logger = createLogger(`simulator:test_executor_metrics`, bindings);
    }
    startRecordingTxSimulation(txLabel) {
        assert(!this.currentTxLabel, 'Cannot start recording tx simulation when another is live');
        assert(!this.txMetrics.has(txLabel), 'Cannot start recording metrics for tx with duplicate label');
        this.txMetrics.set(txLabel, createEmptyTxMetrics());
        this.currentTxLabel = txLabel;
        this.txTimer = new Timer();
    }
    stopRecordingTxSimulation(txLabel, gasUsed, revertedCode) {
        assert(this.currentTxLabel === txLabel, 'Cannot stop recording metrics for tx when another is live');
        const txMetrics = this.txMetrics.get(txLabel);
        // total duration of tx
        txMetrics.totalDurationMs = this.txTimer.ms();
        this.logger.debug(`Public TX simulation of ${txLabel} took ${txMetrics.totalDurationMs}ms`);
        // add manaUsed across all enqueued calls
        txMetrics.manaUsed = gasUsed?.publicGas.l2Gas;
        // add totalInstructionsExecuted across all enqueued calls
        txMetrics.totalInstructionsExecuted = sum(txMetrics.enqueuedCalls.map((call)=>call.totalInstructionsExecuted));
        txMetrics.revertedCode = revertedCode;
        this.currentTxLabel = undefined;
    }
    recordEnqueuedCallSimulation(fnName, durationMs, manaUsed, totalInstructionsExecuted) {
        this.#recordEnqueuedCallSimulation(fnName, durationMs, manaUsed, totalInstructionsExecuted, false);
    }
    recordEnqueuedCallSimulationFailure(fnName, durationMs, manaUsed, totalInstructionsExecuted) {
        this.#recordEnqueuedCallSimulation(fnName, durationMs, manaUsed, totalInstructionsExecuted, true);
    }
    #recordEnqueuedCallSimulation(fnName, durationMs, manaUsed, totalInstructionsExecuted, reverted) {
        assert(this.currentTxLabel, 'Cannot record enqueued call simulation when no tx is live');
        const txMetrics = this.txMetrics.get(this.currentTxLabel);
        txMetrics.enqueuedCalls.push({
            fnName,
            durationMs,
            manaUsed,
            totalInstructionsExecuted: totalInstructionsExecuted,
            reverted
        });
    }
    recordPrivateEffectsInsertion(durationUs, type) {
        assert(this.currentTxLabel, 'Cannot record private effects insertion when no tx is live');
        const txMetrics = this.txMetrics.get(this.currentTxLabel);
        if (type === 'revertible') {
            assert(txMetrics.revertiblePrivateInsertionsUs === undefined, 'Cannot RE-record revertible insertions of private effects');
            txMetrics.revertiblePrivateInsertionsUs = durationUs;
        } else {
            assert(txMetrics.nonRevertiblePrivateInsertionsUs === undefined, 'Cannot RE-record non-revertible insertions of private effects');
            txMetrics.nonRevertiblePrivateInsertionsUs = durationUs;
        }
    }
    recordProverMetrics(txLabel, metrics) {
        if (!this.txMetrics.has(txLabel)) {
            this.txMetrics.set(txLabel, createEmptyTxMetrics());
        }
        const txMetrics = this.txMetrics.get(txLabel);
        for (const [key, value] of Object.entries(metrics)){
            if (key in txMetrics) {
                txMetrics[key] = value;
            }
        }
    }
    prettyPrint(filter = 0) {
        this.logger.info(this.toPrettyString(filter));
    }
    toPrettyString(filter = 0) {
        let pretty = '';
        //pretty += H_LINE + '\n';
        pretty += `${H1}Public TX Simulation Metrics (${PublicTxMetricsFilter[filter]})\n`;
        for (const [txLabel, txMetrics] of this.txMetrics.entries()){
            //pretty += H_LINE + '\n';
            pretty += `${H2}TX Label: ${txLabel}\n`;
            if (filter == 2 || filter === 1 || filter === 0) {
                pretty += `${INDENT0}Total duration: ${fmtNum(txMetrics.totalDurationMs, 'ms')}\n`;
            }
            if (filter === 1 || filter === 0) {
                const manaUsed = txMetrics.manaUsed || 0;
                pretty += `${INDENT0}Total mana used: ${fmtNum(manaUsed)}\n`;
                const manaPerSecond = Math.round(manaUsed * 1000 / txMetrics.totalDurationMs);
                pretty += `${INDENT0}Mana per second: ${fmtNum(manaPerSecond)}\n`;
            }
            if (filter === 3 || filter === 1 || filter === 0) {
                pretty += `${INDENT0}Total instructions executed: ${fmtNum(txMetrics.totalInstructionsExecuted)}\n`;
            }
            if (filter === 2 || filter === 0) {
                pretty += `${INDENT0}Private insertions:\n`;
                pretty += `${INDENT1}Non-revertible: ${fmtNum(txMetrics.nonRevertiblePrivateInsertionsUs / 1_000, 'ms')}\n`;
                pretty += `${INDENT1}Revertible: ${fmtNum(txMetrics.revertiblePrivateInsertionsUs / 1_000, 'ms')}\n`;
            }
            if (filter === 4 || filter === 0) {
                let provingPretty = '';
                if (txMetrics.proverSimulationStepMs !== undefined) {
                    provingPretty += `${INDENT1}Simulation (all): ${fmtNum(txMetrics.proverSimulationStepMs, 'ms')}\n`;
                }
                if (txMetrics.proverProvingStepMs !== undefined) {
                    provingPretty += `${INDENT1}Proving (all): ${fmtNum(txMetrics.proverProvingStepMs, 'ms')}\n`;
                }
                if (txMetrics.proverTraceGenerationStepMs !== undefined) {
                    provingPretty += `${INDENT1}Trace generation (all): ${fmtNum(txMetrics.proverTraceGenerationStepMs, 'ms')}\n`;
                }
                if (txMetrics.traceGenerationInteractionsMs !== undefined) {
                    provingPretty += `${INDENT1}Trace generation interactions: ${fmtNum(txMetrics.traceGenerationInteractionsMs, 'ms')}\n`;
                }
                if (txMetrics.traceGenerationTracesMs !== undefined) {
                    provingPretty += `${INDENT1}Trace generation traces: ${fmtNum(txMetrics.traceGenerationTracesMs, 'ms')}\n`;
                }
                if (txMetrics.provingSumcheckMs !== undefined) {
                    provingPretty += `${INDENT1}Sumcheck: ${fmtNum(txMetrics.provingSumcheckMs, 'ms')}\n`;
                }
                if (txMetrics.provingPcsMs !== undefined) {
                    provingPretty += `${INDENT1}PCS: ${fmtNum(txMetrics.provingPcsMs, 'ms')}\n`;
                }
                if (txMetrics.provingLogDerivativeInverseMs !== undefined) {
                    provingPretty += `${INDENT1}Log derivative inverse: ${fmtNum(txMetrics.provingLogDerivativeInverseMs, 'ms')}\n`;
                }
                if (txMetrics.provingLogDerivativeInverseCommitmentsMs !== undefined) {
                    provingPretty += `${INDENT1}Log derivative inverse commitments: ${fmtNum(txMetrics.provingLogDerivativeInverseCommitmentsMs, 'ms')}\n`;
                }
                if (txMetrics.provingWireCommitmentsMs !== undefined) {
                    provingPretty += `${INDENT1}Wire commitments: ${fmtNum(txMetrics.provingWireCommitmentsMs, 'ms')}\n`;
                }
                if (provingPretty.length > 0) {
                    pretty += `${INDENT0}Proving:\n${provingPretty}`;
                }
            }
            if (filter !== 1) {
                // totals exclude enqueued calls
                pretty += this.#enqueuedCallsToPrettyString(txMetrics, filter);
            }
            if (txMetrics.revertedCode !== undefined && !txMetrics.revertedCode.isOK()) {
                pretty += `${INDENT0}Reverted code: ${txMetrics.revertedCode?.getDescription()}\n`;
            }
            pretty += H_LINE + '\n';
        }
        return pretty;
    }
    #enqueuedCallsToPrettyString(txMetrics, filter) {
        let pretty = '';
        pretty += `${INDENT0}Enqueued public calls:\n`;
        for (const enqueuedCall of txMetrics.enqueuedCalls){
            pretty += `${INDENT1}**Fn: ${enqueuedCall.fnName}**\n`;
            if (filter === 2 || filter === 0) {
                pretty += `${INDENT2}Duration: ${fmtNum(enqueuedCall.durationMs, 'ms')}\n`;
            }
            if (filter === 0) {
                pretty += `${INDENT2}Mana used: ${fmtNum(enqueuedCall.manaUsed)}\n`;
                const manaPerSecond = Math.round(enqueuedCall.manaUsed * 1000 / enqueuedCall.durationMs);
                pretty += `${INDENT2}Mana per second: ${fmtNum(manaPerSecond)}\n`;
            }
            if (filter === 3 || filter === 0) {
                pretty += `${INDENT2}Instructions executed: ${fmtNum(enqueuedCall.totalInstructionsExecuted)}\n`;
            }
            if (enqueuedCall.reverted) {
                pretty += `${INDENT2}Reverted!\n`;
            }
        }
        return pretty;
    }
    toJSON(indent = 2) {
        return JSON.stringify(Object.fromEntries(this.txMetrics.entries()), null, indent);
    }
    toGithubActionBenchmarkJSON(indent = 2) {
        const metricsInfo = {
            totalInstructionsExecuted: {
                name: 'totalInstructionsExecuted',
                unit: '#instructions',
                category: 3
            },
            totalDurationMs: {
                name: 'totalDurationMs',
                unit: 'ms',
                category: 2
            },
            manaUsed: {
                name: 'manaUsed',
                unit: 'mana',
                category: 1
            },
            nonRevertiblePrivateInsertionsUs: {
                name: 'nonRevertiblePrivateInsertionsUs',
                unit: 'us',
                category: 2
            },
            revertiblePrivateInsertionsUs: {
                name: 'revertiblePrivateInsertionsUs',
                unit: 'us',
                category: 2
            },
            proverSimulationStepMs: {
                name: 'proverSimulationStepMs',
                unit: 'ms',
                category: 4
            },
            proverProvingStepMs: {
                name: 'proverProvingStepMs',
                unit: 'ms',
                category: 4
            },
            proverTraceGenerationStepMs: {
                name: 'proverTraceGenerationStepMs',
                unit: 'ms',
                category: 4
            },
            traceGenerationInteractionsMs: {
                name: 'traceGenerationInteractionsMs',
                unit: 'ms',
                category: 4
            },
            traceGenerationTracesMs: {
                name: 'traceGenerationTracesMs',
                unit: 'ms',
                category: 4
            },
            provingSumcheckMs: {
                name: 'provingSumcheckMs',
                unit: 'ms',
                category: 4
            },
            provingPcsMs: {
                name: 'provingPcsMs',
                unit: 'ms',
                category: 4
            },
            provingLogDerivativeInverseMs: {
                name: 'provingLogDerivativeInverseMs',
                unit: 'ms',
                category: 4
            },
            provingLogDerivativeInverseCommitmentsMs: {
                name: 'provingLogDerivativeInverseCommitmentsMs',
                unit: 'ms',
                category: 4
            },
            provingWireCommitmentsMs: {
                name: 'provingWireCommitmentsMs',
                unit: 'ms',
                category: 4
            }
        };
        const data = [];
        for (const [txLabel, txMetrics] of this.txMetrics.entries()){
            for (const [key, value] of Object.entries(txMetrics)){
                if (value !== undefined && key in metricsInfo) {
                    data.push({
                        name: `${txLabel}/${metricsInfo[key].name}`,
                        value: value,
                        unit: metricsInfo[key].unit
                    });
                }
            }
        }
        return JSON.stringify(data, null, indent);
    }
}
function fmtNum(num, unit) {
    return `\`${num.toLocaleString()}${unit ? ` ${unit}` : ''}\``;
}
