// @aztec/stdlib/dest/rollup/block_constant_data.js
import { Fr } from '@aztec/foundation/curves/bn254';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { AppendOnlyTreeSnapshot } from '../trees/append_only_tree_snapshot.js';
import { GlobalVariables } from '../tx/global_variables.js';
/**
 * Constants that are the same for the entire block.
 */ export class BlockConstantData {
    lastArchive;
    l1ToL2TreeSnapshot;
    vkTreeRoot;
    protocolContractsHash;
    globalVariables;
    proverId;
    constructor(/** Archive tree snapshot at the very beginning of the entire rollup. */ lastArchive, /**
     * L1-to-L2 message tree snapshot after this block lands.
     * For the first block in a checkpoint, this should be the snapshot after inserting the new l1-to-l2 message subtree
     * into the last l1-to-l2 tree snapshot in `last_archive`.
     * For subsequent blocks, this should match the snapshot of the previous block.
     */ l1ToL2TreeSnapshot, /** Root of the verification key tree. */ vkTreeRoot, /** Hash of the protocol contracts list. */ protocolContractsHash, /** Global variables for the block. */ globalVariables, /** Identifier of the prover. */ proverId){
        this.lastArchive = lastArchive;
        this.l1ToL2TreeSnapshot = l1ToL2TreeSnapshot;
        this.vkTreeRoot = vkTreeRoot;
        this.protocolContractsHash = protocolContractsHash;
        this.globalVariables = globalVariables;
        this.proverId = proverId;
    }
    static from(fields) {
        return new BlockConstantData(...BlockConstantData.getFields(fields));
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new BlockConstantData(reader.readObject(AppendOnlyTreeSnapshot), reader.readObject(AppendOnlyTreeSnapshot), Fr.fromBuffer(reader), Fr.fromBuffer(reader), reader.readObject(GlobalVariables), Fr.fromBuffer(reader));
    }
    static getFields(fields) {
        return [
            fields.lastArchive,
            fields.l1ToL2TreeSnapshot,
            fields.vkTreeRoot,
            fields.protocolContractsHash,
            fields.globalVariables,
            fields.proverId
        ];
    }
    static empty() {
        return new BlockConstantData(AppendOnlyTreeSnapshot.empty(), AppendOnlyTreeSnapshot.empty(), Fr.ZERO, Fr.ZERO, GlobalVariables.empty(), Fr.ZERO);
    }
    toBuffer() {
        return serializeToBuffer(...BlockConstantData.getFields(this));
    }
}
