// @aztec/stdlib/dest/rollup/epoch_constant_data.js
import { Fr } from '@aztec/foundation/curves/bn254';
import { BufferReader, serializeToBuffer, serializeToFields } from '@aztec/foundation/serialize';
/**
 * Constants that are the same for the entire epoch.
 */ export class EpochConstantData {
    chainId;
    version;
    vkTreeRoot;
    protocolContractsHash;
    proverId;
    constructor(/**
     * ChainId of the rollup.
     */ chainId, /**
     * Version of the rollup.
     */ version, /**
     * Root of the verification key tree.
     */ vkTreeRoot, /**
     * Hash of the protocol contracts list.
     */ protocolContractsHash, /**
     * Identifier of the prover of the epoch.
     */ proverId){
        this.chainId = chainId;
        this.version = version;
        this.vkTreeRoot = vkTreeRoot;
        this.protocolContractsHash = protocolContractsHash;
        this.proverId = proverId;
    }
    static from(fields) {
        return new EpochConstantData(...EpochConstantData.getFields(fields));
    }
    static getFields(fields) {
        return [
            fields.chainId,
            fields.version,
            fields.vkTreeRoot,
            fields.protocolContractsHash,
            fields.proverId
        ];
    }
    toFields() {
        return serializeToFields(...EpochConstantData.getFields(this));
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new EpochConstantData(Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader), Fr.fromBuffer(reader));
    }
    toBuffer() {
        return serializeToBuffer(...EpochConstantData.getFields(this));
    }
    static empty() {
        return new EpochConstantData(Fr.ZERO, Fr.ZERO, Fr.ZERO, Fr.ZERO, Fr.ZERO);
    }
}
