// @aztec/stdlib/dest/rollup/private_tx_base_rollup_private_inputs.js
import { bufferSchemaFor } from '@aztec/foundation/schemas';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { bufferToHex, hexToBuffer } from '@aztec/foundation/string';
import { PrivateToRollupKernelCircuitPublicInputs } from '../kernel/private_to_rollup_kernel_circuit_public_inputs.js';
import { ProofData } from '../proofs/proof_data.js';
import { PrivateBaseRollupHints } from './base_rollup_hints.js';
export class PrivateTxBaseRollupPrivateInputs {
    hidingKernelProofData;
    hints;
    constructor(hidingKernelProofData, hints){
        this.hidingKernelProofData = hidingKernelProofData;
        this.hints = hints;
    }
    static from(fields) {
        return new PrivateTxBaseRollupPrivateInputs(...PrivateTxBaseRollupPrivateInputs.getFields(fields));
    }
    static getFields(fields) {
        return [
            fields.hidingKernelProofData,
            fields.hints
        ];
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PrivateTxBaseRollupPrivateInputs(ProofData.fromBuffer(reader, PrivateToRollupKernelCircuitPublicInputs), reader.readObject(PrivateBaseRollupHints));
    }
    toBuffer() {
        return serializeToBuffer(...PrivateTxBaseRollupPrivateInputs.getFields(this));
    }
    static fromString(str) {
        return PrivateTxBaseRollupPrivateInputs.fromBuffer(hexToBuffer(str));
    }
    toString() {
        return bufferToHex(this.toBuffer());
    }
    /** Returns a buffer representation for JSON serialization. */ toJSON() {
        return this.toBuffer();
    }
    /** Creates an instance from a hex string. */ static get schema() {
        return bufferSchemaFor(PrivateTxBaseRollupPrivateInputs);
    }
}
