// @aztec/stdlib/dest/rollup/public_chonk_verifier_private_inputs.js
import { Fr } from '@aztec/foundation/curves/bn254';
import { bufferSchemaFor } from '@aztec/foundation/schemas';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { bufferToHex, hexToBuffer } from '@aztec/foundation/string';
import { PrivateToPublicKernelCircuitPublicInputs } from '../kernel/private_to_public_kernel_circuit_public_inputs.js';
import { ProofData } from '../proofs/proof_data.js';
// CHONK_VERIFIER: Recursively verifies Chonk (Client Honk) proofs in circuits
export class PublicChonkVerifierPrivateInputs {
    hidingKernelProofData;
    proverId;
    constructor(hidingKernelProofData, proverId){
        this.hidingKernelProofData = hidingKernelProofData;
        this.proverId = proverId;
    }
    static from(fields) {
        return new PublicChonkVerifierPrivateInputs(...PublicChonkVerifierPrivateInputs.getFields(fields));
    }
    static getFields(fields) {
        return [
            fields.hidingKernelProofData,
            fields.proverId
        ];
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PublicChonkVerifierPrivateInputs(ProofData.fromBuffer(reader, PrivateToPublicKernelCircuitPublicInputs), Fr.fromBuffer(reader));
    }
    toBuffer() {
        return serializeToBuffer(...PublicChonkVerifierPrivateInputs.getFields(this));
    }
    static fromString(str) {
        return PublicChonkVerifierPrivateInputs.fromBuffer(hexToBuffer(str));
    }
    toString() {
        return bufferToHex(this.toBuffer());
    }
    /** Returns a representation for JSON serialization. */ toJSON() {
        return this.toBuffer();
    }
    /** Creates an instance from a string. */ static get schema() {
        return bufferSchemaFor(PublicChonkVerifierPrivateInputs);
    }
}
