// @aztec/stdlib/dest/rollup/public_tx_base_rollup_private_inputs.js
import { bufferSchemaFor } from '@aztec/foundation/schemas';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { bufferToHex, hexToBuffer } from '@aztec/foundation/string';
import { AvmCircuitPublicInputs } from '../avm/avm_circuit_public_inputs.js';
import { ProofData, ProofDataForFixedVk } from '../proofs/proof_data.js';
import { PublicBaseRollupHints } from './base_rollup_hints.js';
import { PublicChonkVerifierPublicInputs } from './public_chonk_verifier_public_inputs.js';
export class PublicTxBaseRollupPrivateInputs {
    publicChonkVerifierProofData;
    avmProofData;
    hints;
    constructor(publicChonkVerifierProofData, avmProofData, hints){
        this.publicChonkVerifierProofData = publicChonkVerifierProofData;
        this.avmProofData = avmProofData;
        this.hints = hints;
    }
    static from(fields) {
        return new PublicTxBaseRollupPrivateInputs(...PublicTxBaseRollupPrivateInputs.getFields(fields));
    }
    static getFields(fields) {
        return [
            fields.publicChonkVerifierProofData,
            fields.avmProofData,
            fields.hints
        ];
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        return new PublicTxBaseRollupPrivateInputs(ProofData.fromBuffer(reader, PublicChonkVerifierPublicInputs), ProofDataForFixedVk.fromBuffer(reader, AvmCircuitPublicInputs), reader.readObject(PublicBaseRollupHints));
    }
    toBuffer() {
        return serializeToBuffer(...PublicTxBaseRollupPrivateInputs.getFields(this));
    }
    static fromString(str) {
        return PublicTxBaseRollupPrivateInputs.fromBuffer(hexToBuffer(str));
    }
    toString() {
        return bufferToHex(this.toBuffer());
    }
    /** Returns a representation for JSON serialization. */ toJSON() {
        return this.toBuffer();
    }
    /** Creates an instance from a string. */ static get schema() {
        return bufferSchemaFor(PublicTxBaseRollupPrivateInputs);
    }
}
