// @aztec/simulator/dest/public/public_tx_simulator/factories.js
import { PublicSimulatorConfig } from '@aztec/stdlib/avm';
import { TelemetryCppPublicTxSimulator } from './cpp_public_tx_simulator.js';
import { DumpingCppPublicTxSimulator } from './dumping_cpp_public_tx_simulator.js';
/**
 * Creates a public tx simulator for block building.
 * Uses DumpingCppPublicTxSimulator if DUMP_AVM_INPUTS_TO_DIR env var is set (for CI/testing avm circuit),
 * otherwise uses TelemetryCppPublicTxSimulator (for production).
 */ export function createPublicTxSimulatorForBlockBuilding(merkleTree, contractsDB, globalVariables, telemetryClient, bindings) {
    const config = PublicSimulatorConfig.from({
        skipFeeEnforcement: false,
        collectDebugLogs: false,
        collectHints: false,
        collectPublicInputs: false,
        collectStatistics: false,
        collectCallMetadata: false
    });
    const dumpDir = process.env.DUMP_AVM_INPUTS_TO_DIR;
    if (dumpDir) {
        // must collect hints and PIs for dumping
        const dumpingConfig = {
            ...config,
            collectHints: true,
            collectPublicInputs: true
        };
        return new DumpingCppPublicTxSimulator(merkleTree, contractsDB, globalVariables, dumpingConfig, dumpDir, bindings);
    }
    return new TelemetryCppPublicTxSimulator(merkleTree, contractsDB, globalVariables, telemetryClient, config, bindings);
}
