// @aztec/aztec.js/dest/contract/protocol_contracts/auth-registry.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { ProtocolContractAddress } from '@aztec/protocol-contracts';
import { FunctionType } from '@aztec/stdlib/abi';
import { ContractBase } from '../contract_base.js';
const AuthRegistryContractArtifact = {
    name: 'AuthRegistry',
    functions: [
        {
            ...{
                functionType: FunctionType.PUBLIC,
                name: 'public_dispatch',
                isOnlySelf: false,
                isStatic: false,
                isInitializer: false,
                parameters: [
                    {
                        name: 'selector',
                        type: {
                            kind: 'field'
                        },
                        visibility: 'private'
                    }
                ],
                returnTypes: [],
                errorTypes: {
                    '361444214588792908': {
                        error_kind: 'string',
                        string: 'attempt to multiply with overflow'
                    },
                    '1998584279744703196': {
                        error_kind: 'string',
                        string: 'attempt to subtract with overflow'
                    },
                    '5525725861591553972': {
                        error_kind: 'string',
                        string: 'rejecting all'
                    },
                    '11691217184967371519': {
                        error_kind: 'string',
                        string: 'Function is_reject_all can only be called statically'
                    },
                    '12511970388699677811': {
                        error_kind: 'fmtstring',
                        length: 27,
                        item_types: [
                            {
                                kind: 'field'
                            }
                        ]
                    },
                    '13455385521185560676': {
                        error_kind: 'string',
                        string: 'Storage slot 0 not allowed. Storage slots must start from 1.'
                    },
                    '13929691561421487185': {
                        error_kind: 'string',
                        string: 'Function _set_authorized can only be called by the same contract'
                    },
                    '14990209321349310352': {
                        error_kind: 'string',
                        string: 'attempt to add with overflow'
                    },
                    '15764276373176857197': {
                        error_kind: 'string',
                        string: 'Stack too deep'
                    },
                    '16431471497789672479': {
                        error_kind: 'string',
                        string: 'Index out of bounds'
                    },
                    '17089945683942782951': {
                        error_kind: 'string',
                        string: 'unauthorized'
                    },
                    '17312933876399746100': {
                        error_kind: 'string',
                        string: 'Function is_consumable can only be called statically'
                    }
                }
            },
            bytecode: Buffer.from([]),
            debugSymbols: ''
        },
        {
            ...{
                functionType: FunctionType.PRIVATE,
                name: 'set_authorized_private',
                isOnlySelf: false,
                isStatic: false,
                isInitializer: false,
                parameters: [
                    {
                        name: 'approver',
                        type: {
                            kind: 'struct',
                            fields: [
                                {
                                    name: 'inner',
                                    type: {
                                        kind: 'field'
                                    }
                                }
                            ],
                            path: 'aztec::protocol_types::address::aztec_address::AztecAddress'
                        },
                        visibility: 'private'
                    },
                    {
                        name: 'message_hash',
                        type: {
                            kind: 'field'
                        },
                        visibility: 'private'
                    },
                    {
                        name: 'authorize',
                        type: {
                            kind: 'boolean'
                        },
                        visibility: 'private'
                    }
                ],
                returnTypes: [],
                errorTypes: {
                    '7555607922535724711': {
                        error_kind: 'string',
                        string: 'Preimage mismatch'
                    },
                    '9894212961085021188': {
                        error_kind: 'string',
                        string: 'Message not authorized by account'
                    },
                    '14990209321349310352': {
                        error_kind: 'string',
                        string: 'attempt to add with overflow'
                    },
                    '15764276373176857197': {
                        error_kind: 'string',
                        string: 'Stack too deep'
                    }
                }
            },
            bytecode: Buffer.from([]),
            debugSymbols: ''
        },
        {
            ...{
                functionType: FunctionType.UTILITY,
                name: 'utility_is_consumable',
                isOnlySelf: false,
                isStatic: false,
                isInitializer: false,
                parameters: [
                    {
                        name: 'on_behalf_of',
                        type: {
                            kind: 'struct',
                            fields: [
                                {
                                    name: 'inner',
                                    type: {
                                        kind: 'field'
                                    }
                                }
                            ],
                            path: 'aztec::protocol_types::address::aztec_address::AztecAddress'
                        },
                        visibility: 'private'
                    },
                    {
                        name: 'message_hash',
                        type: {
                            kind: 'field'
                        },
                        visibility: 'private'
                    }
                ],
                returnTypes: [
                    {
                        kind: 'boolean'
                    }
                ],
                errorTypes: {
                    '361444214588792908': {
                        error_kind: 'string',
                        string: 'attempt to multiply with overflow'
                    },
                    '1998584279744703196': {
                        error_kind: 'string',
                        string: 'attempt to subtract with overflow'
                    },
                    '13455385521185560676': {
                        error_kind: 'string',
                        string: 'Storage slot 0 not allowed. Storage slots must start from 1.'
                    },
                    '14990209321349310352': {
                        error_kind: 'string',
                        string: 'attempt to add with overflow'
                    },
                    '15764276373176857197': {
                        error_kind: 'string',
                        string: 'Stack too deep'
                    },
                    '16431471497789672479': {
                        error_kind: 'string',
                        string: 'Index out of bounds'
                    }
                }
            },
            bytecode: Buffer.from([]),
            debugSymbols: ''
        }
    ],
    nonDispatchPublicFunctions: [
        {
            functionType: FunctionType.PUBLIC,
            name: '_set_authorized',
            isOnlySelf: true,
            isStatic: false,
            isInitializer: false,
            parameters: [
                {
                    name: 'approver',
                    type: {
                        kind: 'struct',
                        fields: [
                            {
                                name: 'inner',
                                type: {
                                    kind: 'field'
                                }
                            }
                        ],
                        path: 'aztec::protocol_types::address::aztec_address::AztecAddress'
                    },
                    visibility: 'private'
                },
                {
                    name: 'message_hash',
                    type: {
                        kind: 'field'
                    },
                    visibility: 'private'
                },
                {
                    name: 'authorize',
                    type: {
                        kind: 'boolean'
                    },
                    visibility: 'private'
                }
            ],
            returnTypes: [],
            errorTypes: {
                '361444214588792908': {
                    error_kind: 'string',
                    string: 'attempt to multiply with overflow'
                },
                '1998584279744703196': {
                    error_kind: 'string',
                    string: 'attempt to subtract with overflow'
                },
                '13455385521185560676': {
                    error_kind: 'string',
                    string: 'Storage slot 0 not allowed. Storage slots must start from 1.'
                },
                '13929691561421487185': {
                    error_kind: 'string',
                    string: 'Function _set_authorized can only be called by the same contract'
                },
                '14990209321349310352': {
                    error_kind: 'string',
                    string: 'attempt to add with overflow'
                },
                '15764276373176857197': {
                    error_kind: 'string',
                    string: 'Stack too deep'
                },
                '16431471497789672479': {
                    error_kind: 'string',
                    string: 'Index out of bounds'
                }
            }
        },
        {
            functionType: FunctionType.PUBLIC,
            name: 'consume',
            isOnlySelf: false,
            isStatic: false,
            isInitializer: false,
            parameters: [
                {
                    name: 'on_behalf_of',
                    type: {
                        kind: 'struct',
                        fields: [
                            {
                                name: 'inner',
                                type: {
                                    kind: 'field'
                                }
                            }
                        ],
                        path: 'aztec::protocol_types::address::aztec_address::AztecAddress'
                    },
                    visibility: 'private'
                },
                {
                    name: 'inner_hash',
                    type: {
                        kind: 'field'
                    },
                    visibility: 'private'
                }
            ],
            returnTypes: [
                {
                    kind: 'field'
                }
            ],
            errorTypes: {
                '361444214588792908': {
                    error_kind: 'string',
                    string: 'attempt to multiply with overflow'
                },
                '1998584279744703196': {
                    error_kind: 'string',
                    string: 'attempt to subtract with overflow'
                },
                '5525725861591553972': {
                    error_kind: 'string',
                    string: 'rejecting all'
                },
                '13455385521185560676': {
                    error_kind: 'string',
                    string: 'Storage slot 0 not allowed. Storage slots must start from 1.'
                },
                '14990209321349310352': {
                    error_kind: 'string',
                    string: 'attempt to add with overflow'
                },
                '15764276373176857197': {
                    error_kind: 'string',
                    string: 'Stack too deep'
                },
                '16431471497789672479': {
                    error_kind: 'string',
                    string: 'Index out of bounds'
                },
                '17089945683942782951': {
                    error_kind: 'string',
                    string: 'unauthorized'
                }
            }
        },
        {
            functionType: FunctionType.PUBLIC,
            name: 'is_consumable',
            isOnlySelf: false,
            isStatic: true,
            isInitializer: false,
            parameters: [
                {
                    name: 'on_behalf_of',
                    type: {
                        kind: 'struct',
                        fields: [
                            {
                                name: 'inner',
                                type: {
                                    kind: 'field'
                                }
                            }
                        ],
                        path: 'aztec::protocol_types::address::aztec_address::AztecAddress'
                    },
                    visibility: 'private'
                },
                {
                    name: 'message_hash',
                    type: {
                        kind: 'field'
                    },
                    visibility: 'private'
                }
            ],
            returnTypes: [
                {
                    kind: 'boolean'
                }
            ],
            errorTypes: {
                '361444214588792908': {
                    error_kind: 'string',
                    string: 'attempt to multiply with overflow'
                },
                '1998584279744703196': {
                    error_kind: 'string',
                    string: 'attempt to subtract with overflow'
                },
                '13455385521185560676': {
                    error_kind: 'string',
                    string: 'Storage slot 0 not allowed. Storage slots must start from 1.'
                },
                '14990209321349310352': {
                    error_kind: 'string',
                    string: 'attempt to add with overflow'
                },
                '15764276373176857197': {
                    error_kind: 'string',
                    string: 'Stack too deep'
                },
                '16431471497789672479': {
                    error_kind: 'string',
                    string: 'Index out of bounds'
                },
                '17312933876399746100': {
                    error_kind: 'string',
                    string: 'Function is_consumable can only be called statically'
                }
            }
        },
        {
            functionType: FunctionType.PUBLIC,
            name: 'is_reject_all',
            isOnlySelf: false,
            isStatic: true,
            isInitializer: false,
            parameters: [
                {
                    name: 'on_behalf_of',
                    type: {
                        kind: 'struct',
                        fields: [
                            {
                                name: 'inner',
                                type: {
                                    kind: 'field'
                                }
                            }
                        ],
                        path: 'aztec::protocol_types::address::aztec_address::AztecAddress'
                    },
                    visibility: 'private'
                }
            ],
            returnTypes: [
                {
                    kind: 'boolean'
                }
            ],
            errorTypes: {
                '11691217184967371519': {
                    error_kind: 'string',
                    string: 'Function is_reject_all can only be called statically'
                },
                '13455385521185560676': {
                    error_kind: 'string',
                    string: 'Storage slot 0 not allowed. Storage slots must start from 1.'
                },
                '15764276373176857197': {
                    error_kind: 'string',
                    string: 'Stack too deep'
                }
            }
        },
        {
            functionType: FunctionType.PUBLIC,
            name: 'set_authorized',
            isOnlySelf: false,
            isStatic: false,
            isInitializer: false,
            parameters: [
                {
                    name: 'message_hash',
                    type: {
                        kind: 'field'
                    },
                    visibility: 'private'
                },
                {
                    name: 'authorize',
                    type: {
                        kind: 'boolean'
                    },
                    visibility: 'private'
                }
            ],
            returnTypes: [],
            errorTypes: {
                '361444214588792908': {
                    error_kind: 'string',
                    string: 'attempt to multiply with overflow'
                },
                '1998584279744703196': {
                    error_kind: 'string',
                    string: 'attempt to subtract with overflow'
                },
                '13455385521185560676': {
                    error_kind: 'string',
                    string: 'Storage slot 0 not allowed. Storage slots must start from 1.'
                },
                '14990209321349310352': {
                    error_kind: 'string',
                    string: 'attempt to add with overflow'
                },
                '15764276373176857197': {
                    error_kind: 'string',
                    string: 'Stack too deep'
                },
                '16431471497789672479': {
                    error_kind: 'string',
                    string: 'Index out of bounds'
                }
            }
        },
        {
            functionType: FunctionType.PUBLIC,
            name: 'set_reject_all',
            isOnlySelf: false,
            isStatic: false,
            isInitializer: false,
            parameters: [
                {
                    name: 'reject',
                    type: {
                        kind: 'boolean'
                    },
                    visibility: 'private'
                }
            ],
            returnTypes: [],
            errorTypes: {
                '13455385521185560676': {
                    error_kind: 'string',
                    string: 'Storage slot 0 not allowed. Storage slots must start from 1.'
                },
                '15764276373176857197': {
                    error_kind: 'string',
                    string: 'Stack too deep'
                }
            }
        }
    ],
    outputs: {
        structs: {},
        globals: {}
    },
    storageLayout: {},
    fileMap: {}
};
export class AuthRegistryContract extends ContractBase {
    constructor(wallet){
        super(ProtocolContractAddress.AuthRegistry, AuthRegistryContractArtifact, wallet);
    }
    static at(wallet) {
        return new AuthRegistryContract(wallet);
    }
}
