// @aztec/aztec.js/dest/contract/protocol_contracts/contract-instance-registry.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { ProtocolContractAddress } from '@aztec/protocol-contracts';
import { FunctionType } from '@aztec/stdlib/abi';
import { ContractBase } from '../contract_base.js';
const ContractInstanceRegistryContractArtifact = {
    name: 'ContractInstanceRegistry',
    functions: [
        {
            ...{
                functionType: FunctionType.PUBLIC,
                name: 'public_dispatch',
                isOnlySelf: false,
                isStatic: false,
                isInitializer: false,
                parameters: [
                    {
                        name: 'selector',
                        type: {
                            kind: 'field'
                        },
                        visibility: 'private'
                    }
                ],
                returnTypes: [],
                errorTypes: {
                    '361444214588792908': {
                        error_kind: 'string',
                        string: 'attempt to multiply with overflow'
                    },
                    '1998584279744703196': {
                        error_kind: 'string',
                        string: 'attempt to subtract with overflow'
                    },
                    '5755400808989454547': {
                        error_kind: 'string',
                        string: 'Function get_update_delay can only be called statically'
                    },
                    '6804164082532189961': {
                        error_kind: 'string',
                        string: 'New update delay is too low'
                    },
                    '12511970388699677811': {
                        error_kind: 'fmtstring',
                        length: 27,
                        item_types: [
                            {
                                kind: 'field'
                            }
                        ]
                    },
                    '12579274401768182412': {
                        error_kind: 'string',
                        string: 'New contract class is not registered'
                    },
                    '13455385521185560676': {
                        error_kind: 'string',
                        string: 'Storage slot 0 not allowed. Storage slots must start from 1.'
                    },
                    '14990209321349310352': {
                        error_kind: 'string',
                        string: 'attempt to add with overflow'
                    },
                    '15353471697975175405': {
                        error_kind: 'string',
                        string: 'msg.sender is not deployed'
                    },
                    '15764276373176857197': {
                        error_kind: 'string',
                        string: 'Stack too deep'
                    },
                    '16431471497789672479': {
                        error_kind: 'string',
                        string: 'Index out of bounds'
                    }
                }
            },
            bytecode: Buffer.from([]),
            debugSymbols: ''
        },
        {
            ...{
                functionType: FunctionType.PRIVATE,
                name: 'publish_for_public_execution',
                isOnlySelf: false,
                isStatic: false,
                isInitializer: false,
                parameters: [
                    {
                        name: 'salt',
                        type: {
                            kind: 'field'
                        },
                        visibility: 'private'
                    },
                    {
                        name: 'contract_class_id',
                        type: {
                            kind: 'struct',
                            fields: [
                                {
                                    name: 'inner',
                                    type: {
                                        kind: 'field'
                                    }
                                }
                            ],
                            path: 'aztec::protocol_types::contract_class_id::ContractClassId'
                        },
                        visibility: 'private'
                    },
                    {
                        name: 'initialization_hash',
                        type: {
                            kind: 'field'
                        },
                        visibility: 'private'
                    },
                    {
                        name: 'public_keys',
                        type: {
                            kind: 'struct',
                            fields: [
                                {
                                    name: 'npk_m',
                                    type: {
                                        kind: 'struct',
                                        fields: [
                                            {
                                                name: 'inner',
                                                type: {
                                                    kind: 'struct',
                                                    fields: [
                                                        {
                                                            name: 'x',
                                                            type: {
                                                                kind: 'field'
                                                            }
                                                        },
                                                        {
                                                            name: 'y',
                                                            type: {
                                                                kind: 'field'
                                                            }
                                                        },
                                                        {
                                                            name: 'is_infinite',
                                                            type: {
                                                                kind: 'boolean'
                                                            }
                                                        }
                                                    ],
                                                    path: 'std::embedded_curve_ops::EmbeddedCurvePoint'
                                                }
                                            }
                                        ],
                                        path: 'aztec::protocol_types::public_keys::NpkM'
                                    }
                                },
                                {
                                    name: 'ivpk_m',
                                    type: {
                                        kind: 'struct',
                                        fields: [
                                            {
                                                name: 'inner',
                                                type: {
                                                    kind: 'struct',
                                                    fields: [
                                                        {
                                                            name: 'x',
                                                            type: {
                                                                kind: 'field'
                                                            }
                                                        },
                                                        {
                                                            name: 'y',
                                                            type: {
                                                                kind: 'field'
                                                            }
                                                        },
                                                        {
                                                            name: 'is_infinite',
                                                            type: {
                                                                kind: 'boolean'
                                                            }
                                                        }
                                                    ],
                                                    path: 'std::embedded_curve_ops::EmbeddedCurvePoint'
                                                }
                                            }
                                        ],
                                        path: 'aztec::protocol_types::public_keys::IvpkM'
                                    }
                                },
                                {
                                    name: 'ovpk_m',
                                    type: {
                                        kind: 'struct',
                                        fields: [
                                            {
                                                name: 'inner',
                                                type: {
                                                    kind: 'struct',
                                                    fields: [
                                                        {
                                                            name: 'x',
                                                            type: {
                                                                kind: 'field'
                                                            }
                                                        },
                                                        {
                                                            name: 'y',
                                                            type: {
                                                                kind: 'field'
                                                            }
                                                        },
                                                        {
                                                            name: 'is_infinite',
                                                            type: {
                                                                kind: 'boolean'
                                                            }
                                                        }
                                                    ],
                                                    path: 'std::embedded_curve_ops::EmbeddedCurvePoint'
                                                }
                                            }
                                        ],
                                        path: 'aztec::protocol_types::public_keys::OvpkM'
                                    }
                                },
                                {
                                    name: 'tpk_m',
                                    type: {
                                        kind: 'struct',
                                        fields: [
                                            {
                                                name: 'inner',
                                                type: {
                                                    kind: 'struct',
                                                    fields: [
                                                        {
                                                            name: 'x',
                                                            type: {
                                                                kind: 'field'
                                                            }
                                                        },
                                                        {
                                                            name: 'y',
                                                            type: {
                                                                kind: 'field'
                                                            }
                                                        },
                                                        {
                                                            name: 'is_infinite',
                                                            type: {
                                                                kind: 'boolean'
                                                            }
                                                        }
                                                    ],
                                                    path: 'std::embedded_curve_ops::EmbeddedCurvePoint'
                                                }
                                            }
                                        ],
                                        path: 'aztec::protocol_types::public_keys::TpkM'
                                    }
                                }
                            ],
                            path: 'aztec::protocol_types::public_keys::PublicKeys'
                        },
                        visibility: 'private'
                    },
                    {
                        name: 'universal_deploy',
                        type: {
                            kind: 'boolean'
                        },
                        visibility: 'private'
                    }
                ],
                returnTypes: [],
                errorTypes: {
                    '718532354304118388': {
                        error_kind: 'string',
                        string: 'Point not on curve'
                    },
                    '4135474769840782447': {
                        error_kind: 'string',
                        string: 'Point at infinity should have canonical representation (0 0)'
                    },
                    '12469291177396340830': {
                        error_kind: 'string',
                        string: 'call to assert_max_bit_size'
                    },
                    '14990209321349310352': {
                        error_kind: 'string',
                        string: 'attempt to add with overflow'
                    },
                    '15764276373176857197': {
                        error_kind: 'string',
                        string: 'Stack too deep'
                    }
                }
            },
            bytecode: Buffer.from([]),
            debugSymbols: ''
        }
    ],
    nonDispatchPublicFunctions: [
        {
            functionType: FunctionType.PUBLIC,
            name: 'get_update_delay',
            isOnlySelf: false,
            isStatic: true,
            isInitializer: false,
            parameters: [],
            returnTypes: [
                {
                    kind: 'integer',
                    sign: 'unsigned',
                    width: 64
                }
            ],
            errorTypes: {
                '5755400808989454547': {
                    error_kind: 'string',
                    string: 'Function get_update_delay can only be called statically'
                },
                '13455385521185560676': {
                    error_kind: 'string',
                    string: 'Storage slot 0 not allowed. Storage slots must start from 1.'
                },
                '15764276373176857197': {
                    error_kind: 'string',
                    string: 'Stack too deep'
                }
            }
        },
        {
            functionType: FunctionType.PUBLIC,
            name: 'set_update_delay',
            isOnlySelf: false,
            isStatic: false,
            isInitializer: false,
            parameters: [
                {
                    name: 'new_update_delay',
                    type: {
                        kind: 'integer',
                        sign: 'unsigned',
                        width: 64
                    },
                    visibility: 'private'
                }
            ],
            returnTypes: [],
            errorTypes: {
                '361444214588792908': {
                    error_kind: 'string',
                    string: 'attempt to multiply with overflow'
                },
                '1998584279744703196': {
                    error_kind: 'string',
                    string: 'attempt to subtract with overflow'
                },
                '6804164082532189961': {
                    error_kind: 'string',
                    string: 'New update delay is too low'
                },
                '13455385521185560676': {
                    error_kind: 'string',
                    string: 'Storage slot 0 not allowed. Storage slots must start from 1.'
                },
                '14990209321349310352': {
                    error_kind: 'string',
                    string: 'attempt to add with overflow'
                },
                '15353471697975175405': {
                    error_kind: 'string',
                    string: 'msg.sender is not deployed'
                },
                '15764276373176857197': {
                    error_kind: 'string',
                    string: 'Stack too deep'
                },
                '16431471497789672479': {
                    error_kind: 'string',
                    string: 'Index out of bounds'
                }
            }
        },
        {
            functionType: FunctionType.PUBLIC,
            name: 'update',
            isOnlySelf: false,
            isStatic: false,
            isInitializer: false,
            parameters: [
                {
                    name: 'new_contract_class_id',
                    type: {
                        kind: 'struct',
                        fields: [
                            {
                                name: 'inner',
                                type: {
                                    kind: 'field'
                                }
                            }
                        ],
                        path: 'aztec::protocol_types::contract_class_id::ContractClassId'
                    },
                    visibility: 'private'
                }
            ],
            returnTypes: [],
            errorTypes: {
                '361444214588792908': {
                    error_kind: 'string',
                    string: 'attempt to multiply with overflow'
                },
                '1998584279744703196': {
                    error_kind: 'string',
                    string: 'attempt to subtract with overflow'
                },
                '12579274401768182412': {
                    error_kind: 'string',
                    string: 'New contract class is not registered'
                },
                '13455385521185560676': {
                    error_kind: 'string',
                    string: 'Storage slot 0 not allowed. Storage slots must start from 1.'
                },
                '14990209321349310352': {
                    error_kind: 'string',
                    string: 'attempt to add with overflow'
                },
                '15353471697975175405': {
                    error_kind: 'string',
                    string: 'msg.sender is not deployed'
                },
                '15764276373176857197': {
                    error_kind: 'string',
                    string: 'Stack too deep'
                },
                '16431471497789672479': {
                    error_kind: 'string',
                    string: 'Index out of bounds'
                }
            }
        }
    ],
    outputs: {
        structs: {},
        globals: {}
    },
    storageLayout: {},
    fileMap: {}
};
export class ContractInstanceRegistryContract extends ContractBase {
    constructor(wallet){
        super(ProtocolContractAddress.ContractInstanceRegistry, ContractInstanceRegistryContractArtifact, wallet);
    }
    static at(wallet) {
        return new ContractInstanceRegistryContract(wallet);
    }
}
