// @aztec/aztec.js/dest/contract/protocol_contracts/contract-class-registry.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { ProtocolContractAddress } from '@aztec/protocol-contracts';
import { FunctionType } from '@aztec/stdlib/abi';
import { ContractBase } from '../contract_base.js';
const ContractClassRegistryContractArtifact = {
    name: 'ContractClassRegistry',
    functions: [
        {
            ...{
                functionType: FunctionType.PRIVATE,
                name: 'publish',
                isOnlySelf: false,
                isStatic: false,
                isInitializer: false,
                parameters: [
                    {
                        name: 'artifact_hash',
                        type: {
                            kind: 'field'
                        },
                        visibility: 'private'
                    },
                    {
                        name: 'private_functions_root',
                        type: {
                            kind: 'field'
                        },
                        visibility: 'private'
                    },
                    {
                        name: 'public_bytecode_commitment',
                        type: {
                            kind: 'field'
                        },
                        visibility: 'private'
                    }
                ],
                returnTypes: [],
                errorTypes: {
                    '361444214588792908': {
                        error_kind: 'string',
                        string: 'attempt to multiply with overflow'
                    },
                    '922421554006670918': {
                        error_kind: 'string',
                        string: 'num_items is greater than MAX_ITEMS'
                    },
                    '1998584279744703196': {
                        error_kind: 'string',
                        string: 'attempt to subtract with overflow'
                    },
                    '9475115977882098926': {
                        error_kind: 'string',
                        string: 'Extra bytes in the last field of the bytecode'
                    },
                    '10965409165809799877': {
                        error_kind: 'string',
                        string: 'Given in_len to absorb is larger than the input array len'
                    },
                    '12469291177396340830': {
                        error_kind: 'string',
                        string: 'call to assert_max_bit_size'
                    },
                    '14990209321349310352': {
                        error_kind: 'string',
                        string: 'attempt to add with overflow'
                    },
                    '15669273164684982983': {
                        error_kind: 'string',
                        string: 'num_permutations is greater than MAX_PERMUTATIONS'
                    },
                    '15764276373176857197': {
                        error_kind: 'string',
                        string: 'Stack too deep'
                    },
                    '15835548349546956319': {
                        error_kind: 'string',
                        string: 'Field failed to decompose into specified 32 limbs'
                    },
                    '16431471497789672479': {
                        error_kind: 'string',
                        string: 'Index out of bounds'
                    },
                    '16450579948625159707': {
                        error_kind: 'string',
                        string: 'Sponge must be in cache_size=1'
                    },
                    '17655676068928457687': {
                        error_kind: 'string',
                        string: 'Reader did not read all data'
                    }
                }
            },
            bytecode: Buffer.from([]),
            debugSymbols: ''
        },
        {
            ...{
                functionType: FunctionType.PUBLIC,
                name: 'public_dispatch',
                isOnlySelf: false,
                isStatic: false,
                isInitializer: false,
                parameters: [
                    {
                        name: 'selector',
                        type: {
                            kind: 'field'
                        },
                        visibility: 'private'
                    }
                ],
                returnTypes: [],
                errorTypes: {
                    '1335198680857977525': {
                        error_kind: 'string',
                        string: 'No public functions'
                    }
                }
            },
            bytecode: Buffer.from([]),
            debugSymbols: ''
        }
    ],
    nonDispatchPublicFunctions: [],
    outputs: {
        structs: {},
        globals: {}
    },
    storageLayout: {},
    fileMap: {}
};
export class ContractClassRegistryContract extends ContractBase {
    constructor(wallet){
        super(ProtocolContractAddress.ContractClassRegistry, ContractClassRegistryContractArtifact, wallet);
    }
    static at(wallet) {
        return new ContractClassRegistryContract(wallet);
    }
}
