// @aztec/aztec.js/dest/contract/protocol_contracts/public-checks.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { ProtocolContractAddress } from '@aztec/protocol-contracts';
import { FunctionType } from '@aztec/stdlib/abi';
import { ContractBase } from '../contract_base.js';
const PublicChecksContractArtifact = {
    name: 'PublicChecks',
    functions: [
        {
            ...{
                functionType: FunctionType.UTILITY,
                name: 'process_message',
                isOnlySelf: false,
                isStatic: false,
                isInitializer: false,
                parameters: [
                    {
                        name: 'message_ciphertext',
                        type: {
                            kind: 'struct',
                            fields: [
                                {
                                    name: 'storage',
                                    type: {
                                        kind: 'array',
                                        length: 15,
                                        type: {
                                            kind: 'field'
                                        }
                                    }
                                },
                                {
                                    name: 'len',
                                    type: {
                                        kind: 'integer',
                                        sign: 'unsigned',
                                        width: 32
                                    }
                                }
                            ],
                            path: 'std::collections::bounded_vec::BoundedVec'
                        },
                        visibility: 'private'
                    },
                    {
                        name: 'message_context',
                        type: {
                            kind: 'struct',
                            fields: [
                                {
                                    name: 'tx_hash',
                                    type: {
                                        kind: 'field'
                                    }
                                },
                                {
                                    name: 'unique_note_hashes_in_tx',
                                    type: {
                                        kind: 'struct',
                                        fields: [
                                            {
                                                name: 'storage',
                                                type: {
                                                    kind: 'array',
                                                    length: 64,
                                                    type: {
                                                        kind: 'field'
                                                    }
                                                }
                                            },
                                            {
                                                name: 'len',
                                                type: {
                                                    kind: 'integer',
                                                    sign: 'unsigned',
                                                    width: 32
                                                }
                                            }
                                        ],
                                        path: 'std::collections::bounded_vec::BoundedVec'
                                    }
                                },
                                {
                                    name: 'first_nullifier_in_tx',
                                    type: {
                                        kind: 'field'
                                    }
                                },
                                {
                                    name: 'recipient',
                                    type: {
                                        kind: 'struct',
                                        fields: [
                                            {
                                                name: 'inner',
                                                type: {
                                                    kind: 'field'
                                                }
                                            }
                                        ],
                                        path: 'aztec::protocol_types::address::aztec_address::AztecAddress'
                                    }
                                }
                            ],
                            path: 'aztec::messages::processing::message_context::MessageContext'
                        },
                        visibility: 'private'
                    }
                ],
                returnTypes: [],
                errorTypes: {
                    '361444214588792908': {
                        error_kind: 'string',
                        string: 'attempt to multiply with overflow'
                    },
                    '992401946138144806': {
                        error_kind: 'string',
                        string: 'Attempted to read past end of BoundedVec'
                    },
                    '1998584279744703196': {
                        error_kind: 'string',
                        string: 'attempt to subtract with overflow'
                    },
                    '2967937905572420042': {
                        error_kind: 'fmtstring',
                        length: 61,
                        item_types: [
                            {
                                kind: 'field'
                            },
                            {
                                kind: 'field'
                            }
                        ]
                    },
                    '3330370348214585450': {
                        error_kind: 'fmtstring',
                        length: 48,
                        item_types: [
                            {
                                kind: 'field'
                            },
                            {
                                kind: 'field'
                            }
                        ]
                    },
                    '3670003311596808700': {
                        error_kind: 'fmtstring',
                        length: 77,
                        item_types: [
                            {
                                kind: 'integer',
                                sign: 'unsigned',
                                width: 32
                            }
                        ]
                    },
                    '4261968856572588300': {
                        error_kind: 'string',
                        string: 'Value does not fit in field'
                    },
                    '4440399188109668273': {
                        error_kind: 'string',
                        string: 'Input length must be a multiple of 32'
                    },
                    '5417577161503694006': {
                        error_kind: 'fmtstring',
                        length: 56,
                        item_types: [
                            {
                                kind: 'field'
                            }
                        ]
                    },
                    '9791669845391776238': {
                        error_kind: 'string',
                        string: '0 has a square root; you cannot claim it is not square'
                    },
                    '10135509984888824963': {
                        error_kind: 'fmtstring',
                        length: 58,
                        item_types: [
                            {
                                kind: 'field'
                            }
                        ]
                    },
                    '10791800398362570014': {
                        error_kind: 'string',
                        string: 'extend_from_bounded_vec out of bounds'
                    },
                    '11692359521570349358': {
                        error_kind: 'fmtstring',
                        length: 40,
                        item_types: []
                    },
                    '12469291177396340830': {
                        error_kind: 'string',
                        string: 'call to assert_max_bit_size'
                    },
                    '12913276134398371456': {
                        error_kind: 'string',
                        string: 'push out of bounds'
                    },
                    '13557316507370296400': {
                        error_kind: 'fmtstring',
                        length: 130,
                        item_types: [
                            {
                                kind: 'integer',
                                sign: 'unsigned',
                                width: 32
                            }
                        ]
                    },
                    '14938672389828944159': {
                        error_kind: 'fmtstring',
                        length: 146,
                        item_types: [
                            {
                                kind: 'integer',
                                sign: 'unsigned',
                                width: 32
                            }
                        ]
                    },
                    '14990209321349310352': {
                        error_kind: 'string',
                        string: 'attempt to add with overflow'
                    },
                    '15764276373176857197': {
                        error_kind: 'string',
                        string: 'Stack too deep'
                    },
                    '16431471497789672479': {
                        error_kind: 'string',
                        string: 'Index out of bounds'
                    },
                    '17531474008201752295': {
                        error_kind: 'fmtstring',
                        length: 133,
                        item_types: [
                            {
                                kind: 'integer',
                                sign: 'unsigned',
                                width: 32
                            }
                        ]
                    }
                }
            },
            bytecode: Buffer.from([]),
            debugSymbols: ''
        },
        {
            ...{
                functionType: FunctionType.PUBLIC,
                name: 'public_dispatch',
                isOnlySelf: false,
                isStatic: false,
                isInitializer: false,
                parameters: [
                    {
                        name: 'selector',
                        type: {
                            kind: 'field'
                        },
                        visibility: 'private'
                    }
                ],
                returnTypes: [],
                errorTypes: {
                    '3445904810750103705': {
                        error_kind: 'string',
                        string: 'Block number mismatch.'
                    },
                    '6989792215391248978': {
                        error_kind: 'string',
                        string: 'Function check_timestamp can only be called statically'
                    },
                    '7259463707251703666': {
                        error_kind: 'string',
                        string: 'Function check_block_number can only be called statically'
                    },
                    '12511970388699677811': {
                        error_kind: 'fmtstring',
                        length: 27,
                        item_types: [
                            {
                                kind: 'field'
                            }
                        ]
                    },
                    '13950240626967988787': {
                        error_kind: 'string',
                        string: 'Timestamp mismatch.'
                    },
                    '15015303112667403944': {
                        error_kind: 'fmtstring',
                        length: 17,
                        item_types: []
                    },
                    '15764276373176857197': {
                        error_kind: 'string',
                        string: 'Stack too deep'
                    }
                }
            },
            bytecode: Buffer.from([]),
            debugSymbols: ''
        },
        {
            ...{
                functionType: FunctionType.UTILITY,
                name: 'sync_state',
                isOnlySelf: false,
                isStatic: false,
                isInitializer: false,
                parameters: [],
                returnTypes: [],
                errorTypes: {
                    '361444214588792908': {
                        error_kind: 'string',
                        string: 'attempt to multiply with overflow'
                    },
                    '992401946138144806': {
                        error_kind: 'string',
                        string: 'Attempted to read past end of BoundedVec'
                    },
                    '1998584279744703196': {
                        error_kind: 'string',
                        string: 'attempt to subtract with overflow'
                    },
                    '2967937905572420042': {
                        error_kind: 'fmtstring',
                        length: 61,
                        item_types: [
                            {
                                kind: 'field'
                            },
                            {
                                kind: 'field'
                            }
                        ]
                    },
                    '3330370348214585450': {
                        error_kind: 'fmtstring',
                        length: 48,
                        item_types: [
                            {
                                kind: 'field'
                            },
                            {
                                kind: 'field'
                            }
                        ]
                    },
                    '3670003311596808700': {
                        error_kind: 'fmtstring',
                        length: 77,
                        item_types: [
                            {
                                kind: 'integer',
                                sign: 'unsigned',
                                width: 32
                            }
                        ]
                    },
                    '4261968856572588300': {
                        error_kind: 'string',
                        string: 'Value does not fit in field'
                    },
                    '4440399188109668273': {
                        error_kind: 'string',
                        string: 'Input length must be a multiple of 32'
                    },
                    '5417577161503694006': {
                        error_kind: 'fmtstring',
                        length: 56,
                        item_types: [
                            {
                                kind: 'field'
                            }
                        ]
                    },
                    '9791669845391776238': {
                        error_kind: 'string',
                        string: '0 has a square root; you cannot claim it is not square'
                    },
                    '9885968605480832328': {
                        error_kind: 'string',
                        string: 'Attempted to read past the length of a CapsuleArray'
                    },
                    '10135509984888824963': {
                        error_kind: 'fmtstring',
                        length: 58,
                        item_types: [
                            {
                                kind: 'field'
                            }
                        ]
                    },
                    '10791800398362570014': {
                        error_kind: 'string',
                        string: 'extend_from_bounded_vec out of bounds'
                    },
                    '11021520179822076911': {
                        error_kind: 'string',
                        string: 'Attempted to delete past the length of a CapsuleArray'
                    },
                    '11692359521570349358': {
                        error_kind: 'fmtstring',
                        length: 40,
                        item_types: []
                    },
                    '12327971061804302172': {
                        error_kind: 'fmtstring',
                        length: 98,
                        item_types: []
                    },
                    '12469291177396340830': {
                        error_kind: 'string',
                        string: 'call to assert_max_bit_size'
                    },
                    '12913276134398371456': {
                        error_kind: 'string',
                        string: 'push out of bounds'
                    },
                    '13557316507370296400': {
                        error_kind: 'fmtstring',
                        length: 130,
                        item_types: [
                            {
                                kind: 'integer',
                                sign: 'unsigned',
                                width: 32
                            }
                        ]
                    },
                    '14938672389828944159': {
                        error_kind: 'fmtstring',
                        length: 146,
                        item_types: [
                            {
                                kind: 'integer',
                                sign: 'unsigned',
                                width: 32
                            }
                        ]
                    },
                    '14990209321349310352': {
                        error_kind: 'string',
                        string: 'attempt to add with overflow'
                    },
                    '15764276373176857197': {
                        error_kind: 'string',
                        string: 'Stack too deep'
                    },
                    '16431471497789672479': {
                        error_kind: 'string',
                        string: 'Index out of bounds'
                    },
                    '17531474008201752295': {
                        error_kind: 'fmtstring',
                        length: 133,
                        item_types: [
                            {
                                kind: 'integer',
                                sign: 'unsigned',
                                width: 32
                            }
                        ]
                    },
                    '17655676068928457687': {
                        error_kind: 'string',
                        string: 'Reader did not read all data'
                    }
                }
            },
            bytecode: Buffer.from([]),
            debugSymbols: ''
        }
    ],
    nonDispatchPublicFunctions: [
        {
            functionType: FunctionType.PUBLIC,
            name: 'check_block_number',
            isOnlySelf: false,
            isStatic: true,
            isInitializer: false,
            parameters: [
                {
                    name: 'operation',
                    type: {
                        kind: 'integer',
                        sign: 'unsigned',
                        width: 8
                    },
                    visibility: 'private'
                },
                {
                    name: 'value',
                    type: {
                        kind: 'integer',
                        sign: 'unsigned',
                        width: 32
                    },
                    visibility: 'private'
                }
            ],
            returnTypes: [],
            errorTypes: {
                '3445904810750103705': {
                    error_kind: 'string',
                    string: 'Block number mismatch.'
                },
                '7259463707251703666': {
                    error_kind: 'string',
                    string: 'Function check_block_number can only be called statically'
                },
                '15015303112667403944': {
                    error_kind: 'fmtstring',
                    length: 17,
                    item_types: []
                },
                '15764276373176857197': {
                    error_kind: 'string',
                    string: 'Stack too deep'
                }
            }
        },
        {
            functionType: FunctionType.PUBLIC,
            name: 'check_timestamp',
            isOnlySelf: false,
            isStatic: true,
            isInitializer: false,
            parameters: [
                {
                    name: 'operation',
                    type: {
                        kind: 'integer',
                        sign: 'unsigned',
                        width: 8
                    },
                    visibility: 'private'
                },
                {
                    name: 'value',
                    type: {
                        kind: 'integer',
                        sign: 'unsigned',
                        width: 64
                    },
                    visibility: 'private'
                }
            ],
            returnTypes: [],
            errorTypes: {
                '6989792215391248978': {
                    error_kind: 'string',
                    string: 'Function check_timestamp can only be called statically'
                },
                '13950240626967988787': {
                    error_kind: 'string',
                    string: 'Timestamp mismatch.'
                },
                '15015303112667403944': {
                    error_kind: 'fmtstring',
                    length: 17,
                    item_types: []
                },
                '15764276373176857197': {
                    error_kind: 'string',
                    string: 'Stack too deep'
                }
            }
        }
    ],
    outputs: {
        structs: {},
        globals: {}
    },
    storageLayout: {},
    fileMap: {}
};
export class PublicChecksContract extends ContractBase {
    constructor(wallet){
        super(ProtocolContractAddress.PublicChecks, PublicChecksContractArtifact, wallet);
    }
    static at(wallet) {
        return new PublicChecksContract(wallet);
    }
}
