// @aztec/bb.js/dest/node/crs/browser/cached_net_crs.js
import { NetCrs, NetGrumpkinCrs } from '../net_crs.js';
import { get, set } from 'idb-keyval';
/**
 * Downloader for CRS from the web or local.
 */
export class CachedNetCrs {
    numPoints;
    g1Data;
    g2Data;
    constructor(numPoints) {
        this.numPoints = numPoints;
    }
    static async new(numPoints) {
        const crs = new CachedNetCrs(numPoints);
        await crs.init();
        return crs;
    }
    /**
     * Download the data.
     */
    async init() {
        // Check if data is in IndexedDB
        const g1Data = await get('g1Data');
        const g2Data = await get('g2Data');
        const netCrs = new NetCrs(this.numPoints);
        const g1DataLength = this.numPoints * 64;
        if (!g1Data || g1Data.length < g1DataLength) {
            this.g1Data = await netCrs.downloadG1Data();
            await set('g1Data', this.g1Data);
        }
        else {
            this.g1Data = g1Data;
        }
        if (!g2Data) {
            this.g2Data = await netCrs.downloadG2Data();
            await set('g2Data', this.g2Data);
        }
        else {
            this.g2Data = g2Data;
        }
    }
    /**
     * G1 points data for prover key.
     * @returns The points data.
     */
    getG1Data() {
        return this.g1Data;
    }
    /**
     * G2 points data for verification key.
     * @returns The points data.
     */
    getG2Data() {
        return this.g2Data;
    }
}
/**
 * Downloader for CRS from the web or local.
 */
export class CachedNetGrumpkinCrs {
    numPoints;
    g1Data;
    constructor(numPoints) {
        this.numPoints = numPoints;
    }
    static async new(numPoints) {
        const crs = new CachedNetGrumpkinCrs(numPoints);
        await crs.init();
        return crs;
    }
    /**
     * Download the data.
     */
    async init() {
        // Check if data is in IndexedDB
        const g1Data = await get('grumpkinG1Data');
        const netGrumpkinCrs = new NetGrumpkinCrs(this.numPoints);
        const g1DataLength = this.numPoints * 64;
        if (!g1Data || g1Data.length < g1DataLength) {
            this.g1Data = await netGrumpkinCrs.downloadG1Data();
            await set('grumpkinG1Data', this.g1Data);
        }
        else {
            this.g1Data = g1Data;
        }
    }
    /**
     * G1 points data for prover key.
     * @returns The points data.
     */
    getG1Data() {
        return this.g1Data;
    }
}
