// @aztec/world-state/dest/native/message.js
import { BlockNumber } from '@aztec/foundation/branded-types';
import { Fr } from '@aztec/foundation/curves/bn254';
import { AppendOnlyTreeSnapshot, MerkleTreeId } from '@aztec/stdlib/trees';
export var WorldStateMessageType = /*#__PURE__*/ function(WorldStateMessageType) {
    WorldStateMessageType[WorldStateMessageType["GET_TREE_INFO"] = 100] = "GET_TREE_INFO";
    WorldStateMessageType[WorldStateMessageType["GET_STATE_REFERENCE"] = 101] = "GET_STATE_REFERENCE";
    WorldStateMessageType[WorldStateMessageType["GET_INITIAL_STATE_REFERENCE"] = 102] = "GET_INITIAL_STATE_REFERENCE";
    WorldStateMessageType[WorldStateMessageType["GET_LEAF_VALUE"] = 103] = "GET_LEAF_VALUE";
    WorldStateMessageType[WorldStateMessageType["GET_LEAF_PREIMAGE"] = 104] = "GET_LEAF_PREIMAGE";
    WorldStateMessageType[WorldStateMessageType["GET_SIBLING_PATH"] = 105] = "GET_SIBLING_PATH";
    WorldStateMessageType[WorldStateMessageType["GET_BLOCK_NUMBERS_FOR_LEAF_INDICES"] = 106] = "GET_BLOCK_NUMBERS_FOR_LEAF_INDICES";
    WorldStateMessageType[WorldStateMessageType["FIND_LEAF_INDICES"] = 107] = "FIND_LEAF_INDICES";
    WorldStateMessageType[WorldStateMessageType["FIND_LOW_LEAF"] = 108] = "FIND_LOW_LEAF";
    WorldStateMessageType[WorldStateMessageType["FIND_SIBLING_PATHS"] = 109] = "FIND_SIBLING_PATHS";
    WorldStateMessageType[WorldStateMessageType["APPEND_LEAVES"] = 110] = "APPEND_LEAVES";
    WorldStateMessageType[WorldStateMessageType["BATCH_INSERT"] = 111] = "BATCH_INSERT";
    WorldStateMessageType[WorldStateMessageType["SEQUENTIAL_INSERT"] = 112] = "SEQUENTIAL_INSERT";
    WorldStateMessageType[WorldStateMessageType["UPDATE_ARCHIVE"] = 113] = "UPDATE_ARCHIVE";
    WorldStateMessageType[WorldStateMessageType["COMMIT"] = 114] = "COMMIT";
    WorldStateMessageType[WorldStateMessageType["ROLLBACK"] = 115] = "ROLLBACK";
    WorldStateMessageType[WorldStateMessageType["SYNC_BLOCK"] = 116] = "SYNC_BLOCK";
    WorldStateMessageType[WorldStateMessageType["CREATE_FORK"] = 117] = "CREATE_FORK";
    WorldStateMessageType[WorldStateMessageType["DELETE_FORK"] = 118] = "DELETE_FORK";
    WorldStateMessageType[WorldStateMessageType["FINALIZE_BLOCKS"] = 119] = "FINALIZE_BLOCKS";
    WorldStateMessageType[WorldStateMessageType["UNWIND_BLOCKS"] = 120] = "UNWIND_BLOCKS";
    WorldStateMessageType[WorldStateMessageType["REMOVE_HISTORICAL_BLOCKS"] = 121] = "REMOVE_HISTORICAL_BLOCKS";
    WorldStateMessageType[WorldStateMessageType["GET_STATUS"] = 122] = "GET_STATUS";
    WorldStateMessageType[WorldStateMessageType["CREATE_CHECKPOINT"] = 123] = "CREATE_CHECKPOINT";
    WorldStateMessageType[WorldStateMessageType["COMMIT_CHECKPOINT"] = 124] = "COMMIT_CHECKPOINT";
    WorldStateMessageType[WorldStateMessageType["REVERT_CHECKPOINT"] = 125] = "REVERT_CHECKPOINT";
    WorldStateMessageType[WorldStateMessageType["COMMIT_ALL_CHECKPOINTS"] = 126] = "COMMIT_ALL_CHECKPOINTS";
    WorldStateMessageType[WorldStateMessageType["REVERT_ALL_CHECKPOINTS"] = 127] = "REVERT_ALL_CHECKPOINTS";
    WorldStateMessageType[WorldStateMessageType["COPY_STORES"] = 128] = "COPY_STORES";
    WorldStateMessageType[WorldStateMessageType["CLOSE"] = 999] = "CLOSE";
    return WorldStateMessageType;
}({});
export function buildEmptyDBStats() {
    return {
        name: '',
        numDataItems: 0n,
        totalUsedSize: 0n
    };
}
export function buildEmptyTreeDBStats() {
    return {
        mapSize: 0n,
        physicalFileSize: 0n,
        blocksDBStats: buildEmptyDBStats(),
        nodesDBStats: buildEmptyDBStats(),
        leafIndicesDBStats: buildEmptyDBStats(),
        leafKeysDBStats: buildEmptyDBStats(),
        leafPreimagesDBStats: buildEmptyDBStats(),
        blockIndicesDBStats: buildEmptyDBStats()
    };
}
export function buildEmptyTreeMeta() {
    return {
        name: '',
        depth: 0,
        size: 0n,
        committedSize: 0n,
        unfinalizedBlockHeight: BlockNumber.ZERO,
        finalizedBlockHeight: BlockNumber.ZERO,
        oldestHistoricBlock: BlockNumber.ZERO,
        root: Fr.ZERO,
        initialRoot: Fr.ZERO,
        initialSize: 0n
    };
}
export function buildEmptyWorldStateMeta() {
    return {
        noteHashTreeMeta: buildEmptyTreeMeta(),
        messageTreeMeta: buildEmptyTreeMeta(),
        publicDataTreeMeta: buildEmptyTreeMeta(),
        nullifierTreeMeta: buildEmptyTreeMeta(),
        archiveTreeMeta: buildEmptyTreeMeta()
    };
}
export function buildEmptyWorldStateDBStats() {
    return {
        noteHashTreeStats: buildEmptyTreeDBStats(),
        archiveTreeStats: buildEmptyTreeDBStats(),
        messageTreeStats: buildEmptyTreeDBStats(),
        publicDataTreeStats: buildEmptyTreeDBStats(),
        nullifierTreeStats: buildEmptyTreeDBStats()
    };
}
export function buildEmptyWorldStateSummary() {
    return {
        unfinalizedBlockNumber: BlockNumber.ZERO,
        finalizedBlockNumber: BlockNumber.ZERO,
        oldestHistoricalBlock: BlockNumber.ZERO,
        treesAreSynched: true
    };
}
export function buildEmptyWorldStateStatusFull() {
    return {
        meta: buildEmptyWorldStateMeta(),
        dbStats: buildEmptyWorldStateDBStats(),
        summary: buildEmptyWorldStateSummary()
    };
}
export function sanitizeSummary(summary) {
    summary.finalizedBlockNumber = BlockNumber.fromBigInt(BigInt(summary.finalizedBlockNumber));
    summary.unfinalizedBlockNumber = BlockNumber.fromBigInt(BigInt(summary.unfinalizedBlockNumber));
    summary.oldestHistoricalBlock = BlockNumber.fromBigInt(BigInt(summary.oldestHistoricalBlock));
    return summary;
}
export function sanitizeDBStats(stats) {
    stats.numDataItems = BigInt(stats.numDataItems);
    stats.totalUsedSize = BigInt(stats.totalUsedSize);
    return stats;
}
export function sanitizeMeta(meta) {
    meta.committedSize = BigInt(meta.committedSize);
    meta.finalizedBlockHeight = BlockNumber.fromBigInt(BigInt(meta.finalizedBlockHeight));
    meta.initialSize = BigInt(meta.initialSize);
    meta.oldestHistoricBlock = BlockNumber.fromBigInt(BigInt(meta.oldestHistoricBlock));
    meta.size = BigInt(meta.size);
    meta.unfinalizedBlockHeight = BlockNumber.fromBigInt(BigInt(meta.unfinalizedBlockHeight));
    return meta;
}
export function sanitizeTreeDBStats(stats) {
    stats.blocksDBStats = sanitizeDBStats(stats.blocksDBStats);
    stats.leafIndicesDBStats = sanitizeDBStats(stats.leafIndicesDBStats);
    stats.leafPreimagesDBStats = sanitizeDBStats(stats.leafPreimagesDBStats);
    stats.blockIndicesDBStats = sanitizeDBStats(stats.blockIndicesDBStats);
    stats.nodesDBStats = sanitizeDBStats(stats.nodesDBStats);
    stats.mapSize = BigInt(stats.mapSize);
    stats.physicalFileSize = BigInt(stats.physicalFileSize);
    return stats;
}
export function sanitizeWorldStateDBStats(stats) {
    stats.archiveTreeStats = sanitizeTreeDBStats(stats.archiveTreeStats);
    stats.messageTreeStats = sanitizeTreeDBStats(stats.messageTreeStats);
    stats.noteHashTreeStats = sanitizeTreeDBStats(stats.noteHashTreeStats);
    stats.nullifierTreeStats = sanitizeTreeDBStats(stats.nullifierTreeStats);
    stats.publicDataTreeStats = sanitizeTreeDBStats(stats.publicDataTreeStats);
    return stats;
}
export function sanitizeWorldStateTreeMeta(meta) {
    meta.archiveTreeMeta = sanitizeMeta(meta.archiveTreeMeta);
    meta.messageTreeMeta = sanitizeMeta(meta.messageTreeMeta);
    meta.noteHashTreeMeta = sanitizeMeta(meta.noteHashTreeMeta);
    meta.nullifierTreeMeta = sanitizeMeta(meta.nullifierTreeMeta);
    meta.publicDataTreeMeta = sanitizeMeta(meta.publicDataTreeMeta);
    return meta;
}
export function sanitizeFullStatus(status) {
    status.dbStats = sanitizeWorldStateDBStats(status.dbStats);
    status.summary = sanitizeSummary(status.summary);
    status.meta = sanitizeWorldStateTreeMeta(status.meta);
    return status;
}
export function isWithForkId(body) {
    return body && 'forkId' in body;
}
export function isWithRevision(body) {
    return body && 'revision' in body;
}
export function isWithCanonical(body) {
    return body && 'canonical' in body;
}
export function treeStateReferenceToSnapshot([root, size]) {
    return new AppendOnlyTreeSnapshot(Fr.fromBuffer(root), Number(size));
}
export function treeStateReference(snapshot) {
    return [
        snapshot.root.toBuffer(),
        BigInt(snapshot.nextAvailableLeafIndex)
    ];
}
export function blockStateReference(state) {
    return new Map([
        [
            MerkleTreeId.NULLIFIER_TREE,
            treeStateReference(state.partial.nullifierTree)
        ],
        [
            MerkleTreeId.NOTE_HASH_TREE,
            treeStateReference(state.partial.noteHashTree)
        ],
        [
            MerkleTreeId.PUBLIC_DATA_TREE,
            treeStateReference(state.partial.publicDataTree)
        ],
        [
            MerkleTreeId.L1_TO_L2_MESSAGE_TREE,
            treeStateReference(state.l1ToL2MessageTree)
        ]
    ]);
}
