// @aztec/stdlib/dest/world-state/world_state_revision.js
import { z } from 'zod';
export class WorldStateRevision {
    forkId;
    blockNumber;
    includeUncommitted;
    constructor(forkId, blockNumber, includeUncommitted){
        this.forkId = forkId;
        this.blockNumber = blockNumber;
        this.includeUncommitted = includeUncommitted;
    }
    toString() {
        return `WorldStateRevision(forkId: ${this.forkId}, blockNumber: ${this.blockNumber}, includeUncommitted: ${this.includeUncommitted})`;
    }
    static empty() {
        return new WorldStateRevision(0, 0, false);
    }
    static get schema() {
        return z.object({
            forkId: z.number(),
            blockNumber: z.number(),
            includeUncommitted: z.boolean()
        }).transform(({ forkId, blockNumber, includeUncommitted })=>new WorldStateRevision(forkId, blockNumber, includeUncommitted));
    }
}
export class WorldStateRevisionWithHandle extends WorldStateRevision {
    handle;
    constructor(forkId, blockNumber, includeUncommitted, handle){
        super(forkId, blockNumber, includeUncommitted), this.handle = handle;
    }
    toWorldStateRevision() {
        return new WorldStateRevision(this.forkId, this.blockNumber, this.includeUncommitted);
    }
    static fromWorldStateRevision(revision, handle) {
        return new WorldStateRevisionWithHandle(revision.forkId, revision.blockNumber, revision.includeUncommitted, handle);
    }
    static get schema() {
        return z.object({
            forkId: z.number(),
            blockNumber: z.number(),
            includeUncommitted: z.boolean(),
            handle: z.any()
        }).transform(({ forkId, blockNumber, includeUncommitted, handle })=>new WorldStateRevisionWithHandle(forkId, blockNumber, includeUncommitted, handle));
    }
}
