// @aztec/noir-protocol-circuits-types/dest/types/index.js
/* Autogenerated file, do not edit! */ /* eslint-disable */ import { Noir } from "@aztec/noir-noir_js";
export async function ParityBase(inputs, ParityBase_circuit, foreignCallHandler) {
    const program = new Noir(ParityBase_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function ParityRoot(inputs, ParityRoot_circuit, foreignCallHandler) {
    const program = new Noir(ParityRoot_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function PrivateKernelInit(tx_request, vk_tree_root, protocol_contracts, private_call, is_private_only, first_nullifier_hint, revertible_counter_hint, app_public_inputs, PrivateKernelInit_circuit, foreignCallHandler) {
    const program = new Noir(PrivateKernelInit_circuit);
    const args = {
        tx_request,
        vk_tree_root,
        protocol_contracts,
        private_call,
        is_private_only,
        first_nullifier_hint,
        revertible_counter_hint,
        app_public_inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function PrivateKernelInner(previous_kernel, previous_kernel_public_inputs, private_call, app_public_inputs, PrivateKernelInner_circuit, foreignCallHandler) {
    const program = new Noir(PrivateKernelInner_circuit);
    const args = {
        previous_kernel,
        previous_kernel_public_inputs,
        private_call,
        app_public_inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function PrivateKernelReset(previous_kernel, previous_kernel_public_inputs, padded_side_effects, hints, PrivateKernelReset_circuit, foreignCallHandler) {
    const program = new Noir(PrivateKernelReset_circuit);
    const args = {
        previous_kernel,
        previous_kernel_public_inputs,
        padded_side_effects,
        hints
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function PrivateKernelTail(previous_kernel, previous_kernel_public_inputs, expiration_timestamp_upper_bound, PrivateKernelTail_circuit, foreignCallHandler) {
    const program = new Noir(PrivateKernelTail_circuit);
    const args = {
        previous_kernel,
        previous_kernel_public_inputs,
        expiration_timestamp_upper_bound
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function PrivateKernelTailToPublic(previous_kernel, previous_kernel_public_inputs, padded_side_effect_amounts, expiration_timestamp_upper_bound, PrivateKernelTailToPublic_circuit, foreignCallHandler) {
    const program = new Noir(PrivateKernelTailToPublic_circuit);
    const args = {
        previous_kernel,
        previous_kernel_public_inputs,
        padded_side_effect_amounts,
        expiration_timestamp_upper_bound
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function HidingKernelToRollup(previous_kernel_public_inputs, previous_kernel_vk_data, HidingKernelToRollup_circuit, foreignCallHandler) {
    const program = new Noir(HidingKernelToRollup_circuit);
    const args = {
        previous_kernel_public_inputs,
        previous_kernel_vk_data
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function HidingKernelToPublic(previous_kernel_public_inputs, previous_kernel_vk_data, HidingKernelToPublic_circuit, foreignCallHandler) {
    const program = new Noir(HidingKernelToPublic_circuit);
    const args = {
        previous_kernel_public_inputs,
        previous_kernel_vk_data
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function ChonkVerifierPublic(inputs, ChonkVerifierPublic_circuit, foreignCallHandler) {
    const program = new Noir(ChonkVerifierPublic_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function RollupTxBasePrivate(inputs, RollupTxBasePrivate_circuit, foreignCallHandler) {
    const program = new Noir(RollupTxBasePrivate_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function RollupTxBasePublic(inputs, RollupTxBasePublic_circuit, foreignCallHandler) {
    const program = new Noir(RollupTxBasePublic_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function RollupTxMerge(inputs, RollupTxMerge_circuit, foreignCallHandler) {
    const program = new Noir(RollupTxMerge_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function RollupBlockRoot(inputs, RollupBlockRoot_circuit, foreignCallHandler) {
    const program = new Noir(RollupBlockRoot_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function RollupBlockRootSingleTx(inputs, RollupBlockRootSingleTx_circuit, foreignCallHandler) {
    const program = new Noir(RollupBlockRootSingleTx_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function RollupBlockRootFirst(inputs, RollupBlockRootFirst_circuit, foreignCallHandler) {
    const program = new Noir(RollupBlockRootFirst_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function RollupBlockRootFirstSingleTx(inputs, RollupBlockRootFirstSingleTx_circuit, foreignCallHandler) {
    const program = new Noir(RollupBlockRootFirstSingleTx_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function RollupBlockRootFirstEmptyTx(inputs, RollupBlockRootFirstEmptyTx_circuit, foreignCallHandler) {
    const program = new Noir(RollupBlockRootFirstEmptyTx_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function RollupBlockMerge(inputs, RollupBlockMerge_circuit, foreignCallHandler) {
    const program = new Noir(RollupBlockMerge_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function RollupCheckpointRoot(inputs, RollupCheckpointRoot_circuit, foreignCallHandler) {
    const program = new Noir(RollupCheckpointRoot_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function RollupCheckpointRootSingleBlock(inputs, RollupCheckpointRootSingleBlock_circuit, foreignCallHandler) {
    const program = new Noir(RollupCheckpointRootSingleBlock_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function RollupCheckpointMerge(inputs, RollupCheckpointMerge_circuit, foreignCallHandler) {
    const program = new Noir(RollupCheckpointMerge_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function RollupRoot(inputs, RollupRoot_circuit, foreignCallHandler) {
    const program = new Noir(RollupRoot_circuit);
    const args = {
        inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function TsTypes(_nullifier, _private_log_data, _log_hash, _private_call_request, _public_call_request, _key_validation_request, _note_hash_leaf_preimage, _nullifier_leaf_preimage, _l2_to_l1_message, _counted, _avm_circuit_public_inputs, TsTypes_circuit, foreignCallHandler) {
    const program = new Noir(TsTypes_circuit);
    const args = {
        _nullifier,
        _private_log_data,
        _log_hash,
        _private_call_request,
        _public_call_request,
        _key_validation_request,
        _note_hash_leaf_preimage,
        _nullifier_leaf_preimage,
        _l2_to_l1_message,
        _counted,
        _avm_circuit_public_inputs
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
export async function PrivateKernelReset_32_4_32_4_4_4_4_4_4(previous_kernel, previous_kernel_public_inputs, padded_side_effects, hints, PrivateKernelReset_32_4_32_4_4_4_4_4_4_circuit, foreignCallHandler) {
    const program = new Noir(PrivateKernelReset_32_4_32_4_4_4_4_4_4_circuit);
    const args = {
        previous_kernel,
        previous_kernel_public_inputs,
        padded_side_effects,
        hints
    };
    const { returnValue } = await program.execute(args, foreignCallHandler);
    return returnValue;
}
