// @aztec/protocol-contracts/dest/provider/bundle.js
import { AuthRegistryArtifact } from '../auth-registry/index.js';
import { ContractClassRegistryArtifact } from '../class-registry/index.js';
import { FeeJuiceArtifact } from '../fee-juice/index.js';
import { ContractInstanceRegistryArtifact } from '../instance-registry/index.js';
import { makeProtocolContract } from '../make_protocol_contract.js';
import { MultiCallEntrypointArtifact } from '../multi-call-entrypoint/index.js';
import { PublicChecksArtifact } from '../public-checks/index.js';
export const ProtocolContractArtifact = {
    AuthRegistry: AuthRegistryArtifact,
    ContractInstanceRegistry: ContractInstanceRegistryArtifact,
    ContractClassRegistry: ContractClassRegistryArtifact,
    MultiCallEntrypoint: MultiCallEntrypointArtifact,
    FeeJuice: FeeJuiceArtifact,
    PublicChecks: PublicChecksArtifact
};
export class BundledProtocolContractsProvider {
    getProtocolContractArtifact(name) {
        return makeProtocolContract(name, ProtocolContractArtifact[name]);
    }
}
