// @aztec/blob-lib/dest/batched_blob.js
import { FinalBlobAccumulator } from './circuit_types/index.js';
/**
 * A class to represent the result from accumulating blobs in an epoch using BatchedBlobAccumulator.
 */ export class BatchedBlob {
    blobCommitmentsHash;
    z;
    y;
    commitment;
    q;
    constructor(/** Hash of Cs (to link to L1 blob hashes). */ blobCommitmentsHash, /** Challenge point z such that p_i(z) = y_i. */ z, /** Evaluation y, linear combination of all evaluations y_i = p_i(z) with gamma. */ y, /** Commitment C, linear combination of all commitments C_i = [p_i] with gamma. */ commitment, /** KZG opening 'proof' Q (commitment to the quotient poly.), linear combination of all blob kzg 'proofs' Q_i with gamma. */ q){
        this.blobCommitmentsHash = blobCommitmentsHash;
        this.z = z;
        this.y = y;
        this.commitment = commitment;
        this.q = q;
    }
    toFinalBlobAccumulator() {
        return new FinalBlobAccumulator(this.blobCommitmentsHash, this.z, this.y, this.commitment);
    }
}
