// @aztec/aztec.js/dest/contract/interaction_options.js
/**
 * Constant for explicitly not waiting for transaction confirmation.
 * We use this instead of false to avoid confusion with falsy checks.
 */ export const NO_WAIT = 'NO_WAIT';
/**
 * Transforms and cleans up the higher level SendInteractionOptions defined by the interaction into
 * SendOptions, which are the ones that can be serialized and forwarded to the wallet
 * @param options - The send interaction options with optional wait parameter
 * @returns The send options to forward to the wallet
 */ export function toSendOptions(options) {
    return {
        ...options,
        fee: {
            // If a payment method that includes gas settings was used,
            // try to reuse as much as possible while still allowing
            // manual override. CAREFUL: this can cause mismatches during proving
            gasSettings: {
                ...options.fee?.paymentMethod?.getGasSettings(),
                ...options.fee?.gasSettings
            }
        },
        wait: options.wait
    };
}
/**
 * Transforms and cleans up the higher level SimulateInteractionOptions defined by the interaction into
 * SimulateOptions, which are the ones that can be serialized and forwarded to the wallet
 */ export function toSimulateOptions(options) {
    return {
        ...options,
        fee: {
            // If a payment method that includes gas settings was used,
            // try to reuse as much as possible while still allowing
            // manual override. CAREFUL: this can cause mismatches during proving
            gasSettings: {
                ...options.fee?.paymentMethod?.getGasSettings(),
                ...options.fee?.gasSettings
            },
            estimateGas: options.fee?.estimateGas,
            estimatedGasPadding: options.fee?.estimatedGasPadding
        }
    };
}
/**
 * Transforms and cleans up the higher level ProfileInteractionOptions defined by the interaction into
 * ProfileOptions, which are the ones that can be serialized and forwarded to the wallet
 */ export function toProfileOptions(options) {
    return {
        ...toSimulateOptions(options),
        profileMode: options.profileMode,
        skipProofGeneration: options.skipProofGeneration
    };
}
