// @aztec/ethereum/dest/contracts/empire_base.js
import { Signature } from '@aztec/foundation/eth-signature';
import { EmpireBaseAbi } from '@aztec/l1-artifacts/EmpireBaseAbi';
import { encodeFunctionData } from 'viem';
export function encodeSignal(payload) {
    return encodeFunctionData({
        abi: EmpireBaseAbi,
        functionName: 'signal',
        args: [
            payload
        ]
    });
}
export function encodeSignalWithSignature(payload, signature) {
    return encodeFunctionData({
        abi: EmpireBaseAbi,
        functionName: 'signalWithSig',
        args: [
            payload,
            signature.toViemSignature()
        ]
    });
}
/**
 * Signs a signal proposal using EIP-712 typed data for use with signalWithSig
 * @param walletClient - The viem wallet client to sign with
 * @param payload - The payload address to signal
 * @param verifyingContract - The address of the EmpireBase contract
 * @param chainId - The chain ID where the contract is deployed
 * @param account - The account to sign with (optional if hoisted on wallet client)
 * @returns The EIP-712 signature
 */ export async function signSignalWithSig(signer, payload, slot, instance, verifyingContract, chainId) {
    const domain = {
        name: 'EmpireBase',
        version: '1',
        chainId,
        verifyingContract
    };
    const types = {
        EIP712Domain: [
            {
                name: 'name',
                type: 'string'
            },
            {
                name: 'version',
                type: 'string'
            },
            {
                name: 'chainId',
                type: 'uint256'
            },
            {
                name: 'verifyingContract',
                type: 'address'
            }
        ],
        Signal: [
            {
                name: 'payload',
                type: 'address'
            },
            {
                name: 'slot',
                type: 'uint256'
            },
            {
                name: 'instance',
                type: 'address'
            }
        ]
    };
    const message = {
        payload,
        slot: BigInt(slot),
        instance
    };
    const typedData = {
        domain,
        types,
        primaryType: 'Signal',
        message
    };
    return Signature.fromString(await signer(typedData));
}
