// @aztec/ethereum/dest/contracts/empire_slashing_proposer.js
import { SlotNumber } from '@aztec/foundation/branded-types';
import { EthAddress } from '@aztec/foundation/eth-address';
import { createLogger } from '@aztec/foundation/log';
import { retryUntil } from '@aztec/foundation/retry';
import { EmpireSlashingProposerAbi } from '@aztec/l1-artifacts/EmpireSlashingProposerAbi';
import EventEmitter from 'events';
import { encodeFunctionData, getContract } from 'viem';
import { FormattedViemError, tryExtractEvent } from '../utils.js';
import { encodeSignal, encodeSignalWithSignature, signSignalWithSig } from './empire_base.js';
export class ProposalAlreadyExecutedError extends Error {
    constructor(round){
        super(`Proposal already executed: ${round}`);
    }
}
export class EmpireSlashingProposerContract extends EventEmitter {
    client;
    logger;
    proposer;
    type;
    constructor(client, address){
        super(), this.client = client, this.logger = createLogger('ethereum:contracts:empire_slashing_proposer'), this.type = 'empire';
        this.proposer = getContract({
            address: typeof address === 'string' ? address : address.toString(),
            abi: EmpireSlashingProposerAbi,
            client
        });
    }
    get address() {
        return EthAddress.fromString(this.proposer.address);
    }
    getQuorumSize() {
        return this.proposer.read.QUORUM_SIZE();
    }
    getRoundSize() {
        return this.proposer.read.ROUND_SIZE();
    }
    getLifetimeInRounds() {
        return this.proposer.read.LIFETIME_IN_ROUNDS();
    }
    getExecutionDelayInRounds() {
        return this.proposer.read.EXECUTION_DELAY_IN_ROUNDS();
    }
    getCurrentRound() {
        return this.proposer.read.getCurrentRound();
    }
    computeRound(slot) {
        return this.proposer.read.computeRound([
            BigInt(slot)
        ]);
    }
    getInstance() {
        return this.proposer.read.getInstance();
    }
    async getRoundInfo(rollupAddress, round) {
        const result = await this.proposer.read.getRoundData([
            rollupAddress,
            round
        ]);
        const [signalCount, quorum] = await Promise.all([
            this.proposer.read.signalCount([
                rollupAddress,
                round,
                result.payloadWithMostSignals
            ]),
            this.getQuorumSize()
        ]);
        return {
            lastSignalSlot: SlotNumber.fromBigInt(result.lastSignalSlot),
            payloadWithMostSignals: result.payloadWithMostSignals,
            quorumReached: signalCount >= quorum,
            executed: result.executed
        };
    }
    getPayloadSignals(rollupAddress, round, payload) {
        return this.proposer.read.signalCount([
            rollupAddress,
            round,
            payload
        ]);
    }
    createSignalRequest(payload) {
        return {
            to: this.address.toString(),
            abi: EmpireSlashingProposerAbi,
            data: encodeSignal(payload)
        };
    }
    async createSignalRequestWithSignature(payload, slot, chainId, signerAddress, signer) {
        const signature = await signSignalWithSig(signer, payload, slot, await this.getInstance(), this.address.toString(), chainId);
        return {
            to: this.address.toString(),
            abi: EmpireSlashingProposerAbi,
            data: encodeSignalWithSignature(payload, signature)
        };
    }
    /** Checks if a payload was ever submitted to governance via submitRoundWinner. */ async hasPayloadBeenProposed(payload, fromBlock) {
        const events = await this.proposer.getEvents.PayloadSubmitted({
            payload
        }, {
            fromBlock,
            strict: true
        });
        return events.length > 0;
    }
    listenToSubmittablePayloads(callback) {
        return this.proposer.watchEvent.PayloadSubmittable({}, {
            strict: true,
            onLogs: (logs)=>{
                for (const log of logs){
                    const { payload, round } = log.args;
                    if (payload && round) {
                        callback({
                            payload,
                            round
                        });
                    }
                }
            }
        });
    }
    listenToPayloadSubmitted(callback) {
        return this.proposer.watchEvent.PayloadSubmitted({}, {
            onLogs: (logs)=>{
                for (const log of logs){
                    const { payload, round } = log.args;
                    if (round && payload) {
                        callback({
                            round,
                            payload
                        });
                    }
                }
            }
        });
    }
    listenToSignalCasted(callback) {
        return this.proposer.watchEvent.SignalCast({}, {
            onLogs: (logs)=>{
                for (const log of logs){
                    const { round, payload, signaler } = log.args;
                    if (round && payload && signaler) {
                        callback({
                            round,
                            payload,
                            signaler
                        });
                    }
                }
            }
        });
    }
    /** Creates an L1TxRequest to submit the round winner for the given round. */ buildExecuteRoundRequest(round) {
        return {
            to: this.address.toString(),
            abi: EmpireSlashingProposerAbi,
            data: encodeFunctionData({
                abi: EmpireSlashingProposerAbi,
                functionName: 'submitRoundWinner',
                args: [
                    round
                ]
            })
        };
    }
    /** Tries to extract a PayloadSubmitted event from the given logs. */ tryExtractPayloadSubmittedEvent(logs) {
        return tryExtractEvent(logs, this.address.toString(), EmpireSlashingProposerAbi, 'PayloadSubmitted');
    }
    /**
   * Wait for a round to be reached.
   *
   * @param round - The round to wait for.
   * @param pollingIntervalSeconds - The interval in seconds to poll for the round.
   * @returns True if the round was reached, false otherwise.
   */ waitForRound(round, pollingIntervalSeconds = 1) {
        return retryUntil(async ()=>{
            const currentRound = await this.proposer.read.getCurrentRound().catch((e)=>{
                this.logger.error('Error getting current round', e);
                return undefined;
            });
            return currentRound !== undefined && currentRound >= round;
        }, `Waiting for round ${round} to be reached`, 0, pollingIntervalSeconds).catch(()=>false);
    }
    async executeRound(txUtils, round) {
        if (typeof round === 'number') {
            round = BigInt(round);
        }
        const request = this.buildExecuteRoundRequest(round);
        const response = await txUtils.sendAndMonitorTransaction(request, {
            // Gas estimation is way off for this, likely because we are creating the contract/selector to call
            // for the actual slashing dynamically.
            gasLimitBufferPercentage: 50
        }).catch((err)=>{
            if (err instanceof FormattedViemError && err.message.includes('ProposalAlreadyExecuted')) {
                throw new ProposalAlreadyExecutedError(round);
            }
            throw err;
        });
        if (response.receipt.status === 'reverted') {
            const args = {
                abi: EmpireSlashingProposerAbi,
                functionName: 'submitRoundWinner',
                args: [
                    round
                ],
                address: this.address.toString()
            };
            const error = await txUtils.tryGetErrorFromRevertedTx(request.data, args, undefined, []);
            if (error?.includes('ProposalAlreadyExecuted')) {
                throw new ProposalAlreadyExecutedError(round);
            }
            const errorMessage = `Failed to execute round ${round}, TxHash: ${response.receipt.transactionHash}, Error: ${error ?? 'Unknown error'}`;
            throw new Error(errorMessage);
        }
        return response;
    }
}
