// @aztec/ethereum/dest/contracts/fee_asset_handler.js
import { EthAddress } from '@aztec/foundation/eth-address';
import { FeeAssetHandlerAbi } from '@aztec/l1-artifacts/FeeAssetHandlerAbi';
import { encodeFunctionData, getContract } from 'viem';
export class FeeAssetHandlerContract {
    client;
    address;
    constructor(client, address){
        this.client = client;
        if (address instanceof EthAddress) {
            address = address.toString();
        }
        this.address = EthAddress.fromString(address);
    }
    async getOwner() {
        const contract = getContract({
            abi: FeeAssetHandlerAbi,
            address: this.address.toString(),
            client: this.client
        });
        return EthAddress.fromString(await contract.read.owner());
    }
    getMintAmount() {
        const contract = getContract({
            abi: FeeAssetHandlerAbi,
            address: this.address.toString(),
            client: this.client
        });
        return contract.read.mintAmount();
    }
    mint(txUtils, recipient) {
        if (recipient instanceof EthAddress) {
            recipient = recipient.toString();
        }
        return txUtils.sendAndMonitorTransaction({
            to: this.address.toString(),
            abi: FeeAssetHandlerAbi,
            data: encodeFunctionData({
                abi: FeeAssetHandlerAbi,
                functionName: 'mint',
                args: [
                    recipient
                ]
            })
        });
    }
    setMintAmount(txUtils, amount) {
        return txUtils.sendAndMonitorTransaction({
            to: this.address.toString(),
            abi: FeeAssetHandlerAbi,
            data: encodeFunctionData({
                abi: FeeAssetHandlerAbi,
                functionName: 'setMintAmount',
                args: [
                    amount
                ]
            })
        });
    }
}
