// @aztec/ethereum/dest/contracts/gse.js
import { EthAddress } from '@aztec/foundation/eth-address';
import { GSEAbi } from '@aztec/l1-artifacts/GSEAbi';
import { bn254 } from '@noble/curves/bn254';
import { getContract } from 'viem';
export class GSEContract {
    client;
    address;
    gse;
    constructor(client, address){
        this.client = client;
        if (address instanceof EthAddress) {
            address = address.toString();
        }
        this.address = EthAddress.fromString(address);
        this.gse = getContract({
            address,
            abi: GSEAbi,
            client
        });
    }
    async getOwner() {
        return EthAddress.fromString(await this.gse.read.owner());
    }
    async getGovernance() {
        return EthAddress.fromString(await this.gse.read.getGovernance());
    }
    getAttestersFromIndicesAtTime(instance, ts, indices) {
        if (instance instanceof EthAddress) {
            instance = instance.toString();
        }
        return this.gse.read.getAttestersFromIndicesAtTime([
            instance,
            ts,
            indices
        ]);
    }
    async getRegistrationDigest(publicKey) {
        const affinePublicKey = publicKey.toAffine();
        const g1PointDigest = await this.gse.read.getRegistrationDigest([
            {
                x: affinePublicKey.x,
                y: affinePublicKey.y
            }
        ]);
        return bn254.G1.ProjectivePoint.fromAffine(g1PointDigest);
    }
    async makeRegistrationTuple(privateKey) {
        const publicKeyG1 = bn254.G1.ProjectivePoint.BASE.multiply(privateKey);
        const digest = await this.getRegistrationDigest(publicKeyG1);
        const signature = digest.multiply(privateKey);
        const publicKeyG2 = bn254.G2.ProjectivePoint.BASE.multiply(privateKey);
        const publicKeyG1Affine = publicKeyG1.toAffine();
        const signatureAffine = signature.toAffine();
        const publicKeyG2Affine = publicKeyG2.toAffine();
        return {
            publicKeyInG1: {
                x: publicKeyG1Affine.x,
                y: publicKeyG1Affine.y
            },
            publicKeyInG2: {
                x0: publicKeyG2Affine.x.c0,
                x1: publicKeyG2Affine.x.c1,
                y0: publicKeyG2Affine.y.c0,
                y1: publicKeyG2Affine.y.c1
            },
            proofOfPossession: {
                x: signatureAffine.x,
                y: signatureAffine.y
            }
        };
    }
}
