// @aztec/ethereum/dest/contracts/inbox.js
import { CheckpointNumber } from '@aztec/foundation/branded-types';
import { Buffer16, Buffer32 } from '@aztec/foundation/buffer';
import { Fr } from '@aztec/foundation/curves/bn254';
import { EthAddress } from '@aztec/foundation/eth-address';
import { InboxAbi } from '@aztec/l1-artifacts/InboxAbi';
import { getContract } from 'viem';
import { getPublicClient } from '../client.js';
import { checkBlockTag } from './utils.js';
export class InboxContract {
    client;
    inbox;
    static getFromL1ContractsValues(deployL1ContractsValues) {
        const { l1Client, l1ContractAddresses: { inboxAddress } } = deployL1ContractsValues;
        return new InboxContract(l1Client, inboxAddress.toString());
    }
    static getFromConfig(config) {
        const client = getPublicClient(config);
        const address = config.l1Contracts.inboxAddress.toString();
        return new InboxContract(client, address);
    }
    constructor(client, address){
        this.client = client;
        if (address instanceof EthAddress) {
            address = address.toString();
        }
        this.inbox = getContract({
            address,
            abi: InboxAbi,
            client
        });
    }
    get address() {
        return this.inbox.address;
    }
    getContract() {
        return this.inbox;
    }
    async getLag(opts = {}) {
        await checkBlockTag(opts.blockNumber, this.client);
        return await this.inbox.read.LAG(opts);
    }
    async getState(opts = {}) {
        await checkBlockTag(opts.blockNumber, this.client);
        const state = await this.inbox.read.getState(opts);
        return {
            totalMessagesInserted: state.totalMessagesInserted,
            messagesRollingHash: Buffer16.fromString(state.rollingHash),
            treeInProgress: state.inProgress
        };
    }
    /** Fetches MessageSent events within the given block range. */ async getMessageSentEvents(fromBlock, toBlock) {
        const logs = await this.inbox.getEvents.MessageSent({}, {
            fromBlock,
            toBlock
        });
        return logs.filter((log)=>log.blockNumber >= fromBlock && log.blockNumber <= toBlock).map((log)=>this.mapMessageSentLog(log));
    }
    /** Fetches MessageSent events for a specific message hash within the given block range. */ async getMessageSentEventByHash(hash, fromBlock, toBlock) {
        const logs = await this.inbox.getEvents.MessageSent({
            hash
        }, {
            fromBlock,
            toBlock
        });
        return logs.map((log)=>this.mapMessageSentLog(log));
    }
    mapMessageSentLog(log) {
        return {
            l1BlockNumber: log.blockNumber,
            l1BlockHash: Buffer32.fromString(log.blockHash),
            l1TransactionHash: log.transactionHash,
            args: {
                index: log.args.index,
                leaf: Fr.fromString(log.args.hash),
                checkpointNumber: CheckpointNumber.fromBigInt(log.args.checkpointNumber),
                rollingHash: Buffer16.fromString(log.args.rollingHash)
            }
        };
    }
}
