// @aztec/ethereum/dest/contracts/outbox.js
import { EthAddress } from '@aztec/foundation/eth-address';
import { OutboxAbi } from '@aztec/l1-artifacts/OutboxAbi';
import { encodeAbiParameters, getContract, hexToBigInt, keccak256 } from 'viem';
import { getPublicClient } from '../client.js';
import { isExtendedClient } from '../types.js';
export class OutboxContract {
    client;
    outbox;
    static getFromL1ContractsValues(deployL1ContractsValues) {
        const { l1Client, l1ContractAddresses: { outboxAddress } } = deployL1ContractsValues;
        return new OutboxContract(l1Client, outboxAddress.toString());
    }
    static getFromConfig(config) {
        const client = getPublicClient(config);
        const address = config.l1Contracts.outboxAddress.toString();
        return new OutboxContract(client, address);
    }
    static getEpochRootStorageSlot(epoch) {
        return hexToBigInt(keccak256(encodeAbiParameters([
            {
                type: 'uint256'
            },
            {
                type: 'uint256'
            }
        ], [
            BigInt(epoch),
            0n
        ])));
    }
    constructor(client, address){
        this.client = client;
        if (address instanceof EthAddress) {
            address = address.toString();
        }
        this.outbox = getContract({
            address,
            abi: OutboxAbi,
            client
        });
    }
    get address() {
        return this.outbox.address;
    }
    getContract() {
        return this.outbox;
    }
    hasMessageBeenConsumedAtEpoch(epoch, leafId) {
        return this.outbox.read.hasMessageBeenConsumedAtEpoch([
            BigInt(epoch),
            leafId
        ]);
    }
    getRootData(epoch) {
        return this.outbox.read.getRootData([
            BigInt(epoch)
        ]);
    }
    consume(message, epoch, leafIndex, path) {
        const wallet = this.assertWallet();
        return wallet.write.consume([
            message,
            BigInt(epoch),
            leafIndex,
            path
        ]);
    }
    async getMessageConsumedEvents(l1BlockHash) {
        const events = await this.outbox.getEvents.MessageConsumed({}, {
            blockHash: l1BlockHash,
            strict: true
        });
        return events.map((event)=>({
                epoch: event.args.epoch,
                root: event.args.root,
                messageHash: event.args.messageHash,
                leafId: event.args.leafId
            }));
    }
    assertWallet() {
        if (!isExtendedClient(this.client)) {
            throw new Error('Wallet client is required for this operation');
        }
        return this.outbox;
    }
}
