// @aztec/bb.js/dest/node-cjs/bb_backends/wasm.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BarretenbergWasmAsyncBackend = exports.BarretenbergWasmSyncBackend = void 0;
const index_js_1 = require("../barretenberg_wasm/barretenberg_wasm_main/index.js");
const index_js_2 = require("../barretenberg_wasm/index.js");
const index_js_3 = require("../barretenberg_wasm/barretenberg_wasm_main/factory/node/index.js");
const index_js_4 = require("../barretenberg_wasm/helpers/index.js");
const comlink_1 = require("comlink");
/**
 * Synchronous WASM backend that wraps BarretenbergWasmMain.
 * Encapsulates all WASM initialization and memory management.
 */
class BarretenbergWasmSyncBackend {
    wasm;
    constructor(wasm) {
        this.wasm = wasm;
    }
    /**
     * Create and initialize a synchronous WASM backend.
     * @param wasmPath Optional path to WASM files
     * @param logger Optional logging function
     */
    static async new(wasmPath, logger) {
        const wasm = new index_js_1.BarretenbergWasmMain();
        const { module, threads } = await (0, index_js_2.fetchModuleAndThreads)(1, wasmPath, logger);
        await wasm.init(module, threads, logger);
        return new BarretenbergWasmSyncBackend(wasm);
    }
    call(inputBuffer) {
        return this.wasm.cbindCall('bbapi', inputBuffer);
    }
    destroy() {
        // BarretenbergWasmMain has async destroy, but for sync API we call it without awaiting
        // This is consistent with the synchronous semantics expected by the caller
        void this.wasm.destroy();
    }
}
exports.BarretenbergWasmSyncBackend = BarretenbergWasmSyncBackend;
/**
 * Asynchronous WASM backend that supports both direct WASM and worker-based modes.
 *
 * Worker mode (default): Runs WASM on a worker thread to avoid blocking the main thread. Used in browsers.
 * Direct mode: Runs WASM directly on the calling thread. Used by node.js for better performance.
 */
class BarretenbergWasmAsyncBackend {
    wasm;
    worker;
    constructor(wasm, worker) {
        this.wasm = wasm;
        this.worker = worker;
    }
    /**
     * Create and initialize an asynchronous WASM backend.
     * @param options.threads Number of threads (defaults to hardware max, up to 32 for parallel proving)
     * @param options.wasmPath Optional path to WASM files
     * @param options.logger Optional logging function
     * @param options.memory Optional initial and maximum memory configuration
     * @param options.useWorker Run on worker thread (default: true for browser safety)
     */
    static async new(options = {}) {
        // Default to worker mode for browser safety
        const useWorker = options.useWorker ?? true;
        if (useWorker) {
            // Worker-based mode: runs on worker thread (browser-safe)
            const worker = await (0, index_js_3.createMainWorker)();
            const wasm = (0, index_js_4.getRemoteBarretenbergWasm)(worker);
            const { module, threads } = await (0, index_js_2.fetchModuleAndThreads)(options.threads, options.wasmPath, options.logger);
            await wasm.init(module, threads, (0, comlink_1.proxy)(options.logger ?? (() => { })), options.memory?.initial, options.memory?.maximum);
            return new BarretenbergWasmAsyncBackend(wasm, worker);
        }
        else {
            // Direct mode: runs on calling thread (faster but blocks thread)
            const wasm = new index_js_1.BarretenbergWasmMain();
            const { module, threads } = await (0, index_js_2.fetchModuleAndThreads)(options.threads, options.wasmPath, options.logger);
            await wasm.init(module, threads, options.logger, options.memory?.initial, options.memory?.maximum);
            return new BarretenbergWasmAsyncBackend(wasm);
        }
    }
    async call(inputBuffer) {
        return this.wasm.cbindCall('bbapi', inputBuffer);
    }
    async destroy() {
        await this.wasm.destroy();
        if (this.worker) {
            await this.worker.terminate();
        }
    }
}
exports.BarretenbergWasmAsyncBackend = BarretenbergWasmAsyncBackend;
