// @aztec/noir-protocol-circuits-types/dest/execution/client.js
import { MEGA_VK_LENGTH_IN_FIELDS } from '@aztec/constants';
import { pushTestData } from '@aztec/foundation/testing';
import { abiDecode, abiEncode } from '@aztec/noir-noirc_abi';
import { mapPaddedSideEffectAmountsToNoir, mapPaddedSideEffectsToNoir, mapPrivateCallDataToNoir, mapPrivateCircuitPublicInputsToNoir, mapPrivateKernelCircuitPublicInputsFromNoir, mapPrivateKernelCircuitPublicInputsToNoir, mapPrivateKernelDataToNoir, mapPrivateKernelResetHintsToNoir, mapPrivateKernelTailCircuitPublicInputsForPublicFromNoir, mapPrivateKernelTailCircuitPublicInputsForRollupFromNoir, mapTxRequestToNoir } from '../conversion/client.js';
import { mapFieldToNoir, mapNumberToNoir, mapPrivateToPublicKernelCircuitPublicInputsToNoir, mapPrivateToRollupKernelCircuitPublicInputsToNoir, mapProtocolContractsToNoir, mapU64ToNoir, mapVkDataToNoir } from '../conversion/common.js';
/* eslint-disable camelcase */ /**
 * Converts the inputs of the private kernel init circuit into a witness map
 * @param inputs - The private kernel inputs.
 * @returns The witness map
 */ export function convertPrivateKernelInitInputsToWitnessMapWithAbi(privateKernelInitCircuitPrivateInputs, privateKernelInitAbi) {
    const mapped = {
        tx_request: mapTxRequestToNoir(privateKernelInitCircuitPrivateInputs.txRequest),
        vk_tree_root: mapFieldToNoir(privateKernelInitCircuitPrivateInputs.vkTreeRoot),
        protocol_contracts: mapProtocolContractsToNoir(privateKernelInitCircuitPrivateInputs.protocolContracts),
        private_call: mapPrivateCallDataToNoir(privateKernelInitCircuitPrivateInputs.privateCall),
        is_private_only: privateKernelInitCircuitPrivateInputs.isPrivateOnly,
        first_nullifier_hint: mapFieldToNoir(privateKernelInitCircuitPrivateInputs.firstNullifierHint),
        revertible_counter_hint: mapNumberToNoir(privateKernelInitCircuitPrivateInputs.revertibleCounterHint),
        app_public_inputs: mapPrivateCircuitPublicInputsToNoir(privateKernelInitCircuitPrivateInputs.privateCall.publicInputs)
    };
    pushTestData('private-kernel-init', mapped);
    const initialWitnessMap = abiEncode(privateKernelInitAbi, mapped);
    return initialWitnessMap;
}
/**
 * Converts the inputs of the private kernel inner circuit into a witness map
 * @param inputs - The private kernel inputs.
 * @returns The witness map
 */ export function convertPrivateKernelInnerInputsToWitnessMapWithAbi(privateKernelInnerCircuitPrivateInputs, privateKernelInnerAbi) {
    const mapped = {
        previous_kernel: mapPrivateKernelDataToNoir(privateKernelInnerCircuitPrivateInputs.previousKernel),
        previous_kernel_public_inputs: mapPrivateKernelCircuitPublicInputsToNoir(privateKernelInnerCircuitPrivateInputs.previousKernel.publicInputs),
        private_call: mapPrivateCallDataToNoir(privateKernelInnerCircuitPrivateInputs.privateCall),
        app_public_inputs: mapPrivateCircuitPublicInputsToNoir(privateKernelInnerCircuitPrivateInputs.privateCall.publicInputs)
    };
    pushTestData('private-kernel-inner', mapped);
    const initialWitnessMap = abiEncode(privateKernelInnerAbi, mapped);
    return initialWitnessMap;
}
/**
 * Converts the inputs of the private kernel reset circuit into a witness map
 * @param inputs - The private kernel inputs.
 * @returns The witness map
 */ export function convertPrivateKernelResetInputsToWitnessMapWithAbi(privateKernelResetCircuitPrivateInputs, resetAbi) {
    const mapped = {
        previous_kernel: mapPrivateKernelDataToNoir(privateKernelResetCircuitPrivateInputs.previousKernel),
        previous_kernel_public_inputs: mapPrivateKernelCircuitPublicInputsToNoir(privateKernelResetCircuitPrivateInputs.previousKernel.publicInputs),
        padded_side_effects: mapPaddedSideEffectsToNoir(privateKernelResetCircuitPrivateInputs.paddedSideEffects),
        hints: mapPrivateKernelResetHintsToNoir(privateKernelResetCircuitPrivateInputs.hints)
    };
    const initialWitnessMap = abiEncode(resetAbi, mapped);
    return initialWitnessMap;
}
/**
 * Converts the inputs of the private kernel tail circuit into a witness map
 * @param inputs - The private kernel inputs.
 * @returns The witness map
 */ export function convertPrivateKernelTailInputsToWitnessMapWithAbi(privateKernelTailCircuitPrivateInputs, privateKernelTailAbi) {
    const mapped = {
        previous_kernel: mapPrivateKernelDataToNoir(privateKernelTailCircuitPrivateInputs.previousKernel),
        previous_kernel_public_inputs: mapPrivateKernelCircuitPublicInputsToNoir(privateKernelTailCircuitPrivateInputs.previousKernel.publicInputs),
        expiration_timestamp_upper_bound: mapU64ToNoir(privateKernelTailCircuitPrivateInputs.expirationTimestampUpperBound)
    };
    pushTestData('private-kernel-tail', mapped);
    const initialWitnessMap = abiEncode(privateKernelTailAbi, mapped);
    return initialWitnessMap;
}
/**
 * Converts the inputs of the private kernel tail to public circuit into a witness map
 * @param inputs - The private kernel inputs.
 * @returns The witness map
 */ export function convertPrivateKernelTailToPublicInputsToWitnessMapWithAbi(privateKernelTailToPublicCircuitPrivateInputs, privateKernelTailToPublicAbi) {
    const mapped = {
        previous_kernel: mapPrivateKernelDataToNoir(privateKernelTailToPublicCircuitPrivateInputs.previousKernel),
        previous_kernel_public_inputs: mapPrivateKernelCircuitPublicInputsToNoir(privateKernelTailToPublicCircuitPrivateInputs.previousKernel.publicInputs),
        padded_side_effect_amounts: mapPaddedSideEffectAmountsToNoir(privateKernelTailToPublicCircuitPrivateInputs.paddedSideEffectAmounts),
        expiration_timestamp_upper_bound: mapU64ToNoir(privateKernelTailToPublicCircuitPrivateInputs.expirationTimestampUpperBound)
    };
    pushTestData('private-kernel-tail-to-public', mapped);
    const initialWitnessMap = abiEncode(privateKernelTailToPublicAbi, mapped);
    return initialWitnessMap;
}
export function convertHidingKernelToRollupInputsToWitnessMapWithAbi(inputs, abi) {
    const mapped = {
        previous_kernel_public_inputs: mapPrivateToRollupKernelCircuitPublicInputsToNoir(inputs.previousKernelPublicInputs),
        previous_kernel_vk_data: mapVkDataToNoir(inputs.previousKernelVkData, MEGA_VK_LENGTH_IN_FIELDS)
    };
    return abiEncode(abi, mapped);
}
export function convertHidingKernelPublicInputsToWitnessMapWithAbi(inputs, abi) {
    const mapped = {
        previous_kernel_public_inputs: mapPrivateToPublicKernelCircuitPublicInputsToNoir(inputs.previousKernelPublicInputs),
        previous_kernel_vk_data: mapVkDataToNoir(inputs.previousKernelVkData, MEGA_VK_LENGTH_IN_FIELDS)
    };
    return abiEncode(abi, mapped);
}
/**
 * Converts the outputs of the private kernel init circuit from a witness map.
 * @param outputs - The private kernel outputs as a witness map.
 * @returns The public inputs.
 */ export function convertPrivateKernelInitOutputsFromWitnessMapWithAbi(outputs, privateKernelInitAbi) {
    // Decode the witness map into two fields, the return values and the inputs
    const decodedInputs = abiDecode(privateKernelInitAbi, outputs);
    // Cast the inputs as the return type
    const returnType = decodedInputs.return_value;
    return mapPrivateKernelCircuitPublicInputsFromNoir(returnType);
}
/**
 * Converts the outputs of the private kernel inner circuit from a witness map.
 * @param outputs - The private kernel outputs as a witness map.
 * @returns The public inputs.
 */ export function convertPrivateKernelInnerOutputsFromWitnessMapWithAbi(outputs, privateKernelInnerAbi) {
    // Decode the witness map into two fields, the return values and the inputs
    const decodedInputs = abiDecode(privateKernelInnerAbi, outputs);
    // Cast the inputs as the return type
    const returnType = decodedInputs.return_value;
    return mapPrivateKernelCircuitPublicInputsFromNoir(returnType);
}
/**
 * Converts the outputs of the private kernel reset circuit from a witness map.
 * @param outputs - The private kernel outputs as a witness map.
 * @returns The public inputs.
 */ export function convertPrivateKernelResetOutputsFromWitnessMapWithAbi(outputs, resetAbi) {
    // Decode the witness map into two fields, the return values and the inputs
    const decodedInputs = abiDecode(resetAbi, outputs);
    // Cast the inputs as the return type
    const returnType = decodedInputs.return_value;
    return mapPrivateKernelCircuitPublicInputsFromNoir(returnType);
}
/**
 * Converts the outputs of the private kernel tail circuit from a witness map.
 * @param outputs - The private kernel outputs as a witness map.
 * @returns The public inputs.
 */ export function convertPrivateKernelTailOutputsFromWitnessMapWithAbi(outputs, privateKernelTailAbi) {
    // Decode the witness map into two fields, the return values and the inputs
    const decodedInputs = abiDecode(privateKernelTailAbi, outputs);
    // Cast the inputs as the return type
    const returnType = decodedInputs.return_value;
    return mapPrivateKernelTailCircuitPublicInputsForRollupFromNoir(returnType);
}
/**
 * Converts the outputs of the private kernel tail for public circuit from a witness map.
 * @param outputs - The private kernel outputs as a witness map.
 * @returns The public inputs.
 */ export function convertPrivateKernelTailForPublicOutputsFromWitnessMapWithAbi(outputs, privateKernelTailToPublicAbi) {
    // Decode the witness map into two fields, the return values and the inputs
    const decodedInputs = abiDecode(privateKernelTailToPublicAbi, outputs);
    // Cast the inputs as the return type
    const returnType = decodedInputs.return_value;
    return mapPrivateKernelTailCircuitPublicInputsForPublicFromNoir(returnType);
}
