// @aztec/noir-protocol-circuits-types/dest/execution/server.js
import { pushTestData } from '@aztec/foundation/testing';
import { abiDecode, abiEncode } from '@aztec/noir-noirc_abi';
import { ServerCircuitArtifacts, SimulatedServerCircuitArtifacts } from '../artifacts/server.js';
import { mapProtocolArtifactNameToCircuitName } from '../artifacts/types.js';
import { mapBlockMergeRollupPrivateInputsToNoir, mapBlockRollupPublicInputsFromNoir, mapBlockRootEmptyTxFirstRollupPrivateInputsToNoir, mapBlockRootFirstRollupPrivateInputsToNoir, mapBlockRootRollupPrivateInputsToNoir, mapBlockRootSingleTxFirstRollupPrivateInputsToNoir, mapBlockRootSingleTxRollupPrivateInputsToNoir, mapCheckpointMergeRollupPrivateInputsToNoir, mapCheckpointRollupPublicInputsFromNoir, mapCheckpointRootRollupPrivateInputsToNoir, mapCheckpointRootSingleBlockRollupPrivateInputsToNoir, mapParityBasePrivateInputsToNoir, mapParityPublicInputsFromNoir, mapParityRootPrivateInputsToNoir, mapPrivateTxBaseRollupPrivateInputsToNoir, mapPublicChonkVerifierPrivateInputsToNoir, mapPublicChonkVerifierPublicInputsFromNoir, mapPublicTxBaseRollupPrivateInputsToNoir, mapRootRollupPrivateInputsToNoir, mapRootRollupPublicInputsFromNoir, mapTxMergeRollupPrivateInputsToNoir, mapTxRollupPublicInputsFromNoir } from '../conversion/server.js';
export { mapAvmCircuitPublicInputsToNoir } from '../conversion/server.js';
/**
 * Converts the inputs of the base parity circuit into a witness map.
 * @param inputs - The base parity inputs.
 * @returns The witness map
 */ export function convertParityBasePrivateInputsToWitnessMap(inputs, simulated = false) {
    return convertPrivateInputsToWitnessMap('ParityBaseArtifact', mapParityBasePrivateInputsToNoir(inputs), simulated);
}
/**
 * Converts the inputs of the root parity circuit into a witness map.
 * @param inputs - The root parity inputs.
 * @returns The witness map
 */ export function convertParityRootPrivateInputsToWitnessMap(inputs, simulated = false) {
    return convertPrivateInputsToWitnessMap('ParityRootArtifact', mapParityRootPrivateInputsToNoir(inputs), simulated);
}
export function convertPublicChonkVerifierPrivateInputsToWitnessMap(inputs, simulated = false) {
    return convertPrivateInputsToWitnessMap('PublicChonkVerifier', mapPublicChonkVerifierPrivateInputsToNoir(inputs), simulated);
}
export function convertPrivateTxBaseRollupPrivateInputsToWitnessMap(inputs, simulated = false) {
    return convertPrivateInputsToWitnessMap('PrivateTxBaseRollupArtifact', mapPrivateTxBaseRollupPrivateInputsToNoir(inputs), simulated);
}
export function convertPublicTxBaseRollupPrivateInputsToWitnessMap(inputs, simulated = false) {
    return convertPrivateInputsToWitnessMap('PublicTxBaseRollupArtifact', mapPublicTxBaseRollupPrivateInputsToNoir(inputs), simulated);
}
/**
 * Converts the inputs of the merge rollup circuit into a witness map.
 * @param inputs - The merge rollup inputs.
 * @returns The witness map
 */ export function convertTxMergeRollupPrivateInputsToWitnessMap(inputs, simulated = false) {
    return convertPrivateInputsToWitnessMap('TxMergeRollupArtifact', mapTxMergeRollupPrivateInputsToNoir(inputs), simulated);
}
export function convertBlockRootFirstRollupPrivateInputsToWitnessMap(inputs, simulated = false) {
    return convertPrivateInputsToWitnessMap('BlockRootFirstRollupArtifact', mapBlockRootFirstRollupPrivateInputsToNoir(inputs), simulated);
}
export function convertBlockRootSingleTxFirstRollupPrivateInputsToWitnessMap(inputs, simulated = false) {
    return convertPrivateInputsToWitnessMap('BlockRootSingleTxFirstRollupArtifact', mapBlockRootSingleTxFirstRollupPrivateInputsToNoir(inputs), simulated);
}
export function convertBlockRootEmptyTxFirstRollupPrivateInputsToWitnessMap(inputs, simulated = false) {
    return convertPrivateInputsToWitnessMap('BlockRootEmptyTxFirstRollupArtifact', mapBlockRootEmptyTxFirstRollupPrivateInputsToNoir(inputs), simulated);
}
export function convertBlockRootRollupPrivateInputsToWitnessMap(inputs, simulated = false) {
    return convertPrivateInputsToWitnessMap('BlockRootRollupArtifact', mapBlockRootRollupPrivateInputsToNoir(inputs), simulated);
}
export function convertBlockRootSingleTxRollupPrivateInputsToWitnessMap(inputs, simulated = false) {
    return convertPrivateInputsToWitnessMap('BlockRootSingleTxRollupArtifact', mapBlockRootSingleTxRollupPrivateInputsToNoir(inputs), simulated);
}
export function convertBlockMergeRollupPrivateInputsToWitnessMap(inputs, simulated = false) {
    return convertPrivateInputsToWitnessMap('BlockMergeRollupArtifact', mapBlockMergeRollupPrivateInputsToNoir(inputs), simulated);
}
export function convertCheckpointRootRollupPrivateInputsToWitnessMap(inputs, simulated = false) {
    return convertPrivateInputsToWitnessMap('CheckpointRootRollupArtifact', mapCheckpointRootRollupPrivateInputsToNoir(inputs), simulated);
}
export function convertCheckpointRootSingleBlockRollupPrivateInputsToWitnessMap(inputs, simulated = false) {
    return convertPrivateInputsToWitnessMap('CheckpointRootSingleBlockRollupArtifact', mapCheckpointRootSingleBlockRollupPrivateInputsToNoir(inputs), simulated);
}
export function convertCheckpointPaddingRollupPrivateInputsToWitnessMap(_inputs, _simulated = false) {
    // Checkpoint padding does not have any private inputs.
    return new Map();
}
export function convertCheckpointMergeRollupPrivateInputsToWitnessMap(inputs, simulated = false) {
    return convertPrivateInputsToWitnessMap('CheckpointMergeRollupArtifact', mapCheckpointMergeRollupPrivateInputsToNoir(inputs), simulated);
}
/**
 * Converts the inputs of the root rollup circuit into a witness map.
 * @param inputs - The root rollup inputs.
 * @returns The witness map
 */ export function convertRootRollupPrivateInputsToWitnessMap(inputs) {
    return convertPrivateInputsToWitnessMap('RootRollupArtifact', mapRootRollupPrivateInputsToNoir(inputs));
}
export function convertPublicChonkVerifierOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('PublicChonkVerifier', outputs, simulated);
    return mapPublicChonkVerifierPublicInputsFromNoir(publicInputs);
}
export function convertPrivateTxBaseRollupOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('PrivateTxBaseRollupArtifact', outputs, simulated);
    return mapTxRollupPublicInputsFromNoir(publicInputs);
}
export function convertPublicTxBaseRollupOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('PublicTxBaseRollupArtifact', outputs, simulated);
    return mapTxRollupPublicInputsFromNoir(publicInputs);
}
export function convertTxMergeRollupOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('TxMergeRollupArtifact', outputs, simulated);
    return mapTxRollupPublicInputsFromNoir(publicInputs);
}
export function convertBlockRootFirstRollupOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('BlockRootFirstRollupArtifact', outputs, simulated);
    return mapBlockRollupPublicInputsFromNoir(publicInputs);
}
export function convertBlockRootSingleTxFirstRollupOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('BlockRootSingleTxFirstRollupArtifact', outputs, simulated);
    return mapBlockRollupPublicInputsFromNoir(publicInputs);
}
export function convertBlockRootEmptyTxFirstRollupOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('BlockRootEmptyTxFirstRollupArtifact', outputs, simulated);
    return mapBlockRollupPublicInputsFromNoir(publicInputs);
}
export function convertBlockRootRollupOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('BlockRootRollupArtifact', outputs, simulated);
    return mapBlockRollupPublicInputsFromNoir(publicInputs);
}
export function convertBlockRootSingleTxRollupOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('BlockRootSingleTxRollupArtifact', outputs, simulated);
    return mapBlockRollupPublicInputsFromNoir(publicInputs);
}
export function convertBlockMergeRollupOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('BlockMergeRollupArtifact', outputs, simulated);
    return mapBlockRollupPublicInputsFromNoir(publicInputs);
}
export function convertCheckpointRootRollupOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('CheckpointRootRollupArtifact', outputs, simulated);
    return mapCheckpointRollupPublicInputsFromNoir(publicInputs);
}
export function convertCheckpointRootSingleBlockRollupOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('CheckpointRootSingleBlockRollupArtifact', outputs, simulated);
    return mapCheckpointRollupPublicInputsFromNoir(publicInputs);
}
export function convertCheckpointPaddingRollupOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('CheckpointPaddingRollupArtifact', outputs, simulated);
    return mapCheckpointRollupPublicInputsFromNoir(publicInputs);
}
export function convertCheckpointMergeRollupOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('CheckpointMergeRollupArtifact', outputs, simulated);
    return mapCheckpointRollupPublicInputsFromNoir(publicInputs);
}
/**
 * Converts the outputs of the root rollup circuit from a witness map.
 * @param outputs - The root rollup outputs as a witness map.
 * @returns The public inputs.
 */ export function convertRootRollupOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('RootRollupArtifact', outputs, simulated);
    return mapRootRollupPublicInputsFromNoir(publicInputs);
}
/**
 * Converts the outputs of the base parity circuit from a witness map.
 * @param outputs - The base parity outputs as a witness map.
 * @returns The public inputs.
 */ export function convertParityBaseOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('ParityBaseArtifact', outputs, simulated);
    return mapParityPublicInputsFromNoir(publicInputs);
}
/**
 * Converts the outputs of the root parity circuit from a witness map.
 * @param outputs - The root parity outputs as a witness map.
 * @returns The public inputs.
 */ export function convertParityRootOutputsFromWitnessMap(outputs, simulated = false) {
    const publicInputs = convertOutputsFromWitnessMap('ParityRootArtifact', outputs, simulated);
    return mapParityPublicInputsFromNoir(publicInputs);
}
function convertPrivateInputsToWitnessMap(artifactName, inputs, simulated = false) {
    const circuitName = mapProtocolArtifactNameToCircuitName(artifactName);
    pushTestData(circuitName, {
        inputs
    });
    const abi = (simulated ? SimulatedServerCircuitArtifacts : ServerCircuitArtifacts)[artifactName].abi;
    return abiEncode(abi, {
        inputs
    });
}
function convertOutputsFromWitnessMap(artifactName, outputs, simulated) {
    const abi = (simulated ? SimulatedServerCircuitArtifacts : ServerCircuitArtifacts)[artifactName].abi;
    // Decode the witness map into two fields, the return values and the inputs.
    const decoded = abiDecode(abi, outputs);
    // Cast the return value as the return type.
    return decoded.return_value;
}
