// @aztec/pxe/dest/storage/note_store/stored_note.js
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { NoteDao } from '@aztec/stdlib/note';
export class StoredNote {
    noteDao;
    scopes;
    _nullifiedAt;
    constructor(noteDao, scopes, _nullifiedAt = undefined){
        this.noteDao = noteDao;
        this.scopes = scopes;
        this._nullifiedAt = _nullifiedAt;
    }
    static fromBuffer(buffer) {
        const reader = BufferReader.asReader(buffer);
        const noteDao = NoteDao.fromBuffer(reader);
        const scopes = reader.readVector({
            fromBuffer: (r)=>r.readString()
        });
        const nullifiedAtRaw = reader.readNumber();
        const nullifiedAt = nullifiedAtRaw === 0 ? undefined : nullifiedAtRaw;
        return new StoredNote(noteDao, new Set(scopes), nullifiedAt);
    }
    toBuffer() {
        const scopesArray = [
            ...this.scopes
        ];
        return serializeToBuffer(this.noteDao, scopesArray.length, ...scopesArray, this._nullifiedAt ?? 0);
    }
    addScope(scope) {
        this.scopes.add(scope);
    }
    markAsNullified(blockNumber) {
        this._nullifiedAt = blockNumber;
    }
    markAsActive() {
        this._nullifiedAt = undefined;
    }
    isNullified() {
        return this._nullifiedAt !== undefined;
    }
    get nullifiedAt() {
        return this._nullifiedAt;
    }
}
