// @aztec/noir-protocol-circuits-types/dest/utils/private_kernel_reset.js
import { pushTestData } from '@aztec/foundation/testing';
import { privateKernelResetDimensionNames } from '@aztec/stdlib/kernel';
import { mapPaddedSideEffectsToNoir, mapPrivateKernelCircuitPublicInputsToNoir, mapPrivateKernelDataToNoir, mapPrivateKernelResetHintsToNoir } from '../conversion/client.js';
import { PrivateKernelResetArtifactFileNames } from '../private_kernel_reset_types.js';
export function createPrivateKernelResetTag(dimensions) {
    return privateKernelResetDimensionNames.map((name)=>dimensions[name]).join('_');
}
export function getPrivateKernelResetArtifactName(dimensions) {
    const tag = createPrivateKernelResetTag(dimensions);
    const name = `PrivateKernelResetArtifact_${tag}`;
    if (!PrivateKernelResetArtifactFileNames[name]) {
        throw new Error(`Unknown private reset artifact: ${name}`);
    }
    return name;
}
/**
 * TODO: This is a hack so we can write full reset inputs to a Prover.toml.
 * Ideally we remove it in favour of adding a test that runs a full reset.
 */ export function updateResetCircuitSampleInputs(privateKernelResetCircuitPrivateInputs) {
    /* eslint-disable camelcase */ const inputs = {
        previous_kernel: mapPrivateKernelDataToNoir(privateKernelResetCircuitPrivateInputs.previousKernel),
        previous_kernel_public_inputs: mapPrivateKernelCircuitPublicInputsToNoir(privateKernelResetCircuitPrivateInputs.previousKernel.publicInputs),
        padded_side_effects: mapPaddedSideEffectsToNoir(privateKernelResetCircuitPrivateInputs.paddedSideEffects),
        hints: mapPrivateKernelResetHintsToNoir(privateKernelResetCircuitPrivateInputs.hints)
    };
    pushTestData('private-kernel-reset', inputs);
}
