// @aztec/kv-store/dest/indexeddb/singleton.js
import { hash } from 'ohash';
/**
 * Stores a single value in IndexedDB.
 */ export class IndexedDBAztecSingleton {
    #_db;
    #rootDB;
    #container;
    #slot;
    constructor(rootDB, name){
        this.#rootDB = rootDB;
        this.#container = `singleton:${name}`;
        this.#slot = `singleton:${name}:value`;
    }
    set db(db) {
        this.#_db = db;
    }
    get db() {
        return this.#_db ? this.#_db : this.#rootDB.transaction('data', 'readwrite').store;
    }
    async getAsync() {
        const data = await this.db.get(this.#slot);
        return data?.value;
    }
    async set(val) {
        const result = await this.db.put({
            container: this.#container,
            slot: this.#slot,
            key: [
                this.#slot
            ],
            keyCount: 1,
            value: val,
            hash: hash(val)
        });
        return result !== undefined;
    }
    async delete() {
        await this.db.delete(this.#slot);
        return true;
    }
}
