// @aztec/ethereum/dest/chain.js
import { foundry } from 'viem/chains';
import { AZTEC_TEST_CHAIN_ID } from './constants.js';
/**
 * Helper function to create an instance of Aztec Chain from an rpc url and chain id.
 * @param rpcUrls - The rpc url of the chain or a chain identifier (e.g. 'http://localhost:8545')
 * @param chainId - The chain id of the chain or a chain identifier (e.g. 1337)
 */ export function createEthereumChain(rpcUrls, _chainId) {
    let chainId;
    if (typeof _chainId === 'string') {
        chainId = +_chainId;
    } else {
        chainId = _chainId;
    }
    if (chainId) {
        return {
            chainInfo: {
                id: chainId,
                name: 'Ethereum',
                rpcUrls: {
                    default: {
                        http: rpcUrls
                    }
                },
                nativeCurrency: {
                    decimals: 18,
                    name: 'Ether',
                    symbol: 'ETH'
                }
            },
            rpcUrls
        };
    } else {
        return {
            chainInfo: foundry,
            rpcUrls
        };
    }
}
/**
 * Helper function to determine if a chain id is an instance of Anvil
 */ export function isAnvilTestChain(_chainId) {
    let chainId;
    if (typeof _chainId === 'string') {
        chainId = +_chainId;
    } else {
        chainId = _chainId;
    }
    const testChains = [
        foundry.id,
        AZTEC_TEST_CHAIN_ID
    ];
    return testChains.includes(chainId);
}
