// @aztec/ethereum/dest/l1_artifacts.js
import { CoinIssuerAbi, CoinIssuerBytecode, DateGatedRelayerAbi, DateGatedRelayerBytecode, EmpireSlasherDeploymentExtLibAbi, EmpireSlasherDeploymentExtLibBytecode, EmpireSlashingProposerAbi, EmpireSlashingProposerBytecode, FeeAssetHandlerAbi, FeeAssetHandlerBytecode, FeeJuicePortalAbi, FeeJuicePortalBytecode, GSEAbi, GSEBytecode, GovernanceAbi, GovernanceBytecode, GovernanceProposerAbi, GovernanceProposerBytecode, HonkVerifierAbi, HonkVerifierBytecode, InboxAbi, InboxBytecode, MockVerifierAbi, MockVerifierBytecode, MockZKPassportVerifierAbi, MockZKPassportVerifierBytecode, MultiAdderAbi, MultiAdderBytecode, OutboxAbi, OutboxBytecode, RegisterNewRollupVersionPayloadAbi, RegisterNewRollupVersionPayloadBytecode, RegistryAbi, RegistryBytecode, RewardDistributorAbi, RewardDistributorBytecode, RewardExtLibAbi, RewardExtLibBytecode, RollupAbi, RollupBytecode, RollupLinkReferences, RollupOperationsExtLibAbi, RollupOperationsExtLibBytecode, SlashFactoryAbi, SlashFactoryBytecode, SlasherAbi, SlasherBytecode, StakingAssetHandlerAbi, StakingAssetHandlerBytecode, TallySlasherDeploymentExtLibAbi, TallySlasherDeploymentExtLibBytecode, TallySlashingProposerAbi, TallySlashingProposerBytecode, TestERC20Abi, TestERC20Bytecode, ValidatorOperationsExtLibAbi, ValidatorOperationsExtLibBytecode, ValidatorSelectionLibAbi, ValidatorSelectionLibBytecode } from '@aztec/l1-artifacts';
export const RegistryArtifact = {
    name: 'Registry',
    contractAbi: RegistryAbi,
    contractBytecode: RegistryBytecode
};
export const InboxArtifact = {
    name: 'Inbox',
    contractAbi: InboxAbi,
    contractBytecode: InboxBytecode
};
export const OutboxArtifact = {
    name: 'Outbox',
    contractAbi: OutboxAbi,
    contractBytecode: OutboxBytecode
};
export const RollupArtifact = {
    name: 'Rollup',
    contractAbi: RollupAbi,
    contractBytecode: RollupBytecode,
    libraries: {
        linkReferences: RollupLinkReferences,
        libraryCode: {
            ValidatorSelectionLib: {
                name: 'ValidatorSelectionLib',
                contractAbi: ValidatorSelectionLibAbi,
                contractBytecode: ValidatorSelectionLibBytecode
            },
            RollupOperationsExtLib: {
                name: 'RollupOperationsExtLib',
                contractAbi: RollupOperationsExtLibAbi,
                contractBytecode: RollupOperationsExtLibBytecode
            },
            ValidatorOperationsExtLib: {
                name: 'ValidatorOperationsExtLib',
                contractAbi: ValidatorOperationsExtLibAbi,
                contractBytecode: ValidatorOperationsExtLibBytecode
            },
            RewardExtLib: {
                name: 'RewardExtLib',
                contractAbi: RewardExtLibAbi,
                contractBytecode: RewardExtLibBytecode
            },
            TallySlasherDeploymentExtLib: {
                name: 'TallySlasherDeploymentExtLib',
                contractAbi: TallySlasherDeploymentExtLibAbi,
                contractBytecode: TallySlasherDeploymentExtLibBytecode
            },
            EmpireSlasherDeploymentExtLib: {
                name: 'EmpireSlasherDeploymentExtLib',
                contractAbi: EmpireSlasherDeploymentExtLibAbi,
                contractBytecode: EmpireSlasherDeploymentExtLibBytecode
            }
        }
    }
};
export const StakingAssetArtifact = {
    name: 'StakingAsset',
    contractAbi: TestERC20Abi,
    contractBytecode: TestERC20Bytecode
};
export const FeeAssetArtifact = {
    name: 'FeeAsset',
    contractAbi: TestERC20Abi,
    contractBytecode: TestERC20Bytecode
};
export const FeeJuicePortalArtifact = {
    name: 'FeeJuicePortal',
    contractAbi: FeeJuicePortalAbi,
    contractBytecode: FeeJuicePortalBytecode
};
export const RewardDistributorArtifact = {
    name: 'RewardDistributor',
    contractAbi: RewardDistributorAbi,
    contractBytecode: RewardDistributorBytecode
};
export const CoinIssuerArtifact = {
    name: 'CoinIssuer',
    contractAbi: CoinIssuerAbi,
    contractBytecode: CoinIssuerBytecode
};
export const DateGatedRelayerArtifact = {
    name: 'DateGatedRelayer',
    contractAbi: DateGatedRelayerAbi,
    contractBytecode: DateGatedRelayerBytecode
};
export const GovernanceProposerArtifact = {
    name: 'GovernanceProposer',
    contractAbi: GovernanceProposerAbi,
    contractBytecode: GovernanceProposerBytecode
};
export const GovernanceArtifact = {
    name: 'Governance',
    contractAbi: GovernanceAbi,
    contractBytecode: GovernanceBytecode
};
export const SlasherArtifact = {
    name: 'Slasher',
    contractAbi: SlasherAbi,
    contractBytecode: SlasherBytecode
};
export const EmpireSlashingProposerArtifact = {
    name: 'EmpireSlashingProposer',
    contractAbi: EmpireSlashingProposerAbi,
    contractBytecode: EmpireSlashingProposerBytecode
};
export const TallySlashingProposerArtifact = {
    name: 'TallySlashingProposer',
    contractAbi: TallySlashingProposerAbi,
    contractBytecode: TallySlashingProposerBytecode
};
export const SlashFactoryArtifact = {
    name: 'SlashFactory',
    contractAbi: SlashFactoryAbi,
    contractBytecode: SlashFactoryBytecode
};
export const RegisterNewRollupVersionPayloadArtifact = {
    name: 'RegisterNewRollupVersionPayload',
    contractAbi: RegisterNewRollupVersionPayloadAbi,
    contractBytecode: RegisterNewRollupVersionPayloadBytecode
};
export const FeeAssetHandlerArtifact = {
    name: 'FeeAssetHandler',
    contractAbi: FeeAssetHandlerAbi,
    contractBytecode: FeeAssetHandlerBytecode
};
export const StakingAssetHandlerArtifact = {
    name: 'StakingAssetHandler',
    contractAbi: StakingAssetHandlerAbi,
    contractBytecode: StakingAssetHandlerBytecode
};
export const MultiAdderArtifact = {
    name: 'MultiAdder',
    contractAbi: MultiAdderAbi,
    contractBytecode: MultiAdderBytecode
};
export const GSEArtifact = {
    name: 'GSE',
    contractAbi: GSEAbi,
    contractBytecode: GSEBytecode
};
// Verifier artifacts
export const HonkVerifierArtifact = {
    name: 'HonkVerifier',
    contractAbi: HonkVerifierAbi,
    contractBytecode: HonkVerifierBytecode
};
export const MockVerifierArtifact = {
    name: 'MockVerifier',
    contractAbi: MockVerifierAbi,
    contractBytecode: MockVerifierBytecode
};
export const MockZkPassportVerifierArtifact = {
    name: 'MockZkPassportVerifier',
    contractAbi: MockZKPassportVerifierAbi,
    contractBytecode: MockZKPassportVerifierBytecode
};
// Re-export the verifier artifacts for backwards compatibility
export const l1ArtifactsVerifiers = {
    honkVerifier: HonkVerifierArtifact
};
// Re-export the mock verifiers for backwards compatibility
export const mockVerifiers = {
    mockVerifier: MockVerifierArtifact,
    mockZkPassportVerifier: MockZkPassportVerifierArtifact
};
