// @aztec/ethereum/dest/l1_reader.js
import { getConfigFromMappings, numberConfigHelper } from '@aztec/foundation/config';
import { l1ContractAddressesMapping } from './l1_contract_addresses.js';
export const l1ReaderConfigMappings = {
    l1Contracts: {
        description: 'The deployed L1 contract addresses',
        nested: l1ContractAddressesMapping
    },
    l1ChainId: {
        env: 'L1_CHAIN_ID',
        parseEnv: (val)=>+val,
        description: 'The chain ID of the ethereum host.'
    },
    l1RpcUrls: {
        env: 'ETHEREUM_HOSTS',
        description: 'List of URLs of Ethereum RPC nodes that services will connect to (comma separated).',
        parseEnv: (val)=>val.split(',').map((url)=>url.trim()),
        defaultValue: []
    },
    l1DebugRpcUrls: {
        env: 'ETHEREUM_DEBUG_HOSTS',
        description: 'The RPC Url of the ethereum debug host for trace and debug methods.',
        parseEnv: (val)=>val.split(',').map((url)=>url.trim()),
        defaultValue: []
    },
    viemPollingIntervalMS: {
        env: 'L1_READER_VIEM_POLLING_INTERVAL_MS',
        description: 'The polling interval viem uses in ms',
        ...numberConfigHelper(1_000)
    }
};
export function getL1ReaderConfigFromEnv() {
    return getConfigFromMappings(l1ReaderConfigMappings);
}
