// @aztec/bb.js/dest/node-cjs/cbind/generated/curve_constants.js
"use strict";
/**
 * Curve constants generated from barretenberg native binary.
 * DO NOT EDIT - This file is auto-generated by generate.ts
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.SECP256R1_G1_GENERATOR = exports.SECP256R1_FQ_MODULUS = exports.SECP256R1_FR_MODULUS = exports.SECP256K1_G1_GENERATOR = exports.SECP256K1_FQ_MODULUS = exports.SECP256K1_FR_MODULUS = exports.GRUMPKIN_G1_GENERATOR = exports.GRUMPKIN_FQ_MODULUS = exports.GRUMPKIN_FR_MODULUS = exports.BN254_G2_GENERATOR = exports.BN254_G1_GENERATOR = exports.BN254_FQ_MODULUS = exports.BN254_FR_MODULUS = void 0;
/**
 * BN254 curve constants
 */
exports.BN254_FR_MODULUS = 21888242871839275222246405745257275088548364400416034343698204186575808495617n;
exports.BN254_FQ_MODULUS = 21888242871839275222246405745257275088696311157297823662689037894645226208583n;
exports.BN254_G1_GENERATOR = {
    x: new Uint8Array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]),
    y: new Uint8Array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2]),
};
exports.BN254_G2_GENERATOR = {
    x: [new Uint8Array([24, 0, 222, 239, 18, 31, 30, 118, 66, 106, 0, 102, 94, 92, 68, 121, 103, 67, 34, 212, 247, 94, 218, 221, 70, 222, 189, 92, 217, 146, 246, 237]), new Uint8Array([25, 142, 147, 147, 146, 13, 72, 58, 114, 96, 191, 183, 49, 251, 93, 37, 241, 170, 73, 51, 53, 169, 231, 18, 151, 228, 133, 183, 174, 243, 18, 194])],
    y: [new Uint8Array([18, 200, 94, 165, 219, 140, 109, 235, 74, 171, 113, 128, 141, 203, 64, 143, 227, 209, 231, 105, 12, 67, 211, 123, 76, 230, 204, 1, 102, 250, 125, 170]), new Uint8Array([9, 6, 137, 208, 88, 95, 240, 117, 236, 158, 153, 173, 105, 12, 51, 149, 188, 75, 49, 51, 112, 179, 142, 243, 85, 172, 218, 220, 209, 34, 151, 91])],
};
/**
 * Grumpkin curve constants
 */
exports.GRUMPKIN_FR_MODULUS = 21888242871839275222246405745257275088696311157297823662689037894645226208583n;
exports.GRUMPKIN_FQ_MODULUS = 21888242871839275222246405745257275088548364400416034343698204186575808495617n;
exports.GRUMPKIN_G1_GENERATOR = {
    x: new Uint8Array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1]),
    y: new Uint8Array([0, 0, 0, 0, 0, 0, 0, 2, 207, 19, 94, 117, 6, 164, 93, 99, 45, 39, 13, 69, 241, 24, 18, 148, 131, 63, 196, 141, 130, 63, 39, 44]),
};
/**
 * Secp256k1 curve constants
 */
exports.SECP256K1_FR_MODULUS = 115792089237316195423570985008687907852837564279074904382605163141518161494337n;
exports.SECP256K1_FQ_MODULUS = 115792089237316195423570985008687907853269984665640564039457584007908834671663n;
exports.SECP256K1_G1_GENERATOR = {
    x: new Uint8Array([121, 190, 102, 126, 249, 220, 187, 172, 85, 160, 98, 149, 206, 135, 11, 7, 2, 155, 252, 219, 45, 206, 40, 217, 89, 242, 129, 91, 22, 248, 23, 152]),
    y: new Uint8Array([72, 58, 218, 119, 38, 163, 196, 101, 93, 164, 251, 252, 14, 17, 8, 168, 253, 23, 180, 72, 166, 133, 84, 25, 156, 71, 208, 143, 251, 16, 212, 184]),
};
/**
 * Secp256r1 curve constants
 */
exports.SECP256R1_FR_MODULUS = 115792089210356248762697446949407573529996955224135760342422259061068512044369n;
exports.SECP256R1_FQ_MODULUS = 115792089210356248762697446949407573530086143415290314195533631308867097853951n;
exports.SECP256R1_G1_GENERATOR = {
    x: new Uint8Array([107, 23, 209, 242, 225, 44, 66, 71, 248, 188, 230, 229, 99, 164, 64, 242, 119, 3, 125, 129, 45, 235, 51, 160, 244, 161, 57, 69, 216, 152, 194, 150]),
    y: new Uint8Array([79, 227, 66, 226, 254, 26, 127, 155, 142, 231, 235, 74, 124, 15, 158, 22, 43, 206, 51, 87, 107, 49, 94, 206, 203, 182, 64, 104, 55, 191, 81, 245]),
};
