// @aztec/stdlib/dest/proofs/proof_data.js
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
import { RecursiveProof } from '../proofs/recursive_proof.js';
import { VkData } from '../vks/index.js';
/**
 * Represents the data of a recursive proof.
 */ export class ProofData {
    publicInputs;
    proof;
    vkData;
    constructor(publicInputs, proof, vkData){
        this.publicInputs = publicInputs;
        this.proof = proof;
        this.vkData = vkData;
    }
    toBuffer() {
        return serializeToBuffer(this.publicInputs, this.proof, this.vkData);
    }
    static fromBuffer(buffer, publicInputs) {
        const reader = BufferReader.asReader(buffer);
        return new ProofData(reader.readObject(publicInputs), RecursiveProof.fromBuffer(reader), reader.readObject(VkData));
    }
}
/**
 * Represents the data of a recursive proof for a circuit with a fixed verification key.
 */ export class ProofDataForFixedVk {
    publicInputs;
    proof;
    constructor(publicInputs, proof){
        this.publicInputs = publicInputs;
        this.proof = proof;
    }
    toBuffer() {
        return serializeToBuffer(this.publicInputs, this.proof);
    }
    static fromBuffer(buffer, publicInputs) {
        const reader = BufferReader.asReader(buffer);
        return new ProofDataForFixedVk(reader.readObject(publicInputs), RecursiveProof.fromBuffer(reader));
    }
}
