// @aztec/stdlib/dest/proofs/proving_request_type.js
export var ProvingRequestType = /*#__PURE__*/ function(ProvingRequestType) {
    ProvingRequestType[ProvingRequestType["PUBLIC_VM"] = 0] = "PUBLIC_VM";
    ProvingRequestType[ProvingRequestType["PUBLIC_CHONK_VERIFIER"] = 1] = "PUBLIC_CHONK_VERIFIER";
    ProvingRequestType[ProvingRequestType["PRIVATE_TX_BASE_ROLLUP"] = 2] = "PRIVATE_TX_BASE_ROLLUP";
    ProvingRequestType[ProvingRequestType["PUBLIC_TX_BASE_ROLLUP"] = 3] = "PUBLIC_TX_BASE_ROLLUP";
    ProvingRequestType[ProvingRequestType["TX_MERGE_ROLLUP"] = 4] = "TX_MERGE_ROLLUP";
    ProvingRequestType[ProvingRequestType["BLOCK_ROOT_FIRST_ROLLUP"] = 5] = "BLOCK_ROOT_FIRST_ROLLUP";
    ProvingRequestType[ProvingRequestType["BLOCK_ROOT_SINGLE_TX_FIRST_ROLLUP"] = 6] = "BLOCK_ROOT_SINGLE_TX_FIRST_ROLLUP";
    ProvingRequestType[ProvingRequestType["BLOCK_ROOT_EMPTY_TX_FIRST_ROLLUP"] = 7] = "BLOCK_ROOT_EMPTY_TX_FIRST_ROLLUP";
    ProvingRequestType[ProvingRequestType["BLOCK_ROOT_ROLLUP"] = 8] = "BLOCK_ROOT_ROLLUP";
    ProvingRequestType[ProvingRequestType["BLOCK_ROOT_SINGLE_TX_ROLLUP"] = 9] = "BLOCK_ROOT_SINGLE_TX_ROLLUP";
    ProvingRequestType[ProvingRequestType["BLOCK_MERGE_ROLLUP"] = 10] = "BLOCK_MERGE_ROLLUP";
    ProvingRequestType[ProvingRequestType["CHECKPOINT_ROOT_ROLLUP"] = 11] = "CHECKPOINT_ROOT_ROLLUP";
    ProvingRequestType[ProvingRequestType["CHECKPOINT_ROOT_SINGLE_BLOCK_ROLLUP"] = 12] = "CHECKPOINT_ROOT_SINGLE_BLOCK_ROLLUP";
    ProvingRequestType[ProvingRequestType["CHECKPOINT_PADDING_ROLLUP"] = 13] = "CHECKPOINT_PADDING_ROLLUP";
    ProvingRequestType[ProvingRequestType["CHECKPOINT_MERGE_ROLLUP"] = 14] = "CHECKPOINT_MERGE_ROLLUP";
    ProvingRequestType[ProvingRequestType["ROOT_ROLLUP"] = 15] = "ROOT_ROLLUP";
    ProvingRequestType[ProvingRequestType["PARITY_BASE"] = 16] = "PARITY_BASE";
    ProvingRequestType[ProvingRequestType["PARITY_ROOT"] = 17] = "PARITY_ROOT";
    return ProvingRequestType;
}({});
