// @aztec/noir-protocol-circuits-types/dest/artifacts/client/bundle.js
import HidingKernelToPublicJson from '../../../artifacts/hiding_kernel_to_public.json' with {
    type: 'json'
};
import HidingKernelToRollupJson from '../../../artifacts/hiding_kernel_to_rollup.json' with {
    type: 'json'
};
import PrivateKernelInitJson from '../../../artifacts/private_kernel_init.json' with {
    type: 'json'
};
import PrivateKernelInitSimulatedJson from '../../../artifacts/private_kernel_init_simulated.json' with {
    type: 'json'
};
import PrivateKernelInnerJson from '../../../artifacts/private_kernel_inner.json' with {
    type: 'json'
};
import PrivateKernelInnerSimulatedJson from '../../../artifacts/private_kernel_inner_simulated.json' with {
    type: 'json'
};
import PrivateKernelTailJson from '../../../artifacts/private_kernel_tail.json' with {
    type: 'json'
};
import PrivateKernelTailSimulatedJson from '../../../artifacts/private_kernel_tail_simulated.json' with {
    type: 'json'
};
import PrivateKernelTailToPublicJson from '../../../artifacts/private_kernel_tail_to_public.json' with {
    type: 'json'
};
import PrivateKernelTailToPublicSimulatedJson from '../../../artifacts/private_kernel_tail_to_public_simulated.json' with {
    type: 'json'
};
import { PrivateKernelResetArtifacts, PrivateKernelResetSimulatedArtifacts } from '../../private_kernel_reset_data.js';
import { ClientCircuitVks } from '../vks/client.js';
export const ClientCircuitArtifacts = {
    PrivateKernelInitArtifact: PrivateKernelInitJson,
    PrivateKernelInnerArtifact: PrivateKernelInnerJson,
    PrivateKernelTailArtifact: PrivateKernelTailJson,
    PrivateKernelTailToPublicArtifact: PrivateKernelTailToPublicJson,
    HidingKernelToRollup: HidingKernelToRollupJson,
    HidingKernelToPublic: HidingKernelToPublicJson,
    ...PrivateKernelResetArtifacts
};
export const SimulatedClientCircuitArtifacts = {
    PrivateKernelInitArtifact: PrivateKernelInitSimulatedJson,
    PrivateKernelInnerArtifact: PrivateKernelInnerSimulatedJson,
    PrivateKernelTailArtifact: PrivateKernelTailSimulatedJson,
    PrivateKernelTailToPublicArtifact: PrivateKernelTailToPublicSimulatedJson,
    HidingKernelToRollup: HidingKernelToRollupJson,
    HidingKernelToPublic: HidingKernelToPublicJson,
    ...PrivateKernelResetSimulatedArtifacts
};
export class BundleArtifactProvider {
    getClientCircuitArtifactByName(artifact) {
        return Promise.resolve({
            ...ClientCircuitArtifacts[artifact],
            name: artifact.replace('Artifact', '')
        });
    }
    getSimulatedClientCircuitArtifactByName(artifact) {
        return Promise.resolve({
            ...SimulatedClientCircuitArtifacts[artifact],
            name: artifact.replace('Artifact', '')
        });
    }
    getCircuitVkByName(artifact) {
        return Promise.resolve(ClientCircuitVks[artifact]);
    }
}
