// @aztec/stdlib/dest/contract/contract_instance_update.js
import { Fr } from '@aztec/foundation/curves/bn254';
import { BufferReader, serializeToBuffer } from '@aztec/foundation/serialize';
export class SerializableContractInstanceUpdate {
    prevContractClassId;
    newContractClassId;
    timestampOfChange;
    constructor(instance){
        this.prevContractClassId = instance.prevContractClassId;
        this.newContractClassId = instance.newContractClassId;
        this.timestampOfChange = instance.timestampOfChange;
    }
    toBuffer() {
        return serializeToBuffer(this.prevContractClassId, this.newContractClassId, this.timestampOfChange);
    }
    static fromBuffer(bufferOrReader) {
        const reader = BufferReader.asReader(bufferOrReader);
        return new SerializableContractInstanceUpdate({
            prevContractClassId: reader.readObject(Fr),
            newContractClassId: reader.readObject(Fr),
            timestampOfChange: reader.readObject(Fr).toBigInt()
        });
    }
    static random(opts = {}) {
        return new SerializableContractInstanceUpdate({
            prevContractClassId: Fr.random(),
            newContractClassId: Fr.random(),
            timestampOfChange: BigInt(Math.floor(Math.random() * 1000)),
            ...opts
        });
    }
    static default() {
        return new SerializableContractInstanceUpdate({
            prevContractClassId: Fr.zero(),
            newContractClassId: Fr.zero(),
            timestampOfChange: 0n
        });
    }
}
