// @aztec/stdlib/dest/contract/contract_deployment_data.js
import { ContractClassLog, PrivateLog } from '@aztec/stdlib/logs';
import { z } from 'zod';
/**
 * Class containing contract class logs and private logs which are both
 * relevant for contract registrations and deployments.
 */ export class ContractDeploymentData {
    contractClassLogs;
    privateLogs;
    constructor(contractClassLogs, privateLogs){
        this.contractClassLogs = contractClassLogs;
        this.privateLogs = privateLogs;
    }
    getContractClassLogs() {
        return this.contractClassLogs;
    }
    getPrivateLogs() {
        return this.privateLogs;
    }
    static from(args) {
        return new ContractDeploymentData(args.contractClassLogs, args.privateLogs);
    }
    static empty() {
        return new ContractDeploymentData([], []);
    }
    static get schema() {
        return z.object({
            contractClassLogs: z.array(ContractClassLog.schema),
            privateLogs: z.array(PrivateLog.schema)
        }).transform(ContractDeploymentData.from);
    }
    /**
   * Creates a ContractDeploymentData from a plain object without Zod validation.
   * This method is optimized for performance and skips validation, making it suitable
   * for deserializing trusted data (e.g., from C++ via MessagePack).
   * @param obj - Plain object containing ContractDeploymentData fields
   * @returns A ContractDeploymentData instance
   */ static fromPlainObject(obj) {
        if (obj instanceof ContractDeploymentData) {
            return obj;
        }
        return new ContractDeploymentData(obj.contractClassLogs.map((log)=>ContractClassLog.fromPlainObject(log)), obj.privateLogs.map((log)=>PrivateLog.fromPlainObject(log)));
    }
}
/**
 * Class containing both revertible and non-revertible registration/deployment data.
 */ export class AllContractDeploymentData {
    nonRevertibleContractDeploymentData;
    revertibleContractDeploymentData;
    constructor(nonRevertibleContractDeploymentData, revertibleContractDeploymentData){
        this.nonRevertibleContractDeploymentData = nonRevertibleContractDeploymentData;
        this.revertibleContractDeploymentData = revertibleContractDeploymentData;
    }
    getNonRevertibleContractDeploymentData() {
        return this.nonRevertibleContractDeploymentData;
    }
    getRevertibleContractDeploymentData() {
        return this.revertibleContractDeploymentData;
    }
    /**
   * Extracts all contract registration/deployment data from a tx separated by revertibility.
   * This includes contract class logs and private logs.
   *
   * This method handles both private-only transactions and transactions with public calls,
   * properly splitting logs between revertible and non-revertible categories.
   * @param tx - The transaction to extract data from
   * @returns The extracted deployment data separated by revertibility
   */ static fromTx(tx) {
        const hasPublicCalls = !!tx.data.forPublic;
        // Extract contract class logs from the tx
        const allClassLogs = tx.getContractClassLogs();
        let nonRevertibleClassLogs;
        let revertibleClassLogs;
        if (hasPublicCalls) {
            // Transactions with public calls can have both revertible and non-revertible contract class logs
            // Split the logs up here based on revertibility
            nonRevertibleClassLogs = tx.getSplitContractClassLogs(/*revertible=*/ false);
            revertibleClassLogs = tx.getSplitContractClassLogs(/*revertible=*/ true);
        } else {
            // Private-only tx: all logs are non-revertible
            nonRevertibleClassLogs = allClassLogs;
            revertibleClassLogs = [];
        }
        // Extract contract instance logs from the transaction's private logs
        let nonRevertibleInstanceLogs;
        let revertibleInstanceLogs;
        if (hasPublicCalls) {
            // Transactions with public calls can have both revertible and non-revertible contract instance logs
            // Split the logs up here based on revertibility
            nonRevertibleInstanceLogs = tx.data.forPublic.nonRevertibleAccumulatedData.privateLogs.filter((l)=>!l.isEmpty());
            revertibleInstanceLogs = tx.data.forPublic.revertibleAccumulatedData.privateLogs.filter((l)=>!l.isEmpty());
        } else {
            // Private-only tx: use logs from the `forRollup` member of the tx
            // For private-only txs, all logs are non-revertible
            nonRevertibleInstanceLogs = tx.data.forRollup.end.privateLogs.filter((l)=>!l.isEmpty());
            revertibleInstanceLogs = [];
        }
        return new AllContractDeploymentData(new ContractDeploymentData(nonRevertibleClassLogs, nonRevertibleInstanceLogs), new ContractDeploymentData(revertibleClassLogs, revertibleInstanceLogs));
    }
}
