// @aztec/stdlib/dest/interfaces/aztec-node-admin.js
import { createSafeJsonRpcClient, defaultFetch } from '@aztec/foundation/json-rpc/client';
import { z } from 'zod';
import { OffenseSchema, SlashPayloadRoundSchema } from '../slashing/index.js';
import { getVersioningResponseHandler } from '../versioning/index.js';
import { ArchiverSpecificConfigSchema } from './archiver.js';
import { SequencerConfigSchema } from './configs.js';
import { ProverConfigSchema } from './prover-client.js';
import { SlasherConfigSchema } from './slasher.js';
import { ValidatorClientFullConfigSchema } from './validator.js';
export const AztecNodeAdminConfigSchema = SequencerConfigSchema.merge(ProverConfigSchema).merge(SlasherConfigSchema).merge(ValidatorClientFullConfigSchema.omit({
    l1Contracts: true
})).merge(ArchiverSpecificConfigSchema.pick({
    archiverPollingIntervalMS: true,
    archiverBatchSize: true,
    skipValidateCheckpointAttestations: true
})).merge(z.object({
    maxPendingTxCount: z.number()
}));
export const AztecNodeAdminApiSchema = {
    getConfig: z.function().returns(AztecNodeAdminConfigSchema),
    setConfig: z.function().args(AztecNodeAdminConfigSchema.partial()).returns(z.void()),
    startSnapshotUpload: z.function().args(z.string()).returns(z.void()),
    rollbackTo: z.function().args(z.number()).returns(z.void()),
    pauseSync: z.function().returns(z.void()),
    resumeSync: z.function().returns(z.void()),
    getSlashPayloads: z.function().returns(z.array(SlashPayloadRoundSchema)),
    getSlashOffenses: z.function().args(z.union([
        z.bigint(),
        z.literal('all'),
        z.literal('current')
    ])).returns(z.array(OffenseSchema)),
    reloadKeystore: z.function().returns(z.void())
};
export function createAztecNodeAdminClient(url, versions = {}, fetch = defaultFetch, apiKey) {
    return createSafeJsonRpcClient(url, AztecNodeAdminApiSchema, {
        namespaceMethods: 'nodeAdmin',
        fetch,
        onResponse: getVersioningResponseHandler(versions),
        ...apiKey ? {
            extraHeaders: {
                'x-api-key': apiKey
            }
        } : {}
    });
}
