// @aztec/stdlib/dest/interfaces/configs.js
import { z } from 'zod';
import { schemas, zodFor } from '../schemas/index.js';
import { AllowedElementSchema } from './allowed_element.js';
export const SequencerConfigSchema = zodFor()(z.object({
    sequencerPollingIntervalMS: z.number().optional(),
    maxTxsPerBlock: z.number().optional(),
    minValidTxsPerBlock: z.number().optional(),
    minTxsPerBlock: z.number().optional(),
    maxL2BlockGas: z.number().optional(),
    publishTxsWithProposals: z.boolean().optional(),
    maxDABlockGas: z.number().optional(),
    coinbase: schemas.EthAddress.optional(),
    feeRecipient: schemas.AztecAddress.optional(),
    acvmWorkingDirectory: z.string().optional(),
    acvmBinaryPath: z.string().optional(),
    txPublicSetupAllowList: z.array(AllowedElementSchema).optional(),
    maxBlockSizeInBytes: z.number().optional(),
    governanceProposerPayload: schemas.EthAddress.optional(),
    l1PublishingTime: z.number().optional(),
    enforceTimeTable: z.boolean().optional(),
    fakeProcessingDelayPerTxMs: z.number().optional(),
    fakeThrowAfterProcessingTxCount: z.number().optional(),
    attestationPropagationTime: z.number().optional(),
    skipCollectingAttestations: z.boolean().optional(),
    skipInvalidateBlockAsProposer: z.boolean().optional(),
    secondsBeforeInvalidatingBlockAsCommitteeMember: z.number(),
    secondsBeforeInvalidatingBlockAsNonCommitteeMember: z.number(),
    broadcastInvalidBlockProposal: z.boolean().optional(),
    injectFakeAttestation: z.boolean().optional(),
    fishermanMode: z.boolean().optional(),
    shuffleAttestationOrdering: z.boolean().optional(),
    blockDurationMs: z.number().positive().optional(),
    buildCheckpointIfEmpty: z.boolean().optional(),
    skipPushProposedBlocksToArchiver: z.boolean().optional(),
    minBlocksForCheckpoint: z.number().positive().optional(),
    skipPublishingCheckpointsPercent: z.number().gte(0).lte(100).optional()
}));
