// @aztec/stdlib/dest/interfaces/slasher.js
import { schemas, zodFor } from '@aztec/foundation/schemas';
import { z } from 'zod';
export const SlasherConfigSchema = zodFor()(z.object({
    slashOverridePayload: schemas.EthAddress.optional(),
    slashMinPenaltyPercentage: z.number(),
    slashMaxPenaltyPercentage: z.number(),
    slashValidatorsAlways: z.array(schemas.EthAddress),
    slashValidatorsNever: z.array(schemas.EthAddress),
    slashPrunePenalty: schemas.BigInt,
    slashDataWithholdingPenalty: schemas.BigInt,
    slashInactivityTargetPercentage: z.number(),
    slashInactivityConsecutiveEpochThreshold: z.number(),
    slashInactivityPenalty: schemas.BigInt,
    slashProposeInvalidAttestationsPenalty: schemas.BigInt,
    slashDuplicateProposalPenalty: schemas.BigInt,
    slashDuplicateAttestationPenalty: schemas.BigInt,
    slashAttestDescendantOfInvalidPenalty: schemas.BigInt,
    slashUnknownPenalty: schemas.BigInt,
    slashOffenseExpirationRounds: z.number(),
    slashMaxPayloadSize: z.number(),
    slashGracePeriodL2Slots: z.number(),
    slashBroadcastedInvalidBlockPenalty: schemas.BigInt,
    slashExecuteRoundsLookBack: z.number(),
    slashSelfAllowed: z.boolean().optional()
}));
