// @aztec/stdlib/dest/interfaces/validator.js
import { schemas, zodFor } from '@aztec/foundation/schemas';
import { ValidatorHASignerConfigSchema } from '@aztec/validator-ha-signer/config';
import { z } from 'zod';
import { AllowedElementSchema } from './allowed_element.js';
export const ValidatorClientConfigSchema = zodFor()(ValidatorHASignerConfigSchema.extend({
    validatorAddresses: z.array(schemas.EthAddress).optional(),
    disableValidator: z.boolean(),
    disabledValidators: z.array(schemas.EthAddress),
    attestationPollingIntervalMs: z.number().min(0),
    validatorReexecute: z.boolean(),
    alwaysReexecuteBlockProposals: z.boolean().optional(),
    fishermanMode: z.boolean().optional(),
    skipCheckpointProposalValidation: z.boolean().optional(),
    skipPushProposedBlocksToArchiver: z.boolean().optional(),
    attestToEquivocatedProposals: z.boolean().optional()
}));
export const ValidatorClientFullConfigSchema = zodFor()(ValidatorClientConfigSchema.extend({
    txPublicSetupAllowList: z.array(AllowedElementSchema).optional(),
    broadcastInvalidBlockProposal: z.boolean().optional(),
    slashBroadcastedInvalidBlockPenalty: schemas.BigInt,
    slashDuplicateProposalPenalty: schemas.BigInt,
    slashDuplicateAttestationPenalty: schemas.BigInt,
    disableTransactions: z.boolean().optional()
}));
