// @aztec/telemetry-client/dest/config.js
import { booleanConfigHelper, getConfigFromMappings } from '@aztec/foundation/config';
export const telemetryClientConfigMappings = {
    metricsCollectorUrl: {
        env: 'OTEL_EXPORTER_OTLP_METRICS_ENDPOINT',
        description: 'The URL of the telemetry collector for metrics',
        parseEnv: (val)=>val && new URL(val)
    },
    tracesCollectorUrl: {
        env: 'OTEL_EXPORTER_OTLP_TRACES_ENDPOINT',
        description: 'The URL of the telemetry collector for traces',
        parseEnv: (val)=>val && new URL(val)
    },
    logsCollectorUrl: {
        env: 'OTEL_EXPORTER_OTLP_LOGS_ENDPOINT',
        description: 'The URL of the telemetry collector for logs',
        parseEnv: (val)=>val && new URL(val)
    },
    otelCollectIntervalMs: {
        env: 'OTEL_COLLECT_INTERVAL_MS',
        description: 'The interval at which to collect metrics',
        defaultValue: 60000,
        parseEnv: (val)=>parseInt(val)
    },
    otelExportTimeoutMs: {
        env: 'OTEL_EXPORT_TIMEOUT_MS',
        description: 'The timeout for exporting metrics',
        defaultValue: 30000,
        parseEnv: (val)=>parseInt(val)
    },
    otelExcludeMetrics: {
        env: 'OTEL_EXCLUDE_METRICS',
        description: 'A list of metric prefixes to exclude from export',
        parseEnv: (val)=>val ? val.split(',').map((s)=>s.trim()).filter((s)=>s.length > 0) : [],
        defaultValue: []
    },
    otelIncludeMetrics: {
        env: 'OTEL_INCLUDE_METRICS',
        description: 'A list of metric prefixes to include in export (ignored if OTEL_EXCLUDE_METRICS is set)',
        parseEnv: (val)=>val ? val.split(',').map((s)=>s.trim()).filter((s)=>s.length > 0) : [],
        defaultValue: []
    },
    publicMetricsCollectorUrl: {
        env: 'PUBLIC_OTEL_EXPORTER_OTLP_METRICS_ENDPOINT',
        description: 'A URL to publish a subset of metrics for public consumption',
        parseEnv: (val)=>val && new URL(val)
    },
    publicMetricsCollectFrom: {
        env: 'PUBLIC_OTEL_COLLECT_FROM',
        description: 'The role types to collect metrics from',
        parseEnv: (val)=>val ? val.split(',').map((s)=>s.trim()).filter((s)=>s.length > 0) : [],
        defaultValue: []
    },
    publicIncludeMetrics: {
        env: 'PUBLIC_OTEL_INCLUDE_METRICS',
        description: 'A list of metric prefixes to publicly export',
        parseEnv: (val)=>val ? val.split(',').map((s)=>s.trim()).filter((s)=>s.length > 0) : [],
        defaultValue: []
    },
    publicMetricsOptOut: {
        env: 'PUBLIC_OTEL_OPT_OUT',
        description: 'Whether to opt out of sharing optional telemetry',
        ...booleanConfigHelper(true)
    }
};
export function getConfigEnvVars() {
    return getConfigFromMappings(telemetryClientConfigMappings);
}
