// @aztec/telemetry-client/dest/l1_metrics.js
import { formatEther } from 'viem';
import { L1_SENDER } from './attributes.js';
import { L1_BALANCE_ETH, L1_BLOB_BASE_FEE_WEI, L1_BLOCK_HEIGHT, L1_GAS_PRICE_WEI } from './metrics.js';
export class L1Metrics {
    meter;
    client;
    l1BlockHeight;
    l1BalanceEth;
    gasPriceWei;
    blobBaseFeeWei;
    addresses;
    constructor(meter, client, addresses){
        this.meter = meter;
        this.client = client;
        this.observe = async (observer)=>{
            const blockNumber = await this.client.getBlockNumber();
            const ethBalances = await Promise.all(this.addresses.map((address)=>this.client.getBalance({
                    address,
                    blockNumber
                })));
            const gasPrice = await this.client.getGasPrice();
            const blobFee = await this.client.getBlobBaseFee();
            observer.observe(this.l1BlockHeight, Number(blockNumber));
            observer.observe(this.gasPriceWei, Number(gasPrice));
            observer.observe(this.blobBaseFeeWei, Number(blobFee));
            for(let i = 0; i < ethBalances.length; i++){
                const wei = ethBalances[i];
                const address = this.addresses[i];
                const eth = parseFloat(formatEther(wei, 'wei'));
                observer.observe(this.l1BalanceEth, eth, {
                    [L1_SENDER]: address
                });
            }
        };
        this.l1BlockHeight = meter.createObservableGauge(L1_BLOCK_HEIGHT);
        this.l1BalanceEth = meter.createObservableGauge(L1_BALANCE_ETH);
        this.gasPriceWei = meter.createObservableGauge(L1_GAS_PRICE_WEI);
        this.blobBaseFeeWei = meter.createObservableGauge(L1_BLOB_BASE_FEE_WEI);
        this.addresses = addresses.map((addr)=>addr.toString());
    }
    start() {
        this.meter.addBatchObservableCallback(this.observe, [
            this.l1BalanceEth,
            this.l1BlockHeight
        ]);
    }
    stop() {
        this.meter.removeBatchObservableCallback(this.observe, [
            this.l1BalanceEth,
            this.l1BlockHeight
        ]);
    }
    observe;
}
