// @aztec/telemetry-client/dest/lmdb_metrics.js
import * as Metrics from './metrics.js';
export class LmdbMetrics {
    attributes;
    getStats;
    dbMapSize;
    dbPhysicalFileSize;
    dbUsedSize;
    dbNumItems;
    constructor(meter, attributes, getStats){
        this.attributes = attributes;
        this.getStats = getStats;
        this.recordDBMetrics = async (observable)=>{
            if (!this.getStats) {
                return;
            }
            const metrics = await this.getStats();
            observable.observe(this.dbMapSize, metrics.mappingSize, this.attributes);
            observable.observe(this.dbPhysicalFileSize, metrics.physicalFileSize, this.attributes);
            observable.observe(this.dbNumItems, metrics.numItems, this.attributes);
            observable.observe(this.dbUsedSize, metrics.actualSize, this.attributes);
        };
        this.dbMapSize = meter.createObservableGauge(Metrics.DB_MAP_SIZE);
        this.dbPhysicalFileSize = meter.createObservableGauge(Metrics.DB_PHYSICAL_FILE_SIZE);
        this.dbUsedSize = meter.createObservableGauge(Metrics.DB_USED_SIZE);
        this.dbNumItems = meter.createObservableGauge(Metrics.DB_NUM_ITEMS);
        meter.addBatchObservableCallback(this.recordDBMetrics, [
            this.dbMapSize,
            this.dbPhysicalFileSize,
            this.dbUsedSize,
            this.dbNumItems
        ]);
    }
    recordDBMetrics;
}
