// @aztec/telemetry-client/dest/noop.js
import { createNoopMeter } from '@opentelemetry/api';
/** A no-op meter that implements our custom Meter interface */ class NoopMeter {
    otelMeter = createNoopMeter();
    createGauge(_metric) {
        return this.otelMeter.createGauge('');
    }
    createObservableGauge(_metric) {
        return this.otelMeter.createObservableGauge('');
    }
    createHistogram(_metric, _extraOptions) {
        return this.otelMeter.createHistogram('');
    }
    createUpDownCounter(_metric) {
        return this.otelMeter.createUpDownCounter('');
    }
    createObservableUpDownCounter(_metric) {
        return this.otelMeter.createObservableUpDownCounter('');
    }
    addBatchObservableCallback(callback, observables) {
        this.otelMeter.addBatchObservableCallback(callback, observables);
    }
    removeBatchObservableCallback(callback, observables) {
        this.otelMeter.removeBatchObservableCallback(callback, observables);
    }
}
export class NoopTelemetryClient {
    meter = new NoopMeter();
    setExportedPublicTelemetry(_prefixes) {}
    setPublicTelemetryCollectFrom(_roles) {}
    getMeter() {
        return this.meter;
    }
    getTracer() {
        return new NoopTracer();
    }
    stop() {
        return Promise.resolve();
    }
    flush() {
        return Promise.resolve();
    }
    isEnabled() {
        return false;
    }
    getTraceContext() {
        return undefined;
    }
    extractPropagatedContext(_traceContext) {
        return undefined;
    }
}
// @opentelemetry/api internally uses NoopTracer and NoopSpan but they're not exported
// make our own versions
// https://github.com/open-telemetry/opentelemetry-js/issues/4518#issuecomment-2179405444
export class NoopTracer {
    startSpan() {
        return new NoopSpan();
    }
    startActiveSpan(_name, ...args) {
        // there are three different signatures for startActiveSpan, grab the function, we don't care about the rest
        const fn = args.find((arg)=>typeof arg === 'function');
        return fn(new NoopSpan());
    }
}
class NoopSpan {
    recording = true;
    addEvent() {
        return this;
    }
    addLink() {
        return this;
    }
    addLinks() {
        return this;
    }
    end() {
        this.recording = false;
    }
    isRecording() {
        return this.recording;
    }
    recordException() {
        return;
    }
    setAttribute() {
        return this;
    }
    setAttributes() {
        return this;
    }
    setStatus() {
        return this;
    }
    spanContext() {
        return {
            spanId: '',
            traceId: '',
            traceFlags: 0
        };
    }
    updateName() {
        return this;
    }
}
