// @aztec/telemetry-client/dest/otel_filter_metric_exporter.js
import { ExportResultCode } from '@opentelemetry/core';
import { AZTEC_NODE_ROLE } from './attributes.js';
export class OtelFilterMetricExporter {
    exporter;
    metricPrefix;
    filter;
    constructor(exporter, metricPrefix, filter = 'deny'){
        this.exporter = exporter;
        this.metricPrefix = metricPrefix;
        this.filter = filter;
        if (exporter.selectAggregation) {
            this.selectAggregation = exporter.selectAggregation.bind(exporter);
        }
        if (exporter.selectAggregationTemporality) {
            this.selectAggregationTemporality = exporter.selectAggregationTemporality.bind(exporter);
        }
    }
    export(metrics, resultCallback) {
        const filteredMetrics = {
            resource: metrics.resource,
            scopeMetrics: metrics.scopeMetrics.map(({ scope, metrics })=>({
                    scope,
                    metrics: this.filterMetrics(metrics)
                })).filter(({ metrics })=>metrics.length > 0)
        };
        this.exporter.export(filteredMetrics, resultCallback);
    }
    filterMetrics(metrics) {
        return metrics.filter((metric)=>{
            const matched = this.metricPrefix.some((prefix)=>metric.descriptor.name.startsWith(prefix));
            if (this.filter === 'deny') {
                return !matched;
            }
            if (this.filter === 'allow') {
                return matched;
            }
        });
    }
    forceFlush() {
        return this.exporter.forceFlush();
    }
    shutdown() {
        return this.exporter.shutdown();
    }
    setMetricPrefixes(metrics) {
        this.metricPrefix = metrics;
    }
}
export class PublicOtelFilterMetricExporter extends OtelFilterMetricExporter {
    allowedRoles;
    constructor(allowedRoles, exporter, metricPrefix){
        super(exporter, metricPrefix, 'allow'), this.allowedRoles = allowedRoles;
    }
    export(metrics, resultCallback) {
        const role = String(metrics.resource.attributes[AZTEC_NODE_ROLE] ?? '');
        if (!role || !this.allowedRoles.includes(role)) {
            // noop
            return resultCallback({
                code: ExportResultCode.SUCCESS
            });
        }
        super.export(metrics, resultCallback);
    }
    setAllowedRoles(roles) {
        this.allowedRoles = roles;
    }
}
