// @aztec/stdlib/dest/validators/errors.js
export class ValidatorError extends Error {
    constructor(message){
        super(`Validator Error: ${message}`);
    }
}
export class InvalidValidatorPrivateKeyError extends ValidatorError {
    constructor(){
        super('Invalid validator private key provided');
    }
}
export class AttestationTimeoutError extends ValidatorError {
    collectedCount;
    requiredCount;
    slot;
    constructor(collectedCount, requiredCount, slot){
        super(`Timeout collecting attestations for slot ${slot}: ${collectedCount}/${requiredCount}`), this.collectedCount = collectedCount, this.requiredCount = requiredCount, this.slot = slot;
    }
}
export class TransactionsNotAvailableError extends ValidatorError {
    constructor(txHashes){
        super(`Transactions not available: ${txHashes.join(', ')}`);
    }
}
export class FailedToReExecuteTransactionsError extends ValidatorError {
    constructor(txHashes){
        super(`Failed to re-execute transactions: ${txHashes.join(', ')}`);
    }
}
export class ReExStateMismatchError extends ValidatorError {
    expectedArchiveRoot;
    actualArchiveRoot;
    constructor(expectedArchiveRoot, actualArchiveRoot){
        super('Re-execution state mismatch'), this.expectedArchiveRoot = expectedArchiveRoot, this.actualArchiveRoot = actualArchiveRoot;
    }
}
export class ReExFailedTxsError extends ValidatorError {
    constructor(numFailedTxs){
        super(`Re-execution failed to process ${numFailedTxs} txs`);
    }
}
export class ReExTimeoutError extends ValidatorError {
    constructor(){
        super('Re-execution timed out or failed to process all txs in the proposal');
    }
}
