// @aztec/ethereum/src/generated/l1-contracts-defaults.ts
// Auto-generated from spartan/environments/network-defaults.yml
// Do not edit manually - run yarn generate to regenerate

/** Default L1 contracts configuration values from network-defaults.yml */
export const l1ContractsDefaultEnv = {
  ETHEREUM_SLOT_DURATION: 12,
  AZTEC_SLOT_DURATION: 36,
  AZTEC_EPOCH_DURATION: 32,
  AZTEC_TARGET_COMMITTEE_SIZE: 48,
  AZTEC_LAG_IN_EPOCHS_FOR_VALIDATOR_SET: 2,
  AZTEC_LAG_IN_EPOCHS_FOR_RANDAO: 2,
  AZTEC_ACTIVATION_THRESHOLD: 100000000000000000000,
  AZTEC_EJECTION_THRESHOLD: 50000000000000000000,
  AZTEC_LOCAL_EJECTION_THRESHOLD: 98000000000000000000,
  AZTEC_EXIT_DELAY_SECONDS: 172800,
  AZTEC_INBOX_LAG: 1,
  AZTEC_PROOF_SUBMISSION_EPOCHS: 1,
  AZTEC_MANA_TARGET: 100000000,
  AZTEC_PROVING_COST_PER_MANA: 100,
  AZTEC_INITIAL_ETH_PER_FEE_ASSET: 10000000,
  AZTEC_SLASHER_FLAVOR: 'tally',
  AZTEC_SLASHING_ROUND_SIZE_IN_EPOCHS: 4,
  AZTEC_SLASHING_LIFETIME_IN_ROUNDS: 5,
  AZTEC_SLASHING_EXECUTION_DELAY_IN_ROUNDS: 0,
  AZTEC_SLASHING_OFFSET_IN_ROUNDS: 2,
  AZTEC_SLASHING_VETOER: '0x0000000000000000000000000000000000000000',
  AZTEC_SLASHING_DISABLE_DURATION: 432000,
  AZTEC_SLASH_AMOUNT_SMALL: 10000000000000000000,
  AZTEC_SLASH_AMOUNT_MEDIUM: 20000000000000000000,
  AZTEC_SLASH_AMOUNT_LARGE: 50000000000000000000,
  AZTEC_GOVERNANCE_PROPOSER_ROUND_SIZE: 300,
} as const;
