// @aztec/stdlib/dest/snapshots/types.js
import { schemas, zodFor } from '@aztec/foundation/schemas';
import { z } from 'zod';
export const SnapshotDataKeys = [
    'archiver',
    'nullifier-tree',
    'public-data-tree',
    'note-hash-tree',
    'archive-tree',
    'l1-to-l2-message-tree'
];
export const SnapshotsIndexSchema = zodFor()(z.object({
    l1ChainId: z.number(),
    rollupVersion: z.number(),
    rollupAddress: schemas.EthAddress,
    snapshots: z.array(z.object({
        l2BlockNumber: z.number(),
        l2BlockHash: z.string(),
        l1BlockNumber: z.number(),
        timestamp: z.number(),
        schemaVersions: z.object({
            archiver: z.number(),
            worldState: z.number()
        }),
        dataUrls: z.record(z.enum(SnapshotDataKeys), z.string())// See https://stackoverflow.com/questions/77958464/zod-record-with-required-keys
        .refine((obj)=>SnapshotDataKeys.every((key)=>!!obj[key]))
    }))
}));
export const UploadSnapshotMetadataSchema = zodFor()(z.object({
    l2BlockNumber: z.number(),
    l2BlockHash: z.string(),
    l1BlockNumber: z.number(),
    l1ChainId: z.number(),
    rollupVersion: z.number(),
    rollupAddress: schemas.EthAddress
}));
