// @aztec/aztec.js/dest/authorization/call_authorization_request.js
import { FieldReader } from '@aztec/foundation/serialize';
import { AuthorizationSelector, FunctionSelector } from '@aztec/stdlib/abi';
import { AztecAddress } from '@aztec/stdlib/aztec-address';
/**
 * An authwit request for a function call. Includes the preimage of the data
 * to be signed, as opposed of just the inner hash.
 */ export class CallAuthorizationRequest {
    selector;
    innerHash;
    msgSender;
    functionSelector;
    argsHash;
    args;
    constructor(/**
     * The selector of the authwit type, used to identify it
     * when emitted from `emit_offchain_effect`oracle.
     * Computed as poseidon2("CallAuthwit((Field),(u32),Field)".to_bytes())
     */ selector, /**
     * The inner hash of the authwit, computed as
     * poseidon2([msg_sender, selector, args_hash])
     */ innerHash, /**
     * The address performing the call
     */ msgSender, /**
     * The selector of the function that is to be authorized
     * */ functionSelector, /**
     * The hash of the arguments to the function call,
     */ argsHash, /**
     * The arguments to the function call.
     */ args){
        this.selector = selector;
        this.innerHash = innerHash;
        this.msgSender = msgSender;
        this.functionSelector = functionSelector;
        this.argsHash = argsHash;
        this.args = args;
    }
    static getSelector() {
        return AuthorizationSelector.fromSignature('CallAuthorization((Field),(u32),Field)');
    }
    static async fromFields(fields) {
        const expectedSelector = await CallAuthorizationRequest.getSelector();
        const reader = FieldReader.asReader(fields);
        const selector = AuthorizationSelector.fromField(reader.readField());
        if (!selector.equals(expectedSelector)) {
            throw new Error(`Invalid authorization selector for CallAuthwit: expected ${expectedSelector.toString()}, got ${selector.toString()}`);
        }
        return new CallAuthorizationRequest(selector, reader.readField(), AztecAddress.fromField(reader.readField()), FunctionSelector.fromField(reader.readField()), reader.readField(), reader.readFieldArray(reader.remainingFields()));
    }
}
